{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allstart = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pprint\n",
    "import re\n",
    "from nltk.corpus import wordnet as wn\n",
    "from collections import deque\n",
    "import re\n",
    "from nltk.stem.wordnet import WordNetLemmatizer\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "HPO_FILE = 'data/all_hpo_terms_and_synonyms.txt'\n",
    "PHENOTYPE_FILE = 'data/sample_patient_phenotype.txt'\n",
    "PHENOTYPE_TO_GENE_FILE = 'data/Expanded_OMIM_ALL_FREQUENCIES_phenotype_to_genes_without_synonym.txt'\n",
    "PHENOTYPE_TO_DISEASE_FILE = 'data/Expanded_ALL_SOURCES_ALL_FREQUENCIES_diseases_to_phenotypes.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0100679397583\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "df_genes = pd.read_csv('data/sample_genes.txt', sep = '\\t')\n",
    "CANDIDATE_GENES = pd.unique(df_genes.Gene.values).tolist()\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "OUT_FILE = 'result/ranking_genes.txt'\n",
    "OUT_FILE_DISEASE = 'result/ranking_diseases.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.127974033356\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "hpo_subclass = dict()\n",
    "with open('data/hpo_subclasses.txt', 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        subclasses = parts[1].strip('[]').replace(\"'\", \"\")\n",
    "        subclasses = [_.strip() for _ in subclasses.split(',')]\n",
    "        hpo_subclass[hpoid] = subclasses\n",
    "\n",
    "hpo_levels = dict()\n",
    "level = 0\n",
    "root = 'HP:0000001'\n",
    "curr_level_nodes = [root] \n",
    "while curr_level_nodes:\n",
    "    next_level_nodes = []\n",
    "    for node in curr_level_nodes:\n",
    "        hpo_levels[node] = level\n",
    "        if node in hpo_subclass:\n",
    "            next_level_nodes += hpo_subclass[node]\n",
    "    curr_level_nodes = next_level_nodes\n",
    "    level += 1\n",
    "     \n",
    "\n",
    "hpo_superclass = dict()\n",
    "with open('data/hpo_superclasses.txt', 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        superclasses = parts[1].strip('[]').replace(\"'\", \"\")\n",
    "        superclasses = [_.strip() for _ in superclasses.split(',')]\n",
    "        hpo_superclass[hpoid] = superclasses\n",
    "\n",
    "hpo_name = dict()\n",
    "with open('data/hpo_names.txt', 'rb') as f:\n",
    "    # Skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        hponame = parts[1]\n",
    "        hpo_name[hpoid] = hponame        \n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tree = hpo_superclass\n",
    "def getAllAncestorsBFS(node, tree):\n",
    "    P, Q = [node], deque([node])\n",
    "    while Q:\n",
    "        u = Q.popleft()\n",
    "        if u not in tree:\n",
    "            continue\n",
    "        for v in tree[u]:\n",
    "            if v in P: continue\n",
    "            if v not in tree: continue\n",
    "            P.append(v)\n",
    "            Q.append(v)\n",
    "    # Append All ('HP:0000001') as the highest ancestor \n",
    "    P.append('HP:0000001')\n",
    "    return P \n",
    "\n",
    "def findLowestCommonAncestor(node1, node2, tree):\n",
    "    node1_ancestors = getAllAncestorsBFS(node1, tree) \n",
    "    node2_ancestors = getAllAncestorsBFS(node2, tree)\n",
    "    common_ancestors = [] \n",
    "    for ancestor in node2_ancestors:\n",
    "        if ancestor in node1_ancestors:\n",
    "            name = hpo_name[ancestor]\n",
    "            level = hpo_levels[ancestor]\n",
    "            common_ancestors.append((ancestor, level, name)) \n",
    "    lca = common_ancestors[0]\n",
    "    return lca, common_ancestors\n",
    "\n",
    "## Generate antonyms of a word using NLTK and wordnet\n",
    "def get_antonyms(word):\n",
    "    antonyms = []\n",
    "    for synset in wn.synsets(word):\n",
    "        for lemma in synset.lemmas():\n",
    "            antonyms += lemma.antonyms()\n",
    "            similar_synsets = synset.similar_tos() \n",
    "            for similar_synset in similar_synsets:\n",
    "                for lemma_ in similar_synset.lemmas():\n",
    "                    #print (similar_synset, lemma_, lemma_.name(), \n",
    "                    #      lemma_.antonyms(), similar_synset.similar_tos())\n",
    "                    antonyms += lemma_.antonyms()\n",
    "    direct_antonyms = list(set(antonyms))\n",
    "    # Get similar words of the direct antonyms\n",
    "    for antonym in direct_antonyms:\n",
    "        similar_synsets = antonym.synset().similar_tos()\n",
    "        for synset in similar_synsets:\n",
    "            for lemma in synset.lemmas():\n",
    "                antonyms.append(lemma)\n",
    "    antonyms = list(set(antonyms))\n",
    "    antonyms = set([_.name() for _ in antonyms])\n",
    "    return antonyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.898631811142\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "## Map patient phenotypes to hpo standard terms \n",
    "corner_cases = dict()\n",
    "with open(PHENOTYPE_FILE, 'rb') as infile:\n",
    "    phenos = []\n",
    "    for line in infile:\n",
    "        if line.startswith('#') or not line.strip():\n",
    "            continue\n",
    "        line = line.rstrip()\n",
    "        phenos += line.split(',')\n",
    "        for pheno in phenos:\n",
    "            if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                phenos.append('growth delay')\n",
    "                corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "            if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                phenos.append('developmental delay')\n",
    "                corner_cases['developmental delay'] = pheno.strip()\n",
    "    phenos = [_.strip() for _ in phenos]\n",
    "# print phenos\n",
    "\n",
    "\n",
    "with open(HPO_FILE, 'rb') as infile:\n",
    "    hpo_name2id = dict()\n",
    "    hpo_id2synonyms = dict()\n",
    "    # skip the first line\n",
    "    infile.readline()\n",
    "    for line in infile.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid, hponame = parts[0], parts[1]\n",
    "        hpo_name2id[hponame] = hpoid\n",
    "        hpoid = hpoid.split('-')[0]\n",
    "        if hponame in hpo_id2synonyms:\n",
    "            hpo_id2synonyms[hpoid].append(hponame)\n",
    "        else:\n",
    "            hpo_id2synonyms[hpoid] = [hponame]\n",
    "    hpo_terms = hpo_name2id.keys()  \n",
    "\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.005784034729\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "# Use re for split hpo terms\n",
    "# Use lemmatizer to lemmatize words e.g., disturbances --> distrubance\n",
    "def isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "    # If any word ends with ness, convert to its noun form\n",
    "    words_in_pheno_only |= set([_[0:-4] for _ in words_in_pheno_only if _.endswith('ness')])\n",
    "    words_in_hpo_only |= set([_[0:-4] for _ in words_in_hpo_only if _.endswith('ness')])\n",
    "    # corner case: gain\n",
    "    if 'gain' in words_in_hpo_only:\n",
    "        words_in_hpo_only.add('increased')\n",
    "        words_in_hpo_only.add('steady')\n",
    "    if 'gain' in words_in_pheno_only:\n",
    "        words_in_pheno_only.add('increased')\n",
    "        words_in_pheno_only.add('steady')\n",
    "    # Poor weight gain and weight gain are in opposite meaning\n",
    "    if (words_in_pheno_only == set(['poor']) and words_in_hpo_only == set()) or (words_in_hpo_only == set(['poor']) and words_in_pheno_only == set()):\n",
    "#         print \"Opposite meaning due to word 'poor'\", words_in_pheno_only, words_in_hpo_only\n",
    "        return True\n",
    "    for word in words_in_pheno_only:\n",
    "        try:\n",
    "            antonyms = get_antonyms(word)\n",
    "        except UnicodeDecodeError as e:\n",
    "            continue\n",
    "        if antonyms.intersection(words_in_hpo_only):\n",
    "#             print \"Opposite meaning: \", antonyms.intersection(words_in_hpo_only)\n",
    "            return True  \n",
    "    for word in words_in_hpo_only:\n",
    "        try:\n",
    "            antonyms = get_antonyms(word)\n",
    "        except UnicodeDecodeError as e:\n",
    "            continue\n",
    "        if antonyms.intersection(words_in_pheno_only):\n",
    "#             print \"Opposite meaning: \", antonyms.intersection(words_in_hpo_only)\n",
    "            return True  \n",
    "    return False \n",
    "\n",
    "\n",
    "# If the opposite meanings are found for a hpoid, then its synonyms are all in opposite meanings\n",
    "unmatches_due_to_opposite_meanings = []\n",
    "\n",
    "def map2hpo(pheno):\n",
    "    global unmatches_due_to_opposite_meanings\n",
    "    matches = []\n",
    "    # Use lemmatizer to lemmatize words e.g., disturbances --> distrubance\n",
    "    lemmatizer = WordNetLemmatizer()\n",
    "    for hpo in hpo_terms:\n",
    "        hpoid = hpo_name2id[hpo]\n",
    "        pheno_ = pheno.lower() \n",
    "        hpo_ = hpo.lower()\n",
    "        if hpo_.startswith('obsolete '):\n",
    "            continue \n",
    "        if pheno_ == hpo_:\n",
    "            matches.append((hpoid, hpo, 'EXACT', 1.0)) \n",
    "            continue\n",
    "        pheno_words = [_.strip(\",()'\") for _ in re.split(' |/', pheno_)]\n",
    "        hpo_words = [_.strip(\",()'\") for _ in re.split(' |/', hpo_)]\n",
    "\n",
    "        '''\n",
    "        # This is a corner case. Developmental delay means retarded and growth delay means short. \n",
    "        # But in plain language, they are exchangable.\n",
    "        for pheno_word in pheno_words:\n",
    "            if re.search('development', pheno_word, re.I):\n",
    "                pheno_words.append('growth') \n",
    "            if re.search('growth', pheno_word, re.I):\n",
    "                pheno_words.append('developmental') \n",
    "        '''\n",
    "        try:\n",
    "            pheno_words = [lemmatizer.lemmatize(word) for word in pheno_words]\n",
    "            hpo_words = [lemmatizer.lemmatize(word) for word in hpo_words]\n",
    "        except UnicodeDecodeError as e:\n",
    "            pass\n",
    "        if pheno_words == hpo_words:\n",
    "            matches.append((hpoid, hpo, 'LEMMA_MATCH', 1.0)) \n",
    "            continue\n",
    "        if set(pheno_words) == set(hpo_words):\n",
    "            matches.append((hpoid, hpo, 'REORDER', 1.0)) \n",
    "            continue\n",
    "\n",
    "        common_words = set(pheno_words) & set(hpo_words)\n",
    "        words_in_pheno_only = set(pheno_words) - common_words\n",
    "        words_in_hpo_only = set(hpo_words) - common_words\n",
    "        # check if the pheno and hpo are in opposite meaning\n",
    "        # if any word in pheno is an antonym of a word in hpo, or the other way\n",
    "        # we skip this hpo\n",
    "        sim = float(len(common_words)) / max(len(hpo_words), len(pheno_words))  \n",
    "        if (sim >= 0.3333) and (pheno_ in hpo_) and (len(hpo_words) <= 10 * len(pheno_words)):\n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'SUB', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "            continue\n",
    "        if (sim >= 0.3333) and (hpo_ in pheno_) and (len(pheno_words) <= 10 * len(hpo_words)): \n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'SUPER', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "            continue\n",
    "        if (sim >= 0.5) and len(pheno_words) >= 2:\n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'OVERLAP', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "    matches_res = []\n",
    "    for match in matches:\n",
    "        if match[0].split('-')[0] not in unmatches_due_to_opposite_meanings:\n",
    "            matches_res.append(match)\n",
    "    matches = matches_res\n",
    "    return matches\n",
    "\n",
    "def filterMatchesOnCommonAncestors(matches):\n",
    "    ## e.g., for \"Central hypotonia\", matches is like: \n",
    "    ## [('Central', 'SUPER', 0.5), ('Central hypotonia', 'EXACT', '1.0'), ('Hypotonia', 'SUPER', 0.5)]\n",
    "    ## lca is like ('HP:0005872', 9, 'Brachytelomesophalangy')\n",
    "    ## The function returns like: [('Stereotypical hand wringing', 'HP:0012171', 0.6666666666666666)]\n",
    "    if not matches:\n",
    "        return matches \n",
    "    final_matches = []\n",
    "    # If the match is a disease not a phenotype, then we just add it to the final matches result\n",
    "    matches_tmp = []\n",
    "    for match in matches:\n",
    "        if not match[0].startswith('HP:'):\n",
    "            final_matches.append((match[1], match[0], match[3]))\n",
    "        else:\n",
    "            matches_tmp.append(match)\n",
    "    matches = matches_tmp\n",
    "    if not matches:\n",
    "        return final_matches\n",
    "    best_match_hpo = sorted(matches, key = lambda x: x[3], reverse = True)[0][0]\n",
    "    for match in matches:\n",
    "        term = match[0]\n",
    "        # Remove 'synonym' from the hpoid (term) e.g., HP:0004414-synonym --> HP:0004414\n",
    "        term = term.split('-')[0]\n",
    "        lca, common_ancestors = findLowestCommonAncestor(term, best_match_hpo.split('-')[0], hpo_superclass)\n",
    "        if lca[1] >= 2:\n",
    "            final_matches.append((match[1], match[0], match[3])) \n",
    "    return final_matches    \n",
    "\n",
    "def map2hpoWithPhenoSynonyms(pheno):\n",
    "    global unmatches_due_to_opposite_meanings\n",
    "    matches = map2hpo(pheno)\n",
    "    direct_matches = filterMatchesOnCommonAncestors(matches)  \n",
    "    final_matches = direct_matches \n",
    "    for match in direct_matches:\n",
    "        hponame = match[0]\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if sim >= 0.75:\n",
    "            hposynonyms = hpo_id2synonyms[hpoid]\n",
    "            for synonym in hposynonyms:\n",
    "                if synonym != hponame:\n",
    "#                     print synonym\n",
    "                    matches_ = map2hpo(synonym)\n",
    "                    indirect_matches = filterMatchesOnCommonAncestors(matches_)\n",
    "                    final_matches = final_matches + indirect_matches\n",
    "    final_matches = list(set(final_matches))\n",
    "    matches_res = []\n",
    "    for match in final_matches:\n",
    "        if match[0].split('-')[0] not in unmatches_due_to_opposite_meanings:\n",
    "            matches_res.append(match)\n",
    "    final_matches = matches_res\n",
    "    unmatches_due_to_opposite_meanings = []\n",
    "    return final_matches\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.22824883461\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "hpoid2gene = dict()\n",
    "with open(PHENOTYPE_TO_GENE_FILE, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        geneid, genesymbol, hponame, hpoid = parts[0], parts[1], parts[2], parts[3]\n",
    "        if hpoid in hpoid2gene:\n",
    "            hpoid2gene[hpoid].append(genesymbol)   \n",
    "        else:\n",
    "            hpoid2gene[hpoid] = [genesymbol]\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def map2gene(final_matches):\n",
    "    # Map to genes for each phenotype (one phenotype at one time)\n",
    "    # Pre-process the final_matches by remove '-synonym' from the hpoid and then unique the hpoids\n",
    "    final_matches = [_[1] for _ in final_matches] \n",
    "    final_matches = [_.split('-')[0] for _ in final_matches] \n",
    "    final_matches = list(set(final_matches))\n",
    "    mapped_genes = []\n",
    "    for hpoid in final_matches:\n",
    "        try:\n",
    "            mapped_genes += hpoid2gene[hpoid] \n",
    "        except KeyError:\n",
    "            pass\n",
    "    ## For each phenotype in patient record, we generated multiple phenotype keywords \n",
    "    ## to maximize the mapping to the genes. However, we only count it once for each \n",
    "    ## phenotype even if the multiple keywords lead to multiple mappings.\n",
    "    ## In practice, we don't need to screen through all genes. Only a few gene candidates are examined.\n",
    "    mapped_genes = list(set(mapped_genes) & set(CANDIDATE_GENES))\n",
    "    return mapped_genes\n",
    "\n",
    "def map2geneWithSim(final_matches):\n",
    "    # Map to genes for each phenotype (one phenotype at one time) with similarity information (word len)\n",
    "    # final_matches like: [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "    # map2gene function returns a list while this function returns a dict with genesymbol as key and similarity as value \n",
    "    hpoid_sim = dict() \n",
    "    for match in final_matches:\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if hpoid not in hpoid_sim or sim > hpoid_sim[hpoid]:\n",
    "            hpoid_sim[hpoid] = sim\n",
    "    mapped_genes_score = dict()\n",
    "    for hpoid in hpoid_sim:\n",
    "        try:\n",
    "            mapped_genes = hpoid2gene[hpoid]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        sim = hpoid_sim[hpoid]\n",
    "        for gene in mapped_genes:\n",
    "            # In practice, we don't need to screen through all genes. \n",
    "            # Only a few gene candidates are examined.\n",
    "            if gene not in CANDIDATE_GENES:\n",
    "                continue\n",
    "            if gene not in mapped_genes_score or sim > mapped_genes_score[gene]:\n",
    "                mapped_genes_score[gene] = sim\n",
    "    return mapped_genes_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.259385824203\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "\n",
    "hpoid2disease = dict()\n",
    "numphenosindisease = dict()\n",
    "with open(PHENOTYPE_TO_DISEASE_FILE, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        diseaseid, hpoid, hponame, diseasename = parts[0], parts[1], parts[2], parts[3]\n",
    "        if diseasename in numphenosindisease:\n",
    "            numphenosindisease[diseasename] += 1\n",
    "        else:\n",
    "            numphenosindisease[diseasename] = 1\n",
    "        if hpoid in hpoid2disease:\n",
    "            hpoid2disease[hpoid].append(diseasename)\n",
    "        else:\n",
    "            hpoid2disease[hpoid] = [diseasename]\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def map2disease(final_matches):\n",
    "    # Map to diseases for each phenotype (one phenotype at one time)\n",
    "    # Pre-process the final_matches by remove '-synonym' from the hpoid and then unique the hpoids\n",
    "    final_matches = [_[1] for _ in final_matches] \n",
    "    final_matches = [_.split('-')[0] for _ in final_matches] \n",
    "    final_matches = list(set(final_matches))\n",
    "    mapped_diseases = []\n",
    "    for hpoid in final_matches:\n",
    "        try:\n",
    "            mapped_diseases += hpoid2disease[hpoid] \n",
    "        except KeyError:\n",
    "            pass\n",
    "    ## For each phenotype in patient record, we generated multiple phenotype keywords \n",
    "    ## to maximize the mapping to the genes. However, we only count it once for each \n",
    "    ## phenotype even if the multiple keywords lead to multiple mappings.\n",
    "    ## In practice, we don't need to screen through all genes. Only a few gene candidates are examined.\n",
    "    #mapped_genes = list(set(mapped_genes) & set(CANDIDATE_GENES))\n",
    "    return mapped_diseases\n",
    "\n",
    "def map2diseaseWithSim(final_matches):\n",
    "    # Map to diseases for each phenotype (one phenotype at one time) with similarity information (word len)\n",
    "    # final_matches like: [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "    # map2disease function returns a list while this function returns a dict with diseasename as key and similarity as value \n",
    "    hpoid_sim = dict() \n",
    "    for match in final_matches:\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if hpoid not in hpoid_sim or sim > hpoid_sim[hpoid]:\n",
    "            hpoid_sim[hpoid] = sim\n",
    "    mapped_diseases_score = dict()\n",
    "    for hpoid in hpoid_sim:\n",
    "        try:\n",
    "            mapped_diseases = hpoid2disease[hpoid]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        sim = hpoid_sim[hpoid]\n",
    "        for disease in mapped_diseases:\n",
    "            # In practice, we don't need to screen through all genes. \n",
    "            # Only a few gene candidates are examined.\n",
    "            #if disease not in CANDIDATE_DISEASES:\n",
    "            #    continue\n",
    "            if disease not in mapped_diseases_score or sim > mapped_diseases_score[disease]:\n",
    "                mapped_diseases_score[disease] = sim\n",
    "    return mapped_diseases_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "260.649292946\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "\n",
    "all_mapped_genes = []\n",
    "all_mapped_genes_score = {}\n",
    "gene_phenos = {} \n",
    "all_mapped_diseases = []\n",
    "all_mapped_diseases_score = {}\n",
    "disease_phenos = {} \n",
    "\n",
    "for pheno in phenos:\n",
    "#     print (\"===========================================================================================\")\n",
    "#     pprint.pprint(pheno)\n",
    "    matches =  map2hpo(pheno) \n",
    "    final_matches = map2hpoWithPhenoSynonyms(pheno)\n",
    "#     pprint.pprint(final_matches)\n",
    "    mapped_genes = map2gene(final_matches)\n",
    "#     pprint.pprint(mapped_genes)\n",
    "    all_mapped_genes += mapped_genes\n",
    "    mapped_genes_score = map2geneWithSim(final_matches)\n",
    "    if pheno in corner_cases:\n",
    "        pheno = corner_cases[pheno]\n",
    "    for gene in mapped_genes_score:\n",
    "        if gene not in all_mapped_genes_score:\n",
    "            all_mapped_genes_score[gene] = mapped_genes_score[gene]\n",
    "        else:\n",
    "            all_mapped_genes_score[gene] += mapped_genes_score[gene]\n",
    "        if gene not in gene_phenos:\n",
    "            gene_phenos[gene] = [pheno]\n",
    "        else:\n",
    "            gene_phenos[gene].append(pheno)\n",
    "\n",
    "    mapped_diseases = map2disease(final_matches)\n",
    "#     pprint.pprint(mapped_diseases)\n",
    "    all_mapped_diseases += mapped_diseases\n",
    "    mapped_diseases_score = map2diseaseWithSim(final_matches)\n",
    "    for disease in mapped_diseases_score:\n",
    "        if disease not in all_mapped_diseases_score:\n",
    "            all_mapped_diseases_score[disease] = mapped_diseases_score[disease]\n",
    "        else:\n",
    "            all_mapped_diseases_score[disease] += mapped_diseases_score[disease]\n",
    "        if disease not in disease_phenos:\n",
    "            disease_phenos[disease] = [pheno]\n",
    "        else:\n",
    "            disease_phenos[disease].append(pheno)\n",
    "            \n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0956249237061\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "\n",
    "num_patient_phenos = len(phenos)\n",
    "for disease in all_mapped_diseases_score:\n",
    "    numphenos = numphenosindisease[disease]\n",
    "    disease_score = all_mapped_diseases_score[disease]\n",
    "#     print disease, numphenos, num_patient_phenos, disease_score\n",
    "    # algo to calculate the disease score\n",
    "    all_mapped_diseases_score[disease] = disease_score ** 2 / float(numphenos) / (1 + num_patient_phenos - disease_score)\n",
    "\n",
    "all_mapped_genes_score = [(k,v) for k, v in all_mapped_genes_score.items()]   \n",
    "all_mapped_genes_score = sorted(all_mapped_genes_score, key = lambda pair: pair[1], reverse = True)\n",
    "all_mapped_diseases_score = [(k,v) for k, v in all_mapped_diseases_score.items()]   \n",
    "all_mapped_diseases_score = sorted(all_mapped_diseases_score, key = lambda pair: pair[1], reverse = True)\n",
    "\n",
    "count = Counter(all_mapped_genes)\n",
    "count = sorted(count.items(), key = lambda pair: pair[1], reverse = True)\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Count of matches to phenos: \"\n",
    "#pprint.pprint(count[0:])\n",
    "#print len(count)\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Score of matches to phenos: \"\n",
    "#pprint.pprint(all_mapped_genes_score[0:])\n",
    "#print len(all_mapped_genes_score)\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Matched phenos: \"\n",
    "#pprint.pprint(gene_phenos)\n",
    "\n",
    "#print \"============================================================================================\"\n",
    "#print \"============================================================================================\"\n",
    "count_disease = Counter(all_mapped_diseases)\n",
    "count_disease = sorted(count_disease.items(), key = lambda pair: pair[1], reverse = True)\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Count of matches to phenos: \"\n",
    "#pprint.pprint(count_disease[0:])\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Score of matches to phenos: \"\n",
    "#pprint.pprint(all_mapped_diseases_score[0:])\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"********************************************************************************************\"\n",
    "#print \"Matched phenos: \"\n",
    "#pprint.pprint(disease_phenos)\n",
    "\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.879236936569\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "with open(OUT_FILE, 'wb') as f:\n",
    "    f.write('gene\\tscore\\thits\\n')\n",
    "    for id in xrange(len(all_mapped_genes_score)):\n",
    "        gene, score = all_mapped_genes_score[id]\n",
    "        hits = None\n",
    "        for gene_hits in count:\n",
    "            if gene_hits[0] == gene:\n",
    "                hits = gene_hits[1] \n",
    "                break\n",
    "        if hits: \n",
    "            f.write(gene + '\\t' + str(score) + '\\t' + str(hits) + '\\n')\n",
    "\n",
    "with open(OUT_FILE_DISEASE, 'wb') as f:\n",
    "    f.write('disease\\tscore\\thits\\n')\n",
    "    for id in xrange(len(all_mapped_diseases_score)):\n",
    "        disease, score = all_mapped_diseases_score[id]\n",
    "        hits = None\n",
    "        for disease_hits in count_disease:\n",
    "            if disease_hits[0] == disease:\n",
    "                hits = disease_hits[1] \n",
    "                break\n",
    "        if hits:\n",
    "            f.write(disease + '\\t' + str(score) + '\\t' + str(hits) + '\\n')\n",
    "            \n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "267.042837858\n"
     ]
    }
   ],
   "source": [
    "print(end - allstart)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(phenos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "map2hpo 2.3309879303\n",
      "map2hpoWithPhenoSynonyms 2.36133313179\n",
      "map2gene 4.6968460083e-05\n",
      "map2geneWithSim 5.96046447754e-06\n",
      "mapped_genes_score 9.53674316406e-07\n",
      "map2disease 3.91006469727e-05\n",
      "map2diseaseWithSim 1.50203704834e-05\n",
      "all_mapped_diseases_score 9.53674316406e-07\n"
     ]
    }
   ],
   "source": [
    "all_mapped_genes = []\n",
    "all_mapped_genes_score = {}\n",
    "gene_phenos = {} \n",
    "all_mapped_diseases = []\n",
    "all_mapped_diseases_score = {}\n",
    "disease_phenos = {} \n",
    "\n",
    "for pheno in phenos:\n",
    "\n",
    "    start = time.time()\n",
    "    matches = map2hpo(pheno) \n",
    "    end = time.time()\n",
    "    print 'map2hpo', (end - start)\n",
    "    \n",
    "    start = time.time()\n",
    "    final_matches = map2hpoWithPhenoSynonyms(pheno)\n",
    "    end = time.time()\n",
    "    print 'map2hpoWithPhenoSynonyms', (end - start)\n",
    "    \n",
    "    start = time.time()\n",
    "    mapped_genes = map2gene(final_matches)\n",
    "    end = time.time()\n",
    "    print 'map2gene', (end - start)\n",
    "    \n",
    "    all_mapped_genes += mapped_genes\n",
    "    \n",
    "    start = time.time()\n",
    "    mapped_genes_score = map2geneWithSim(final_matches)\n",
    "    end = time.time()\n",
    "    print 'map2geneWithSim', (end - start)\n",
    "    \n",
    "    start = time.time()\n",
    "    if pheno in corner_cases:\n",
    "        pheno = corner_cases[pheno]\n",
    "    for gene in mapped_genes_score:\n",
    "        if gene not in all_mapped_genes_score:\n",
    "            all_mapped_genes_score[gene] = mapped_genes_score[gene]\n",
    "        else:\n",
    "            all_mapped_genes_score[gene] += mapped_genes_score[gene]\n",
    "        if gene not in gene_phenos:\n",
    "            gene_phenos[gene] = [pheno]\n",
    "        else:\n",
    "            gene_phenos[gene].append(pheno)\n",
    "    end = time.time()\n",
    "    print 'mapped_genes_score', (end - start)\n",
    "    \n",
    "    start = time.time()\n",
    "    mapped_diseases = map2disease(final_matches)\n",
    "    end = time.time()\n",
    "    print 'map2disease', (end - start)\n",
    "    \n",
    "    all_mapped_diseases += mapped_diseases\n",
    "    \n",
    "    start = time.time()\n",
    "    mapped_diseases_score = map2diseaseWithSim(final_matches)\n",
    "    end = time.time()\n",
    "    print 'map2diseaseWithSim', (end - start)\n",
    "    \n",
    "    start = time.time()\n",
    "    for disease in mapped_diseases_score:\n",
    "        if disease not in all_mapped_diseases_score:\n",
    "            all_mapped_diseases_score[disease] = mapped_diseases_score[disease]\n",
    "        else:\n",
    "            all_mapped_diseases_score[disease] += mapped_diseases_score[disease]\n",
    "        if disease not in disease_phenos:\n",
    "            disease_phenos[disease] = [pheno]\n",
    "        else:\n",
    "            disease_phenos[disease].append(pheno)\n",
    "    end = time.time()\n",
    "    print 'all_mapped_diseases_score', (end - start)\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "165.858657837\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "import map_phenotype_to_gene\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "df_genes = pd.read_csv(input_genes, sep = '\\t')\n",
    "CANDIDATE_GENES = pd.unique(df_genes.Gene.values).tolist()\n",
    "\n",
    "corner_cases = dict()\n",
    "with open(input_phenotype, 'rb') as infile:\n",
    "    phenos = []\n",
    "    for line in infile:\n",
    "        if line.startswith('#') or not line.strip():\n",
    "            continue\n",
    "        line = line.rstrip()\n",
    "        phenos += line.split(',')\n",
    "        for pheno in phenos:\n",
    "            if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                phenos.append('growth delay')\n",
    "                corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "            if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                phenos.append('developmental delay')\n",
    "                corner_cases['developmental delay'] = pheno.strip()\n",
    "    phenos = [_.strip() for _ in phenos]\n",
    "\n",
    "ranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\n",
    "end = time.time()\n",
    "print end-start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3322"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ranking_disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# ranking_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# np.unique([i[0] for i in ranking_genes]).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "start = time.time()\n",
    "\n",
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "\n",
    "\n",
    "input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "df_genes = pd.read_csv(input_genes, sep = '\\t', usecols = [0])\n",
    "CANDIDATE_GENES = pd.unique(df_genes.Gene.values).tolist()\n",
    "\n",
    "corner_cases = dict()\n",
    "with open(input_phenotype, 'rb') as infile:\n",
    "    phenos = []\n",
    "    for line in infile:\n",
    "        if line.startswith('#') or not line.strip():\n",
    "            continue\n",
    "        line = line.rstrip()\n",
    "        phenos += line.split(',')\n",
    "        for pheno in phenos:\n",
    "            if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                phenos.append('growth delay')\n",
    "                corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "            if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                phenos.append('developmental delay')\n",
    "                corner_cases['developmental delay'] = pheno.strip()\n",
    "    phenos = [_.strip() for _ in phenos]\n",
    "\n",
    "mv = myvariant.MyVariantInfo()\n",
    "\n",
    "ranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\n",
    "hpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "candidate_vars = []\n",
    "with open(input_genes, 'rb') as f:\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        gene = parts[0]\n",
    "        if gene not in hpo_filtered_genes:\n",
    "            continue\n",
    "        for part in parts:\n",
    "            if re.search(r'_.*:', part):\n",
    "                transcript, variant = part.split(':') \n",
    "        candidate_vars.append((gene, variant, transcript))\n",
    "\n",
    "df_final_res1 = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "end = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "\n",
    "\n",
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "\n",
    "#input \n",
    "input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "df_genes = pd.read_csv(input_genes, sep = '\\t', usecols = [0])\n",
    "CANDIDATE_GENES = pd.unique(df_genes.Gene.values).tolist()\n",
    "\n",
    "corner_cases = dict()\n",
    "with open(input_phenotype, 'rb') as infile:\n",
    "    phenos = []\n",
    "    for line in infile:\n",
    "        if line.startswith('#') or not line.strip():\n",
    "            continue\n",
    "        line = line.rstrip()\n",
    "        phenos += line.split(',')\n",
    "        for pheno in phenos:\n",
    "            if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                phenos.append('growth delay')\n",
    "                corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "            if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                phenos.append('developmental delay')\n",
    "                corner_cases['developmental delay'] = pheno.strip()\n",
    "    phenos = [_.strip() for _ in phenos]\n",
    "\n",
    "\n",
    "# map phenotype to gene\n",
    "mv = myvariant.MyVariantInfo()\n",
    "\n",
    "ranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "# collect variant info\n",
    "hpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "candidate_vars = []\n",
    "with open(input_genes, 'rb') as f:\n",
    "\tf.readline()\n",
    "\tfor line in f.readlines():\n",
    "\t\tline = line.rstrip()\n",
    "\t\tparts = line.split('\\t')\n",
    "\t\tgene = parts[0]\n",
    "\t\tif gene not in hpo_filtered_genes:\n",
    "\t\t\tcontinue\n",
    "\t\tfor part in parts:\n",
    "\t\t\tif re.search(r'_.*:', part):\n",
    "\t\t  \t\ttranscript, variant = part.split(':') \n",
    "\t\tcandidate_vars.append((gene, variant, transcript))\n",
    "\n",
    "final_res = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "t2 = time.time()\n",
    "\n",
    "# pubmed\n",
    "pubmed_query_results = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "\n",
    "\n",
    "t3 = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78.045254945755"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t3-t0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>16.333333</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNA</td>\n",
       "      <td>14.833333</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FANCL</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GNAS</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FANCM</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MEGF8</td>\n",
       "      <td>10.666667</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CAMKMT</td>\n",
       "      <td>9.500000</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WWOX</td>\n",
       "      <td>9.500000</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PRKCSH</td>\n",
       "      <td>9.333333</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>WNT7A</td>\n",
       "      <td>8.500000</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>8.166667</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PGAP3</td>\n",
       "      <td>7.833333</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SMPD1</td>\n",
       "      <td>7.666667</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ITGB6</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>WAS</td>\n",
       "      <td>6.333333</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CBS</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CITED2</td>\n",
       "      <td>5.833333</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NPHP4</td>\n",
       "      <td>5.166667</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>LYST</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NEB</td>\n",
       "      <td>4.833333</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GJC2</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>DYNC1H1</td>\n",
       "      <td>4.333333</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CYP21A2</td>\n",
       "      <td>4.166667</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ABCA4</td>\n",
       "      <td>4.166667</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>RTTN</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>UBAC2</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CACNA1D</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PDSS1</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>EXOSC8</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CLCNKA</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ATP2A2</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HFE</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CDH23</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>TP53</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SHROOM4</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ACACA</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ARMC5</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MYO7A</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>H6PD</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>MPO</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>LIPE</td>\n",
       "      <td>1.666667</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MYH2</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NOD2</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>TMC8</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>ATN1</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MMP1</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>GP1BA</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>GSN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>OSTM1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>F11</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>DNAJB2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>SALL2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NRXN1</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene      score  hits\n",
       "0      FLNA  16.333333    21\n",
       "1      LMNA  14.833333    21\n",
       "2     FANCL  14.000000    17\n",
       "3      GNAS  14.000000    20\n",
       "4     FANCM  13.000000    16\n",
       "5     MEGF8  10.666667    15\n",
       "6    CAMKMT   9.500000    11\n",
       "7      WWOX   9.500000    13\n",
       "8    PRKCSH   9.333333    13\n",
       "9      PLEC   9.000000    13\n",
       "10    WNT7A   8.500000    12\n",
       "11    CCBE1   8.166667    10\n",
       "12    PGAP3   7.833333    10\n",
       "13    SMPD1   7.666667     9\n",
       "14    ITGB6   7.500000    10\n",
       "15      WAS   6.333333    11\n",
       "16      CBS   6.000000     8\n",
       "17   CITED2   5.833333     8\n",
       "18    NPHP4   5.166667     6\n",
       "19     LYST   5.000000     7\n",
       "20      NEB   4.833333     7\n",
       "21     GJC2   4.500000     6\n",
       "22  DYNC1H1   4.333333     7\n",
       "23  CYP21A2   4.166667     5\n",
       "24    ABCA4   4.166667     6\n",
       "25     RTTN   4.000000     5\n",
       "26    UBAC2   4.000000     7\n",
       "27  CACNA1D   3.500000     4\n",
       "28    PDSS1   3.500000     6\n",
       "29   COL4A1   3.500000     4\n",
       "30   EXOSC8   3.500000     4\n",
       "31   CLCNKA   3.333333     5\n",
       "32   ATP2A2   3.000000     5\n",
       "33      HFE   3.000000     5\n",
       "34    CDH23   3.000000     3\n",
       "35     TP53   3.000000     5\n",
       "36  SHROOM4   2.666667     3\n",
       "37    ACACA   2.666667     3\n",
       "38    ARMC5   2.666667     4\n",
       "39    MYO7A   2.500000     3\n",
       "40     H6PD   2.000000     4\n",
       "41      MPO   2.000000     4\n",
       "42     LIPE   1.666667     3\n",
       "43     MYH2   1.500000     2\n",
       "44     NOD2   1.500000     3\n",
       "45     TMC8   1.500000     3\n",
       "46     ATN1   1.500000     2\n",
       "47  CACNA1A   1.500000     2\n",
       "48     MMP1   1.500000     2\n",
       "49    GP1BA   1.000000     2\n",
       "50      GSN   1.000000     2\n",
       "51    OSTM1   1.000000     2\n",
       "52      F11   0.500000     1\n",
       "53   DNAJB2   0.500000     1\n",
       "54    SALL2   0.500000     1\n",
       "55    NRXN1   0.500000     1"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(ranking_genes, columns=['gene','score','hits'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t1 = time.time()\n",
    "\n",
    "input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "import main\n",
    "acmg_result= main.master_function(input_phenotype,input_genes)\n",
    "\n",
    "t2 = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164.84264588356018"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2-t1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CBS</td>\n",
       "      <td>c.833T&gt;C</td>\n",
       "      <td>chr21:g.44483184A&gt;G</td>\n",
       "      <td>7.783641</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM5|PP3|PP5</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXOSC8</td>\n",
       "      <td>c.815G&gt;C</td>\n",
       "      <td>chr13:g.37583420G&gt;C</td>\n",
       "      <td>7.708397</td>\n",
       "      <td>5.125000</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>PS1|PM5|PM1|PP3|PP2|PP5</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEGF8</td>\n",
       "      <td>c.8180G&gt;A</td>\n",
       "      <td>chr19:g.42880569G&gt;A</td>\n",
       "      <td>5.527655</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>10.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CLCNKA</td>\n",
       "      <td>c.910C&gt;T</td>\n",
       "      <td>chr1:g.16354556C&gt;T</td>\n",
       "      <td>5.132180</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PVS1|PP3</td>\n",
       "      <td>3.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGAP3</td>\n",
       "      <td>c.67G&gt;A</td>\n",
       "      <td>chr17:g.37844201C&gt;T</td>\n",
       "      <td>5.083242</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>c.344C&gt;T</td>\n",
       "      <td>chr18:g.57136761G&gt;A</td>\n",
       "      <td>4.154201</td>\n",
       "      <td>1.875000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>8.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GNAS</td>\n",
       "      <td>c.*293+1G&gt;A</td>\n",
       "      <td>chr20:g.57478847G&gt;A</td>\n",
       "      <td>4.062075</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>c.3611C&gt;T</td>\n",
       "      <td>chrX:g.153588552G&gt;A</td>\n",
       "      <td>3.922368</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>16.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RTTN</td>\n",
       "      <td>c.4275G&gt;C</td>\n",
       "      <td>chr18:g.67755252C&gt;G</td>\n",
       "      <td>3.420056</td>\n",
       "      <td>2.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WNT7A</td>\n",
       "      <td>c.299-4G&gt;A</td>\n",
       "      <td>chr3:g.13896304C&gt;T</td>\n",
       "      <td>3.376938</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>8.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CAMKMT</td>\n",
       "      <td>c.493G&gt;T</td>\n",
       "      <td>chr2:g.44934538G&gt;T</td>\n",
       "      <td>3.056788</td>\n",
       "      <td>1.300000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>9.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACACA</td>\n",
       "      <td>c.3496A&gt;G</td>\n",
       "      <td>chr17:g.35564641T&gt;C</td>\n",
       "      <td>3.031660</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WWOX</td>\n",
       "      <td>c.990C&gt;G</td>\n",
       "      <td>chr16:g.78466583C&gt;G</td>\n",
       "      <td>2.939219</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>9.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NPHP4</td>\n",
       "      <td>c.2246T&gt;C</td>\n",
       "      <td>chr1:g.5950986A&gt;G</td>\n",
       "      <td>2.728738</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>5.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WAS</td>\n",
       "      <td>c.995T&gt;C</td>\n",
       "      <td>chrX:g.48547112T&gt;C</td>\n",
       "      <td>2.490538</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>6.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ARMC5</td>\n",
       "      <td>c.2303A&gt;G</td>\n",
       "      <td>chr16:g.31477609A&gt;G</td>\n",
       "      <td>2.273745</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CACNA1D</td>\n",
       "      <td>c.3881C&gt;T</td>\n",
       "      <td>chr3:g.53796059C&gt;T</td>\n",
       "      <td>2.068106</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FANCL</td>\n",
       "      <td>c.979C&gt;G</td>\n",
       "      <td>chr2:g.58388713G&gt;C</td>\n",
       "      <td>2.031038</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>c.1402C&gt;T</td>\n",
       "      <td>chr13:g.110845240G&gt;A</td>\n",
       "      <td>1.880097</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>UBAC2</td>\n",
       "      <td>c.332T&gt;C</td>\n",
       "      <td>chr13:g.99896821T&gt;C</td>\n",
       "      <td>1.810618</td>\n",
       "      <td>1.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H6PD</td>\n",
       "      <td>c.1342G&gt;A</td>\n",
       "      <td>chr1:g.9323894G&gt;A</td>\n",
       "      <td>1.647918</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GP1BA</td>\n",
       "      <td>c.1763T&gt;C</td>\n",
       "      <td>chr17:g.4837662T&gt;C</td>\n",
       "      <td>1.559581</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MPO</td>\n",
       "      <td>c.1348G&gt;A</td>\n",
       "      <td>chr17:g.56352920C&gt;T</td>\n",
       "      <td>1.373265</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OSTM1</td>\n",
       "      <td>c.238C&gt;A</td>\n",
       "      <td>chr6:g.108395618G&gt;T</td>\n",
       "      <td>1.213008</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GSN</td>\n",
       "      <td>c.1663G&gt;A</td>\n",
       "      <td>chr9:g.124088928G&gt;A</td>\n",
       "      <td>1.213008</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TMC8</td>\n",
       "      <td>c.913C&gt;T</td>\n",
       "      <td>chr17:g.76130571C&gt;T</td>\n",
       "      <td>1.030827</td>\n",
       "      <td>1.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LIPE</td>\n",
       "      <td>c.1210C&gt;T</td>\n",
       "      <td>chr19:g.42914668G&gt;A</td>\n",
       "      <td>0.735622</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>1.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SHROOM4</td>\n",
       "      <td>c.3364A&gt;C</td>\n",
       "      <td>chrX:g.50350778T&gt;G</td>\n",
       "      <td>0.649641</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BP4</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TP53</td>\n",
       "      <td>c.558T&gt;A</td>\n",
       "      <td>chr17:g.7578372A&gt;T</td>\n",
       "      <td>0.554518</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SALL2</td>\n",
       "      <td>c.2718G&gt;T</td>\n",
       "      <td>chr14:g.21991144C&gt;A</td>\n",
       "      <td>0.405465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PDSS1</td>\n",
       "      <td>c.1055G&gt;A</td>\n",
       "      <td>chr10:g.27031454G&gt;A</td>\n",
       "      <td>0.376019</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MYO7A</td>\n",
       "      <td>c.687C&gt;T</td>\n",
       "      <td>chr11:g.76868002C&gt;T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM2|BP7|BP6</td>\n",
       "      <td>2.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FANCM</td>\n",
       "      <td>c.2716A&gt;T</td>\n",
       "      <td>chr14:g.45644673A&gt;T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BP1|BP4</td>\n",
       "      <td>13.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NRXN1</td>\n",
       "      <td>c.2242C&gt;A</td>\n",
       "      <td>chr2:g.50765412G&gt;T</td>\n",
       "      <td>-0.101366</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3|BP1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.5438G&gt;A</td>\n",
       "      <td>chr8:g.144999070C&gt;T</td>\n",
       "      <td>-0.287823</td>\n",
       "      <td>-0.125000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BS2|PP3|BP1</td>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ABCA4</td>\n",
       "      <td>c.5693G&gt;A</td>\n",
       "      <td>chr1:g.94476377C&gt;T</td>\n",
       "      <td>-0.410557</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3|PP5</td>\n",
       "      <td>4.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DYNC1H1</td>\n",
       "      <td>c.13808G&gt;T</td>\n",
       "      <td>chr14:g.102516531G&gt;T</td>\n",
       "      <td>-0.627741</td>\n",
       "      <td>-0.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3|BP6</td>\n",
       "      <td>4.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MYH2</td>\n",
       "      <td>c.3884G&gt;A</td>\n",
       "      <td>chr17:g.10430361C&gt;T</td>\n",
       "      <td>-0.687218</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NOD2</td>\n",
       "      <td>c.2377G&gt;A</td>\n",
       "      <td>chr16:g.50746199G&gt;A</td>\n",
       "      <td>-1.145363</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NOD2</td>\n",
       "      <td>c.1292C&gt;T</td>\n",
       "      <td>chr16:g.50745114C&gt;T</td>\n",
       "      <td>-1.145363</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MMP1</td>\n",
       "      <td>c.105+2T&gt;C</td>\n",
       "      <td>chr11:g.102668717A&gt;G</td>\n",
       "      <td>-1.374436</td>\n",
       "      <td>-1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GJC2</td>\n",
       "      <td>c.1193C&gt;T</td>\n",
       "      <td>chr1:g.228346652C&gt;T</td>\n",
       "      <td>-1.917842</td>\n",
       "      <td>-1.125000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HFE</td>\n",
       "      <td>c.845G&gt;A</td>\n",
       "      <td>chr6:g.26093141G&gt;A</td>\n",
       "      <td>-2.079442</td>\n",
       "      <td>-1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS1|PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ATP2A2</td>\n",
       "      <td>c.464-70C&gt;A</td>\n",
       "      <td>chr12:g.110760727C&gt;A</td>\n",
       "      <td>-2.772589</td>\n",
       "      <td>-2.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CDH23</td>\n",
       "      <td>c.3293A&gt;G</td>\n",
       "      <td>chr10:g.73472494A&gt;G</td>\n",
       "      <td>-3.003638</td>\n",
       "      <td>-2.166667</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>BS2|PP3|BP6</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene      variant                    id  final_score  \\\n",
       "0       CBS     c.833T>C   chr21:g.44483184A>G     7.783641   \n",
       "1    EXOSC8     c.815G>C   chr13:g.37583420G>C     7.708397   \n",
       "2     MEGF8    c.8180G>A   chr19:g.42880569G>A     5.527655   \n",
       "3    CLCNKA     c.910C>T    chr1:g.16354556C>T     5.132180   \n",
       "4     PGAP3      c.67G>A   chr17:g.37844201C>T     5.083242   \n",
       "5     CCBE1     c.344C>T   chr18:g.57136761G>A     4.154201   \n",
       "6      GNAS  c.*293+1G>A   chr20:g.57478847G>A     4.062075   \n",
       "7      FLNA    c.3611C>T   chrX:g.153588552G>A     3.922368   \n",
       "8      RTTN    c.4275G>C   chr18:g.67755252C>G     3.420056   \n",
       "9     WNT7A   c.299-4G>A    chr3:g.13896304C>T     3.376938   \n",
       "10   CAMKMT     c.493G>T    chr2:g.44934538G>T     3.056788   \n",
       "11    ACACA    c.3496A>G   chr17:g.35564641T>C     3.031660   \n",
       "12     WWOX     c.990C>G   chr16:g.78466583C>G     2.939219   \n",
       "13    NPHP4    c.2246T>C     chr1:g.5950986A>G     2.728738   \n",
       "14      WAS     c.995T>C    chrX:g.48547112T>C     2.490538   \n",
       "15    ARMC5    c.2303A>G   chr16:g.31477609A>G     2.273745   \n",
       "16  CACNA1D    c.3881C>T    chr3:g.53796059C>T     2.068106   \n",
       "17    FANCL     c.979C>G    chr2:g.58388713G>C     2.031038   \n",
       "18   COL4A1    c.1402C>T  chr13:g.110845240G>A     1.880097   \n",
       "19    UBAC2     c.332T>C   chr13:g.99896821T>C     1.810618   \n",
       "20     H6PD    c.1342G>A     chr1:g.9323894G>A     1.647918   \n",
       "21    GP1BA    c.1763T>C    chr17:g.4837662T>C     1.559581   \n",
       "22      MPO    c.1348G>A   chr17:g.56352920C>T     1.373265   \n",
       "23    OSTM1     c.238C>A   chr6:g.108395618G>T     1.213008   \n",
       "24      GSN    c.1663G>A   chr9:g.124088928G>A     1.213008   \n",
       "25     TMC8     c.913C>T   chr17:g.76130571C>T     1.030827   \n",
       "26     LIPE    c.1210C>T   chr19:g.42914668G>A     0.735622   \n",
       "27  SHROOM4    c.3364A>C    chrX:g.50350778T>G     0.649641   \n",
       "28     TP53     c.558T>A    chr17:g.7578372A>T     0.554518   \n",
       "29    SALL2    c.2718G>T   chr14:g.21991144C>A     0.405465   \n",
       "30    PDSS1    c.1055G>A   chr10:g.27031454G>A     0.376019   \n",
       "31    MYO7A     c.687C>T   chr11:g.76868002C>T     0.000000   \n",
       "32    FANCM    c.2716A>T   chr14:g.45644673A>T     0.000000   \n",
       "33    NRXN1    c.2242C>A    chr2:g.50765412G>T    -0.101366   \n",
       "34     PLEC    c.5438G>A   chr8:g.144999070C>T    -0.287823   \n",
       "35    ABCA4    c.5693G>A    chr1:g.94476377C>T    -0.410557   \n",
       "36  DYNC1H1   c.13808G>T  chr14:g.102516531G>T    -0.627741   \n",
       "37     MYH2    c.3884G>A   chr17:g.10430361C>T    -0.687218   \n",
       "38     NOD2    c.2377G>A   chr16:g.50746199G>A    -1.145363   \n",
       "39     NOD2    c.1292C>T   chr16:g.50745114C>T    -1.145363   \n",
       "40     MMP1   c.105+2T>C  chr11:g.102668717A>G    -1.374436   \n",
       "41     GJC2    c.1193C>T   chr1:g.228346652C>T    -1.917842   \n",
       "42      HFE     c.845G>A    chr6:g.26093141G>A    -2.079442   \n",
       "43   ATP2A2  c.464-70C>A  chr12:g.110760727C>A    -2.772589   \n",
       "44    CDH23    c.3293A>G   chr10:g.73472494A>G    -3.003638   \n",
       "\n",
       "    pathogenicity_score           pathogenicity             hit_criteria  \\\n",
       "0              4.000000       Likely pathogenic          PS1|PM5|PP3|PP5   \n",
       "1              5.125000              Pathogenic  PS1|PM5|PM1|PP3|PP2|PP5   \n",
       "2              2.250000  Uncertain significance              PM1|PM2|PP3   \n",
       "3              3.500000  Uncertain significance                 PVS1|PP3   \n",
       "4              2.333333  Uncertain significance              PM1|PM2|PP3   \n",
       "5              1.875000  Uncertain significance              PM1|PP3|PP2   \n",
       "6              1.500000  Uncertain significance                  PM2|PP3   \n",
       "7              1.375000  Uncertain significance                  PM2|PP3   \n",
       "8              2.125000  Uncertain significance              PM1|PM2|PP3   \n",
       "9              1.500000  Uncertain significance                  PM2|PP3   \n",
       "10             1.300000  Uncertain significance                  PM1|PP3   \n",
       "11             2.333333  Uncertain significance              PM1|PM2|PP3   \n",
       "12             1.250000  Uncertain significance                  PM1|PP3   \n",
       "13             1.500000  Uncertain significance                  PM1|PP3   \n",
       "14             1.250000  Uncertain significance                  PM1|PP3   \n",
       "15             1.750000  Uncertain significance          PM1|PM2|PP3|BP1   \n",
       "16             1.375000  Uncertain significance                  PM1|PP3   \n",
       "17             0.750000  Uncertain significance              PM1|PP3|BP1   \n",
       "18             1.250000  Uncertain significance                  PM1|PP3   \n",
       "19             1.125000  Uncertain significance                  PM1|PP3   \n",
       "20             1.500000  Uncertain significance                  PM1|PP3   \n",
       "21             2.250000  Uncertain significance              PM1|PM2|PP3   \n",
       "22             1.250000  Uncertain significance                  PM1|PP3   \n",
       "23             1.750000  Uncertain significance          PM1|PM2|PP3|BP1   \n",
       "24             1.750000  Uncertain significance              PM1|PP3|PP2   \n",
       "25             1.125000  Uncertain significance                  PM1|PP3   \n",
       "26             0.750000  Uncertain significance              PM1|PP3|BP1   \n",
       "27             0.500000  Uncertain significance                  PM1|BP4   \n",
       "28             0.400000  Uncertain significance                      PP3   \n",
       "29             1.000000           Likely benign          PM1|PM2|BP1|BP4   \n",
       "30             0.250000  Uncertain significance                      PP3   \n",
       "31             0.000000           Likely benign              PM2|BP7|BP6   \n",
       "32             0.000000           Likely benign              PM1|BP1|BP4   \n",
       "33            -0.250000  Uncertain significance                  PP3|BP1   \n",
       "34            -0.125000           Likely benign      PM1|PM2|BS2|PP3|BP1   \n",
       "35            -0.250000  Uncertain significance          PM1|BS2|PP3|PP5   \n",
       "36            -0.375000  Uncertain significance                  PP3|BP6   \n",
       "37            -0.750000  Uncertain significance              PM1|BS2|PP3   \n",
       "38            -1.250000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "39            -1.250000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "40            -1.500000  Uncertain significance                  BS2|PP3   \n",
       "41            -1.125000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "42            -1.500000  Uncertain significance                  BS1|PP3   \n",
       "43            -2.000000  Uncertain significance                      BS2   \n",
       "44            -2.166667           Likely benign              BS2|PP3|BP6   \n",
       "\n",
       "    hpo_hit_score  \n",
       "0        6.000000  \n",
       "1        3.500000  \n",
       "2       10.666667  \n",
       "3        3.333333  \n",
       "4        7.833333  \n",
       "5        8.166667  \n",
       "6       14.000000  \n",
       "7       16.333333  \n",
       "8        4.000000  \n",
       "9        8.500000  \n",
       "10       9.500000  \n",
       "11       2.666667  \n",
       "12       9.500000  \n",
       "13       5.166667  \n",
       "14       6.333333  \n",
       "15       2.666667  \n",
       "16       3.500000  \n",
       "17      14.000000  \n",
       "18       3.500000  \n",
       "19       4.000000  \n",
       "20       2.000000  \n",
       "21       1.000000  \n",
       "22       2.000000  \n",
       "23       1.000000  \n",
       "24       1.000000  \n",
       "25       1.500000  \n",
       "26       1.666667  \n",
       "27       2.666667  \n",
       "28       3.000000  \n",
       "29       0.500000  \n",
       "30       3.500000  \n",
       "31       2.500000  \n",
       "32      13.000000  \n",
       "33       0.500000  \n",
       "34       9.000000  \n",
       "35       4.166667  \n",
       "36       4.333333  \n",
       "37       1.500000  \n",
       "38       1.500000  \n",
       "39       1.500000  \n",
       "40       1.500000  \n",
       "41       4.500000  \n",
       "42       3.000000  \n",
       "43       3.000000  \n",
       "44       3.000000  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acmg_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}