{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_file_path = 'input/input_genes.txt'\n",
    "phenotype_file_path = 'input/input_phenotype.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "        if not input_phenotype:\n",
    "            return '', ''\n",
    "        text = StringIO(unicode(input_phenotype), newline=None)\n",
    "        lines = text.readlines()\n",
    "        lines = [line.strip() for line in lines]\n",
    "        phenos = []\n",
    "        for line in lines:\n",
    "            if not line:\n",
    "                continue\n",
    "            phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "            phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "            phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "            phenos += phenos_each_line\n",
    "\n",
    "        corner_cases = dict()\n",
    "        for pheno in phenos:\n",
    "                if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                        phenos.append('growth delay')\n",
    "                        corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "                if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                        phenos.append('developmental delay')\n",
    "                        corner_cases['developmental delay'] = pheno.strip()\n",
    "        return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "        candidate_vars = []\n",
    "        input_gene = input_gene.split('\\n')\n",
    "        header = input_gene[0]\n",
    "        sniffer = csv.Sniffer()\n",
    "        dialect = sniffer.sniff(header)\n",
    "        delimiter =  dialect.delimiter\n",
    "        field_names = header.split(delimiter)\n",
    "        chrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, 0 \n",
    "\n",
    "        for idx in xrange(len(field_names)):\n",
    "            field = field_names[idx]\n",
    "            if re.match(r'chrom', field, re.I): chrom_idx = idx\n",
    "            if re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "            if re.match(r'ref', field, re.I): ref_idx = idx\n",
    "            if re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "            if re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "        input_gene_list = []\n",
    "        CANDIDATE_GENES = []\n",
    "        for line in input_gene[1:]:\n",
    "                if not line:\n",
    "                        continue\n",
    "                line = line.rstrip()\n",
    "                parts = re.split(r'%s' % delimiter, line)\n",
    "                input_gene_list.append(parts)\n",
    "                gene = parts[gene_idx]\n",
    "                CANDIDATE_GENES.append(gene)\n",
    "                transcript, variant, variant_id = '', '', ''\n",
    "                for part in parts:\n",
    "                    if re.search(r'_.*:c\\.', part):\n",
    "                        transcript, variant = part.split(':')\n",
    "                    if re.search(r'_.*:g\\.', part):\n",
    "                        variant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "                    if re.search(r'chr.*:g\\.', part, re.I):\n",
    "                        variant_id = part\n",
    "                if not variant_id and (chrom_idx and pos_idx and ref_idx and alt_idx):\n",
    "                    chrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "                    variant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "                candidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "        # remove lines in the input file which has wrong number of fields\n",
    "        field_nums = []\n",
    "        for line in input_gene_list:\n",
    "            field_nums.append(len(line))\n",
    "        count = Counter(field_nums)\n",
    "        correct_field_num = count.most_common()[0][0]\n",
    "        correct_input_gene_list = []\n",
    "        for line in input_gene_list:\n",
    "                if len(line) == correct_field_num:\n",
    "                        correct_input_gene_list.append(line)\n",
    "        df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "        return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "        ranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "        # collect variant info\n",
    "        hpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "        tmp_candidate_vars = []\n",
    "        for var in candidate_vars:\n",
    "                if var[0] in hpo_filtered_genes:\n",
    "                        tmp_candidate_vars.append(var)\n",
    "        candidate_vars = tmp_candidate_vars\n",
    "        return ranking_genes, candidate_vars\n",
    "\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/xiaonantest1_g.txt\",'rU').read()\n",
    "input_phenotype = open(\"/Users/Tianqi/Desktop/xiaonantest1_p.txt\",'rU').read()\n",
    "\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "# map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "if phenos:\n",
    "    ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "else:\n",
    "    ranking_genes = []\n",
    "    for gene in CANDIDATE_GENES:\n",
    "        ranking_genes.append((gene, 1.0, 1))\n",
    "\n",
    "# collect variant info\n",
    "mv = myvariant.MyVariantInfo()\n",
    "final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "# pubmed\n",
    "df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "# ACMG\n",
    "df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "ACMG_result = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "# filter variant on phenotype\n",
    "if phenos:\n",
    "    df_final_res = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos)\n",
    "else:\n",
    "    df_final_res = ACMG_result\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "    '''get a valid hgvs name from VCF-style \"chrom, pos, ref, alt\" data.\n",
    "\n",
    "    Example:\n",
    "\n",
    "        >>> myvariant.format_hgvs(\"1\", 35366, \"C\", \"T\")\n",
    "        >>> myvariant.format_hgvs(\"2\", 17142, \"G\", \"GA\")\n",
    "        >>> myvariant.format_hgvs(\"MT\", 8270, \"CACCCCCTCT\", \"C\")\n",
    "        >>> myvariant.format_hgvs(\"X\", 107930849, \"GGA\", \"C\")\n",
    "\n",
    "    '''\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "    if not input_phenotype:\n",
    "        return '', ''\n",
    "    text = StringIO(unicode(input_phenotype), newline=None)\n",
    "    lines = text.readlines()\n",
    "    lines = [line.strip() for line in lines]\n",
    "    phenos = []\n",
    "    for line in lines:\n",
    "        if not line:\n",
    "            continue\n",
    "        phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "        phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "        phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "        phenos += phenos_each_line\n",
    "\n",
    "    corner_cases = dict()\n",
    "    for pheno in phenos:\n",
    "        if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "            phenos.append('growth delay')\n",
    "            corner_cases['growth delay'] = pheno.strip()\n",
    "    for pheno in phenos:\n",
    "        if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "            phenos.append('developmental delay')\n",
    "            corner_cases['developmental delay'] = pheno.strip()\n",
    "    return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "\tcandidate_vars = []\n",
    "\tinput_gene = input_gene.split('\\n')\n",
    "\theader = input_gene[0]\n",
    "\tsniffer = csv.Sniffer()\n",
    "\tdialect = sniffer.sniff(header)\n",
    "\tdelimiter =  dialect.delimiter\n",
    "\tfield_names = header.split(delimiter)\n",
    "\tchrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, 0 \n",
    "\n",
    "\tfor idx in xrange(len(field_names)):\n",
    "\t\tfield = field_names[idx]\n",
    "\t\tif re.match(r'chrom', field, re.I): chrom_idx = idx\n",
    "\t\tif re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "\t\tif re.match(r'ref', field, re.I): ref_idx = idx\n",
    "\t\tif re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "\t\tif re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "\tinput_gene_list = []\n",
    "\tCANDIDATE_GENES = []\n",
    "\tfor line in input_gene[1:]:\n",
    "\t\tif not line:\n",
    "\t\t\tcontinue\n",
    "\t\tline = line.rstrip()\n",
    "\t\tparts = re.split(r'%s' % delimiter, line)\n",
    "\t\tinput_gene_list.append(parts)\n",
    "\t\tgene = parts[gene_idx]\n",
    "\t\tCANDIDATE_GENES.append(gene)\n",
    "\t\ttranscript, variant, variant_id = '', '', ''\n",
    "\t\tfor part in parts:\n",
    "\t\t\tif re.search(r'_.*:c\\.', part):\n",
    "\t\t\t\ttranscript, variant = part.split(':')\n",
    "\t\t\telse:\n",
    "\t\t\t\tif re.search(r'c\\.', part):\n",
    "\t\t\t\t\tvariant = part\n",
    "\t\t\t\tif re.search(r'NM_', part, re.I):\n",
    "\t\t\t\t\ttranscript = part.split(':')[0]\n",
    "\t\t\tif re.search(r'_.*:g\\.', part):\n",
    "\t\t\t\tvariant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "\t\t\tif re.search(r'chr.*:g\\.', part, re.I):\n",
    "\t\t\t\tvariant_id = part\n",
    "\t\tif not variant_id and (chrom_idx and pos_idx and ref_idx and alt_idx):\n",
    "\t\t\tchrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "\t\t\tvariant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "\t\tcandidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "        # remove lines in the input file which has wrong number of fields\n",
    "        field_nums = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tfield_nums.append(len(line))\n",
    "        count = Counter(field_nums)\n",
    "        correct_field_num = count.most_common()[0][0]\n",
    "        correct_input_gene_list = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tif len(line) == correct_field_num:\n",
    "\t\t\t\tcorrect_input_gene_list.append(line)\n",
    "        df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "        return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "\tranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\t# collect variant info\n",
    "\thpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "\ttmp_candidate_vars = []\n",
    "\tfor var in candidate_vars:\n",
    "\t\tif var[0] in hpo_filtered_genes:\n",
    "\t\t\ttmp_candidate_vars.append(var)\n",
    "\tcandidate_vars = tmp_candidate_vars\n",
    "\treturn ranking_genes, candidate_vars\n",
    "\n",
    "# \tstatus_step = \"generating candidate variants ...\" \n",
    "# \traw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "# \tinput_gene = raw_input.raw_input_gene\n",
    "# \tinput_phenotype = raw_input.raw_input_phenotype\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/xiaonantest1_g.txt\",'rU').read()\n",
    "input_phenotype = open(\"/Users/Tianqi/Desktop/xiaonantest1_p.txt\",'rU').read()\n",
    "\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "# map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "\n",
    "if phenos:\n",
    "#     raw_input.status = \"Maping phenotypes to genes\"\n",
    "#     raw_input.save()\n",
    "    ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "else:\n",
    "    ranking_genes = []\n",
    "    for gene in CANDIDATE_GENES:\n",
    "        ranking_genes.append((gene, 1.0, 1))\n",
    "\n",
    "# collect variant info\n",
    "# raw_input.status = \"Annotating variants using genomic databases\"\n",
    "# raw_input.save()\n",
    "mv = myvariant.MyVariantInfo()\n",
    "final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "# pubmed\n",
    "# raw_input.status = \"Searching biomedical literatures\"\n",
    "# raw_input.save()\n",
    "df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "# ACMG\n",
    "# raw_input.status = \"Checking ACMG standard\"\n",
    "# raw_input.save()\n",
    "df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "ACMG_result, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "# filter variant on phenotype\n",
    "\n",
    "if phenos:\n",
    "#     raw_input.status = \"Filtering variants based on phenotypes\"\n",
    "#     raw_input.save()\n",
    "    df_final_res, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos, variant_ACMG_interpretation, variant_ACMG_interpret_chinese)\n",
    "else:\n",
    "    df_final_res = ACMG_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>protein</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "      <th>pheno_match_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>p.Arg49Trp</td>\n",
       "      <td>chr12:g.76741994G&gt;A</td>\n",
       "      <td>9.511010</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM1|PP3|PP5</td>\n",
       "      <td>3.8750</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>1.154249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>p.Gly70Ser</td>\n",
       "      <td>chr11:g.5247914C&gt;T</td>\n",
       "      <td>8.713391</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM5|PM1|PP3|PP5</td>\n",
       "      <td>2.6250</td>\n",
       "      <td>19.213095</td>\n",
       "      <td>1.074401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>p.Lys112Arg</td>\n",
       "      <td>chr17:g.48070945T&gt;C</td>\n",
       "      <td>6.150000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|PM2|PP3|BP1</td>\n",
       "      <td>3.9612</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>c.652A&gt;C</td>\n",
       "      <td>p.Ser218Arg</td>\n",
       "      <td>chr5:g.177419739T&gt;G</td>\n",
       "      <td>5.230000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>SOX3</td>\n",
       "      <td>c.818C&gt;T</td>\n",
       "      <td>p.Ser273Leu</td>\n",
       "      <td>chrX:g.139586408G&gt;A</td>\n",
       "      <td>5.190000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.333333</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene   variant      protein                   id  final_score  \\\n",
       "76   BBS10  c.145C>T   p.Arg49Trp  chr12:g.76741994G>A     9.511010   \n",
       "103    HBB  c.208G>A   p.Gly70Ser   chr11:g.5247914C>T     8.713391   \n",
       "66    DLX3  c.335A>G  p.Lys112Arg  chr17:g.48070945T>C     6.150000   \n",
       "53   PROP1  c.652A>C  p.Ser218Arg  chr5:g.177419739T>G     5.230000   \n",
       "107   SOX3  c.818C>T  p.Ser273Leu  chrX:g.139586408G>A     5.190000   \n",
       "\n",
       "              pathogenicity         hit_criteria  pathogenicity_score  \\\n",
       "76        Likely pathogenic      PS1|PM1|PP3|PP5               3.8750   \n",
       "103  Uncertain significance      PM5|PM1|PP3|PP5               2.6250   \n",
       "66   Uncertain significance  PS1|PM1|PM2|PP3|BP1               3.9612   \n",
       "53   Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "107  Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "\n",
       "     hpo_hit_score  pheno_match_score  \n",
       "76        5.666667           1.154249  \n",
       "103      19.213095           1.074401  \n",
       "66        2.000000           1.000000  \n",
       "53        7.500000           1.000000  \n",
       "107       7.333333           1.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Allele 1</th>\n",
       "      <th>Allele 2</th>\n",
       "      <th>Allele In Scope</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene Profile Report</th>\n",
       "      <th>Gene (gene)</th>\n",
       "      <th>cDNA (cNomen)</th>\n",
       "      <th>Protein (pNomen)</th>\n",
       "      <th>HGVS cDNA-level nomenclature (fullCNomen)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1635536</td>\n",
       "      <td>1635536</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_024011.2</td>\n",
       "      <td></td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>c.1735dupT</td>\n",
       "      <td>p.Y579Lfs*66</td>\n",
       "      <td>NM_024011.2:c.1735dupT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>23713843</td>\n",
       "      <td>23713843</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_003196.2</td>\n",
       "      <td></td>\n",
       "      <td>TCEA3</td>\n",
       "      <td>c.889A&gt;G</td>\n",
       "      <td>p.M297V</td>\n",
       "      <td>NM_003196.2:c.889A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>24882663</td>\n",
       "      <td>24882663</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NM_001010980.4</td>\n",
       "      <td></td>\n",
       "      <td>NCMAP</td>\n",
       "      <td>c.-8+3A&gt;G</td>\n",
       "      <td></td>\n",
       "      <td>NM_001010980.4:c.-8+3A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>26663842</td>\n",
       "      <td>26663842</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_001039775.3</td>\n",
       "      <td></td>\n",
       "      <td>AIM1L</td>\n",
       "      <td>c.3673C&gt;T</td>\n",
       "      <td>p.R1225W</td>\n",
       "      <td>NM_001039775.3:c.3673C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>39896387</td>\n",
       "      <td>39896387</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_012090.5</td>\n",
       "      <td></td>\n",
       "      <td>MACF1</td>\n",
       "      <td>c.10958A&gt;C</td>\n",
       "      <td>p.E3653A</td>\n",
       "      <td>NM_012090.5:c.10958A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start      Stop Reference Allele 1 Allele 2 Allele In Scope  \\\n",
       "0          1   1635536   1635536         .        .        A               A   \n",
       "1          1  23713843  23713843         T        C        T               C   \n",
       "2          1  24882663  24882663         A        A        G               G   \n",
       "3          1  26663842  26663842         G        G        A               A   \n",
       "4          1  39896387  39896387         A        C        A               C   \n",
       "\n",
       "       Transcript Gene Profile Report Gene (gene) cDNA (cNomen)  \\\n",
       "0     NM_024011.2                          CDK11A    c.1735dupT   \n",
       "1     NM_003196.2                           TCEA3      c.889A>G   \n",
       "2  NM_001010980.4                           NCMAP     c.-8+3A>G   \n",
       "3  NM_001039775.3                           AIM1L     c.3673C>T   \n",
       "4     NM_012090.5                           MACF1    c.10958A>C   \n",
       "\n",
       "  Protein (pNomen) HGVS cDNA-level nomenclature (fullCNomen)  \n",
       "0     p.Y579Lfs*66                    NM_024011.2:c.1735dupT  \n",
       "1          p.M297V                      NM_003196.2:c.889A>G  \n",
       "2                                   NM_001010980.4:c.-8+3A>G  \n",
       "3         p.R1225W                  NM_001039775.3:c.3673C>T  \n",
       "4         p.E3653A                    NM_012090.5:c.10958A>C  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'bilateral post-axial polydactyly of hands and feet',\n",
       " u'macrocephaly',\n",
       " u'tall stature',\n",
       " u'post-natal',\n",
       " u'central hypotonia',\n",
       " u'global developmental delay',\n",
       " u'obstructive sleep apnea',\n",
       " u'delayed cns central nervous myelination',\n",
       " u'white matter',\n",
       " u'corpus callosum',\n",
       " u'mildly ectopic neurohypophys posterior pituitary upslanted palpebral fissures',\n",
       " u'epicanthus',\n",
       " u'arched eyebrows',\n",
       " u'craniofacial asymmetry',\n",
       " u'bronchial stenosis',\n",
       " u'macrosomia',\n",
       " u'recurrent infection',\n",
       " u'gastroesophageal reflux',\n",
       " u'supratentorial perivascular space',\n",
       " u'mucopolysaccharidosis plasma urine creatine',\n",
       " u'purine',\n",
       " u'pyrimidine metabolism',\n",
       " u'gross motor',\n",
       " u'language',\n",
       " u'cognitive',\n",
       " u'ventriculomegaly',\n",
       " u'metatarsal bones',\n",
       " u'proximal phalanges',\n",
       " u'obesity',\n",
       " u'wheeze',\n",
       " u'brachycephaly frontal prominence',\n",
       " u'bitemporal narrowing',\n",
       " u'clinodactyly',\n",
       " u'brachydactyly',\n",
       " u'bardet-biedl',\n",
       " 'growth delay',\n",
       " 'developmental delay']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Chromosome',\n",
       " 'Start',\n",
       " 'Stop',\n",
       " 'Reference',\n",
       " 'Allele 1',\n",
       " 'Allele 2',\n",
       " 'Allele In Scope',\n",
       " 'Transcript',\n",
       " 'Gene Profile Report',\n",
       " 'Gene (gene)',\n",
       " 'cDNA (cNomen)',\n",
       " 'Protein (pNomen)',\n",
       " 'HGVS cDNA-level nomenclature (fullCNomen)']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>variant_annotations</td>\n",
       "      <td>Effect: missense_variant.&lt;br/&gt;Protein domain: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;exon: 1.&lt;br/&gt;ExAC MAF: 5.052e-05.&lt;br/&gt;DANN pathogenicity score: 0.999117068756.&lt;br/&gt;FATHMM pathogenicity score: 0.79186.&lt;br/&gt;MetaSVM pathogenicity score: 0.84564.&lt;br/&gt;GERP++ conservation score: 4.27.&lt;br/&gt;Clinvar variation ids: 225010.&lt;br/&gt;Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar review status: criteria provided, single submitter.&lt;br/&gt;Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1 and PM5</td>\n",
       "      <td>Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3 and BS3</td>\n",
       "      <td>Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene   variant             criteria  \\\n",
       "0  BBS10  c.145C>T  variant_annotations   \n",
       "1  BBS10  c.145C>T                 PVS1   \n",
       "2  BBS10  c.145C>T          PS1 and PM5   \n",
       "3  BBS10  c.145C>T          PS3 and BS3   \n",
       "4  BBS10  c.145C>T                  PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     interpretation  \n",
       "0  Effect: missense_variant.<br/>Protein domain: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>exon: 1.<br/>ExAC MAF: 5.052e-05.<br/>DANN pathogenicity score: 0.999117068756.<br/>FATHMM pathogenicity score: 0.79186.<br/>MetaSVM pathogenicity score: 0.84564.<br/>GERP++ conservation score: 4.27.<br/>Clinvar variation ids: 225010.<br/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br/>Clinvar review status: criteria provided, single submitter.<br/>Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.  \n",
       "2                                                                                                                                                                                                                                                                                                             Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpretation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>基因</th>\n",
       "      <th>变异</th>\n",
       "      <th>标准</th>\n",
       "      <th>解读</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: 致病|致病.&lt;br/&gt;Clinvar数据库记录审核状态: 提供标准，单个提交者.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      基因        变异       标准  \\\n",
       "0  BBS10  c.145C>T     变异注释   \n",
       "1  BBS10  c.145C>T     PVS1   \n",
       "2  BBS10  c.145C>T  PS1和PM5   \n",
       "3  BBS10  c.145C>T  PS3和BS3   \n",
       "4  BBS10  c.145C>T      PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               解读  \n",
       "0  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: 致病|致病.<br/>Clinvar数据库记录审核状态: 提供标准，单个提交者.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                  基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpret_chinese.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_final_res, df_genes, phenos, field_names, variant_ACMG_interpretation, variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nput_gene = df_genes.to_json(orient='records')\n",
    "input_phenotype = ', '.join(phenos)\n",
    "result_table = ACMG_result.to_json(orient='records')\n",
    "interpretation = variant_ACMG_interpretation.to_json(orient='records')\n",
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')\n",
    "# logger.info(\"Finish processing data, start writing data to DB in background main task\")\n",
    "\n",
    "\n",
    "# sample = Main_table(\n",
    "#     task_id=raw_input_id,\n",
    "#     input_gene=input_gene,\n",
    "#     input_phenotype=input_phenotype,\n",
    "#     result=result_table,\n",
    "#     interpretation=interpretation,\n",
    "#     interpretation_chinese=interpretation_chinese,\n",
    "#     pub_date=timezone.now(),\n",
    "#     user_name=raw_input.user_name,\n",
    "#     task_name=raw_input.task_name,\n",
    "# )\n",
    "# sample.save()\n",
    "# logger.info(\"Finish writing data to DB in background main task\")\n",
    "# raw_input.status = \"succeed\"\n",
    "# raw_input.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"variant_annotations\",\"interpretation\":\"Effect: missense_variant.<br\\\\/>Protein domain: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>exon: 1.<br\\\\/>ExAC MAF: 5.052e-05.<br\\\\/>DANN pathogenicity score: 0.999117068756.<br\\\\/>FATHMM pathogenicity score: 0.79186.<br\\\\/>MetaSVM pathogenicity score: 0.84564.<br\\\\/>GERP++ conservation score: 4.27.<br\\\\/>Clinvar variation ids: 225010.<br\\\\/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br\\\\/>Clinvar review status: criteria provided, single submitter.<br\\\\/>Pubmed references from Clinvar: [\\'25356970\\', \\'20120035\\', \\'16582908\\', \\'21044901\\', \\'20876674\\', \\'21642631\\', \\'20498079\\', \\'21517826\\', \\'20177705\\', \\'24746959\\', \\'25982971\\'].\"},{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"PVS1\",\"interpretation\":\"Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have dama'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"\\\\u57fa\\\\u56e0\":\"BBS10\",\"\\\\u53d8\\\\u5f02\":\"c.145C>T\",\"\\\\u6807\\\\u51c6\":\"\\\\u53d8\\\\u5f02\\\\u6ce8\\\\u91ca\",\"\\\\u89e3\\\\u8bfb\":\"\\\\u7a81\\\\u53d8\\\\u7c7b\\\\u578b: missense_variant.<br\\\\/>\\\\u86cb\\\\u767d\\\\u529f\\\\u80fd\\\\u533a: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>\\\\u5916\\\\u663e\\\\u5b50: 1.<br\\\\/>ExAC \\\\u6700\\\\u5c0f\\\\u7b49\\\\u4f4d\\\\u57fa\\\\u56e0\\\\u9891\\\\u7387(MAF): 5.052e-05.<br\\\\/>DANN\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.999117068756.<br\\\\/>FATHMM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.79186.<br\\\\/>MetaSVM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.84564.<br\\\\/>GERP++\\\\u5e8f\\\\u5217\\\\u4fdd\\\\u5b88\\\\u6027\\\\u9884\\\\u6d4b\\\\u5206\\\\u6570: 4.27.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93ID: 225010.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684\\\\u53d8\\\\u5f02\\\\u81f4\\\\u75c5\\\\u6027: \\\\u81f4\\\\u75c5|\\\\u81f4\\\\u75c5.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u5ba1\\\\u6838\\\\u72b6\\\\u6001: \\\\u63d0\\\\u4f9b\\\\u6807\\\\u51c6\\\\uff0c\\\\u5355\\\\u4e2a\\\\u63d0\\\\u4ea4\\\\u8005.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684Pubmed\\\\u76f8\\\\u5173\\\\u751f\\\\u7269\\\\u533b\\\\u5b66\\\\u6587\\\\u'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation_chinese[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
