{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_file_path = 'input/input_genes.txt'\n",
    "phenotype_file_path = 'input/input_phenotype.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "    '''get a valid hgvs name from VCF-style \"chrom, pos, ref, alt\" data.\n",
    "\n",
    "    Example:\n",
    "\n",
    "        >>> myvariant.format_hgvs(\"1\", 35366, \"C\", \"T\")\n",
    "        >>> myvariant.format_hgvs(\"2\", 17142, \"G\", \"GA\")\n",
    "        >>> myvariant.format_hgvs(\"MT\", 8270, \"CACCCCCTCT\", \"C\")\n",
    "        >>> myvariant.format_hgvs(\"X\", 107930849, \"GGA\", \"C\")\n",
    "\n",
    "    '''\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "    if not input_phenotype:\n",
    "        return '', ''\n",
    "    text = StringIO(unicode(input_phenotype), newline=None)\n",
    "    lines = text.readlines()\n",
    "    lines = [line.strip() for line in lines]\n",
    "    phenos = []\n",
    "    for line in lines:\n",
    "        if not line:\n",
    "            continue\n",
    "        phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "        phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "        phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "        phenos += phenos_each_line\n",
    "\n",
    "    corner_cases = dict()\n",
    "    for pheno in phenos:\n",
    "        if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "            phenos.append('growth delay')\n",
    "            corner_cases['growth delay'] = pheno.strip()\n",
    "    for pheno in phenos:\n",
    "        if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "            phenos.append('developmental delay')\n",
    "            corner_cases['developmental delay'] = pheno.strip()\n",
    "    return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "\tcandidate_vars = []\n",
    "\tinput_gene = input_gene.split('\\n')\n",
    "\theader = input_gene[0]\n",
    "\tsniffer = csv.Sniffer()\n",
    "\tdialect = sniffer.sniff(header)\n",
    "\tdelimiter =  dialect.delimiter\n",
    "\tfield_names = header.split(delimiter)\n",
    "\tchrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, 0 \n",
    "\n",
    "\tfor idx in xrange(len(field_names)):\n",
    "\t\tfield = field_names[idx]\n",
    "\t\tif re.match(r'chrom', field, re.I): chrom_idx = idx\n",
    "\t\tif re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "\t\tif re.match(r'ref', field, re.I): ref_idx = idx\n",
    "\t\tif re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "\t\tif re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "\tinput_gene_list = []\n",
    "\tCANDIDATE_GENES = []\n",
    "\tfor line in input_gene[1:]:\n",
    "\t\tif not line:\n",
    "\t\t\tcontinue\n",
    "\t\tline = line.rstrip()\n",
    "\t\tparts = re.split(r'%s' % delimiter, line)\n",
    "\t\tinput_gene_list.append(parts)\n",
    "\t\tgene = parts[gene_idx]\n",
    "\t\tCANDIDATE_GENES.append(gene)\n",
    "\t\ttranscript, variant, variant_id = '', '', ''\n",
    "\t\tfor part in parts:\n",
    "\t\t\tif re.search(r'_.*:c\\.', part):\n",
    "\t\t\t\ttranscript, variant = part.split(':')\n",
    "\t\t\telse:\n",
    "\t\t\t\tif re.search(r'c\\.', part):\n",
    "\t\t\t\t\tvariant = part\n",
    "\t\t\t\tif re.search(r'NM_', part, re.I):\n",
    "\t\t\t\t\ttranscript = part.split(':')[0]\n",
    "\t\t\tif re.search(r'_.*:g\\.', part):\n",
    "\t\t\t\tvariant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "\t\t\tif re.search(r'chr.*:g\\.', part, re.I):\n",
    "\t\t\t\tvariant_id = part\n",
    "\t\tif not variant_id and (chrom_idx and pos_idx and ref_idx and alt_idx):\n",
    "\t\t\tchrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "\t\t\tvariant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "\t\tcandidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "        # remove lines in the input file which has wrong number of fields\n",
    "        field_nums = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tfield_nums.append(len(line))\n",
    "        count = Counter(field_nums)\n",
    "        correct_field_num = count.most_common()[0][0]\n",
    "        correct_input_gene_list = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tif len(line) == correct_field_num:\n",
    "\t\t\t\tcorrect_input_gene_list.append(line)\n",
    "        df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "        return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "\tranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\t# collect variant info\n",
    "\thpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "\ttmp_candidate_vars = []\n",
    "\tfor var in candidate_vars:\n",
    "\t\tif var[0] in hpo_filtered_genes:\n",
    "\t\t\ttmp_candidate_vars.append(var)\n",
    "\tcandidate_vars = tmp_candidate_vars\n",
    "\treturn ranking_genes, candidate_vars\n",
    "\n",
    "# \tstatus_step = \"generating candidate variants ...\" \n",
    "# \traw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "# \tinput_gene = raw_input.raw_input_gene\n",
    "# \tinput_phenotype = raw_input.raw_input_phenotype\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/xiaonantest1_g.txt\",'rU').read()\n",
    "input_phenotype = open(\"/Users/Tianqi/Desktop/xiaonantest1_p.txt\",'rU').read()\n",
    "\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "# map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "\n",
    "if phenos:\n",
    "#     raw_input.status = \"Maping phenotypes to genes\"\n",
    "#     raw_input.save()\n",
    "    ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "else:\n",
    "    ranking_genes = []\n",
    "    for gene in CANDIDATE_GENES:\n",
    "        ranking_genes.append((gene, 1.0, 1))\n",
    "\n",
    "# collect variant info\n",
    "# raw_input.status = \"Annotating variants using genomic databases\"\n",
    "# raw_input.save()\n",
    "mv = myvariant.MyVariantInfo()\n",
    "final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "# pubmed\n",
    "# raw_input.status = \"Searching biomedical literatures\"\n",
    "# raw_input.save()\n",
    "df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "# ACMG\n",
    "# raw_input.status = \"Checking ACMG standard\"\n",
    "# raw_input.save()\n",
    "df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "ACMG_result, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "# filter variant on phenotype\n",
    "\n",
    "# if phenos:\n",
    "# #     raw_input.status = \"Filtering variants based on phenotypes\"\n",
    "# #     raw_input.save()\n",
    "#     df_final_res, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos, variant_ACMG_interpretation, variant_ACMG_interpret_chinese)\n",
    "# else:\n",
    "#     df_final_res = ACMG_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>chr12:g.76741994G&gt;A</td>\n",
       "      <td>8.24</td>\n",
       "      <td>3.875000</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM1|PP3|PP5</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>chr11:g.5247914C&gt;T</td>\n",
       "      <td>8.11</td>\n",
       "      <td>2.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM5|PM1|PP3|PP5</td>\n",
       "      <td>19.213095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>chr17:g.48070945T&gt;C</td>\n",
       "      <td>6.15</td>\n",
       "      <td>3.961200</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|PM2|PP3|BP1</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>c.652A&gt;C</td>\n",
       "      <td>chr5:g.177419739T&gt;G</td>\n",
       "      <td>5.23</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SOX3</td>\n",
       "      <td>c.818C&gt;T</td>\n",
       "      <td>chrX:g.139586408G&gt;A</td>\n",
       "      <td>5.19</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CEP152</td>\n",
       "      <td>c.3530A&gt;G</td>\n",
       "      <td>chr15:g.49040744T&gt;C</td>\n",
       "      <td>4.97</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>6.404762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRKDC</td>\n",
       "      <td>c.3278A&gt;T</td>\n",
       "      <td>chr8:g.48813019T&gt;A</td>\n",
       "      <td>4.45</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCDH15</td>\n",
       "      <td>c.4793G&gt;A</td>\n",
       "      <td>chr10:g.55582714C&gt;T</td>\n",
       "      <td>3.99</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>3.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.1712G&gt;A</td>\n",
       "      <td>chr8:g.145007482C&gt;T</td>\n",
       "      <td>3.92</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>c.286G&gt;T</td>\n",
       "      <td>chr11:g.126141532G&gt;T</td>\n",
       "      <td>3.47</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>7.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GLYCTK</td>\n",
       "      <td>c.1313C&gt;T</td>\n",
       "      <td>chr3:g.52326883C&gt;T</td>\n",
       "      <td>3.34</td>\n",
       "      <td>1.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>5.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PIEZO1</td>\n",
       "      <td>c.1387G&gt;A</td>\n",
       "      <td>chr16:g.88802726C&gt;T</td>\n",
       "      <td>3.21</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|PP2</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>COL5A2</td>\n",
       "      <td>c.1976C&gt;T</td>\n",
       "      <td>chr2:g.189927592G&gt;A</td>\n",
       "      <td>3.10</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>5.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MTPAP</td>\n",
       "      <td>c.1533G&gt;C</td>\n",
       "      <td>chr10:g.30602754C&gt;G</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN5A</td>\n",
       "      <td>c.1982G&gt;A</td>\n",
       "      <td>chr3:g.38640450C&gt;T</td>\n",
       "      <td>2.92</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PYGL</td>\n",
       "      <td>c.1900G&gt;C</td>\n",
       "      <td>chr14:g.51378517C&gt;G</td>\n",
       "      <td>2.88</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP5</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>c.1174A&gt;T</td>\n",
       "      <td>chr18:g.57103187T&gt;A</td>\n",
       "      <td>2.82</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>3.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SLC13A5</td>\n",
       "      <td>c.1163A&gt;G</td>\n",
       "      <td>chr17:g.6596475T&gt;C</td>\n",
       "      <td>2.63</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PRSS1</td>\n",
       "      <td>c.502C&gt;A</td>\n",
       "      <td>chr7:g.142460329C&gt;A</td>\n",
       "      <td>2.52</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>c.1472A&gt;G</td>\n",
       "      <td>chr5:g.60817228A&gt;G</td>\n",
       "      <td>2.38</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.11402A&gt;G</td>\n",
       "      <td>chr8:g.144992998T&gt;C</td>\n",
       "      <td>2.24</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>c.1973G&gt;A</td>\n",
       "      <td>chr7:g.55238092G&gt;A</td>\n",
       "      <td>2.16</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|BP4</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CTGF</td>\n",
       "      <td>c.172C&gt;G</td>\n",
       "      <td>chr6:g.132272027G&gt;C</td>\n",
       "      <td>2.10</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>COL13A1</td>\n",
       "      <td>c.770G&gt;A</td>\n",
       "      <td>chr10:g.71658508G&gt;A</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>DNAJB6</td>\n",
       "      <td>c.365T&gt;A</td>\n",
       "      <td>chr7:g.157174958T&gt;A</td>\n",
       "      <td>1.90</td>\n",
       "      <td>1.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ACAD9</td>\n",
       "      <td>c.1414C&gt;T</td>\n",
       "      <td>chr3:g.128627871C&gt;T</td>\n",
       "      <td>1.64</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TGM5</td>\n",
       "      <td>c.23C&gt;T</td>\n",
       "      <td>chr15:g.43552765G&gt;A</td>\n",
       "      <td>1.49</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PCK2</td>\n",
       "      <td>c.1354G&gt;A</td>\n",
       "      <td>chr14:g.24573006G&gt;A</td>\n",
       "      <td>1.46</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TTN</td>\n",
       "      <td>c.57332G&gt;A</td>\n",
       "      <td>chr2:g.179462477C&gt;T</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3|BP1</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CEP135</td>\n",
       "      <td>c.1528C&gt;T</td>\n",
       "      <td>chr4:g.56846363C&gt;T</td>\n",
       "      <td>1.22</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>AICDA</td>\n",
       "      <td>c.428-5_428-4insTT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SOS2</td>\n",
       "      <td>c.2786-6_2786-5insT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>c.211-18_211-17insT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>17.071429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>DCTN1</td>\n",
       "      <td>c.3746C&gt;T</td>\n",
       "      <td>chr2:g.74588717G&gt;A</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3|PP2</td>\n",
       "      <td>3.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>CAT</td>\n",
       "      <td>c.1057-5T&gt;C</td>\n",
       "      <td>chr11:g.34482793T&gt;C</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP4</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SYNE1</td>\n",
       "      <td>c.17203-6delT</td>\n",
       "      <td></td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PFKM</td>\n",
       "      <td>c.1340+6G&gt;T</td>\n",
       "      <td>chr12:g.48533138G&gt;T</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP4</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>PON1</td>\n",
       "      <td>c.380T&gt;G</td>\n",
       "      <td>chr7:g.94940880A&gt;C</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>-0.696000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|BS2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>EYS</td>\n",
       "      <td>c.6079-4_6079-3delTC</td>\n",
       "      <td></td>\n",
       "      <td>-0.84</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ABCC6</td>\n",
       "      <td>c.3207C&gt;T</td>\n",
       "      <td>chr16:g.16259579G&gt;A</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP7</td>\n",
       "      <td>4.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PMS2</td>\n",
       "      <td>c.706-4delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.01</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>4.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>SMARCA2</td>\n",
       "      <td>c.705_707delGCA</td>\n",
       "      <td></td>\n",
       "      <td>-1.06</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>NEB</td>\n",
       "      <td>c.11077-5delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.12</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>c.2626-4delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.18</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>7.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>c.1165-6delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.25</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>9.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>LRP5</td>\n",
       "      <td>c.58_60delCTG</td>\n",
       "      <td></td>\n",
       "      <td>-1.36</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>12.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>WFS1</td>\n",
       "      <td>c.1760G&gt;A</td>\n",
       "      <td>chr4:g.6303282G&gt;A</td>\n",
       "      <td>-1.44</td>\n",
       "      <td>-0.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|BS2|PP3|BP6</td>\n",
       "      <td>7.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SMPD1</td>\n",
       "      <td>c.107_118delTGCTGGCGCTGG</td>\n",
       "      <td></td>\n",
       "      <td>-1.58</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>EIF2AK3</td>\n",
       "      <td>c.61_63delCTG</td>\n",
       "      <td></td>\n",
       "      <td>-1.91</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>10.041667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>KIF1A</td>\n",
       "      <td>c.2751_2753delGGA</td>\n",
       "      <td></td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>12.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>ANK1</td>\n",
       "      <td>c.1096G&gt;A</td>\n",
       "      <td>chr8:g.41577289C&gt;T</td>\n",
       "      <td>-2.61</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP1</td>\n",
       "      <td>5.375000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>KAT6A</td>\n",
       "      <td>c.5597C&gt;G</td>\n",
       "      <td>chr8:g.41790141G&gt;C</td>\n",
       "      <td>-2.81</td>\n",
       "      <td>-1.375000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP1</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>CDH23</td>\n",
       "      <td>c.2263C&gt;T</td>\n",
       "      <td>chr10:g.73453990C&gt;T</td>\n",
       "      <td>-2.91</td>\n",
       "      <td>-1.666667</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2|PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.316-125A&gt;G</td>\n",
       "      <td>chr11:g.5247081T&gt;C</td>\n",
       "      <td>-3.09</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|BS2</td>\n",
       "      <td>19.213095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PCLO</td>\n",
       "      <td>c.1501T&gt;C</td>\n",
       "      <td>chr7:g.82784456_82784486delinsGTGGGGGCTTTGCTGAGCCAGGCTGTTGAGG</td>\n",
       "      <td>-8.51</td>\n",
       "      <td>-4.000000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>PM1|BS1|BA1</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>HLA-DQB1</td>\n",
       "      <td>c.85C&gt;G</td>\n",
       "      <td>chr6:g.32634300G&gt;C</td>\n",
       "      <td>-8.57</td>\n",
       "      <td>-4.500000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>PM1|BS1|BA1|BP4</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ERAP1</td>\n",
       "      <td>c.380G&gt;C</td>\n",
       "      <td>chr5:g.96139250C&gt;G</td>\n",
       "      <td>-8.91</td>\n",
       "      <td>-4.875000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|PP3|BA1</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>chr2:g.128180746C&gt;T</td>\n",
       "      <td>-8.97</td>\n",
       "      <td>-3.500000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>BS2|BP4|BP7|BP6</td>\n",
       "      <td>10.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>chr11:g.62383715G&gt;C</td>\n",
       "      <td>-12.40</td>\n",
       "      <td>-5.500000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|BA1|BP4</td>\n",
       "      <td>6.804762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>chr17:g.3397702C&gt;T</td>\n",
       "      <td>-14.57</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|BA1|BP7|BP6</td>\n",
       "      <td>7.566667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene                   variant  \\\n",
       "0       BBS10                  c.145C>T   \n",
       "1         HBB                  c.208G>A   \n",
       "2        DLX3                  c.335A>G   \n",
       "3       PROP1                  c.652A>C   \n",
       "4        SOX3                  c.818C>T   \n",
       "5      CEP152                 c.3530A>G   \n",
       "6       PRKDC                 c.3278A>T   \n",
       "7      PCDH15                 c.4793G>A   \n",
       "8        PLEC                 c.1712G>A   \n",
       "9     FOXRED1                  c.286G>T   \n",
       "10     GLYCTK                 c.1313C>T   \n",
       "11     PIEZO1                 c.1387G>A   \n",
       "12     COL5A2                 c.1976C>T   \n",
       "13      MTPAP                 c.1533G>C   \n",
       "14      SCN5A                 c.1982G>A   \n",
       "15       PYGL                 c.1900G>C   \n",
       "16      CCBE1                 c.1174A>T   \n",
       "17    SLC13A5                 c.1163A>G   \n",
       "18      PRSS1                  c.502C>A   \n",
       "19     ZSWIM6                 c.1472A>G   \n",
       "20       PLEC                c.11402A>G   \n",
       "21       EGFR                 c.1973G>A   \n",
       "22       CTGF                  c.172C>G   \n",
       "23    COL13A1                  c.770G>A   \n",
       "24     DNAJB6                  c.365T>A   \n",
       "25      ACAD9                 c.1414C>T   \n",
       "26       TGM5                   c.23C>T   \n",
       "27       PCK2                 c.1354G>A   \n",
       "28        TTN                c.57332G>A   \n",
       "29     CEP135                 c.1528C>T   \n",
       "..        ...                       ...   \n",
       "82      AICDA        c.428-5_428-4insTT   \n",
       "83       SOS2       c.2786-6_2786-5insT   \n",
       "84       PTEN       c.211-18_211-17insT   \n",
       "85      DCTN1                 c.3746C>T   \n",
       "86        CAT               c.1057-5T>C   \n",
       "87      SYNE1             c.17203-6delT   \n",
       "88       PFKM               c.1340+6G>T   \n",
       "89       PON1                  c.380T>G   \n",
       "90        EYS      c.6079-4_6079-3delTC   \n",
       "91      ABCC6                 c.3207C>T   \n",
       "92       PMS2               c.706-4delT   \n",
       "93    SMARCA2           c.705_707delGCA   \n",
       "94        NEB             c.11077-5delT   \n",
       "95       TSC1              c.2626-4delT   \n",
       "96      ABCC9              c.1165-6delT   \n",
       "97       LRP5             c.58_60delCTG   \n",
       "98       WFS1                 c.1760G>A   \n",
       "99      SMPD1  c.107_118delTGCTGGCGCTGG   \n",
       "100   EIF2AK3             c.61_63delCTG   \n",
       "101     KIF1A         c.2751_2753delGGA   \n",
       "102      ANK1                 c.1096G>A   \n",
       "103     KAT6A                 c.5597C>G   \n",
       "104     CDH23                 c.2263C>T   \n",
       "105       HBB              c.316-125A>G   \n",
       "106      PCLO                 c.1501T>C   \n",
       "107  HLA-DQB1                   c.85C>G   \n",
       "108     ERAP1                  c.380G>C   \n",
       "109      PROC                  c.399C>T   \n",
       "110    B3GAT3                c.910-3C>G   \n",
       "111      ASPA                  c.693C>T   \n",
       "\n",
       "                                                                id  \\\n",
       "0                                              chr12:g.76741994G>A   \n",
       "1                                               chr11:g.5247914C>T   \n",
       "2                                              chr17:g.48070945T>C   \n",
       "3                                              chr5:g.177419739T>G   \n",
       "4                                              chrX:g.139586408G>A   \n",
       "5                                              chr15:g.49040744T>C   \n",
       "6                                               chr8:g.48813019T>A   \n",
       "7                                              chr10:g.55582714C>T   \n",
       "8                                              chr8:g.145007482C>T   \n",
       "9                                             chr11:g.126141532G>T   \n",
       "10                                              chr3:g.52326883C>T   \n",
       "11                                             chr16:g.88802726C>T   \n",
       "12                                             chr2:g.189927592G>A   \n",
       "13                                             chr10:g.30602754C>G   \n",
       "14                                              chr3:g.38640450C>T   \n",
       "15                                             chr14:g.51378517C>G   \n",
       "16                                             chr18:g.57103187T>A   \n",
       "17                                              chr17:g.6596475T>C   \n",
       "18                                             chr7:g.142460329C>A   \n",
       "19                                              chr5:g.60817228A>G   \n",
       "20                                             chr8:g.144992998T>C   \n",
       "21                                              chr7:g.55238092G>A   \n",
       "22                                             chr6:g.132272027G>C   \n",
       "23                                             chr10:g.71658508G>A   \n",
       "24                                             chr7:g.157174958T>A   \n",
       "25                                             chr3:g.128627871C>T   \n",
       "26                                             chr15:g.43552765G>A   \n",
       "27                                             chr14:g.24573006G>A   \n",
       "28                                             chr2:g.179462477C>T   \n",
       "29                                              chr4:g.56846363C>T   \n",
       "..                                                             ...   \n",
       "82                                                                   \n",
       "83                                                                   \n",
       "84                                                                   \n",
       "85                                              chr2:g.74588717G>A   \n",
       "86                                             chr11:g.34482793T>C   \n",
       "87                                                                   \n",
       "88                                             chr12:g.48533138G>T   \n",
       "89                                              chr7:g.94940880A>C   \n",
       "90                                                                   \n",
       "91                                             chr16:g.16259579G>A   \n",
       "92                                                                   \n",
       "93                                                                   \n",
       "94                                                                   \n",
       "95                                                                   \n",
       "96                                                                   \n",
       "97                                                                   \n",
       "98                                               chr4:g.6303282G>A   \n",
       "99                                                                   \n",
       "100                                                                  \n",
       "101                                                                  \n",
       "102                                             chr8:g.41577289C>T   \n",
       "103                                             chr8:g.41790141G>C   \n",
       "104                                            chr10:g.73453990C>T   \n",
       "105                                             chr11:g.5247081T>C   \n",
       "106  chr7:g.82784456_82784486delinsGTGGGGGCTTTGCTGAGCCAGGCTGTTGAGG   \n",
       "107                                             chr6:g.32634300G>C   \n",
       "108                                             chr5:g.96139250C>G   \n",
       "109                                            chr2:g.128180746C>T   \n",
       "110                                            chr11:g.62383715G>C   \n",
       "111                                             chr17:g.3397702C>T   \n",
       "\n",
       "     final_score  pathogenicity_score           pathogenicity  \\\n",
       "0           8.24             3.875000       Likely pathogenic   \n",
       "1           8.11             2.625000  Uncertain significance   \n",
       "2           6.15             3.961200  Uncertain significance   \n",
       "3           5.23             2.250000  Uncertain significance   \n",
       "4           5.19             2.250000  Uncertain significance   \n",
       "5           4.97             2.250000  Uncertain significance   \n",
       "6           4.45             2.250000  Uncertain significance   \n",
       "7           3.99             2.250000  Uncertain significance   \n",
       "8           3.92             1.750000  Uncertain significance   \n",
       "9           3.47             1.500000  Uncertain significance   \n",
       "10          3.34             1.625000  Uncertain significance   \n",
       "11          3.21             2.750000  Uncertain significance   \n",
       "12          3.10             1.500000  Uncertain significance   \n",
       "13          2.95             2.250000  Uncertain significance   \n",
       "14          2.92             2.500000  Uncertain significance   \n",
       "15          2.88             2.000000  Uncertain significance   \n",
       "16          2.82             1.500000  Uncertain significance   \n",
       "17          2.63             2.250000  Uncertain significance   \n",
       "18          2.52             1.750000  Uncertain significance   \n",
       "19          2.38             1.250000  Uncertain significance   \n",
       "20          2.24             1.000000           Likely benign   \n",
       "21          2.16             1.500000  Uncertain significance   \n",
       "22          2.10             1.250000  Uncertain significance   \n",
       "23          1.98             1.000000  Uncertain significance   \n",
       "24          1.90             1.625000  Uncertain significance   \n",
       "25          1.64             1.250000  Uncertain significance   \n",
       "26          1.49             1.250000  Uncertain significance   \n",
       "27          1.46             1.250000  Uncertain significance   \n",
       "28          1.31             0.750000  Uncertain significance   \n",
       "29          1.22             1.000000           Likely benign   \n",
       "..           ...                  ...                     ...   \n",
       "82          0.00             0.000000  Uncertain significance   \n",
       "83          0.00             0.000000  Uncertain significance   \n",
       "84          0.00             0.000000  Uncertain significance   \n",
       "85         -0.47            -0.250000  Uncertain significance   \n",
       "86         -0.58            -0.500000  Uncertain significance   \n",
       "87         -0.58            -0.500000  Uncertain significance   \n",
       "88         -0.63            -0.500000  Uncertain significance   \n",
       "89         -0.81            -0.696000  Uncertain significance   \n",
       "90         -0.84            -0.500000  Uncertain significance   \n",
       "91         -0.99            -0.500000  Uncertain significance   \n",
       "92         -1.01            -0.500000  Uncertain significance   \n",
       "93         -1.06            -0.500000  Uncertain significance   \n",
       "94         -1.12            -0.500000  Uncertain significance   \n",
       "95         -1.18            -0.500000  Uncertain significance   \n",
       "96         -1.25            -0.500000  Uncertain significance   \n",
       "97         -1.36            -0.500000  Uncertain significance   \n",
       "98         -1.44            -0.625000  Uncertain significance   \n",
       "99         -1.58            -0.750000  Uncertain significance   \n",
       "100        -1.91            -0.750000  Uncertain significance   \n",
       "101        -2.03            -0.750000  Uncertain significance   \n",
       "102        -2.61            -1.250000           Likely benign   \n",
       "103        -2.81            -1.375000           Likely benign   \n",
       "104        -2.91            -1.666667  Uncertain significance   \n",
       "105        -3.09            -1.000000  Uncertain significance   \n",
       "106        -8.51            -4.000000                  Benign   \n",
       "107        -8.57            -4.500000                  Benign   \n",
       "108        -8.91            -4.875000                  Benign   \n",
       "109        -8.97            -3.500000           Likely benign   \n",
       "110       -12.40            -5.500000                  Benign   \n",
       "111       -14.57            -6.250000                  Benign   \n",
       "\n",
       "            hit_criteria  hpo_hit_score  \n",
       "0        PS1|PM1|PP3|PP5       5.666667  \n",
       "1        PM5|PM1|PP3|PP5      19.213095  \n",
       "2    PS1|PM1|PM2|PP3|BP1       2.000000  \n",
       "3            PM1|PM2|PP3       7.500000  \n",
       "4            PM1|PM2|PP3       7.333333  \n",
       "5            PM1|PM2|PP3       6.404762  \n",
       "6            PM1|PM2|PP3       4.500000  \n",
       "7            PM1|PM2|PP3       3.166667  \n",
       "8        PM1|PM2|PP3|BP1       6.666667  \n",
       "9                PM1|PP3       7.400000  \n",
       "10           PM1|PP3|PP2       5.100000  \n",
       "11       PM1|PM2|PP3|PP2       0.500000  \n",
       "12               PM1|PP3       5.166667  \n",
       "13           PM1|PM2|PP3       1.000000  \n",
       "14           PM1|PM2|PP3       0.500000  \n",
       "15           PM1|PP3|PP5       1.500000  \n",
       "16               PM2|PP3       3.833333  \n",
       "17           PM1|PM2|PP3       0.500000  \n",
       "18           PM1|PP3|PP2       1.500000  \n",
       "19               PM1|PP3       4.000000  \n",
       "20       PM1|PM2|BP1|BP4       6.666667  \n",
       "21           PM1|PM2|BP4       1.500000  \n",
       "22               PM1|PP3       2.666667  \n",
       "23           PM1|PP3|BP1       4.500000  \n",
       "24           PM1|PP3|PP2       0.500000  \n",
       "25               PM1|PP3       1.000000  \n",
       "26               PM2|PP3       0.571429  \n",
       "27               PM1|PP3       0.500000  \n",
       "28           PM2|PP3|BP1       3.000000  \n",
       "29       PM1|PM2|BP1|BP4       0.666667  \n",
       "..                   ...            ...  \n",
       "82                             1.500000  \n",
       "83                             7.000000  \n",
       "84                            17.071429  \n",
       "85       PM1|BS2|PP3|PP2       3.833333  \n",
       "86                   BP4       0.500000  \n",
       "87                   BP6       0.500000  \n",
       "88                   BP4       0.833333  \n",
       "89       PS1|PM1|BS2|PP3       0.500000  \n",
       "90                   BP6       2.666667  \n",
       "91                   BP7       4.571429  \n",
       "92                   BP6       4.833333  \n",
       "93                   BP6       5.666667  \n",
       "94                   BP6       6.666667  \n",
       "95                   BP6       7.833333  \n",
       "96                   BP6       9.571429  \n",
       "97                   BP6      12.500000  \n",
       "98   PM1|PM2|BS2|PP3|BP6       7.333333  \n",
       "99                   BP6       5.500000  \n",
       "100                  BP6      10.041667  \n",
       "101                  BP6      12.166667  \n",
       "102      PM1|BS2|PP3|BP1       5.375000  \n",
       "103      PM1|BS2|PP3|BP1       5.000000  \n",
       "104              BS2|PP3       3.000000  \n",
       "105              PM2|BS2      19.213095  \n",
       "106          PM1|BS1|BA1       5.666667  \n",
       "107      PM1|BS1|BA1|BP4       4.000000  \n",
       "108          BS1|PP3|BA1       3.500000  \n",
       "109      BS2|BP4|BP7|BP6      10.266667  \n",
       "110          BS1|BA1|BP4       6.804762  \n",
       "111      BS1|BA1|BP7|BP6       7.566667  \n",
       "\n",
       "[112 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ACMG_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_list = list(ACMG_result['gene'])\n",
    "variant_list = list(ACMG_result['variant'])\n",
    "gene_variant_list = [(gene_list[i], variant_list[i]) for i in range(len(gene_list))]\n",
    "\n",
    "df_variant_ACMG_interpret = pd.DataFrame()\n",
    "df_variant_ACMG_interpret_chinese = pd.DataFrame() \n",
    "for key in gene_variant_list:\n",
    "    tmp_df = pd.DataFrame(variant_ACMG_interpretation[key], columns = ['criteria', 'interpretation'])\n",
    "    tmp_df['gene'] = key[0]\n",
    "    tmp_df['variant'] = key[1]\n",
    "    tmp_df = tmp_df[['gene', 'variant', 'criteria', 'interpretation']] \n",
    "    df_variant_ACMG_interpret = pd.concat([df_variant_ACMG_interpret, tmp_df])\n",
    "\n",
    "    tmp_df = pd.DataFrame(variant_ACMG_interpret_chinese[key], columns = ['criteria', 'interpretation'])\n",
    "    tmp_df['gene'] = key[0]\n",
    "    tmp_df['variant'] = key[1]\n",
    "    tmp_df = tmp_df[['gene', 'variant', 'criteria', 'interpretation']] \n",
    "    df_variant_ACMG_interpret_chinese = pd.concat([df_variant_ACMG_interpret_chinese, tmp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)极低(&lt; 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: Globin|Globin-like|Globin/Protoglobin.&lt;br/&gt;HGVS ID: chr11:g.5247914C&gt;T.&lt;br/&gt;RefSeq ID: rs33947415.&lt;br/&gt;外显子: 2.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.0008402.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.&lt;br/&gt;DANN致病性分数: 0.644724000156.&lt;br/&gt;FATHMM致病性分数: 0.93934.&lt;br/&gt;MetaSVM致病性分数: 0.65183.&lt;br/&gt;GERP++序列保守性预测分数: -10.2.&lt;br/&gt;Clinvar数据库ID: 15138.&lt;br/&gt;Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;HGVS ID: chr17:g.48070945T&gt;C.&lt;br/&gt;外显子: 2.&lt;br/&gt;DANN致病性分数: 0.993674333469.&lt;br/&gt;FATHMM致病性分数: 0.91543.&lt;br/&gt;MetaSVM致病性分数: 0.91694.&lt;br/&gt;GERP++序列保守性预测分数: 3.59.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: splice_region_variant&amp;intron_variant.&lt;br/&gt;HGVS ID: chr11:g.62383715G&gt;C.&lt;br/&gt;RefSeq ID: rs12794886.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.72.&lt;br/&gt;基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.&lt;br/&gt;基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: synonymous_variant.&lt;br/&gt;HGVS ID: chr17:g.3397702C&gt;T.&lt;br/&gt;RefSeq ID: rs12948217.&lt;br/&gt;外显子: 5.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.265550752376.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.21.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.&lt;br/&gt;Clinvar数据库ID: 197621.&lt;br/&gt;Clinvar数据库记录的变异致病性: Benign|Likely benign.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1278 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     variant criteria  \\\n",
       "0    BBS10    c.145C>T     变异注释   \n",
       "1    BBS10    c.145C>T     PVS1   \n",
       "2    BBS10    c.145C>T  PS1和PM5   \n",
       "3    BBS10    c.145C>T  PS3和BS3   \n",
       "4    BBS10    c.145C>T      PS4   \n",
       "5    BBS10    c.145C>T      PM2   \n",
       "6    BBS10    c.145C>T  BA1和BS1   \n",
       "7    BBS10    c.145C>T      BS2   \n",
       "8    BBS10    c.145C>T      PM1   \n",
       "9    BBS10    c.145C>T  PM4和BP3   \n",
       "10   BBS10    c.145C>T  PP2和BP1   \n",
       "11   BBS10    c.145C>T  PP3和BP4   \n",
       "12   BBS10    c.145C>T  PP5和BP6   \n",
       "13   BBS10    c.145C>T      BP7   \n",
       "0      HBB    c.208G>A     变异注释   \n",
       "1      HBB    c.208G>A     PVS1   \n",
       "2      HBB    c.208G>A  PS1和PM5   \n",
       "3      HBB    c.208G>A  PS3和BS3   \n",
       "4      HBB    c.208G>A      PS4   \n",
       "5      HBB    c.208G>A      PM2   \n",
       "6      HBB    c.208G>A  BA1和BS1   \n",
       "7      HBB    c.208G>A      BS2   \n",
       "8      HBB    c.208G>A      PM1   \n",
       "9      HBB    c.208G>A  PM4和BP3   \n",
       "10     HBB    c.208G>A  PP2和BP1   \n",
       "11     HBB    c.208G>A  PP3和BP4   \n",
       "12     HBB    c.208G>A  PP5和BP6   \n",
       "13     HBB    c.208G>A      BP7   \n",
       "0     DLX3    c.335A>G     变异注释   \n",
       "1     DLX3    c.335A>G     PVS1   \n",
       "..     ...         ...      ...   \n",
       "12    PROC    c.399C>T  PP5和BP6   \n",
       "13    PROC    c.399C>T      BP7   \n",
       "0   B3GAT3  c.910-3C>G     变异注释   \n",
       "1   B3GAT3  c.910-3C>G     PVS1   \n",
       "2   B3GAT3  c.910-3C>G  PS1和PM5   \n",
       "3   B3GAT3  c.910-3C>G  PS3和BS3   \n",
       "4   B3GAT3  c.910-3C>G      PS4   \n",
       "5   B3GAT3  c.910-3C>G      PM2   \n",
       "6   B3GAT3  c.910-3C>G  BA1和BS1   \n",
       "7   B3GAT3  c.910-3C>G      BS2   \n",
       "8   B3GAT3  c.910-3C>G      PM1   \n",
       "9   B3GAT3  c.910-3C>G  PM4和BP3   \n",
       "10  B3GAT3  c.910-3C>G  PP2和BP1   \n",
       "11  B3GAT3  c.910-3C>G  PP3和BP4   \n",
       "12  B3GAT3  c.910-3C>G  PP5和BP6   \n",
       "13  B3GAT3  c.910-3C>G      BP7   \n",
       "0     ASPA    c.693C>T     变异注释   \n",
       "1     ASPA    c.693C>T     PVS1   \n",
       "2     ASPA    c.693C>T  PS1和PM5   \n",
       "3     ASPA    c.693C>T  PS3和BS3   \n",
       "4     ASPA    c.693C>T      PS4   \n",
       "5     ASPA    c.693C>T      PM2   \n",
       "6     ASPA    c.693C>T  BA1和BS1   \n",
       "7     ASPA    c.693C>T      BS2   \n",
       "8     ASPA    c.693C>T      PM1   \n",
       "9     ASPA    c.693C>T  PM4和BP3   \n",
       "10    ASPA    c.693C>T  PP2和BP1   \n",
       "11    ASPA    c.693C>T  PP3和BP4   \n",
       "12    ASPA    c.693C>T  PP5和BP6   \n",
       "13    ASPA    c.693C>T      BP7   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            interpretation  \n",
       "0                  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.<br/>Clinvar数据库记录审核状态: criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                         变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异的最小等位基因频率(MAF)极低(< 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0   突变类型: missense_variant.<br/>蛋白功能区: Globin|Globin-like|Globin/Protoglobin.<br/>HGVS ID: chr11:g.5247914C>T.<br/>RefSeq ID: rs33947415.<br/>外显子: 2.<br/>ExAC 最小等位基因频率(MAF): 0.0008402.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.<br/>DANN致病性分数: 0.644724000156.<br/>FATHMM致病性分数: 0.93934.<br/>MetaSVM致病性分数: 0.65183.<br/>GERP++序列保守性预测分数: -10.2.<br/>Clinvar数据库ID: 15138.<br/>Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.<br/>Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                             变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                     突变类型: missense_variant.<br/>HGVS ID: chr17:g.48070945T>C.<br/>外显子: 2.<br/>DANN致病性分数: 0.993674333469.<br/>FATHMM致病性分数: 0.91543.<br/>MetaSVM致病性分数: 0.91694.<br/>GERP++序列保守性预测分数: 3.59.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ...  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                突变类型: splice_region_variant&intron_variant.<br/>HGVS ID: chr11:g.62383715G>C.<br/>RefSeq ID: rs12794886.<br/>1000Genomes 最小等位基因频率(MAF): 0.72.<br/>基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.<br/>基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                        突变类型: synonymous_variant.<br/>HGVS ID: chr17:g.3397702C>T.<br/>RefSeq ID: rs12948217.<br/>外显子: 5.<br/>ExAC 最小等位基因频率(MAF): 0.265550752376.<br/>1000Genomes 最小等位基因频率(MAF): 0.21.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.<br/>Clinvar数据库ID: 197621.<br/>Clinvar数据库记录的变异致病性: Benign|Likely benign.<br/>Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "\n",
       "[1278 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)极低(&lt; 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: Globin|Globin-like|Globin/Protoglobin.&lt;br/&gt;HGVS ID: chr11:g.5247914C&gt;T.&lt;br/&gt;RefSeq ID: rs33947415.&lt;br/&gt;外显子: 2.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.0008402.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.&lt;br/&gt;DANN致病性分数: 0.644724000156.&lt;br/&gt;FATHMM致病性分数: 0.93934.&lt;br/&gt;MetaSVM致病性分数: 0.65183.&lt;br/&gt;GERP++序列保守性预测分数: -10.2.&lt;br/&gt;Clinvar数据库ID: 15138.&lt;br/&gt;Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;HGVS ID: chr17:g.48070945T&gt;C.&lt;br/&gt;外显子: 2.&lt;br/&gt;DANN致病性分数: 0.993674333469.&lt;br/&gt;FATHMM致病性分数: 0.91543.&lt;br/&gt;MetaSVM致病性分数: 0.91694.&lt;br/&gt;GERP++序列保守性预测分数: 3.59.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: splice_region_variant&amp;intron_variant.&lt;br/&gt;HGVS ID: chr11:g.62383715G&gt;C.&lt;br/&gt;RefSeq ID: rs12794886.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.72.&lt;br/&gt;基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.&lt;br/&gt;基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: synonymous_variant.&lt;br/&gt;HGVS ID: chr17:g.3397702C&gt;T.&lt;br/&gt;RefSeq ID: rs12948217.&lt;br/&gt;外显子: 5.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.265550752376.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.21.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.&lt;br/&gt;Clinvar数据库ID: 197621.&lt;br/&gt;Clinvar数据库记录的变异致病性: Benign|Likely benign.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1278 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     variant criteria  \\\n",
       "0    BBS10    c.145C>T     变异注释   \n",
       "1    BBS10    c.145C>T     PVS1   \n",
       "2    BBS10    c.145C>T  PS1和PM5   \n",
       "3    BBS10    c.145C>T  PS3和BS3   \n",
       "4    BBS10    c.145C>T      PS4   \n",
       "5    BBS10    c.145C>T      PM2   \n",
       "6    BBS10    c.145C>T  BA1和BS1   \n",
       "7    BBS10    c.145C>T      BS2   \n",
       "8    BBS10    c.145C>T      PM1   \n",
       "9    BBS10    c.145C>T  PM4和BP3   \n",
       "10   BBS10    c.145C>T  PP2和BP1   \n",
       "11   BBS10    c.145C>T  PP3和BP4   \n",
       "12   BBS10    c.145C>T  PP5和BP6   \n",
       "13   BBS10    c.145C>T      BP7   \n",
       "0      HBB    c.208G>A     变异注释   \n",
       "1      HBB    c.208G>A     PVS1   \n",
       "2      HBB    c.208G>A  PS1和PM5   \n",
       "3      HBB    c.208G>A  PS3和BS3   \n",
       "4      HBB    c.208G>A      PS4   \n",
       "5      HBB    c.208G>A      PM2   \n",
       "6      HBB    c.208G>A  BA1和BS1   \n",
       "7      HBB    c.208G>A      BS2   \n",
       "8      HBB    c.208G>A      PM1   \n",
       "9      HBB    c.208G>A  PM4和BP3   \n",
       "10     HBB    c.208G>A  PP2和BP1   \n",
       "11     HBB    c.208G>A  PP3和BP4   \n",
       "12     HBB    c.208G>A  PP5和BP6   \n",
       "13     HBB    c.208G>A      BP7   \n",
       "0     DLX3    c.335A>G     变异注释   \n",
       "1     DLX3    c.335A>G     PVS1   \n",
       "..     ...         ...      ...   \n",
       "12    PROC    c.399C>T  PP5和BP6   \n",
       "13    PROC    c.399C>T      BP7   \n",
       "0   B3GAT3  c.910-3C>G     变异注释   \n",
       "1   B3GAT3  c.910-3C>G     PVS1   \n",
       "2   B3GAT3  c.910-3C>G  PS1和PM5   \n",
       "3   B3GAT3  c.910-3C>G  PS3和BS3   \n",
       "4   B3GAT3  c.910-3C>G      PS4   \n",
       "5   B3GAT3  c.910-3C>G      PM2   \n",
       "6   B3GAT3  c.910-3C>G  BA1和BS1   \n",
       "7   B3GAT3  c.910-3C>G      BS2   \n",
       "8   B3GAT3  c.910-3C>G      PM1   \n",
       "9   B3GAT3  c.910-3C>G  PM4和BP3   \n",
       "10  B3GAT3  c.910-3C>G  PP2和BP1   \n",
       "11  B3GAT3  c.910-3C>G  PP3和BP4   \n",
       "12  B3GAT3  c.910-3C>G  PP5和BP6   \n",
       "13  B3GAT3  c.910-3C>G      BP7   \n",
       "0     ASPA    c.693C>T     变异注释   \n",
       "1     ASPA    c.693C>T     PVS1   \n",
       "2     ASPA    c.693C>T  PS1和PM5   \n",
       "3     ASPA    c.693C>T  PS3和BS3   \n",
       "4     ASPA    c.693C>T      PS4   \n",
       "5     ASPA    c.693C>T      PM2   \n",
       "6     ASPA    c.693C>T  BA1和BS1   \n",
       "7     ASPA    c.693C>T      BS2   \n",
       "8     ASPA    c.693C>T      PM1   \n",
       "9     ASPA    c.693C>T  PM4和BP3   \n",
       "10    ASPA    c.693C>T  PP2和BP1   \n",
       "11    ASPA    c.693C>T  PP3和BP4   \n",
       "12    ASPA    c.693C>T  PP5和BP6   \n",
       "13    ASPA    c.693C>T      BP7   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            interpretation  \n",
       "0                  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.<br/>Clinvar数据库记录审核状态: criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                         变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异的最小等位基因频率(MAF)极低(< 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0   突变类型: missense_variant.<br/>蛋白功能区: Globin|Globin-like|Globin/Protoglobin.<br/>HGVS ID: chr11:g.5247914C>T.<br/>RefSeq ID: rs33947415.<br/>外显子: 2.<br/>ExAC 最小等位基因频率(MAF): 0.0008402.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.<br/>DANN致病性分数: 0.644724000156.<br/>FATHMM致病性分数: 0.93934.<br/>MetaSVM致病性分数: 0.65183.<br/>GERP++序列保守性预测分数: -10.2.<br/>Clinvar数据库ID: 15138.<br/>Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.<br/>Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                             变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                     突变类型: missense_variant.<br/>HGVS ID: chr17:g.48070945T>C.<br/>外显子: 2.<br/>DANN致病性分数: 0.993674333469.<br/>FATHMM致病性分数: 0.91543.<br/>MetaSVM致病性分数: 0.91694.<br/>GERP++序列保守性预测分数: 3.59.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ...  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                突变类型: splice_region_variant&intron_variant.<br/>HGVS ID: chr11:g.62383715G>C.<br/>RefSeq ID: rs12794886.<br/>1000Genomes 最小等位基因频率(MAF): 0.72.<br/>基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.<br/>基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                        突变类型: synonymous_variant.<br/>HGVS ID: chr17:g.3397702C>T.<br/>RefSeq ID: rs12948217.<br/>外显子: 5.<br/>ExAC 最小等位基因频率(MAF): 0.265550752376.<br/>1000Genomes 最小等位基因频率(MAF): 0.21.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.<br/>Clinvar数据库ID: 197621.<br/>Clinvar数据库记录的变异致病性: Benign|Likely benign.<br/>Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "\n",
       "[1278 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('variant_annotations',\n",
       "  u\"Effect: missense_variant.<br/>Protein domain: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>exon: 1.<br/>ExAC MAF: 5.052e-05.<br/>DANN pathogenicity score: 0.999117068756.<br/>FATHMM pathogenicity score: 0.79186.<br/>MetaSVM pathogenicity score: 0.84564.<br/>GERP++ conservation score: 4.27.<br/>Clinvar variation ids: 225010.<br/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br/>Clinvar review status: criteria provided, single submitter.<br/>Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].\"),\n",
       " ('PVS1',\n",
       "  'Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.'),\n",
       " ('PS1 and PM5',\n",
       "  'Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.'),\n",
       " ('PS3 and BS3',\n",
       "  'Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.'),\n",
       " ('PS4',\n",
       "  'Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.'),\n",
       " ('PM2',\n",
       "  'The variant has an extremely low frequency with MAF < 0.5%, but NOT causes recessive diseases. PM2 is NOT met.'),\n",
       " ('BA1 and BS1',\n",
       "  'Allele frequeny is <= 1%. BA1 is NOT met. \\xe4\\xb8\\x8d\\xe7\\xac\\xa6\\xe5\\x90\\x88BA1\\xe6\\xa0\\x87\\xe5\\x87\\x86. BS1 is NOT met.'),\n",
       " ('BS2',\n",
       "  'The variant is NOT observed in a healthy adult in the 1000 Genomes Project as a heterozygote (for dominant diseases) nor as a homozygote (for recessive diseases). BS2 is NOT met.'),\n",
       " ('PM1',\n",
       "  'The variant has missense effect in a critical protein domain where all missense variants in the domain identified to date have been shown to be pathogenic. PM1 is met.'),\n",
       " ('PM4 and BP3',\n",
       "  'Allele is in a repeat region. PM4 is NOT met. BP3 is NOT met.'),\n",
       " ('PP2 and BP1',\n",
       "  'The variant effect is missense, but neither in a gene that has a low rate of benign missense variation and in which missense variants are a common mechanism of disease nor in a gene for which primarily truncating variants are known to cause disease. PP2 is NOT met. BP1 is NOT met.'),\n",
       " ('PP3 and BP4',\n",
       "  'The cutoffs for computational pathogenicity predictions are DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. PP3 is partially met. BP4 is NOT met.'),\n",
       " ('PP5 and BP6',\n",
       "  'Clinvar reports the variant as pathogenic (Clinvar: 225010). Pubmed references: 25356970, 20120035, 16582908, 21044901, 20876674, 21642631, 20498079, 21517826, 20177705, 24746959, 25982971. PP5 is met. BP6 is NOT met.'),\n",
       " ('BP7', 'Variant effect is not synonymous. BP7 is NOT met.')]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpretation[('BBS10', 'c.145C>T')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>protein</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "      <th>pheno_match_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>p.Arg49Trp</td>\n",
       "      <td>chr12:g.76741994G&gt;A</td>\n",
       "      <td>9.511010</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM1|PP3|PP5</td>\n",
       "      <td>3.8750</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>1.154249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>p.Gly70Ser</td>\n",
       "      <td>chr11:g.5247914C&gt;T</td>\n",
       "      <td>8.713391</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM5|PM1|PP3|PP5</td>\n",
       "      <td>2.6250</td>\n",
       "      <td>19.213095</td>\n",
       "      <td>1.074401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>p.Lys112Arg</td>\n",
       "      <td>chr17:g.48070945T&gt;C</td>\n",
       "      <td>6.150000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|PM2|PP3|BP1</td>\n",
       "      <td>3.9612</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>c.652A&gt;C</td>\n",
       "      <td>p.Ser218Arg</td>\n",
       "      <td>chr5:g.177419739T&gt;G</td>\n",
       "      <td>5.230000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>SOX3</td>\n",
       "      <td>c.818C&gt;T</td>\n",
       "      <td>p.Ser273Leu</td>\n",
       "      <td>chrX:g.139586408G&gt;A</td>\n",
       "      <td>5.190000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.333333</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene   variant      protein                   id  final_score  \\\n",
       "76   BBS10  c.145C>T   p.Arg49Trp  chr12:g.76741994G>A     9.511010   \n",
       "103    HBB  c.208G>A   p.Gly70Ser   chr11:g.5247914C>T     8.713391   \n",
       "66    DLX3  c.335A>G  p.Lys112Arg  chr17:g.48070945T>C     6.150000   \n",
       "53   PROP1  c.652A>C  p.Ser218Arg  chr5:g.177419739T>G     5.230000   \n",
       "107   SOX3  c.818C>T  p.Ser273Leu  chrX:g.139586408G>A     5.190000   \n",
       "\n",
       "              pathogenicity         hit_criteria  pathogenicity_score  \\\n",
       "76        Likely pathogenic      PS1|PM1|PP3|PP5               3.8750   \n",
       "103  Uncertain significance      PM5|PM1|PP3|PP5               2.6250   \n",
       "66   Uncertain significance  PS1|PM1|PM2|PP3|BP1               3.9612   \n",
       "53   Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "107  Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "\n",
       "     hpo_hit_score  pheno_match_score  \n",
       "76        5.666667           1.154249  \n",
       "103      19.213095           1.074401  \n",
       "66        2.000000           1.000000  \n",
       "53        7.500000           1.000000  \n",
       "107       7.333333           1.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Allele 1</th>\n",
       "      <th>Allele 2</th>\n",
       "      <th>Allele In Scope</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene Profile Report</th>\n",
       "      <th>Gene (gene)</th>\n",
       "      <th>cDNA (cNomen)</th>\n",
       "      <th>Protein (pNomen)</th>\n",
       "      <th>HGVS cDNA-level nomenclature (fullCNomen)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1635536</td>\n",
       "      <td>1635536</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_024011.2</td>\n",
       "      <td></td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>c.1735dupT</td>\n",
       "      <td>p.Y579Lfs*66</td>\n",
       "      <td>NM_024011.2:c.1735dupT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>23713843</td>\n",
       "      <td>23713843</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_003196.2</td>\n",
       "      <td></td>\n",
       "      <td>TCEA3</td>\n",
       "      <td>c.889A&gt;G</td>\n",
       "      <td>p.M297V</td>\n",
       "      <td>NM_003196.2:c.889A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>24882663</td>\n",
       "      <td>24882663</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NM_001010980.4</td>\n",
       "      <td></td>\n",
       "      <td>NCMAP</td>\n",
       "      <td>c.-8+3A&gt;G</td>\n",
       "      <td></td>\n",
       "      <td>NM_001010980.4:c.-8+3A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>26663842</td>\n",
       "      <td>26663842</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_001039775.3</td>\n",
       "      <td></td>\n",
       "      <td>AIM1L</td>\n",
       "      <td>c.3673C&gt;T</td>\n",
       "      <td>p.R1225W</td>\n",
       "      <td>NM_001039775.3:c.3673C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>39896387</td>\n",
       "      <td>39896387</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_012090.5</td>\n",
       "      <td></td>\n",
       "      <td>MACF1</td>\n",
       "      <td>c.10958A&gt;C</td>\n",
       "      <td>p.E3653A</td>\n",
       "      <td>NM_012090.5:c.10958A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start      Stop Reference Allele 1 Allele 2 Allele In Scope  \\\n",
       "0          1   1635536   1635536         .        .        A               A   \n",
       "1          1  23713843  23713843         T        C        T               C   \n",
       "2          1  24882663  24882663         A        A        G               G   \n",
       "3          1  26663842  26663842         G        G        A               A   \n",
       "4          1  39896387  39896387         A        C        A               C   \n",
       "\n",
       "       Transcript Gene Profile Report Gene (gene) cDNA (cNomen)  \\\n",
       "0     NM_024011.2                          CDK11A    c.1735dupT   \n",
       "1     NM_003196.2                           TCEA3      c.889A>G   \n",
       "2  NM_001010980.4                           NCMAP     c.-8+3A>G   \n",
       "3  NM_001039775.3                           AIM1L     c.3673C>T   \n",
       "4     NM_012090.5                           MACF1    c.10958A>C   \n",
       "\n",
       "  Protein (pNomen) HGVS cDNA-level nomenclature (fullCNomen)  \n",
       "0     p.Y579Lfs*66                    NM_024011.2:c.1735dupT  \n",
       "1          p.M297V                      NM_003196.2:c.889A>G  \n",
       "2                                   NM_001010980.4:c.-8+3A>G  \n",
       "3         p.R1225W                  NM_001039775.3:c.3673C>T  \n",
       "4         p.E3653A                    NM_012090.5:c.10958A>C  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'bilateral post-axial polydactyly of hands and feet',\n",
       " u'macrocephaly',\n",
       " u'tall stature',\n",
       " u'post-natal',\n",
       " u'central hypotonia',\n",
       " u'global developmental delay',\n",
       " u'obstructive sleep apnea',\n",
       " u'delayed cns central nervous myelination',\n",
       " u'white matter',\n",
       " u'corpus callosum',\n",
       " u'mildly ectopic neurohypophys posterior pituitary upslanted palpebral fissures',\n",
       " u'epicanthus',\n",
       " u'arched eyebrows',\n",
       " u'craniofacial asymmetry',\n",
       " u'bronchial stenosis',\n",
       " u'macrosomia',\n",
       " u'recurrent infection',\n",
       " u'gastroesophageal reflux',\n",
       " u'supratentorial perivascular space',\n",
       " u'mucopolysaccharidosis plasma urine creatine',\n",
       " u'purine',\n",
       " u'pyrimidine metabolism',\n",
       " u'gross motor',\n",
       " u'language',\n",
       " u'cognitive',\n",
       " u'ventriculomegaly',\n",
       " u'metatarsal bones',\n",
       " u'proximal phalanges',\n",
       " u'obesity',\n",
       " u'wheeze',\n",
       " u'brachycephaly frontal prominence',\n",
       " u'bitemporal narrowing',\n",
       " u'clinodactyly',\n",
       " u'brachydactyly',\n",
       " u'bardet-biedl',\n",
       " 'growth delay',\n",
       " 'developmental delay']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Chromosome',\n",
       " 'Start',\n",
       " 'Stop',\n",
       " 'Reference',\n",
       " 'Allele 1',\n",
       " 'Allele 2',\n",
       " 'Allele In Scope',\n",
       " 'Transcript',\n",
       " 'Gene Profile Report',\n",
       " 'Gene (gene)',\n",
       " 'cDNA (cNomen)',\n",
       " 'Protein (pNomen)',\n",
       " 'HGVS cDNA-level nomenclature (fullCNomen)']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>variant_annotations</td>\n",
       "      <td>Effect: missense_variant.&lt;br/&gt;Protein domain: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;exon: 1.&lt;br/&gt;ExAC MAF: 5.052e-05.&lt;br/&gt;DANN pathogenicity score: 0.999117068756.&lt;br/&gt;FATHMM pathogenicity score: 0.79186.&lt;br/&gt;MetaSVM pathogenicity score: 0.84564.&lt;br/&gt;GERP++ conservation score: 4.27.&lt;br/&gt;Clinvar variation ids: 225010.&lt;br/&gt;Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar review status: criteria provided, single submitter.&lt;br/&gt;Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1 and PM5</td>\n",
       "      <td>Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3 and BS3</td>\n",
       "      <td>Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene   variant             criteria  \\\n",
       "0  BBS10  c.145C>T  variant_annotations   \n",
       "1  BBS10  c.145C>T                 PVS1   \n",
       "2  BBS10  c.145C>T          PS1 and PM5   \n",
       "3  BBS10  c.145C>T          PS3 and BS3   \n",
       "4  BBS10  c.145C>T                  PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     interpretation  \n",
       "0  Effect: missense_variant.<br/>Protein domain: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>exon: 1.<br/>ExAC MAF: 5.052e-05.<br/>DANN pathogenicity score: 0.999117068756.<br/>FATHMM pathogenicity score: 0.79186.<br/>MetaSVM pathogenicity score: 0.84564.<br/>GERP++ conservation score: 4.27.<br/>Clinvar variation ids: 225010.<br/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br/>Clinvar review status: criteria provided, single submitter.<br/>Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.  \n",
       "2                                                                                                                                                                                                                                                                                                             Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpretation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>基因</th>\n",
       "      <th>变异</th>\n",
       "      <th>标准</th>\n",
       "      <th>解读</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: 致病|致病.&lt;br/&gt;Clinvar数据库记录审核状态: 提供标准，单个提交者.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      基因        变异       标准  \\\n",
       "0  BBS10  c.145C>T     变异注释   \n",
       "1  BBS10  c.145C>T     PVS1   \n",
       "2  BBS10  c.145C>T  PS1和PM5   \n",
       "3  BBS10  c.145C>T  PS3和BS3   \n",
       "4  BBS10  c.145C>T      PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               解读  \n",
       "0  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: 致病|致病.<br/>Clinvar数据库记录审核状态: 提供标准，单个提交者.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                  基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpret_chinese.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_final_res, df_genes, phenos, field_names, variant_ACMG_interpretation, variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nput_gene = df_genes.to_json(orient='records')\n",
    "input_phenotype = ', '.join(phenos)\n",
    "result_table = ACMG_result.to_json(orient='records')\n",
    "interpretation = variant_ACMG_interpretation.to_json(orient='records')\n",
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')\n",
    "# logger.info(\"Finish processing data, start writing data to DB in background main task\")\n",
    "\n",
    "\n",
    "# sample = Main_table(\n",
    "#     task_id=raw_input_id,\n",
    "#     input_gene=input_gene,\n",
    "#     input_phenotype=input_phenotype,\n",
    "#     result=result_table,\n",
    "#     interpretation=interpretation,\n",
    "#     interpretation_chinese=interpretation_chinese,\n",
    "#     pub_date=timezone.now(),\n",
    "#     user_name=raw_input.user_name,\n",
    "#     task_name=raw_input.task_name,\n",
    "# )\n",
    "# sample.save()\n",
    "# logger.info(\"Finish writing data to DB in background main task\")\n",
    "# raw_input.status = \"succeed\"\n",
    "# raw_input.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"variant_annotations\",\"interpretation\":\"Effect: missense_variant.<br\\\\/>Protein domain: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>exon: 1.<br\\\\/>ExAC MAF: 5.052e-05.<br\\\\/>DANN pathogenicity score: 0.999117068756.<br\\\\/>FATHMM pathogenicity score: 0.79186.<br\\\\/>MetaSVM pathogenicity score: 0.84564.<br\\\\/>GERP++ conservation score: 4.27.<br\\\\/>Clinvar variation ids: 225010.<br\\\\/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br\\\\/>Clinvar review status: criteria provided, single submitter.<br\\\\/>Pubmed references from Clinvar: [\\'25356970\\', \\'20120035\\', \\'16582908\\', \\'21044901\\', \\'20876674\\', \\'21642631\\', \\'20498079\\', \\'21517826\\', \\'20177705\\', \\'24746959\\', \\'25982971\\'].\"},{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"PVS1\",\"interpretation\":\"Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have dama'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"\\\\u57fa\\\\u56e0\":\"BBS10\",\"\\\\u53d8\\\\u5f02\":\"c.145C>T\",\"\\\\u6807\\\\u51c6\":\"\\\\u53d8\\\\u5f02\\\\u6ce8\\\\u91ca\",\"\\\\u89e3\\\\u8bfb\":\"\\\\u7a81\\\\u53d8\\\\u7c7b\\\\u578b: missense_variant.<br\\\\/>\\\\u86cb\\\\u767d\\\\u529f\\\\u80fd\\\\u533a: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>\\\\u5916\\\\u663e\\\\u5b50: 1.<br\\\\/>ExAC \\\\u6700\\\\u5c0f\\\\u7b49\\\\u4f4d\\\\u57fa\\\\u56e0\\\\u9891\\\\u7387(MAF): 5.052e-05.<br\\\\/>DANN\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.999117068756.<br\\\\/>FATHMM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.79186.<br\\\\/>MetaSVM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.84564.<br\\\\/>GERP++\\\\u5e8f\\\\u5217\\\\u4fdd\\\\u5b88\\\\u6027\\\\u9884\\\\u6d4b\\\\u5206\\\\u6570: 4.27.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93ID: 225010.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684\\\\u53d8\\\\u5f02\\\\u81f4\\\\u75c5\\\\u6027: \\\\u81f4\\\\u75c5|\\\\u81f4\\\\u75c5.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u5ba1\\\\u6838\\\\u72b6\\\\u6001: \\\\u63d0\\\\u4f9b\\\\u6807\\\\u51c6\\\\uff0c\\\\u5355\\\\u4e2a\\\\u63d0\\\\u4ea4\\\\u8005.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684Pubmed\\\\u76f8\\\\u5173\\\\u751f\\\\u7269\\\\u533b\\\\u5b66\\\\u6587\\\\u'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation_chinese[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "round(2.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "8/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
