{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_file_path = 'input/input_genes.txt'\n",
    "phenotype_file_path = 'input/input_phenotype.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "too many values to unpack",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-392b0178f70f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[0mdf_hpo_ranking_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mranking_genes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'score'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'hits'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    199\u001b[0m \u001b[0mdf_hpo_ranking_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_hpo_ranking_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'score'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 200\u001b[0;31m \u001b[0mACMG_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant_ACMG_interpretation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant_ACMG_interpret_chinese\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mACMG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGet_ACMG_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_hpo_ranking_genes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariants\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_pubmed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    201\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    202\u001b[0m \u001b[0;31m# filter variant on phenotype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: too many values to unpack"
     ]
    }
   ],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "    '''get a valid hgvs name from VCF-style \"chrom, pos, ref, alt\" data.\n",
    "\n",
    "    Example:\n",
    "\n",
    "        >>> myvariant.format_hgvs(\"1\", 35366, \"C\", \"T\")\n",
    "        >>> myvariant.format_hgvs(\"2\", 17142, \"G\", \"GA\")\n",
    "        >>> myvariant.format_hgvs(\"MT\", 8270, \"CACCCCCTCT\", \"C\")\n",
    "        >>> myvariant.format_hgvs(\"X\", 107930849, \"GGA\", \"C\")\n",
    "\n",
    "    '''\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "    if not input_phenotype:\n",
    "        return '', ''\n",
    "    text = StringIO(unicode(input_phenotype), newline=None)\n",
    "    lines = text.readlines()\n",
    "    lines = [line.strip() for line in lines]\n",
    "    phenos = []\n",
    "    for line in lines:\n",
    "        if not line:\n",
    "            continue\n",
    "        phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "        phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "        phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "        phenos += phenos_each_line\n",
    "\n",
    "    corner_cases = dict()\n",
    "    for pheno in phenos:\n",
    "        if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "            phenos.append('growth delay')\n",
    "            corner_cases['growth delay'] = pheno.strip()\n",
    "    for pheno in phenos:\n",
    "        if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "            phenos.append('developmental delay')\n",
    "            corner_cases['developmental delay'] = pheno.strip()\n",
    "    return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "\tcandidate_vars = []\n",
    "\tinput_gene = input_gene.split('\\n')\n",
    "\theader = input_gene[0]\n",
    "\tsniffer = csv.Sniffer()\n",
    "\tdialect = sniffer.sniff(header)\n",
    "\tdelimiter =  dialect.delimiter\n",
    "\tfield_names = header.split(delimiter)\n",
    "\tchrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, 0 \n",
    "\n",
    "\tfor idx in xrange(len(field_names)):\n",
    "\t\tfield = field_names[idx]\n",
    "\t\tif re.match(r'chrom', field, re.I): chrom_idx = idx\n",
    "\t\tif re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "\t\tif re.match(r'ref', field, re.I): ref_idx = idx\n",
    "\t\tif re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "\t\tif re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "\tinput_gene_list = []\n",
    "\tCANDIDATE_GENES = []\n",
    "\tfor line in input_gene[1:]:\n",
    "\t\tif not line:\n",
    "\t\t\tcontinue\n",
    "\t\tline = line.rstrip()\n",
    "\t\tparts = re.split(r'%s' % delimiter, line)\n",
    "\t\tinput_gene_list.append(parts)\n",
    "\t\tgene = parts[gene_idx]\n",
    "\t\tCANDIDATE_GENES.append(gene)\n",
    "\t\ttranscript, variant, variant_id = '', '', ''\n",
    "\t\tfor part in parts:\n",
    "\t\t\tif re.search(r'_.*:c\\.', part):\n",
    "\t\t\t\ttranscript, variant = part.split(':')\n",
    "\t\t\telse:\n",
    "\t\t\t\tif re.search(r'c\\.', part):\n",
    "\t\t\t\t\tvariant = part\n",
    "\t\t\t\tif re.search(r'NM_', part, re.I):\n",
    "\t\t\t\t\ttranscript = part.split(':')[0]\n",
    "\t\t\tif re.search(r'_.*:g\\.', part):\n",
    "\t\t\t\tvariant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "\t\t\tif re.search(r'chr.*:g\\.', part, re.I):\n",
    "\t\t\t\tvariant_id = part\n",
    "\t\tif not variant_id and (chrom_idx and pos_idx and ref_idx and alt_idx):\n",
    "\t\t\tchrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "\t\t\tvariant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "\t\tcandidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "        # remove lines in the input file which has wrong number of fields\n",
    "        field_nums = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tfield_nums.append(len(line))\n",
    "        count = Counter(field_nums)\n",
    "        correct_field_num = count.most_common()[0][0]\n",
    "        correct_input_gene_list = []\n",
    "        for line in input_gene_list:\n",
    "\t\t\tif len(line) == correct_field_num:\n",
    "\t\t\t\tcorrect_input_gene_list.append(line)\n",
    "        df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "        return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "\tranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\t# collect variant info\n",
    "\thpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "\ttmp_candidate_vars = []\n",
    "\tfor var in candidate_vars:\n",
    "\t\tif var[0] in hpo_filtered_genes:\n",
    "\t\t\ttmp_candidate_vars.append(var)\n",
    "\tcandidate_vars = tmp_candidate_vars\n",
    "\treturn ranking_genes, candidate_vars\n",
    "\n",
    "# \tstatus_step = \"generating candidate variants ...\" \n",
    "# \traw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "# \tinput_gene = raw_input.raw_input_gene\n",
    "# \tinput_phenotype = raw_input.raw_input_phenotype\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/xiaonantest1_g.txt\",'rU').read()\n",
    "input_phenotype = open(\"/Users/Tianqi/Desktop/xiaonantest1_p.txt\",'rU').read()\n",
    "\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "# map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "\n",
    "if phenos:\n",
    "#     raw_input.status = \"Maping phenotypes to genes\"\n",
    "#     raw_input.save()\n",
    "    ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "else:\n",
    "    ranking_genes = []\n",
    "    for gene in CANDIDATE_GENES:\n",
    "        ranking_genes.append((gene, 1.0, 1))\n",
    "\n",
    "# collect variant info\n",
    "# raw_input.status = \"Annotating variants using genomic databases\"\n",
    "# raw_input.save()\n",
    "mv = myvariant.MyVariantInfo()\n",
    "final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "# pubmed\n",
    "# raw_input.status = \"Searching biomedical literatures\"\n",
    "# raw_input.save()\n",
    "df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "# ACMG\n",
    "# raw_input.status = \"Checking ACMG standard\"\n",
    "# raw_input.save()\n",
    "df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "ACMG_result, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "# filter variant on phenotype\n",
    "\n",
    "# if phenos:\n",
    "# #     raw_input.status = \"Filtering variants based on phenotypes\"\n",
    "# #     raw_input.save()\n",
    "#     df_final_res, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos, variant_ACMG_interpretation, variant_ACMG_interpret_chinese)\n",
    "# else:\n",
    "#     df_final_res = ACMG_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "not enough arguments for format string",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-f1eb3f93d918>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m[\u001b[0m\u001b[0;34m\"<a href='https://www.ncbi.nlm.nih.gov/pubmed/%s'> %s </a>\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0mi\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'b'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: not enough arguments for format string"
     ]
    }
   ],
   "source": [
    "[\"<a href='https://www.ncbi.nlm.nih.gov/pubmed/%s'> %s </a>\" %(i,i) for i in ['a','b']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>chr12:g.76741994G&gt;A</td>\n",
       "      <td>8.24</td>\n",
       "      <td>3.875000</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM1|PP3|PP5</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>chr11:g.5247914C&gt;T</td>\n",
       "      <td>8.11</td>\n",
       "      <td>2.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM5|PM1|PP3|PP5</td>\n",
       "      <td>19.213095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>chr17:g.48070945T&gt;C</td>\n",
       "      <td>6.15</td>\n",
       "      <td>3.961200</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|PM2|PP3|BP1</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>c.652A&gt;C</td>\n",
       "      <td>chr5:g.177419739T&gt;G</td>\n",
       "      <td>5.23</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SOX3</td>\n",
       "      <td>c.818C&gt;T</td>\n",
       "      <td>chrX:g.139586408G&gt;A</td>\n",
       "      <td>5.19</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CEP152</td>\n",
       "      <td>c.3530A&gt;G</td>\n",
       "      <td>chr15:g.49040744T&gt;C</td>\n",
       "      <td>4.97</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>6.404762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRKDC</td>\n",
       "      <td>c.3278A&gt;T</td>\n",
       "      <td>chr8:g.48813019T&gt;A</td>\n",
       "      <td>4.45</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCDH15</td>\n",
       "      <td>c.4793G&gt;A</td>\n",
       "      <td>chr10:g.55582714C&gt;T</td>\n",
       "      <td>3.99</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>3.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.1712G&gt;A</td>\n",
       "      <td>chr8:g.145007482C&gt;T</td>\n",
       "      <td>3.92</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>c.286G&gt;T</td>\n",
       "      <td>chr11:g.126141532G&gt;T</td>\n",
       "      <td>3.47</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>7.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GLYCTK</td>\n",
       "      <td>c.1313C&gt;T</td>\n",
       "      <td>chr3:g.52326883C&gt;T</td>\n",
       "      <td>3.34</td>\n",
       "      <td>1.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>5.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PIEZO1</td>\n",
       "      <td>c.1387G&gt;A</td>\n",
       "      <td>chr16:g.88802726C&gt;T</td>\n",
       "      <td>3.21</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|PP2</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>COL5A2</td>\n",
       "      <td>c.1976C&gt;T</td>\n",
       "      <td>chr2:g.189927592G&gt;A</td>\n",
       "      <td>3.10</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>5.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MTPAP</td>\n",
       "      <td>c.1533G&gt;C</td>\n",
       "      <td>chr10:g.30602754C&gt;G</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN5A</td>\n",
       "      <td>c.1982G&gt;A</td>\n",
       "      <td>chr3:g.38640450C&gt;T</td>\n",
       "      <td>2.92</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PYGL</td>\n",
       "      <td>c.1900G&gt;C</td>\n",
       "      <td>chr14:g.51378517C&gt;G</td>\n",
       "      <td>2.88</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP5</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>c.1174A&gt;T</td>\n",
       "      <td>chr18:g.57103187T&gt;A</td>\n",
       "      <td>2.82</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>3.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SLC13A5</td>\n",
       "      <td>c.1163A&gt;G</td>\n",
       "      <td>chr17:g.6596475T&gt;C</td>\n",
       "      <td>2.63</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PRSS1</td>\n",
       "      <td>c.502C&gt;A</td>\n",
       "      <td>chr7:g.142460329C&gt;A</td>\n",
       "      <td>2.52</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>c.1472A&gt;G</td>\n",
       "      <td>chr5:g.60817228A&gt;G</td>\n",
       "      <td>2.38</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.11402A&gt;G</td>\n",
       "      <td>chr8:g.144992998T&gt;C</td>\n",
       "      <td>2.24</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>c.1973G&gt;A</td>\n",
       "      <td>chr7:g.55238092G&gt;A</td>\n",
       "      <td>2.16</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|BP4</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CTGF</td>\n",
       "      <td>c.172C&gt;G</td>\n",
       "      <td>chr6:g.132272027G&gt;C</td>\n",
       "      <td>2.10</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>COL13A1</td>\n",
       "      <td>c.770G&gt;A</td>\n",
       "      <td>chr10:g.71658508G&gt;A</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>DNAJB6</td>\n",
       "      <td>c.365T&gt;A</td>\n",
       "      <td>chr7:g.157174958T&gt;A</td>\n",
       "      <td>1.90</td>\n",
       "      <td>1.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ACAD9</td>\n",
       "      <td>c.1414C&gt;T</td>\n",
       "      <td>chr3:g.128627871C&gt;T</td>\n",
       "      <td>1.64</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TGM5</td>\n",
       "      <td>c.23C&gt;T</td>\n",
       "      <td>chr15:g.43552765G&gt;A</td>\n",
       "      <td>1.49</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PCK2</td>\n",
       "      <td>c.1354G&gt;A</td>\n",
       "      <td>chr14:g.24573006G&gt;A</td>\n",
       "      <td>1.46</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TTN</td>\n",
       "      <td>c.57332G&gt;A</td>\n",
       "      <td>chr2:g.179462477C&gt;T</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3|BP1</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CEP135</td>\n",
       "      <td>c.1528C&gt;T</td>\n",
       "      <td>chr4:g.56846363C&gt;T</td>\n",
       "      <td>1.22</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>AICDA</td>\n",
       "      <td>c.428-5_428-4insTT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SOS2</td>\n",
       "      <td>c.2786-6_2786-5insT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>c.211-18_211-17insT</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td></td>\n",
       "      <td>17.071429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>DCTN1</td>\n",
       "      <td>c.3746C&gt;T</td>\n",
       "      <td>chr2:g.74588717G&gt;A</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3|PP2</td>\n",
       "      <td>3.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>CAT</td>\n",
       "      <td>c.1057-5T&gt;C</td>\n",
       "      <td>chr11:g.34482793T&gt;C</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP4</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SYNE1</td>\n",
       "      <td>c.17203-6delT</td>\n",
       "      <td></td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PFKM</td>\n",
       "      <td>c.1340+6G&gt;T</td>\n",
       "      <td>chr12:g.48533138G&gt;T</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP4</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>PON1</td>\n",
       "      <td>c.380T&gt;G</td>\n",
       "      <td>chr7:g.94940880A&gt;C</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>-0.696000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|BS2|PP3</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>EYS</td>\n",
       "      <td>c.6079-4_6079-3delTC</td>\n",
       "      <td></td>\n",
       "      <td>-0.84</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ABCC6</td>\n",
       "      <td>c.3207C&gt;T</td>\n",
       "      <td>chr16:g.16259579G&gt;A</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP7</td>\n",
       "      <td>4.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PMS2</td>\n",
       "      <td>c.706-4delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.01</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>4.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>SMARCA2</td>\n",
       "      <td>c.705_707delGCA</td>\n",
       "      <td></td>\n",
       "      <td>-1.06</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>NEB</td>\n",
       "      <td>c.11077-5delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.12</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>c.2626-4delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.18</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>7.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ABCC9</td>\n",
       "      <td>c.1165-6delT</td>\n",
       "      <td></td>\n",
       "      <td>-1.25</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>9.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>LRP5</td>\n",
       "      <td>c.58_60delCTG</td>\n",
       "      <td></td>\n",
       "      <td>-1.36</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>12.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>WFS1</td>\n",
       "      <td>c.1760G&gt;A</td>\n",
       "      <td>chr4:g.6303282G&gt;A</td>\n",
       "      <td>-1.44</td>\n",
       "      <td>-0.625000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|BS2|PP3|BP6</td>\n",
       "      <td>7.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SMPD1</td>\n",
       "      <td>c.107_118delTGCTGGCGCTGG</td>\n",
       "      <td></td>\n",
       "      <td>-1.58</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>EIF2AK3</td>\n",
       "      <td>c.61_63delCTG</td>\n",
       "      <td></td>\n",
       "      <td>-1.91</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>10.041667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>KIF1A</td>\n",
       "      <td>c.2751_2753delGGA</td>\n",
       "      <td></td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BP6</td>\n",
       "      <td>12.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>ANK1</td>\n",
       "      <td>c.1096G&gt;A</td>\n",
       "      <td>chr8:g.41577289C&gt;T</td>\n",
       "      <td>-2.61</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP1</td>\n",
       "      <td>5.375000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>KAT6A</td>\n",
       "      <td>c.5597C&gt;G</td>\n",
       "      <td>chr8:g.41790141G&gt;C</td>\n",
       "      <td>-2.81</td>\n",
       "      <td>-1.375000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP1</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>CDH23</td>\n",
       "      <td>c.2263C&gt;T</td>\n",
       "      <td>chr10:g.73453990C&gt;T</td>\n",
       "      <td>-2.91</td>\n",
       "      <td>-1.666667</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2|PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.316-125A&gt;G</td>\n",
       "      <td>chr11:g.5247081T&gt;C</td>\n",
       "      <td>-3.09</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|BS2</td>\n",
       "      <td>19.213095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PCLO</td>\n",
       "      <td>c.1501T&gt;C</td>\n",
       "      <td>chr7:g.82784456_82784486delinsGTGGGGGCTTTGCTGAGCCAGGCTGTTGAGG</td>\n",
       "      <td>-8.51</td>\n",
       "      <td>-4.000000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>PM1|BS1|BA1</td>\n",
       "      <td>5.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>HLA-DQB1</td>\n",
       "      <td>c.85C&gt;G</td>\n",
       "      <td>chr6:g.32634300G&gt;C</td>\n",
       "      <td>-8.57</td>\n",
       "      <td>-4.500000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>PM1|BS1|BA1|BP4</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ERAP1</td>\n",
       "      <td>c.380G&gt;C</td>\n",
       "      <td>chr5:g.96139250C&gt;G</td>\n",
       "      <td>-8.91</td>\n",
       "      <td>-4.875000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|PP3|BA1</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>chr2:g.128180746C&gt;T</td>\n",
       "      <td>-8.97</td>\n",
       "      <td>-3.500000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>BS2|BP4|BP7|BP6</td>\n",
       "      <td>10.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>chr11:g.62383715G&gt;C</td>\n",
       "      <td>-12.40</td>\n",
       "      <td>-5.500000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|BA1|BP4</td>\n",
       "      <td>6.804762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>chr17:g.3397702C&gt;T</td>\n",
       "      <td>-14.57</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>Benign</td>\n",
       "      <td>BS1|BA1|BP7|BP6</td>\n",
       "      <td>7.566667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene                   variant  \\\n",
       "0       BBS10                  c.145C>T   \n",
       "1         HBB                  c.208G>A   \n",
       "2        DLX3                  c.335A>G   \n",
       "3       PROP1                  c.652A>C   \n",
       "4        SOX3                  c.818C>T   \n",
       "5      CEP152                 c.3530A>G   \n",
       "6       PRKDC                 c.3278A>T   \n",
       "7      PCDH15                 c.4793G>A   \n",
       "8        PLEC                 c.1712G>A   \n",
       "9     FOXRED1                  c.286G>T   \n",
       "10     GLYCTK                 c.1313C>T   \n",
       "11     PIEZO1                 c.1387G>A   \n",
       "12     COL5A2                 c.1976C>T   \n",
       "13      MTPAP                 c.1533G>C   \n",
       "14      SCN5A                 c.1982G>A   \n",
       "15       PYGL                 c.1900G>C   \n",
       "16      CCBE1                 c.1174A>T   \n",
       "17    SLC13A5                 c.1163A>G   \n",
       "18      PRSS1                  c.502C>A   \n",
       "19     ZSWIM6                 c.1472A>G   \n",
       "20       PLEC                c.11402A>G   \n",
       "21       EGFR                 c.1973G>A   \n",
       "22       CTGF                  c.172C>G   \n",
       "23    COL13A1                  c.770G>A   \n",
       "24     DNAJB6                  c.365T>A   \n",
       "25      ACAD9                 c.1414C>T   \n",
       "26       TGM5                   c.23C>T   \n",
       "27       PCK2                 c.1354G>A   \n",
       "28        TTN                c.57332G>A   \n",
       "29     CEP135                 c.1528C>T   \n",
       "..        ...                       ...   \n",
       "82      AICDA        c.428-5_428-4insTT   \n",
       "83       SOS2       c.2786-6_2786-5insT   \n",
       "84       PTEN       c.211-18_211-17insT   \n",
       "85      DCTN1                 c.3746C>T   \n",
       "86        CAT               c.1057-5T>C   \n",
       "87      SYNE1             c.17203-6delT   \n",
       "88       PFKM               c.1340+6G>T   \n",
       "89       PON1                  c.380T>G   \n",
       "90        EYS      c.6079-4_6079-3delTC   \n",
       "91      ABCC6                 c.3207C>T   \n",
       "92       PMS2               c.706-4delT   \n",
       "93    SMARCA2           c.705_707delGCA   \n",
       "94        NEB             c.11077-5delT   \n",
       "95       TSC1              c.2626-4delT   \n",
       "96      ABCC9              c.1165-6delT   \n",
       "97       LRP5             c.58_60delCTG   \n",
       "98       WFS1                 c.1760G>A   \n",
       "99      SMPD1  c.107_118delTGCTGGCGCTGG   \n",
       "100   EIF2AK3             c.61_63delCTG   \n",
       "101     KIF1A         c.2751_2753delGGA   \n",
       "102      ANK1                 c.1096G>A   \n",
       "103     KAT6A                 c.5597C>G   \n",
       "104     CDH23                 c.2263C>T   \n",
       "105       HBB              c.316-125A>G   \n",
       "106      PCLO                 c.1501T>C   \n",
       "107  HLA-DQB1                   c.85C>G   \n",
       "108     ERAP1                  c.380G>C   \n",
       "109      PROC                  c.399C>T   \n",
       "110    B3GAT3                c.910-3C>G   \n",
       "111      ASPA                  c.693C>T   \n",
       "\n",
       "                                                                id  \\\n",
       "0                                              chr12:g.76741994G>A   \n",
       "1                                               chr11:g.5247914C>T   \n",
       "2                                              chr17:g.48070945T>C   \n",
       "3                                              chr5:g.177419739T>G   \n",
       "4                                              chrX:g.139586408G>A   \n",
       "5                                              chr15:g.49040744T>C   \n",
       "6                                               chr8:g.48813019T>A   \n",
       "7                                              chr10:g.55582714C>T   \n",
       "8                                              chr8:g.145007482C>T   \n",
       "9                                             chr11:g.126141532G>T   \n",
       "10                                              chr3:g.52326883C>T   \n",
       "11                                             chr16:g.88802726C>T   \n",
       "12                                             chr2:g.189927592G>A   \n",
       "13                                             chr10:g.30602754C>G   \n",
       "14                                              chr3:g.38640450C>T   \n",
       "15                                             chr14:g.51378517C>G   \n",
       "16                                             chr18:g.57103187T>A   \n",
       "17                                              chr17:g.6596475T>C   \n",
       "18                                             chr7:g.142460329C>A   \n",
       "19                                              chr5:g.60817228A>G   \n",
       "20                                             chr8:g.144992998T>C   \n",
       "21                                              chr7:g.55238092G>A   \n",
       "22                                             chr6:g.132272027G>C   \n",
       "23                                             chr10:g.71658508G>A   \n",
       "24                                             chr7:g.157174958T>A   \n",
       "25                                             chr3:g.128627871C>T   \n",
       "26                                             chr15:g.43552765G>A   \n",
       "27                                             chr14:g.24573006G>A   \n",
       "28                                             chr2:g.179462477C>T   \n",
       "29                                              chr4:g.56846363C>T   \n",
       "..                                                             ...   \n",
       "82                                                                   \n",
       "83                                                                   \n",
       "84                                                                   \n",
       "85                                              chr2:g.74588717G>A   \n",
       "86                                             chr11:g.34482793T>C   \n",
       "87                                                                   \n",
       "88                                             chr12:g.48533138G>T   \n",
       "89                                              chr7:g.94940880A>C   \n",
       "90                                                                   \n",
       "91                                             chr16:g.16259579G>A   \n",
       "92                                                                   \n",
       "93                                                                   \n",
       "94                                                                   \n",
       "95                                                                   \n",
       "96                                                                   \n",
       "97                                                                   \n",
       "98                                               chr4:g.6303282G>A   \n",
       "99                                                                   \n",
       "100                                                                  \n",
       "101                                                                  \n",
       "102                                             chr8:g.41577289C>T   \n",
       "103                                             chr8:g.41790141G>C   \n",
       "104                                            chr10:g.73453990C>T   \n",
       "105                                             chr11:g.5247081T>C   \n",
       "106  chr7:g.82784456_82784486delinsGTGGGGGCTTTGCTGAGCCAGGCTGTTGAGG   \n",
       "107                                             chr6:g.32634300G>C   \n",
       "108                                             chr5:g.96139250C>G   \n",
       "109                                            chr2:g.128180746C>T   \n",
       "110                                            chr11:g.62383715G>C   \n",
       "111                                             chr17:g.3397702C>T   \n",
       "\n",
       "     final_score  pathogenicity_score           pathogenicity  \\\n",
       "0           8.24             3.875000       Likely pathogenic   \n",
       "1           8.11             2.625000  Uncertain significance   \n",
       "2           6.15             3.961200  Uncertain significance   \n",
       "3           5.23             2.250000  Uncertain significance   \n",
       "4           5.19             2.250000  Uncertain significance   \n",
       "5           4.97             2.250000  Uncertain significance   \n",
       "6           4.45             2.250000  Uncertain significance   \n",
       "7           3.99             2.250000  Uncertain significance   \n",
       "8           3.92             1.750000  Uncertain significance   \n",
       "9           3.47             1.500000  Uncertain significance   \n",
       "10          3.34             1.625000  Uncertain significance   \n",
       "11          3.21             2.750000  Uncertain significance   \n",
       "12          3.10             1.500000  Uncertain significance   \n",
       "13          2.95             2.250000  Uncertain significance   \n",
       "14          2.92             2.500000  Uncertain significance   \n",
       "15          2.88             2.000000  Uncertain significance   \n",
       "16          2.82             1.500000  Uncertain significance   \n",
       "17          2.63             2.250000  Uncertain significance   \n",
       "18          2.52             1.750000  Uncertain significance   \n",
       "19          2.38             1.250000  Uncertain significance   \n",
       "20          2.24             1.000000           Likely benign   \n",
       "21          2.16             1.500000  Uncertain significance   \n",
       "22          2.10             1.250000  Uncertain significance   \n",
       "23          1.98             1.000000  Uncertain significance   \n",
       "24          1.90             1.625000  Uncertain significance   \n",
       "25          1.64             1.250000  Uncertain significance   \n",
       "26          1.49             1.250000  Uncertain significance   \n",
       "27          1.46             1.250000  Uncertain significance   \n",
       "28          1.31             0.750000  Uncertain significance   \n",
       "29          1.22             1.000000           Likely benign   \n",
       "..           ...                  ...                     ...   \n",
       "82          0.00             0.000000  Uncertain significance   \n",
       "83          0.00             0.000000  Uncertain significance   \n",
       "84          0.00             0.000000  Uncertain significance   \n",
       "85         -0.47            -0.250000  Uncertain significance   \n",
       "86         -0.58            -0.500000  Uncertain significance   \n",
       "87         -0.58            -0.500000  Uncertain significance   \n",
       "88         -0.63            -0.500000  Uncertain significance   \n",
       "89         -0.81            -0.696000  Uncertain significance   \n",
       "90         -0.84            -0.500000  Uncertain significance   \n",
       "91         -0.99            -0.500000  Uncertain significance   \n",
       "92         -1.01            -0.500000  Uncertain significance   \n",
       "93         -1.06            -0.500000  Uncertain significance   \n",
       "94         -1.12            -0.500000  Uncertain significance   \n",
       "95         -1.18            -0.500000  Uncertain significance   \n",
       "96         -1.25            -0.500000  Uncertain significance   \n",
       "97         -1.36            -0.500000  Uncertain significance   \n",
       "98         -1.44            -0.625000  Uncertain significance   \n",
       "99         -1.58            -0.750000  Uncertain significance   \n",
       "100        -1.91            -0.750000  Uncertain significance   \n",
       "101        -2.03            -0.750000  Uncertain significance   \n",
       "102        -2.61            -1.250000           Likely benign   \n",
       "103        -2.81            -1.375000           Likely benign   \n",
       "104        -2.91            -1.666667  Uncertain significance   \n",
       "105        -3.09            -1.000000  Uncertain significance   \n",
       "106        -8.51            -4.000000                  Benign   \n",
       "107        -8.57            -4.500000                  Benign   \n",
       "108        -8.91            -4.875000                  Benign   \n",
       "109        -8.97            -3.500000           Likely benign   \n",
       "110       -12.40            -5.500000                  Benign   \n",
       "111       -14.57            -6.250000                  Benign   \n",
       "\n",
       "            hit_criteria  hpo_hit_score  \n",
       "0        PS1|PM1|PP3|PP5       5.666667  \n",
       "1        PM5|PM1|PP3|PP5      19.213095  \n",
       "2    PS1|PM1|PM2|PP3|BP1       2.000000  \n",
       "3            PM1|PM2|PP3       7.500000  \n",
       "4            PM1|PM2|PP3       7.333333  \n",
       "5            PM1|PM2|PP3       6.404762  \n",
       "6            PM1|PM2|PP3       4.500000  \n",
       "7            PM1|PM2|PP3       3.166667  \n",
       "8        PM1|PM2|PP3|BP1       6.666667  \n",
       "9                PM1|PP3       7.400000  \n",
       "10           PM1|PP3|PP2       5.100000  \n",
       "11       PM1|PM2|PP3|PP2       0.500000  \n",
       "12               PM1|PP3       5.166667  \n",
       "13           PM1|PM2|PP3       1.000000  \n",
       "14           PM1|PM2|PP3       0.500000  \n",
       "15           PM1|PP3|PP5       1.500000  \n",
       "16               PM2|PP3       3.833333  \n",
       "17           PM1|PM2|PP3       0.500000  \n",
       "18           PM1|PP3|PP2       1.500000  \n",
       "19               PM1|PP3       4.000000  \n",
       "20       PM1|PM2|BP1|BP4       6.666667  \n",
       "21           PM1|PM2|BP4       1.500000  \n",
       "22               PM1|PP3       2.666667  \n",
       "23           PM1|PP3|BP1       4.500000  \n",
       "24           PM1|PP3|PP2       0.500000  \n",
       "25               PM1|PP3       1.000000  \n",
       "26               PM2|PP3       0.571429  \n",
       "27               PM1|PP3       0.500000  \n",
       "28           PM2|PP3|BP1       3.000000  \n",
       "29       PM1|PM2|BP1|BP4       0.666667  \n",
       "..                   ...            ...  \n",
       "82                             1.500000  \n",
       "83                             7.000000  \n",
       "84                            17.071429  \n",
       "85       PM1|BS2|PP3|PP2       3.833333  \n",
       "86                   BP4       0.500000  \n",
       "87                   BP6       0.500000  \n",
       "88                   BP4       0.833333  \n",
       "89       PS1|PM1|BS2|PP3       0.500000  \n",
       "90                   BP6       2.666667  \n",
       "91                   BP7       4.571429  \n",
       "92                   BP6       4.833333  \n",
       "93                   BP6       5.666667  \n",
       "94                   BP6       6.666667  \n",
       "95                   BP6       7.833333  \n",
       "96                   BP6       9.571429  \n",
       "97                   BP6      12.500000  \n",
       "98   PM1|PM2|BS2|PP3|BP6       7.333333  \n",
       "99                   BP6       5.500000  \n",
       "100                  BP6      10.041667  \n",
       "101                  BP6      12.166667  \n",
       "102      PM1|BS2|PP3|BP1       5.375000  \n",
       "103      PM1|BS2|PP3|BP1       5.000000  \n",
       "104              BS2|PP3       3.000000  \n",
       "105              PM2|BS2      19.213095  \n",
       "106          PM1|BS1|BA1       5.666667  \n",
       "107      PM1|BS1|BA1|BP4       4.000000  \n",
       "108          BS1|PP3|BA1       3.500000  \n",
       "109      BS2|BP4|BP7|BP6      10.266667  \n",
       "110          BS1|BA1|BP4       6.804762  \n",
       "111      BS1|BA1|BP7|BP6       7.566667  \n",
       "\n",
       "[112 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ACMG_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_list = list(ACMG_result['gene'])\n",
    "variant_list = list(ACMG_result['variant'])\n",
    "gene_variant_list = [(gene_list[i], variant_list[i]) for i in range(len(gene_list))]\n",
    "\n",
    "df_variant_ACMG_interpret = pd.DataFrame()\n",
    "df_variant_ACMG_interpret_chinese = pd.DataFrame() \n",
    "for key in gene_variant_list:\n",
    "    tmp_df = pd.DataFrame(variant_ACMG_interpretation[key], columns = ['criteria', 'interpretation'])\n",
    "    tmp_df['gene'] = key[0]\n",
    "    tmp_df['variant'] = key[1]\n",
    "    tmp_df = tmp_df[['gene', 'variant', 'criteria', 'interpretation']] \n",
    "    df_variant_ACMG_interpret = pd.concat([df_variant_ACMG_interpret, tmp_df])\n",
    "\n",
    "    tmp_df = pd.DataFrame(variant_ACMG_interpret_chinese[key], columns = ['criteria', 'interpretation'])\n",
    "    tmp_df['gene'] = key[0]\n",
    "    tmp_df['variant'] = key[1]\n",
    "    tmp_df = tmp_df[['gene', 'variant', 'criteria', 'interpretation']] \n",
    "    df_variant_ACMG_interpret_chinese = pd.concat([df_variant_ACMG_interpret_chinese, tmp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)极低(&lt; 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: Globin|Globin-like|Globin/Protoglobin.&lt;br/&gt;HGVS ID: chr11:g.5247914C&gt;T.&lt;br/&gt;RefSeq ID: rs33947415.&lt;br/&gt;外显子: 2.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.0008402.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.&lt;br/&gt;DANN致病性分数: 0.644724000156.&lt;br/&gt;FATHMM致病性分数: 0.93934.&lt;br/&gt;MetaSVM致病性分数: 0.65183.&lt;br/&gt;GERP++序列保守性预测分数: -10.2.&lt;br/&gt;Clinvar数据库ID: 15138.&lt;br/&gt;Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;HGVS ID: chr17:g.48070945T&gt;C.&lt;br/&gt;外显子: 2.&lt;br/&gt;DANN致病性分数: 0.993674333469.&lt;br/&gt;FATHMM致病性分数: 0.91543.&lt;br/&gt;MetaSVM致病性分数: 0.91694.&lt;br/&gt;GERP++序列保守性预测分数: 3.59.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: splice_region_variant&amp;intron_variant.&lt;br/&gt;HGVS ID: chr11:g.62383715G&gt;C.&lt;br/&gt;RefSeq ID: rs12794886.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.72.&lt;br/&gt;基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.&lt;br/&gt;基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: synonymous_variant.&lt;br/&gt;HGVS ID: chr17:g.3397702C&gt;T.&lt;br/&gt;RefSeq ID: rs12948217.&lt;br/&gt;外显子: 5.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.265550752376.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.21.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.&lt;br/&gt;Clinvar数据库ID: 197621.&lt;br/&gt;Clinvar数据库记录的变异致病性: Benign|Likely benign.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1278 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     variant criteria  \\\n",
       "0    BBS10    c.145C>T     变异注释   \n",
       "1    BBS10    c.145C>T     PVS1   \n",
       "2    BBS10    c.145C>T  PS1和PM5   \n",
       "3    BBS10    c.145C>T  PS3和BS3   \n",
       "4    BBS10    c.145C>T      PS4   \n",
       "5    BBS10    c.145C>T      PM2   \n",
       "6    BBS10    c.145C>T  BA1和BS1   \n",
       "7    BBS10    c.145C>T      BS2   \n",
       "8    BBS10    c.145C>T      PM1   \n",
       "9    BBS10    c.145C>T  PM4和BP3   \n",
       "10   BBS10    c.145C>T  PP2和BP1   \n",
       "11   BBS10    c.145C>T  PP3和BP4   \n",
       "12   BBS10    c.145C>T  PP5和BP6   \n",
       "13   BBS10    c.145C>T      BP7   \n",
       "0      HBB    c.208G>A     变异注释   \n",
       "1      HBB    c.208G>A     PVS1   \n",
       "2      HBB    c.208G>A  PS1和PM5   \n",
       "3      HBB    c.208G>A  PS3和BS3   \n",
       "4      HBB    c.208G>A      PS4   \n",
       "5      HBB    c.208G>A      PM2   \n",
       "6      HBB    c.208G>A  BA1和BS1   \n",
       "7      HBB    c.208G>A      BS2   \n",
       "8      HBB    c.208G>A      PM1   \n",
       "9      HBB    c.208G>A  PM4和BP3   \n",
       "10     HBB    c.208G>A  PP2和BP1   \n",
       "11     HBB    c.208G>A  PP3和BP4   \n",
       "12     HBB    c.208G>A  PP5和BP6   \n",
       "13     HBB    c.208G>A      BP7   \n",
       "0     DLX3    c.335A>G     变异注释   \n",
       "1     DLX3    c.335A>G     PVS1   \n",
       "..     ...         ...      ...   \n",
       "12    PROC    c.399C>T  PP5和BP6   \n",
       "13    PROC    c.399C>T      BP7   \n",
       "0   B3GAT3  c.910-3C>G     变异注释   \n",
       "1   B3GAT3  c.910-3C>G     PVS1   \n",
       "2   B3GAT3  c.910-3C>G  PS1和PM5   \n",
       "3   B3GAT3  c.910-3C>G  PS3和BS3   \n",
       "4   B3GAT3  c.910-3C>G      PS4   \n",
       "5   B3GAT3  c.910-3C>G      PM2   \n",
       "6   B3GAT3  c.910-3C>G  BA1和BS1   \n",
       "7   B3GAT3  c.910-3C>G      BS2   \n",
       "8   B3GAT3  c.910-3C>G      PM1   \n",
       "9   B3GAT3  c.910-3C>G  PM4和BP3   \n",
       "10  B3GAT3  c.910-3C>G  PP2和BP1   \n",
       "11  B3GAT3  c.910-3C>G  PP3和BP4   \n",
       "12  B3GAT3  c.910-3C>G  PP5和BP6   \n",
       "13  B3GAT3  c.910-3C>G      BP7   \n",
       "0     ASPA    c.693C>T     变异注释   \n",
       "1     ASPA    c.693C>T     PVS1   \n",
       "2     ASPA    c.693C>T  PS1和PM5   \n",
       "3     ASPA    c.693C>T  PS3和BS3   \n",
       "4     ASPA    c.693C>T      PS4   \n",
       "5     ASPA    c.693C>T      PM2   \n",
       "6     ASPA    c.693C>T  BA1和BS1   \n",
       "7     ASPA    c.693C>T      BS2   \n",
       "8     ASPA    c.693C>T      PM1   \n",
       "9     ASPA    c.693C>T  PM4和BP3   \n",
       "10    ASPA    c.693C>T  PP2和BP1   \n",
       "11    ASPA    c.693C>T  PP3和BP4   \n",
       "12    ASPA    c.693C>T  PP5和BP6   \n",
       "13    ASPA    c.693C>T      BP7   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            interpretation  \n",
       "0                  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.<br/>Clinvar数据库记录审核状态: criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                         变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异的最小等位基因频率(MAF)极低(< 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0   突变类型: missense_variant.<br/>蛋白功能区: Globin|Globin-like|Globin/Protoglobin.<br/>HGVS ID: chr11:g.5247914C>T.<br/>RefSeq ID: rs33947415.<br/>外显子: 2.<br/>ExAC 最小等位基因频率(MAF): 0.0008402.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.<br/>DANN致病性分数: 0.644724000156.<br/>FATHMM致病性分数: 0.93934.<br/>MetaSVM致病性分数: 0.65183.<br/>GERP++序列保守性预测分数: -10.2.<br/>Clinvar数据库ID: 15138.<br/>Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.<br/>Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                             变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                     突变类型: missense_variant.<br/>HGVS ID: chr17:g.48070945T>C.<br/>外显子: 2.<br/>DANN致病性分数: 0.993674333469.<br/>FATHMM致病性分数: 0.91543.<br/>MetaSVM致病性分数: 0.91694.<br/>GERP++序列保守性预测分数: 3.59.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ...  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                突变类型: splice_region_variant&intron_variant.<br/>HGVS ID: chr11:g.62383715G>C.<br/>RefSeq ID: rs12794886.<br/>1000Genomes 最小等位基因频率(MAF): 0.72.<br/>基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.<br/>基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                        突变类型: synonymous_variant.<br/>HGVS ID: chr17:g.3397702C>T.<br/>RefSeq ID: rs12948217.<br/>外显子: 5.<br/>ExAC 最小等位基因频率(MAF): 0.265550752376.<br/>1000Genomes 最小等位基因频率(MAF): 0.21.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.<br/>Clinvar数据库ID: 197621.<br/>Clinvar数据库记录的变异致病性: Benign|Likely benign.<br/>Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "\n",
       "[1278 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)极低(&lt; 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: Globin|Globin-like|Globin/Protoglobin.&lt;br/&gt;HGVS ID: chr11:g.5247914C&gt;T.&lt;br/&gt;RefSeq ID: rs33947415.&lt;br/&gt;外显子: 2.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.0008402.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.&lt;br/&gt;DANN致病性分数: 0.644724000156.&lt;br/&gt;FATHMM致病性分数: 0.93934.&lt;br/&gt;MetaSVM致病性分数: 0.65183.&lt;br/&gt;GERP++序列保守性预测分数: -10.2.&lt;br/&gt;Clinvar数据库ID: 15138.&lt;br/&gt;Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.&lt;br/&gt;Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &lt;= 1%. 不符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;HGVS ID: chr17:g.48070945T&gt;C.&lt;br/&gt;外显子: 2.&lt;br/&gt;DANN致病性分数: 0.993674333469.&lt;br/&gt;FATHMM致病性分数: 0.91543.&lt;br/&gt;MetaSVM致病性分数: 0.91694.&lt;br/&gt;GERP++序列保守性预测分数: 3.59.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PROC</td>\n",
       "      <td>c.399C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: splice_region_variant&amp;intron_variant.&lt;br/&gt;HGVS ID: chr11:g.62383715G&gt;C.&lt;br/&gt;RefSeq ID: rs12794886.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.72.&lt;br/&gt;基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.&lt;br/&gt;基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>B3GAT3</td>\n",
       "      <td>c.910-3C&gt;G</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型非同义（沉默）变异. 不符合BP7标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: synonymous_variant.&lt;br/&gt;HGVS ID: chr17:g.3397702C&gt;T.&lt;br/&gt;RefSeq ID: rs12948217.&lt;br/&gt;外显子: 5.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 0.265550752376.&lt;br/&gt;1000Genomes 最小等位基因频率(MAF): 0.21.&lt;br/&gt;Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.&lt;br/&gt;Clinvar数据库ID: 197621.&lt;br/&gt;Clinvar数据库记录的变异致病性: Benign|Likely benign.&lt;br/&gt;Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>[未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM2</td>\n",
       "      <td>此基因变异的最小等位基因频率(MAF)较高(&gt;= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BA1和BS1</td>\n",
       "      <td>等位基因频率 &gt; 5%. 等位基因频率 &gt; 1%. 符合BA1标准. 符合BS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BS2</td>\n",
       "      <td>此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM1</td>\n",
       "      <td>此基因变异不是错义突变.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PM4和BP3</td>\n",
       "      <td>变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP2和BP1</td>\n",
       "      <td>此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP3和BP4</td>\n",
       "      <td>各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>PP5和BP6</td>\n",
       "      <td>Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ASPA</td>\n",
       "      <td>c.693C&gt;T</td>\n",
       "      <td>BP7</td>\n",
       "      <td>变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1278 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene     variant criteria  \\\n",
       "0    BBS10    c.145C>T     变异注释   \n",
       "1    BBS10    c.145C>T     PVS1   \n",
       "2    BBS10    c.145C>T  PS1和PM5   \n",
       "3    BBS10    c.145C>T  PS3和BS3   \n",
       "4    BBS10    c.145C>T      PS4   \n",
       "5    BBS10    c.145C>T      PM2   \n",
       "6    BBS10    c.145C>T  BA1和BS1   \n",
       "7    BBS10    c.145C>T      BS2   \n",
       "8    BBS10    c.145C>T      PM1   \n",
       "9    BBS10    c.145C>T  PM4和BP3   \n",
       "10   BBS10    c.145C>T  PP2和BP1   \n",
       "11   BBS10    c.145C>T  PP3和BP4   \n",
       "12   BBS10    c.145C>T  PP5和BP6   \n",
       "13   BBS10    c.145C>T      BP7   \n",
       "0      HBB    c.208G>A     变异注释   \n",
       "1      HBB    c.208G>A     PVS1   \n",
       "2      HBB    c.208G>A  PS1和PM5   \n",
       "3      HBB    c.208G>A  PS3和BS3   \n",
       "4      HBB    c.208G>A      PS4   \n",
       "5      HBB    c.208G>A      PM2   \n",
       "6      HBB    c.208G>A  BA1和BS1   \n",
       "7      HBB    c.208G>A      BS2   \n",
       "8      HBB    c.208G>A      PM1   \n",
       "9      HBB    c.208G>A  PM4和BP3   \n",
       "10     HBB    c.208G>A  PP2和BP1   \n",
       "11     HBB    c.208G>A  PP3和BP4   \n",
       "12     HBB    c.208G>A  PP5和BP6   \n",
       "13     HBB    c.208G>A      BP7   \n",
       "0     DLX3    c.335A>G     变异注释   \n",
       "1     DLX3    c.335A>G     PVS1   \n",
       "..     ...         ...      ...   \n",
       "12    PROC    c.399C>T  PP5和BP6   \n",
       "13    PROC    c.399C>T      BP7   \n",
       "0   B3GAT3  c.910-3C>G     变异注释   \n",
       "1   B3GAT3  c.910-3C>G     PVS1   \n",
       "2   B3GAT3  c.910-3C>G  PS1和PM5   \n",
       "3   B3GAT3  c.910-3C>G  PS3和BS3   \n",
       "4   B3GAT3  c.910-3C>G      PS4   \n",
       "5   B3GAT3  c.910-3C>G      PM2   \n",
       "6   B3GAT3  c.910-3C>G  BA1和BS1   \n",
       "7   B3GAT3  c.910-3C>G      BS2   \n",
       "8   B3GAT3  c.910-3C>G      PM1   \n",
       "9   B3GAT3  c.910-3C>G  PM4和BP3   \n",
       "10  B3GAT3  c.910-3C>G  PP2和BP1   \n",
       "11  B3GAT3  c.910-3C>G  PP3和BP4   \n",
       "12  B3GAT3  c.910-3C>G  PP5和BP6   \n",
       "13  B3GAT3  c.910-3C>G      BP7   \n",
       "0     ASPA    c.693C>T     变异注释   \n",
       "1     ASPA    c.693C>T     PVS1   \n",
       "2     ASPA    c.693C>T  PS1和PM5   \n",
       "3     ASPA    c.693C>T  PS3和BS3   \n",
       "4     ASPA    c.693C>T      PS4   \n",
       "5     ASPA    c.693C>T      PM2   \n",
       "6     ASPA    c.693C>T  BA1和BS1   \n",
       "7     ASPA    c.693C>T      BS2   \n",
       "8     ASPA    c.693C>T      PM1   \n",
       "9     ASPA    c.693C>T  PM4和BP3   \n",
       "10    ASPA    c.693C>T  PP2和BP1   \n",
       "11    ASPA    c.693C>T  PP3和BP4   \n",
       "12    ASPA    c.693C>T  PP5和BP6   \n",
       "13    ASPA    c.693C>T      BP7   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            interpretation  \n",
       "0                  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: Pathogenic|Pathogenic.<br/>Clinvar数据库记录审核状态: criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                         变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异的最小等位基因频率(MAF)极低(< 0.5%)，但并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异致病(Clinvar IDs: 225010). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0   突变类型: missense_variant.<br/>蛋白功能区: Globin|Globin-like|Globin/Protoglobin.<br/>HGVS ID: chr11:g.5247914C>T.<br/>RefSeq ID: rs33947415.<br/>外显子: 2.<br/>ExAC 最小等位基因频率(MAF): 0.0008402.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.001.<br/>DANN致病性分数: 0.644724000156.<br/>FATHMM致病性分数: 0.93934.<br/>MetaSVM致病性分数: 0.65183.<br/>GERP++序列保守性预测分数: -10.2.<br/>Clinvar数据库ID: 15138.<br/>Clinvar数据库记录的变异致病性: other|Uncertain significance|Likely pathogenic.<br/>Clinvar数据库记录审核状态: no assertion criteria provided|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25113778', '2467892'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                             变异为错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Gly70Asp)，尽管是不同氨基酸改变 (Clinvar数据库参考: 15226). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  等位基因频率 <= 1%. 不符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   此基因变异位于突变热点和/或关键的功能域(例如酶的活性部位)，且在这些区域不存在良性变异. 符合PM1标准.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        变异位点在非重复区域(nonrepeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              此基因变异类型是错义突变，但并非发生在以下两种类型的基因上：(1) 良性错义突变在基因上的频率低，且在这个基因上的错义突变是致病的常见原因；(2) 基因的截短(truncating)变异是致病的主要原因. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 部分符合PP3标准(部分致病性算法预测此变异致病). 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Clinvar报道此基因变异致病(Clinvar IDs: 15138). 符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                     突变类型: missense_variant.<br/>HGVS ID: chr17:g.48070945T>C.<br/>外显子: 2.<br/>DANN致病性分数: 0.993674333469.<br/>FATHMM致病性分数: 0.91543.<br/>MetaSVM致病性分数: 0.91694.<br/>GERP++序列保守性预测分数: 3.59.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ...  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Clinvar报道此基因变异良性(Clinvar IDs: 331103). 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                突变类型: splice_region_variant&intron_variant.<br/>HGVS ID: chr11:g.62383715G>C.<br/>RefSeq ID: rs12794886.<br/>1000Genomes 最小等位基因频率(MAF): 0.72.<br/>基于随机森林算法的dbscSNV剪接效应预测分数: 0.004.<br/>基于AdaBoost算法的dbscSNV剪接效应预测分数: 1.42005417025298E-5.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)不是已知的致病机制. 此变异具有有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异非错义突变. 未发现之前报道导致相同氨基酸改变的致病基因变异. 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              未在数据库中发现关于此基因变异致病性的报道. 不符合PP5标准. 不符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                变异类型非同义（沉默）变异. 不符合BP7标准.  \n",
       "0                                                                                                                                        突变类型: synonymous_variant.<br/>HGVS ID: chr17:g.3397702C>T.<br/>RefSeq ID: rs12948217.<br/>外显子: 5.<br/>ExAC 最小等位基因频率(MAF): 0.265550752376.<br/>1000Genomes 最小等位基因频率(MAF): 0.21.<br/>Exome Sequencing Project(ESP) 6500 最小等位基因频率(MAF): 0.289.<br/>Clinvar数据库ID: 197621.<br/>Clinvar数据库记录的变异致病性: Benign|Likely benign.<br/>Clinvar数据库记录审核状态: criteria provided, multiple submitters, no conflicts|criteria provided, single submitter.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['8023850'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异非错义突变. 此错义突变发生在的氨基酸残基在之前发现过致病错义突变(p.Tyr231Cys)，尽管是不同氨基酸改变 (Clinvar数据库参考: 2612). 未发现之前报道导致相同氨基酸改变的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 不符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        [未发现针对此基因变异的完善的体内或体外功能性研究., 不符合PS3标准., 不符合BS3标准.]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     此基因变异的最小等位基因频率(MAF)较高(>= 0.5%)，且并不引发隐性遗传病. 不符合PM2标准.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              等位基因频率 > 5%. 等位基因频率 > 1%. 符合BA1标准. 符合BS1标准.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            此基因变异在健康人中既不以隐性(纯合子)也不以显性(杂合子)状态存在. 不符合BS2标准.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             此基因变异不是错义突变.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            变异位点在重复区域(repeat region). 不符合PM4标准. 不符合BP3标准.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      此基因变异类型不是错义突变. 不符合PP2标准. 不符合BP1标准.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 各致病性算法的致病分数阈值是：DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. 不符合PP3标准. 不符合BP4标准.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Clinvar报道此基因变异良性(Clinvar IDs: 197621). 生物医学参考文献Pubmed: 8023850. 不符合PP5标准. 符合BP6标准.  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 变异类型为同义（沉默）变异，剪接预测算法预测此变异对剪接序列没有影响，核苷酸不是高度保守的. 符合BP7标准.  \n",
       "\n",
       "[1278 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('variant_annotations',\n",
       "  u\"Effect: missense_variant.<br/>Protein domain: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>exon: 1.<br/>ExAC MAF: 5.052e-05.<br/>DANN pathogenicity score: 0.999117068756.<br/>FATHMM pathogenicity score: 0.79186.<br/>MetaSVM pathogenicity score: 0.84564.<br/>GERP++ conservation score: 4.27.<br/>Clinvar variation ids: 225010.<br/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br/>Clinvar review status: criteria provided, single submitter.<br/>Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].\"),\n",
       " ('PVS1',\n",
       "  'Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.'),\n",
       " ('PS1 and PM5',\n",
       "  'Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.'),\n",
       " ('PS3 and BS3',\n",
       "  'Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.'),\n",
       " ('PS4',\n",
       "  'Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.'),\n",
       " ('PM2',\n",
       "  'The variant has an extremely low frequency with MAF < 0.5%, but NOT causes recessive diseases. PM2 is NOT met.'),\n",
       " ('BA1 and BS1',\n",
       "  'Allele frequeny is <= 1%. BA1 is NOT met. \\xe4\\xb8\\x8d\\xe7\\xac\\xa6\\xe5\\x90\\x88BA1\\xe6\\xa0\\x87\\xe5\\x87\\x86. BS1 is NOT met.'),\n",
       " ('BS2',\n",
       "  'The variant is NOT observed in a healthy adult in the 1000 Genomes Project as a heterozygote (for dominant diseases) nor as a homozygote (for recessive diseases). BS2 is NOT met.'),\n",
       " ('PM1',\n",
       "  'The variant has missense effect in a critical protein domain where all missense variants in the domain identified to date have been shown to be pathogenic. PM1 is met.'),\n",
       " ('PM4 and BP3',\n",
       "  'Allele is in a repeat region. PM4 is NOT met. BP3 is NOT met.'),\n",
       " ('PP2 and BP1',\n",
       "  'The variant effect is missense, but neither in a gene that has a low rate of benign missense variation and in which missense variants are a common mechanism of disease nor in a gene for which primarily truncating variants are known to cause disease. PP2 is NOT met. BP1 is NOT met.'),\n",
       " ('PP3 and BP4',\n",
       "  'The cutoffs for computational pathogenicity predictions are DANN:0.96, FATHMM:0.81415, MetaSVM:0.83357, GERP++:2.0, dbscSNV_rf_score:0.6, dbscSNV_ada_score:0.6. PP3 is partially met. BP4 is NOT met.'),\n",
       " ('PP5 and BP6',\n",
       "  'Clinvar reports the variant as pathogenic (Clinvar: 225010). Pubmed references: 25356970, 20120035, 16582908, 21044901, 20876674, 21642631, 20498079, 21517826, 20177705, 24746959, 25982971. PP5 is met. BP6 is NOT met.'),\n",
       " ('BP7', 'Variant effect is not synonymous. BP7 is NOT met.')]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpretation[('BBS10', 'c.145C>T')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>protein</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "      <th>pheno_match_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>p.Arg49Trp</td>\n",
       "      <td>chr12:g.76741994G&gt;A</td>\n",
       "      <td>9.511010</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM1|PP3|PP5</td>\n",
       "      <td>3.8750</td>\n",
       "      <td>5.666667</td>\n",
       "      <td>1.154249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>HBB</td>\n",
       "      <td>c.208G&gt;A</td>\n",
       "      <td>p.Gly70Ser</td>\n",
       "      <td>chr11:g.5247914C&gt;T</td>\n",
       "      <td>8.713391</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM5|PM1|PP3|PP5</td>\n",
       "      <td>2.6250</td>\n",
       "      <td>19.213095</td>\n",
       "      <td>1.074401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>DLX3</td>\n",
       "      <td>c.335A&gt;G</td>\n",
       "      <td>p.Lys112Arg</td>\n",
       "      <td>chr17:g.48070945T&gt;C</td>\n",
       "      <td>6.150000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PS1|PM1|PM2|PP3|BP1</td>\n",
       "      <td>3.9612</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>c.652A&gt;C</td>\n",
       "      <td>p.Ser218Arg</td>\n",
       "      <td>chr5:g.177419739T&gt;G</td>\n",
       "      <td>5.230000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>SOX3</td>\n",
       "      <td>c.818C&gt;T</td>\n",
       "      <td>p.Ser273Leu</td>\n",
       "      <td>chrX:g.139586408G&gt;A</td>\n",
       "      <td>5.190000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.2500</td>\n",
       "      <td>7.333333</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene   variant      protein                   id  final_score  \\\n",
       "76   BBS10  c.145C>T   p.Arg49Trp  chr12:g.76741994G>A     9.511010   \n",
       "103    HBB  c.208G>A   p.Gly70Ser   chr11:g.5247914C>T     8.713391   \n",
       "66    DLX3  c.335A>G  p.Lys112Arg  chr17:g.48070945T>C     6.150000   \n",
       "53   PROP1  c.652A>C  p.Ser218Arg  chr5:g.177419739T>G     5.230000   \n",
       "107   SOX3  c.818C>T  p.Ser273Leu  chrX:g.139586408G>A     5.190000   \n",
       "\n",
       "              pathogenicity         hit_criteria  pathogenicity_score  \\\n",
       "76        Likely pathogenic      PS1|PM1|PP3|PP5               3.8750   \n",
       "103  Uncertain significance      PM5|PM1|PP3|PP5               2.6250   \n",
       "66   Uncertain significance  PS1|PM1|PM2|PP3|BP1               3.9612   \n",
       "53   Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "107  Uncertain significance          PM1|PM2|PP3               2.2500   \n",
       "\n",
       "     hpo_hit_score  pheno_match_score  \n",
       "76        5.666667           1.154249  \n",
       "103      19.213095           1.074401  \n",
       "66        2.000000           1.000000  \n",
       "53        7.500000           1.000000  \n",
       "107       7.333333           1.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Allele 1</th>\n",
       "      <th>Allele 2</th>\n",
       "      <th>Allele In Scope</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene Profile Report</th>\n",
       "      <th>Gene (gene)</th>\n",
       "      <th>cDNA (cNomen)</th>\n",
       "      <th>Protein (pNomen)</th>\n",
       "      <th>HGVS cDNA-level nomenclature (fullCNomen)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1635536</td>\n",
       "      <td>1635536</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_024011.2</td>\n",
       "      <td></td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>c.1735dupT</td>\n",
       "      <td>p.Y579Lfs*66</td>\n",
       "      <td>NM_024011.2:c.1735dupT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>23713843</td>\n",
       "      <td>23713843</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_003196.2</td>\n",
       "      <td></td>\n",
       "      <td>TCEA3</td>\n",
       "      <td>c.889A&gt;G</td>\n",
       "      <td>p.M297V</td>\n",
       "      <td>NM_003196.2:c.889A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>24882663</td>\n",
       "      <td>24882663</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NM_001010980.4</td>\n",
       "      <td></td>\n",
       "      <td>NCMAP</td>\n",
       "      <td>c.-8+3A&gt;G</td>\n",
       "      <td></td>\n",
       "      <td>NM_001010980.4:c.-8+3A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>26663842</td>\n",
       "      <td>26663842</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>NM_001039775.3</td>\n",
       "      <td></td>\n",
       "      <td>AIM1L</td>\n",
       "      <td>c.3673C&gt;T</td>\n",
       "      <td>p.R1225W</td>\n",
       "      <td>NM_001039775.3:c.3673C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>39896387</td>\n",
       "      <td>39896387</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NM_012090.5</td>\n",
       "      <td></td>\n",
       "      <td>MACF1</td>\n",
       "      <td>c.10958A&gt;C</td>\n",
       "      <td>p.E3653A</td>\n",
       "      <td>NM_012090.5:c.10958A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start      Stop Reference Allele 1 Allele 2 Allele In Scope  \\\n",
       "0          1   1635536   1635536         .        .        A               A   \n",
       "1          1  23713843  23713843         T        C        T               C   \n",
       "2          1  24882663  24882663         A        A        G               G   \n",
       "3          1  26663842  26663842         G        G        A               A   \n",
       "4          1  39896387  39896387         A        C        A               C   \n",
       "\n",
       "       Transcript Gene Profile Report Gene (gene) cDNA (cNomen)  \\\n",
       "0     NM_024011.2                          CDK11A    c.1735dupT   \n",
       "1     NM_003196.2                           TCEA3      c.889A>G   \n",
       "2  NM_001010980.4                           NCMAP     c.-8+3A>G   \n",
       "3  NM_001039775.3                           AIM1L     c.3673C>T   \n",
       "4     NM_012090.5                           MACF1    c.10958A>C   \n",
       "\n",
       "  Protein (pNomen) HGVS cDNA-level nomenclature (fullCNomen)  \n",
       "0     p.Y579Lfs*66                    NM_024011.2:c.1735dupT  \n",
       "1          p.M297V                      NM_003196.2:c.889A>G  \n",
       "2                                   NM_001010980.4:c.-8+3A>G  \n",
       "3         p.R1225W                  NM_001039775.3:c.3673C>T  \n",
       "4         p.E3653A                    NM_012090.5:c.10958A>C  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'bilateral post-axial polydactyly of hands and feet',\n",
       " u'macrocephaly',\n",
       " u'tall stature',\n",
       " u'post-natal',\n",
       " u'central hypotonia',\n",
       " u'global developmental delay',\n",
       " u'obstructive sleep apnea',\n",
       " u'delayed cns central nervous myelination',\n",
       " u'white matter',\n",
       " u'corpus callosum',\n",
       " u'mildly ectopic neurohypophys posterior pituitary upslanted palpebral fissures',\n",
       " u'epicanthus',\n",
       " u'arched eyebrows',\n",
       " u'craniofacial asymmetry',\n",
       " u'bronchial stenosis',\n",
       " u'macrosomia',\n",
       " u'recurrent infection',\n",
       " u'gastroesophageal reflux',\n",
       " u'supratentorial perivascular space',\n",
       " u'mucopolysaccharidosis plasma urine creatine',\n",
       " u'purine',\n",
       " u'pyrimidine metabolism',\n",
       " u'gross motor',\n",
       " u'language',\n",
       " u'cognitive',\n",
       " u'ventriculomegaly',\n",
       " u'metatarsal bones',\n",
       " u'proximal phalanges',\n",
       " u'obesity',\n",
       " u'wheeze',\n",
       " u'brachycephaly frontal prominence',\n",
       " u'bitemporal narrowing',\n",
       " u'clinodactyly',\n",
       " u'brachydactyly',\n",
       " u'bardet-biedl',\n",
       " 'growth delay',\n",
       " 'developmental delay']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Chromosome',\n",
       " 'Start',\n",
       " 'Stop',\n",
       " 'Reference',\n",
       " 'Allele 1',\n",
       " 'Allele 2',\n",
       " 'Allele In Scope',\n",
       " 'Transcript',\n",
       " 'Gene Profile Report',\n",
       " 'Gene (gene)',\n",
       " 'cDNA (cNomen)',\n",
       " 'Protein (pNomen)',\n",
       " 'HGVS cDNA-level nomenclature (fullCNomen)']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>criteria</th>\n",
       "      <th>interpretation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>variant_annotations</td>\n",
       "      <td>Effect: missense_variant.&lt;br/&gt;Protein domain: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;exon: 1.&lt;br/&gt;ExAC MAF: 5.052e-05.&lt;br/&gt;DANN pathogenicity score: 0.999117068756.&lt;br/&gt;FATHMM pathogenicity score: 0.79186.&lt;br/&gt;MetaSVM pathogenicity score: 0.84564.&lt;br/&gt;GERP++ conservation score: 4.27.&lt;br/&gt;Clinvar variation ids: 225010.&lt;br/&gt;Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.&lt;br/&gt;Clinvar review status: criteria provided, single submitter.&lt;br/&gt;Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1 and PM5</td>\n",
       "      <td>Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3 and BS3</td>\n",
       "      <td>Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene   variant             criteria  \\\n",
       "0  BBS10  c.145C>T  variant_annotations   \n",
       "1  BBS10  c.145C>T                 PVS1   \n",
       "2  BBS10  c.145C>T          PS1 and PM5   \n",
       "3  BBS10  c.145C>T          PS3 and BS3   \n",
       "4  BBS10  c.145C>T                  PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     interpretation  \n",
       "0  Effect: missense_variant.<br/>Protein domain: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>exon: 1.<br/>ExAC MAF: 5.052e-05.<br/>DANN pathogenicity score: 0.999117068756.<br/>FATHMM pathogenicity score: 0.79186.<br/>MetaSVM pathogenicity score: 0.84564.<br/>GERP++ conservation score: 4.27.<br/>Clinvar variation ids: 225010.<br/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br/>Clinvar review status: criteria provided, single submitter.<br/>Pubmed references from Clinvar: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have damaging splicing effect. PVS1 is NOT met.  \n",
       "2                                                                                                                                                                                                                                                                                                             Variant effect is missense. Same amino acid change as a previously established pathogenic variant regardless of nucleotide change (Clinvar references: 225010). Not find missense change at an amino acid residue where a different missense change determined to be pathogenic has been seen before. The variant does NOT have damaging splicing effect. PS1 is met. PM5 is NOT met.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Not find well-established functional studies on this variant. PS3 is NOT met. BS3 is NOT met.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Relative risk is smaller than 5. The prevalence of the variant in affected individuals is NOT significantly increased compared with the prevalence in controls. PS4 is NOT met.  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpretation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>基因</th>\n",
       "      <th>变异</th>\n",
       "      <th>标准</th>\n",
       "      <th>解读</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>变异注释</td>\n",
       "      <td>突变类型: missense_variant.&lt;br/&gt;蛋白功能区: GroEL-like equatorial domain.&lt;br/&gt;HGVS ID: chr12:g.76741994G&gt;A.&lt;br/&gt;RefSeq ID: rs768933093.&lt;br/&gt;外显子: 1.&lt;br/&gt;ExAC 最小等位基因频率(MAF): 5.052e-05.&lt;br/&gt;DANN致病性分数: 0.999117068756.&lt;br/&gt;FATHMM致病性分数: 0.79186.&lt;br/&gt;MetaSVM致病性分数: 0.84564.&lt;br/&gt;GERP++序列保守性预测分数: 4.27.&lt;br/&gt;Clinvar数据库ID: 225010.&lt;br/&gt;Clinvar数据库记录的变异致病性: 致病|致病.&lt;br/&gt;Clinvar数据库记录审核状态: 提供标准，单个提交者.&lt;br/&gt;Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PVS1</td>\n",
       "      <td>基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS1和PM5</td>\n",
       "      <td>变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS3和BS3</td>\n",
       "      <td>未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBS10</td>\n",
       "      <td>c.145C&gt;T</td>\n",
       "      <td>PS4</td>\n",
       "      <td>具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      基因        变异       标准  \\\n",
       "0  BBS10  c.145C>T     变异注释   \n",
       "1  BBS10  c.145C>T     PVS1   \n",
       "2  BBS10  c.145C>T  PS1和PM5   \n",
       "3  BBS10  c.145C>T  PS3和BS3   \n",
       "4  BBS10  c.145C>T      PS4   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               解读  \n",
       "0  突变类型: missense_variant.<br/>蛋白功能区: GroEL-like equatorial domain.<br/>HGVS ID: chr12:g.76741994G>A.<br/>RefSeq ID: rs768933093.<br/>外显子: 1.<br/>ExAC 最小等位基因频率(MAF): 5.052e-05.<br/>DANN致病性分数: 0.999117068756.<br/>FATHMM致病性分数: 0.79186.<br/>MetaSVM致病性分数: 0.84564.<br/>GERP++序列保守性预测分数: 4.27.<br/>Clinvar数据库ID: 225010.<br/>Clinvar数据库记录的变异致病性: 致病|致病.<br/>Clinvar数据库记录审核状态: 提供标准，单个提交者.<br/>Clinvar数据库记录的Pubmed相关生物医学文献: ['25356970', '20120035', '16582908', '21044901', '20876674', '21642631', '20498079', '21517826', '20177705', '24746959', '25982971'].  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                  基因变异类型不是无效变异(null variant). 变异位点所在基因的功能丢失(loss of function)是已知的致病机制. 此变异不具有害的剪接效应(splicing effect). 不符合PVS1标准.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                变异为错义突变. 之前报道导致相同氨基酸改变的基因变异被证明是致病的 (Clinvar数据库参考: 225010). 未发现导致相同氨基酸残基错义突变（不同氨基酸改变）导致的致病基因变异. 此变异不具有害的剪接效应(splicing effect). 符合PS1标准. 不符合PM5标准.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   未发现针对此基因变异的完善的体内或体外功能性研究. 不符合PS3标准. 不符合BS3标准.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        具有此基因变异的人群患病率并未显著升高，相对风险(Relative Risk)小于5.0.. 不符合PS4标准.  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_ACMG_interpret_chinese.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_final_res, df_genes, phenos, field_names, variant_ACMG_interpretation, variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nput_gene = df_genes.to_json(orient='records')\n",
    "input_phenotype = ', '.join(phenos)\n",
    "result_table = ACMG_result.to_json(orient='records')\n",
    "interpretation = variant_ACMG_interpretation.to_json(orient='records')\n",
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')\n",
    "# logger.info(\"Finish processing data, start writing data to DB in background main task\")\n",
    "\n",
    "\n",
    "# sample = Main_table(\n",
    "#     task_id=raw_input_id,\n",
    "#     input_gene=input_gene,\n",
    "#     input_phenotype=input_phenotype,\n",
    "#     result=result_table,\n",
    "#     interpretation=interpretation,\n",
    "#     interpretation_chinese=interpretation_chinese,\n",
    "#     pub_date=timezone.now(),\n",
    "#     user_name=raw_input.user_name,\n",
    "#     task_name=raw_input.task_name,\n",
    "# )\n",
    "# sample.save()\n",
    "# logger.info(\"Finish writing data to DB in background main task\")\n",
    "# raw_input.status = \"succeed\"\n",
    "# raw_input.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"variant_annotations\",\"interpretation\":\"Effect: missense_variant.<br\\\\/>Protein domain: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>exon: 1.<br\\\\/>ExAC MAF: 5.052e-05.<br\\\\/>DANN pathogenicity score: 0.999117068756.<br\\\\/>FATHMM pathogenicity score: 0.79186.<br\\\\/>MetaSVM pathogenicity score: 0.84564.<br\\\\/>GERP++ conservation score: 4.27.<br\\\\/>Clinvar variation ids: 225010.<br\\\\/>Pathogenicity reported by Clinvar: Pathogenic|Pathogenic.<br\\\\/>Clinvar review status: criteria provided, single submitter.<br\\\\/>Pubmed references from Clinvar: [\\'25356970\\', \\'20120035\\', \\'16582908\\', \\'21044901\\', \\'20876674\\', \\'21642631\\', \\'20498079\\', \\'21517826\\', \\'20177705\\', \\'24746959\\', \\'25982971\\'].\"},{\"gene\":\"BBS10\",\"variant\":\"c.145C>T\",\"criteria\":\"PVS1\",\"interpretation\":\"Variant effect NOT in null variant type. Allele in a gene where loss of function (LOF) is a known mechanism of disease. The variant does NOT have dama'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"\\\\u57fa\\\\u56e0\":\"BBS10\",\"\\\\u53d8\\\\u5f02\":\"c.145C>T\",\"\\\\u6807\\\\u51c6\":\"\\\\u53d8\\\\u5f02\\\\u6ce8\\\\u91ca\",\"\\\\u89e3\\\\u8bfb\":\"\\\\u7a81\\\\u53d8\\\\u7c7b\\\\u578b: missense_variant.<br\\\\/>\\\\u86cb\\\\u767d\\\\u529f\\\\u80fd\\\\u533a: GroEL-like equatorial domain.<br\\\\/>HGVS ID: chr12:g.76741994G>A.<br\\\\/>RefSeq ID: rs768933093.<br\\\\/>\\\\u5916\\\\u663e\\\\u5b50: 1.<br\\\\/>ExAC \\\\u6700\\\\u5c0f\\\\u7b49\\\\u4f4d\\\\u57fa\\\\u56e0\\\\u9891\\\\u7387(MAF): 5.052e-05.<br\\\\/>DANN\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.999117068756.<br\\\\/>FATHMM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.79186.<br\\\\/>MetaSVM\\\\u81f4\\\\u75c5\\\\u6027\\\\u5206\\\\u6570: 0.84564.<br\\\\/>GERP++\\\\u5e8f\\\\u5217\\\\u4fdd\\\\u5b88\\\\u6027\\\\u9884\\\\u6d4b\\\\u5206\\\\u6570: 4.27.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93ID: 225010.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684\\\\u53d8\\\\u5f02\\\\u81f4\\\\u75c5\\\\u6027: \\\\u81f4\\\\u75c5|\\\\u81f4\\\\u75c5.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u5ba1\\\\u6838\\\\u72b6\\\\u6001: \\\\u63d0\\\\u4f9b\\\\u6807\\\\u51c6\\\\uff0c\\\\u5355\\\\u4e2a\\\\u63d0\\\\u4ea4\\\\u8005.<br\\\\/>Clinvar\\\\u6570\\\\u636e\\\\u5e93\\\\u8bb0\\\\u5f55\\\\u7684Pubmed\\\\u76f8\\\\u5173\\\\u751f\\\\u7269\\\\u533b\\\\u5b66\\\\u6587\\\\u'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpretation_chinese[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "interpretation_chinese = variant_ACMG_interpret_chinese.to_json(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "round(2.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "8/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "aa = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"<a href='https://www.ncbi.nlm.nih.gov/clinvar/variation//'>  </a>\"]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"<a href='https://www.ncbi.nlm.nih.gov/clinvar/variation/%s/'> %s </a>\" %(i,i) for i in aa.split('|')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if bb:\n",
    "    b=1+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'b' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-3b5d5c371295>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'b' is not defined"
     ]
    }
   ],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bb = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"<a href='https://www.ncbi.nlm.nih.gov/pubmed/%s'> %s </a>\" %(i,i) for i in bb]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "from io import StringIO\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "input_gene = open(\"/Users/Tianqi/Desktop/vcfsample.txt\",'rU').read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_vars = []\n",
    "input_gene = input_gene.split('\\n')\n",
    "\n",
    "for line in input_gene:\n",
    "    if line and line[:2] != \"##\":\n",
    "        header = line\n",
    "        sniffer = csv.Sniffer()\n",
    "        dialect = sniffer.sniff(header)\n",
    "        delimiter =  dialect.delimiter\n",
    "        field_names = header.split(delimiter)\n",
    "        break\n",
    "        \n",
    "chrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, None\n",
    "\n",
    "for idx in xrange(len(field_names)):\n",
    "    field = field_names[idx]\n",
    "    if re.match(r'chrom|#chrom', field, re.I): chrom_idx = idx\n",
    "    if re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "    if re.match(r'ref', field, re.I): ref_idx = idx\n",
    "    if re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "    if re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "input_gene_list = []\n",
    "CANDIDATE_GENES = []\n",
    "for line in input_gene[1:]:\n",
    "    if not line:\n",
    "        continue\n",
    "    line = line.rstrip()\n",
    "    parts = re.split(r'%s' % delimiter, line)\n",
    "    if line.startswith('#'):\n",
    "        continue\n",
    "    input_gene_list.append(parts)\n",
    "    gene, transcript, variant, variant_id = '', '', '', ''\n",
    "    if gene_idx:\n",
    "        gene = parts[gene_idx]\n",
    "        CANDIDATE_GENES.append(gene)\n",
    "    for part in parts:\n",
    "        if re.search(r'_.*:c\\.', part):\n",
    "            transcript, variant = part.split(':')\n",
    "        else:\n",
    "            if re.search(r'c\\.', part):\n",
    "                variant = part\n",
    "            if re.search(r'NM_', part, re.I):\n",
    "                transcript = part.split(':')[0]\n",
    "        if re.search(r'_.*:g\\.', part):\n",
    "            variant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "        if re.search(r'chr.*:g\\.', part, re.I):\n",
    "            variant_id = part\n",
    "    if not variant_id and (chrom_idx is not None and pos_idx is not None and ref_idx is not None and alt_idx is not None):\n",
    "        chrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "        variant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "    candidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "# remove lines in the input file which has wrong number of fields\n",
    "field_nums = []\n",
    "for line in input_gene_list:\n",
    "    field_nums.append(len(line))\n",
    "count = Counter(field_nums)\n",
    "correct_field_num = count.most_common()[0][0]\n",
    "correct_input_gene_list = []\n",
    "for line in input_gene_list:\n",
    "    if len(line) == correct_field_num:\n",
    "        correct_input_gene_list.append(line)\n",
    "df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "      <th>NA00002</th>\n",
       "      <th>NA00003</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>14370</td>\n",
       "      <td>rs6054257</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>29</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=3;DP=14;AF=0.5;DB;H2</td>\n",
       "      <td>GT:GQ:DP:HQ</td>\n",
       "      <td>0|0:48:1:51,51</td>\n",
       "      <td>1|0:48:8:51,51</td>\n",
       "      <td>1/1:43:5:.,.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS         ID REF ALT QUAL FILTER                     INFO  \\\n",
       "0     20  14370  rs6054257   G   A   29   PASS  NS=3;DP=14;AF=0.5;DB;H2   \n",
       "\n",
       "        FORMAT         NA00001         NA00002       NA00003  \n",
       "0  GT:GQ:DP:HQ  0|0:48:1:51,51  1|0:48:8:51,51  1/1:43:5:.,.  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('', '', '', 'chr20:g.14370G>A')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "can't multiply sequence by non-int of type 'float'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-bcfd7c265b7d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    223\u001b[0m     \u001b[0mdf_hpo_ranking_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mranking_genes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'score'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'hits'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0mdf_hpo_ranking_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_hpo_ranking_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'score'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m     \u001b[0mACMG_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant_ACMG_interpretation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant_ACMG_interpret_chinese\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_variant_ACMG_interpret\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_variant_ACMG_interpret_chinese\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mACMG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGet_ACMG_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_hpo_ranking_genes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariants\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_pubmed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m     \u001b[0;31m# filter variant on phenotype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/Keep_Learning/bio-nlp/deepbrain/deepb/ACMG.pyc\u001b[0m in \u001b[0;36mGet_ACMG_result\u001b[0;34m(df_hpo_ranking_genes, variants, df_pubmed)\u001b[0m\n\u001b[1;32m   1019\u001b[0m                                    \u001b[0mPP1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPP2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPP3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPP4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mPP5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBA1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBS1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBS2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBS3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBS4\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1020\u001b[0m     \t\t\t\t   BP1, BP2, BP3, BP4, BP5, BP6, BP7])\n\u001b[0;32m-> 1021\u001b[0;31m         \u001b[0mclassification_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mACMG_weighted_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhit_criteria\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclassify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvariant_ACMG_score\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1022\u001b[0m         \u001b[0mvariant_ACMG_result\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclassification_result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1023\u001b[0m         \u001b[0mvariant_ACMG_weighted_score\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mACMG_weighted_score\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/Keep_Learning/bio-nlp/deepbrain/deepb/ACMG.pyc\u001b[0m in \u001b[0;36mclassify\u001b[0;34m(ACMG_score_list)\u001b[0m\n\u001b[1;32m    859\u001b[0m         \u001b[0mscore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mACMG_score\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcriteria\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    860\u001b[0m         \u001b[0mweight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetWeight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcriteria\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 861\u001b[0;31m         \u001b[0mscore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscore\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mweight\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    862\u001b[0m         \u001b[0mACMG_weighted_score\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mscore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    863\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mscore\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mhit_criteria\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcriteria\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: can't multiply sequence by non-int of type 'float'"
     ]
    }
   ],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "    '''get a valid hgvs name from VCF-style \"chrom, pos, ref, alt\" data.\n",
    "\n",
    "    Example:\n",
    "\n",
    "        >>> myvariant.format_hgvs(\"1\", 35366, \"C\", \"T\")\n",
    "        >>> myvariant.format_hgvs(\"2\", 17142, \"G\", \"GA\")\n",
    "        >>> myvariant.format_hgvs(\"MT\", 8270, \"CACCCCCTCT\", \"C\")\n",
    "        >>> myvariant.format_hgvs(\"X\", 107930849, \"GGA\", \"C\")\n",
    "\n",
    "    '''\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "    if not input_phenotype:\n",
    "        return '', ''\n",
    "    text = StringIO(unicode(input_phenotype), newline=None)\n",
    "    lines = text.readlines()\n",
    "    lines = [line.strip() for line in lines]\n",
    "    phenos = []\n",
    "    for line in lines:\n",
    "        if not line:\n",
    "            continue\n",
    "        phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "        phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "        phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "        phenos += phenos_each_line\n",
    "\n",
    "    corner_cases = dict()\n",
    "    for pheno in phenos:\n",
    "        if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "            phenos.append('growth delay')\n",
    "            corner_cases['growth delay'] = pheno.strip()\n",
    "    for pheno in phenos:\n",
    "        if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "            phenos.append('developmental delay')\n",
    "            corner_cases['developmental delay'] = pheno.strip()\n",
    "    return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "    candidate_vars = []\n",
    "    input_gene = input_gene.split('\\n')\n",
    "\n",
    "    for line in input_gene:\n",
    "        if line and line[:2] != \"##\":\n",
    "            header = line\n",
    "            sniffer = csv.Sniffer()\n",
    "            dialect = sniffer.sniff(header)\n",
    "            delimiter =  dialect.delimiter\n",
    "            field_names = header.split(delimiter)\n",
    "            break\n",
    "\n",
    "    chrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, None\n",
    "\n",
    "    for idx in xrange(len(field_names)):\n",
    "        field = field_names[idx]\n",
    "        if re.match(r'chrom|#chrom', field, re.I): chrom_idx = idx\n",
    "        if re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "        if re.match(r'ref', field, re.I): ref_idx = idx\n",
    "        if re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "        if re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "    input_gene_list = []\n",
    "    CANDIDATE_GENES = []\n",
    "    for line in input_gene[1:]:\n",
    "        if not line:\n",
    "            continue\n",
    "        line = line.rstrip()\n",
    "        parts = re.split(r'%s' % delimiter, line)\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        input_gene_list.append(parts)\n",
    "        gene, transcript, variant, variant_id = '', '', '', ''\n",
    "        if gene_idx:\n",
    "            gene = parts[gene_idx]\n",
    "            CANDIDATE_GENES.append(gene)\n",
    "        for part in parts:\n",
    "            if re.search(r'_.*:c\\.', part):\n",
    "                transcript, variant = part.split(':')\n",
    "            else:\n",
    "                if re.search(r'c\\.', part):\n",
    "                    variant = part\n",
    "                if re.search(r'NM_', part, re.I):\n",
    "                    transcript = part.split(':')[0]\n",
    "            if re.search(r'_.*:g\\.', part):\n",
    "                variant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "            if re.search(r'chr.*:g\\.', part, re.I):\n",
    "                variant_id = part\n",
    "        if not variant_id and (chrom_idx is not None and pos_idx is not None and ref_idx is not None and alt_idx is not None):\n",
    "            chrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "            variant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "        candidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "    # remove lines in the input file which has wrong number of fields\n",
    "    field_nums = []\n",
    "    for line in input_gene_list:\n",
    "        field_nums.append(len(line))\n",
    "    count = Counter(field_nums)\n",
    "    correct_field_num = count.most_common()[0][0]\n",
    "    correct_input_gene_list = []\n",
    "    for line in input_gene_list:\n",
    "        if len(line) == correct_field_num:\n",
    "            correct_input_gene_list.append(line)\n",
    "    df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "    return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "\tranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "\t# collect variant info\n",
    "\thpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "\ttmp_candidate_vars = []\n",
    "\tfor var in candidate_vars:\n",
    "\t\tif var[0] in hpo_filtered_genes:\n",
    "\t\t\ttmp_candidate_vars.append(var)\n",
    "\tcandidate_vars = tmp_candidate_vars\n",
    "\treturn ranking_genes, candidate_vars\n",
    "\n",
    "# def master_function(raw_input_id):\n",
    "# \tstatus_step = \"generating candidate variants ...\" \n",
    "# \traw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "# \tinput_gene = raw_input.raw_input_gene\n",
    "# \tinput_phenotype = raw_input.raw_input_phenotype\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/vcfsample.txt\",'rU').read()\n",
    "input_phenotype = ''\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "if not CANDIDATE_GENES:\n",
    "    # collect variant info\n",
    "#     raw_input.status = \"Annotating variants using genomic databases\"\n",
    "#     raw_input.save()\n",
    "    final_res, variants = collectVariantInfo.get_variants_from_vcf(candidate_vars)\n",
    "    CANDIDATE_GENES = [_[0] for _ in final_res]\n",
    "    # map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "    if phenos:\n",
    "#         raw_input.status = \"Maping phenotypes to genes\"\n",
    "#         raw_input.save()\n",
    "        ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "    else:\n",
    "        ranking_genes = []\n",
    "        for gene in CANDIDATE_GENES:\n",
    "            ranking_genes.append((gene, 1.0, 1))\n",
    "else:\n",
    "    # map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "    if phenos:\n",
    "#         raw_input.status = \"Maping phenotypes to genes\"\n",
    "#         raw_input.save()\n",
    "        ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "    else:\n",
    "        ranking_genes = []\n",
    "        for gene in CANDIDATE_GENES:\n",
    "            ranking_genes.append((gene, 1.0, 1))\n",
    "    # collect variant info\n",
    "#     raw_input.status = \"Annotating variants using genomic databases\"\n",
    "#     raw_input.save()\n",
    "    final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "if final_res == [] and variants == defaultdict(dict):\n",
    "#     return None, df_genes, phenos, field_names, None, None\n",
    "    1\n",
    "else:\n",
    "    # pubmed\n",
    "    # raw_input.status = \"Searching biomedical literatures\"\n",
    "    # raw_input.save()\n",
    "    df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "    # ACMG\n",
    "    # raw_input.status = \"Checking ACMG standard\"\n",
    "    # raw_input.save()\n",
    "    df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "    df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "    ACMG_result, variant_ACMG_interpretation, variant_ACMG_interpret_chinese, df_variant_ACMG_interpret, df_variant_ACMG_interpret_chinese = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "    # filter variant on phenotype\n",
    "\n",
    "    if phenos:\n",
    "    #     raw_input.status = \"Filtering variants based on phenotypes\"\n",
    "    #     raw_input.save()\n",
    "        df_final_res, variant_ACMG_interpretation, variant_ACMG_interpret_chinese = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos, variant_ACMG_interpretation, variant_ACMG_interpret_chinese)\n",
    "    #     return df_final_res, df_genes, phenos, field_names, variant_ACMG_interpretation, variant_ACMG_interpret_chinese\n",
    "\n",
    "    # else:\n",
    "    #     return ACMG_result, df_genes, phenos, field_names, df_variant_ACMG_interpret, df_variant_ACMG_interpret_chinese"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final_res, variants = collectVariantInfo.get_variants_from_vcf(candidate_vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('', '', '', 'chr20:g.14370G>A')]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_vars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(u'PTPN11', u'c.923A>G', u'p.Asn308Ser')]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(dict,\n",
       "            {(u'PTPN11', u'c.923A>G'): {'alt': u'G',\n",
       "              'clinvar_pathogenicity': u'Pathogenic|Pathogenic|Pathogenic|Pathogenic',\n",
       "              'clinvar_pmids': [],\n",
       "              'clinvar_review_status': '',\n",
       "              'clinvar_variation_ids': '',\n",
       "              'dann': '0.998400682813',\n",
       "              'dbscSNV_ada_score': '',\n",
       "              'dbscSNV_rf_score': '',\n",
       "              'effect': u'missense_variant',\n",
       "              'exon': u'8',\n",
       "              'fathmm': '0.99101',\n",
       "              'gene': u'PTPN11',\n",
       "              'gerp++': '5.64',\n",
       "              'id': u'chr12:g.112915524A>G',\n",
       "              'interpro_domain': [u'PTP type protein phosphatase|Protein-tyrosine phosphatase-like'],\n",
       "              'maf_1000g': '',\n",
       "              'maf_esp6500': '',\n",
       "              'maf_exac': '',\n",
       "              'metasvm': '0.99878',\n",
       "              'protein': u'p.Asn308Ser',\n",
       "              'ref': u'A',\n",
       "              'rsid': u'rs121918455',\n",
       "              'transcript': u'NM_002834.3',\n",
       "              'variant': u'c.923A>G'}})"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "('CD59', 'c.146delA')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-4318910d39b6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpubmed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqueryPubmedDB\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfinal_res\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/Tianqi/Keep_Learning/bio-nlp/deepbrain/deepb/pubmed.pyc\u001b[0m in \u001b[0;36mqueryPubmedDB\u001b[0;34m(candidate_vars)\u001b[0m\n\u001b[1;32m     48\u001b[0m       \u001b[0mvariant\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgeneprotein2var\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein_variant\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein_variant\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 50\u001b[0;31m       \u001b[0mvariant\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprotein_variant\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenevar2protein\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein_variant\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     51\u001b[0m     \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotein\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpmid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtitle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mjournal\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0myear\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mimpact_factor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mabstract\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpathogenicity_score\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m   \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'Gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Variant'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Protein'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PMID'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Title'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Journal'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Year'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Impact_Factor'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Abstract'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'pathogenicity_score'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ('CD59', 'c.146delA')"
     ]
    }
   ],
   "source": [
    "pubmed.queryPubmedDB(final_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA00001</th>\n",
       "      <th>NA00002</th>\n",
       "      <th>NA00003</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>112915524</td>\n",
       "      <td>rs6054257</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>29</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=3;DP=14;AF=0.5;DB;H2</td>\n",
       "      <td>GT:GQ:DP:HQ</td>\n",
       "      <td>0|0:48:1:51,51</td>\n",
       "      <td>1|0:48:8:51,51</td>\n",
       "      <td>1/1:43:5:.,.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM        POS         ID REF ALT QUAL FILTER                     INFO  \\\n",
       "0     12  112915524  rs6054257   A   G   29   PASS  NS=3;DP=14;AF=0.5;DB;H2   \n",
       "\n",
       "        FORMAT         NA00001         NA00002       NA00003  \n",
       "0  GT:GQ:DP:HQ  0|0:48:1:51,51  1|0:48:8:51,51  1/1:43:5:.,.  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a = '[{\"#CHROM\":\"16\",\"POS\":\"50745926\",\"ID\":\"rs2066844\",\"REF\":\"C\",\"ALT\":\"T\",\"QUAL\":\"80\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=14;AF=0.5;DB;H2\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:48:1:51,51\",\"NA00002\":\"1|0:48:8:51,51\",\"NA00003\":\"1\\/1:43:5:.,.\"},{\"#CHROM\":\"20\",\"POS\":\"14370\",\"ID\":\"rs6054257\",\"REF\":\"G\",\"ALT\":\"A\",\"QUAL\":\"29\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=14;AF=0.5;DB;H2\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:48:1:51,51\",\"NA00002\":\"1|0:48:8:51,51\",\"NA00003\":\"1\\/1:43:5:.,.\"},{\"#CHROM\":\"20\",\"POS\":\"17330\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\"A\",\"QUAL\":\"3\",\"FILTER\":\"q10\",\"INFO\":\"NS=3;DP=11;AF=0.017\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:49:3:58,50\",\"NA00002\":\"0|1:3:5:65,3\",\"NA00003\":\"0\\/0:41:3\"},{\"#CHROM\":\"20\",\"POS\":\"1110696\",\"ID\":\"rs6040355\",\"REF\":\"A\",\"ALT\":\"G,T\",\"QUAL\":\"67\",\"FILTER\":\"PASS\",\"INFO\":\"NS=2;DP=10;AF=0.333,0.667;AA=T;DB\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"1|2:21:6:23,27\",\"NA00002\":\"2|1:2:0:18,2\",\"NA00003\":\"2\\/2:35:4\"},{\"#CHROM\":\"20\",\"POS\":\"1230237\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\"G\",\"QUAL\":\"47\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=13;AA=T\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:54:7:56,60\",\"NA00002\":\"0|0:48:4:51,51\",\"NA00003\":\"0\\/0:61:2\"},{\"#CHROM\":\"20\",\"POS\":\"1230288\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\".\",\"QUAL\":\"50\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=13;AA=T\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:54:7:56,60\",\"NA00002\":\"0|0:48:4:51,51\",\"NA00003\":\"0\\/0:61:2\"},{\"#CHROM\":\"20\",\"POS\":\"1234567\",\"ID\":\"microsat1\",\"REF\":\"GTCT\",\"ALT\":\"G,GTACT\",\"QUAL\":\"50\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=9;AA=G\",\"FORMAT\":\"GT:GQ:DP\",\"NA00001\":\"0\\/1:35:4\",\"NA00002\":\"0\\/2:17:2\",\"NA00003\":\"1\\/1:40:3\"}]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['[{\"#CHROM\":\"16\",\"POS\":\"50745926\",\"ID\":\"rs2066844\",\"REF\":\"C\",\"ALT\":\"T\",\"QUAL\":\"80\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=14;AF=0.5;DB;H2\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:48:1:51,51\",\"NA00002\":\"1|0:48:8:51,51\",\"NA00003\":\"1\\\\/1:43:5:.,.\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"14370\",\"ID\":\"rs6054257\",\"REF\":\"G\",\"ALT\":\"A\",\"QUAL\":\"29\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=14;AF=0.5;DB;H2\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:48:1:51,51\",\"NA00002\":\"1|0:48:8:51,51\",\"NA00003\":\"1\\\\/1:43:5:.,.\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"17330\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\"A\",\"QUAL\":\"3\",\"FILTER\":\"q10\",\"INFO\":\"NS=3;DP=11;AF=0.017\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:49:3:58,50\",\"NA00002\":\"0|1:3:5:65,3\",\"NA00003\":\"0\\\\/0:41:3\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"1110696\",\"ID\":\"rs6040355\",\"REF\":\"A\",\"ALT\":\"G,T\",\"QUAL\":\"67\",\"FILTER\":\"PASS\",\"INFO\":\"NS=2;DP=10;AF=0.333,0.667;AA=T;DB\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"1|2:21:6:23,27\",\"NA00002\":\"2|1:2:0:18,2\",\"NA00003\":\"2\\\\/2:35:4\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"1230237\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\"G\",\"QUAL\":\"47\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=13;AA=T\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:54:7:56,60\",\"NA00002\":\"0|0:48:4:51,51\",\"NA00003\":\"0\\\\/0:61:2\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"1230288\",\"ID\":\".\",\"REF\":\"T\",\"ALT\":\".\",\"QUAL\":\"50\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=13;AA=T\",\"FORMAT\":\"GT:GQ:DP:HQ\",\"NA00001\":\"0|0:54:7:56,60\",\"NA00002\":\"0|0:48:4:51,51\",\"NA00003\":\"0\\\\/0:61:2\"',\n",
       " '\"#CHROM\":\"20\",\"POS\":\"1234567\",\"ID\":\"microsat1\",\"REF\":\"GTCT\",\"ALT\":\"G,GTACT\",\"QUAL\":\"50\",\"FILTER\":\"PASS\",\"INFO\":\"NS=3;DP=9;AA=G\",\"FORMAT\":\"GT:GQ:DP\",\"NA00001\":\"0\\\\/1:35:4\",\"NA00002\":\"0\\\\/2:17:2\",\"NA00003\":\"1\\\\/1:40:3\"}]']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.split(\"},{\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "b='[{\"Gene\":\"AARS\",\"Transcript\":\"NM_001605.2\",\"cDNA (cNomen)\":\"c.1786-5T>C\",\"Protein (pNomen)\":\"\",\"Target Position\":\"chr16:70293094-70293094\"},{\"Gene\":\"AP4E1\",\"Transcript\":\"NM_007347.4\",\"cDNA (cNomen)\":\"c.2553A>T\",\"Protein (pNomen)\":\"p.Glu851Asp\",\"Target Position\":\"chr15:51289729-51289729\"},{\"Gene\":\"ARFGEF2\",\"Transcript\":\"NM_006420.2\",\"cDNA (cNomen)\":\"c.3120C>T\",\"Protein (pNomen)\":\"p.Leu1040Leu\",\"Target Position\":\"chr20:47611134-47611134\"},{\"Gene\":\"ARL2BP\",\"Transcript\":\"NM_012106.3\",\"cDNA (cNomen)\":\"c.207C>T\",\"Protein (pNomen)\":\"p.Tyr69Tyr\",\"Target Position\":\"chr16:57282555-57282555\"},{\"Gene\":\"ASPM\",\"Transcript\":\"NM_018136.4\",\"cDNA (cNomen)\":\"c.7812A>T\",\"Protein (pNomen)\":\"p.Lys2604Asn\",\"Target Position\":\"chr1:197070569-197070569\"},{\"Gene\":\"CD96\",\"Transcript\":\"NM_198196.2\",\"cDNA (cNomen)\":\"c.1295A>T\",\"Protein (pNomen)\":\"p.Gln432Leu\",\"Target Position\":\"chr3:111342667-111342667\"},{\"Gene\":\"CHAMP1\",\"Transcript\":\"NM_001164144.1\",\"cDNA (cNomen)\":\"c.449_451delCTC\",\"Protein (pNomen)\":\"p.Pro150del\",\"Target Position\":\"chr13:115089762-115089766\"},{\"Gene\":\"CREBBP\",\"Transcript\":\"NM_004380.2\",\"cDNA (cNomen)\":\"c.5354G>A\",\"Protein (pNomen)\":\"p.Cys1785Tyr\",\"Target Position\":\"chr16:3779694-3779694\"},{\"Gene\":\"DYNC1H1\",\"Transcript\":\"NM_001376.4\",\"cDNA (cNomen)\":\"c.634A>G\",\"Protein (pNomen)\":\"p.Met212Val\",\"Target Position\":\"chr14:102446171-102446171\"},{\"Gene\":\"DYNC2H1\",\"Transcript\":\"NM_001080463.1\",\"cDNA (cNomen)\":\"c.2479A>G\",\"Protein (pNomen)\":\"p.Ile827Val\",\"Target Position\":\"chr11:103006582-103006582\"},{\"Gene\":\"ERCC6L2\",\"Transcript\":\"NM_020207.4\",\"cDNA (cNomen)\":\"c.4123A>C\",\"Protein (pNomen)\":\"p.Thr1375Pro\",\"Target Position\":\"chr9:98774922-98774922\"},{\"Gene\":\"FRAS1\",\"Transcript\":\"NM_025074.6\",\"cDNA (cNomen)\":\"c.3065A>C\",\"Protein (pNomen)\":\"p.Lys1022Thr\",\"Target Position\":\"chr4:79295319-79295319\"},{\"Gene\":\"GBA2\",\"Transcript\":\"NM_020944.2\",\"cDNA (cNomen)\":\"c.379C>T\",\"Protein (pNomen)\":\"p.Arg127Trp\",\"Target Position\":\"chr9:35744684-35744684\"},{\"Gene\":\"KAT6B\",\"Transcript\":\"NM_012330.3\",\"cDNA (cNomen)\":\"c.4097_4105dupAAGAGGAAG\",\"Protein (pNomen)\":\"p.Glu1366_Glu1368dup\",\"Target Position\":\"chr10:76788659-76788660\"},{\"Gene\":\"KLLN\",\"Transcript\":\"NM_001126049.1\",\"cDNA (cNomen)\":\"c.17C>T\",\"Protein (pNomen)\":\"p.Pro6Leu\",\"Target Position\":\"chr10:89622228-89622228\"},{\"Gene\":\"LARP7\",\"Transcript\":\"NM_016648.3\",\"cDNA (cNomen)\":\"c.-2-4A>G\",\"Protein (pNomen)\":\"\",\"Target Position\":\"chr4:113565820-113565820\"},{\"Gene\":\"LIFR\",\"Transcript\":\"NM_002310.5\",\"cDNA (cNomen)\":\"c.46G>A\",\"Protein (pNomen)\":\"p.Asp16Asn\",\"Target Position\":\"chr5:38530704-38530704\"},{\"Gene\":\"MC1R\",\"Transcript\":\"NM_002386.3\",\"cDNA (cNomen)\":\"c.892T>C\",\"Protein (pNomen)\":\"p.Tyr298His\",\"Target Position\":\"chr16:89986558-89986558\"},{\"Gene\":\"MCM4\",\"Transcript\":\"NM_005914.3\",\"cDNA (cNomen)\":\"c.2102G>A\",\"Protein (pNomen)\":\"p.Arg701Gln\",\"Target Position\":\"chr8:48885590-48885590\"},{\"Gene\":\"NBAS\",\"Transcript\":\"NM_015909.3\",\"cDNA (cNomen)\":\"c.5110C>T\",\"Protein (pNomen)\":\"p.His1704Tyr\",\"Target Position\":\"chr2:15427225-15427225\"},{\"Gene\":\"NBAS\",\"Transcript\":\"NM_015909.3\",\"cDNA (cNomen)\":\"c.2290C>T\",\"Protein (pNomen)\":\"p.Pro764Ser\",\"Target Position\":\"chr2:15601374-15601374\"},{\"Gene\":\"NFKB1\",\"Transcript\":\"NM_003998.3\",\"cDNA (cNomen)\":\"c.1601G>A\",\"Protein (pNomen)\":\"p.Arg534His\",\"Target Position\":\"chr4:103518782-103518782\"},{\"Gene\":\"PAH\",\"Transcript\":\"NM_000277.1\",\"cDNA (cNomen)\":\"c.1208C>T\",\"Protein (pNomen)\":\"p.Ala403Val\",\"Target Position\":\"chr12:103234285-103234285\"},{\"Gene\":\"PDE6C\",\"Transcript\":\"NM_006204.3\",\"cDNA (cNomen)\":\"c.2087C>T\",\"Protein (pNomen)\":\"p.Thr696Met\",\"Target Position\":\"chr10:95418708-95418708\"},{\"Gene\":\"PIGN\",\"Transcript\":\"NM_176787.4\",\"cDNA (cNomen)\":\"c.1372G>A\",\"Protein (pNomen)\":\"p.Ala458Thr\",\"Target Position\":\"chr18:59780429-59780429\"},{\"Gene\":\"POC1B\",\"Transcript\":\"NM_172240.2\",\"cDNA (cNomen)\":\"c.179A>G\",\"Protein (pNomen)\":\"p.Lys60Arg\",\"Target Position\":\"chr12:89891041-89891041\"},{\"Gene\":\"PRF1\",\"Transcript\":\"NM_001083116.1\",\"cDNA (cNomen)\":\"c.1528T>C\",\"Protein (pNomen)\":\"p.Cys510Arg\",\"Target Position\":\"chr10:72357949-72357949\"},{\"Gene\":\"PROP1\",\"Transcript\":\"NM_006261.4\",\"cDNA (cNomen)\":\"c.652A>C\",\"Protein (pNomen)\":\"p.Ser218Arg\",\"Target Position\":\"chr5:177419739-177419739\"},{\"Gene\":\"RELN\",\"Transcript\":\"NM_005045.3\",\"cDNA (cNomen)\":\"c.8795C>A\",\"Protein (pNomen)\":\"p.Ser2932Tyr\",\"Target Position\":\"chr7:103138572-103138572\"},{\"Gene\":\"RYR1\",\"Transcript\":\"NM_000540.2\",\"cDNA (cNomen)\":\"c.8959G>C\",\"Protein (pNomen)\":\"p.Glu2987Gln\",\"Target Position\":\"chr19:39001164-39001164\"},{\"Gene\":\"SH3PXD2B\",\"Transcript\":\"NM_001017995.2\",\"cDNA (cNomen)\":\"c.921G>C\",\"Protein (pNomen)\":\"p.Gln307His\",\"Target Position\":\"chr5:171777458-171777458\"},{\"Gene\":\"SLC39A4\",\"Transcript\":\"NM_130849.3\",\"cDNA (cNomen)\":\"c.1126G>A\",\"Protein (pNomen)\":\"p.Ala376Thr\",\"Target Position\":\"chr8:145639669-145639669\"},{\"Gene\":\"SLC4A4\",\"Transcript\":\"NM_003759.3\",\"cDNA (cNomen)\":\"c.2542T>C\",\"Protein (pNomen)\":\"p.Phe848Leu\",\"Target Position\":\"chr4:72413417-72413417\"},{\"Gene\":\"SMPD1\",\"Transcript\":\"NM_000543.4\",\"cDNA (cNomen)\":\"c.108_109insGCG\",\"Protein (pNomen)\":\"p.Val36_Leu37insAla\",\"Target Position\":\"chr11:6411935-6411936\"},{\"Gene\":\"SRCAP\",\"Transcript\":\"NM_006662.2\",\"cDNA (cNomen)\":\"c.7853A>G\",\"Protein (pNomen)\":\"p.Asn2618Ser\",\"Target Position\":\"chr16:30749214-30749214\"},{\"Gene\":\"TCOF1\",\"Transcript\":\"NM_001135243.1\",\"cDNA (cNomen)\":\"c.925G>C\",\"Protein (pNomen)\":\"p.Gly309Arg\",\"Target Position\":\"chr5:149753791-149753791\"},{\"Gene\":\"TG\",\"Transcript\":\"NM_003235.4\",\"cDNA (cNomen)\":\"c.2222C>T\",\"Protein (pNomen)\":\"p.Thr741Met\",\"Target Position\":\"chr8:133900274-133900274\"},{\"Gene\":\"TMPRSS15\",\"Transcript\":\"NM_002772.2\",\"cDNA (cNomen)\":\"c.687T>G\",\"Protein (pNomen)\":\"p.Phe229Leu\",\"Target Position\":\"chr21:19737543-19737543\"},{\"Gene\":\"UBR1\",\"Transcript\":\"NM_174916.2\",\"cDNA (cNomen)\":\"c.3959T>C\",\"Protein (pNomen)\":\"p.Leu1320Pro\",\"Target Position\":\"chr15:43281055-43281055\"},{\"Gene\":\"ZNF513\",\"Transcript\":\"NM_144631.5\",\"cDNA (cNomen)\":\"c.901G>A\",\"Protein (pNomen)\":\"p.Gly301Arg\",\"Target Position\":\"chr2:27601137-27601137\"}]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene', 'Transcript', 'cDNA (cNomen)', 'Protein (pNomen)', 'Target Position']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i.split(\":\")[0][1:-1] for i in b.split(\"},{\")[0][2:].split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHRO',\n",
       " 'O',\n",
       " '',\n",
       " 'E',\n",
       " 'L',\n",
       " 'UA',\n",
       " 'ILTE',\n",
       " 'NF',\n",
       " 'ORMA',\n",
       " 'A0000',\n",
       " 'A0000',\n",
       " 'A0000']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i.split('\":\"')[0][1:-1] for i in a.split(\"},{\")[0][2:].split('\",\"')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHRO',\n",
       " 'O',\n",
       " '',\n",
       " 'E',\n",
       " 'L',\n",
       " 'UA',\n",
       " 'ILTE',\n",
       " 'NF',\n",
       " 'ORMA',\n",
       " 'A0000',\n",
       " 'A0000',\n",
       " 'A0000']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_gene_field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " 'FORMAT',\n",
       " 'NA00001',\n",
       " '1',\n",
       " 'NA00002',\n",
       " '1',\n",
       " 'NA00003',\n",
       " '']"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i.split(\":\")[0][1:-1] for i in a.split(\"},{\")[0][2:].split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " 'FORMAT',\n",
       " 'NA00001',\n",
       " 'NA00002',\n",
       " 'NA00003']"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i.split('\":\"')[0] for i in a.split(\"},{\")[0][3:].split('\",\"')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Gene\":\"AARS\",\"Transcript\":\"NM_001605.2\",\"cDNA (cNomen)\":\"c.1786-5T>C\",\"Protein (pNomen)\":\"\",\"Target Position\":\"chr16:70293094-70293094\"'"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b.split(\"},{\")[0][3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from langdetect import detect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'zh-cn'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detect(u\"任务1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from oauth2client.service_account import ServiceAccountCredentials\n",
    "\n",
    "# scopes = ['https://www.googleapis.com/auth/translateservice.admin']\n",
    "\n",
    "credentials = ServiceAccountCredentials.from_json_keyfile_name(\n",
    "    '/Users/Tianqi/Keep_Learning/google_api/translate-7696ad4e40b3.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from httplib2 import Http\n",
    "\n",
    "http_auth = credentials.authorize(Http())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:googleapiclient.discovery_cache:file_cache is unavailable when using oauth2client >= 4.0.0\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/Tianqi/anaconda/lib/python2.7/site-packages/googleapiclient/discovery_cache/__init__.py\", line 41, in autodetect\n",
      "    from . import file_cache\n",
      "  File \"/Users/Tianqi/anaconda/lib/python2.7/site-packages/googleapiclient/discovery_cache/file_cache.py\", line 41, in <module>\n",
      "    'file_cache is unavailable when using oauth2client >= 4.0.0')\n",
      "ImportError: file_cache is unavailable when using oauth2client >= 4.0.0\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot instantiate ctype 'EVP_MD_CTX' of unknown size",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-f54006658921>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mapiclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdiscovery\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mbuild\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mservice\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbuild\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'translate'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'v2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhttp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhttp_auth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/_helpers.pyc\u001b[0m in \u001b[0;36mpositional_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    131\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0mpositional_parameters_enforcement\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mPOSITIONAL_WARNING\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m                     \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mwrapped\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    134\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpositional_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/googleapiclient/discovery.pyc\u001b[0m in \u001b[0;36mbuild\u001b[0;34m(serviceName, version, http, discoveryServiceUrl, developerKey, model, requestBuilder, credentials, cache_discovery, cache)\u001b[0m\n\u001b[1;32m    226\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m       content = _retrieve_discovery_doc(\n\u001b[0;32m--> 228\u001b[0;31m         requested_url, discovery_http, cache_discovery, cache)\n\u001b[0m\u001b[1;32m    229\u001b[0m       return build_from_document(content, base=discovery_url, http=http,\n\u001b[1;32m    230\u001b[0m           \u001b[0mdeveloperKey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdeveloperKey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequestBuilder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequestBuilder\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/googleapiclient/discovery.pyc\u001b[0m in \u001b[0;36m_retrieve_discovery_doc\u001b[0;34m(url, http, cache_discovery, cache)\u001b[0m\n\u001b[1;32m    273\u001b[0m   \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'URL being requested: GET %s'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mactual_url\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 275\u001b[0;31m   \u001b[0mresp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mactual_url\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    276\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m400\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/transport.pyc\u001b[0m in \u001b[0;36mnew_request\u001b[0;34m(uri, method, body, headers, redirections, connection_type)\u001b[0m\n\u001b[1;32m    157\u001b[0m             _LOGGER.info('Attempting refresh to obtain '\n\u001b[1;32m    158\u001b[0m                          'initial access_token')\n\u001b[0;32m--> 159\u001b[0;31m             \u001b[0mcredentials\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_refresh\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0morig_request_method\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m         \u001b[0;31m# Clone and modify the request headers to add the appropriate\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/client.pyc\u001b[0m in \u001b[0;36m_refresh\u001b[0;34m(self, http)\u001b[0m\n\u001b[1;32m    747\u001b[0m         \"\"\"\n\u001b[1;32m    748\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 749\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_refresh_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhttp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    750\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    751\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire_lock\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/client.pyc\u001b[0m in \u001b[0;36m_do_refresh_request\u001b[0;34m(self, http)\u001b[0m\n\u001b[1;32m    772\u001b[0m             \u001b[0mHttpAccessTokenRefreshError\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mWhen\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mrefresh\u001b[0m \u001b[0mfails\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    773\u001b[0m         \"\"\"\n\u001b[0;32m--> 774\u001b[0;31m         \u001b[0mbody\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_generate_refresh_request_body\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    775\u001b[0m         \u001b[0mheaders\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_generate_refresh_request_headers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    776\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/client.pyc\u001b[0m in \u001b[0;36m_generate_refresh_request_body\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1482\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1483\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_generate_refresh_request_body\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1484\u001b[0;31m         \u001b[0massertion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_generate_assertion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1485\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1486\u001b[0m         body = urllib.parse.urlencode({\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/service_account.pyc\u001b[0m in \u001b[0;36m_generate_assertion\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    383\u001b[0m         \u001b[0mpayload\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m         return crypt.make_signed_jwt(self._signer, payload,\n\u001b[0;32m--> 385\u001b[0;31m                                      key_id=self._private_key_id)\n\u001b[0m\u001b[1;32m    386\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    387\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0msign_blob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblob\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/crypt.pyc\u001b[0m in \u001b[0;36mmake_signed_jwt\u001b[0;34m(signer, payload, key_id)\u001b[0m\n\u001b[1;32m     95\u001b[0m     \u001b[0msigning_input\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mb'.'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msegments\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m     \u001b[0msignature\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msigner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msigning_input\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m     \u001b[0msegments\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_helpers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_urlsafe_b64encode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msignature\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/oauth2client/_openssl_crypt.pyc\u001b[0m in \u001b[0;36msign\u001b[0;34m(self, message)\u001b[0m\n\u001b[1;32m     95\u001b[0m         \"\"\"\n\u001b[1;32m     96\u001b[0m         \u001b[0mmessage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_helpers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_to_bytes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcrypto\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'sha256'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/OpenSSL/crypto.pyc\u001b[0m in \u001b[0;36msign\u001b[0;34m(pkey, data, digest)\u001b[0m\n\u001b[1;32m   2401\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"No such digest method\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2402\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2403\u001b[0;31m     \u001b[0mmd_ctx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_ffi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"EVP_MD_CTX*\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2404\u001b[0m     \u001b[0mmd_ctx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_ffi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmd_ctx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_lib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEVP_MD_CTX_cleanup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2405\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot instantiate ctype 'EVP_MD_CTX' of unknown size"
     ]
    }
   ],
   "source": [
    "from apiclient.discovery import build\n",
    "service = build('translate', 'v2', http=http_auth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "service = build('translate', 'v2',\n",
    "        developerKey='e37b6a3f5406868430b2b58d7b65825e4cb21082')\n",
    "service.translations().list(\n",
    "  source='en',\n",
    "  target='fr',\n",
    "  q=['flower', 'car']\n",
    ").execute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:google.auth._default:No project ID could be determined from the Cloud SDK configuration. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n"
     ]
    },
    {
     "ename": "Forbidden",
     "evalue": "403 Daily Limit Exceeded (GET https://translation.googleapis.com/language/translate/v2?target=ru&q=Hello%2C+world%21)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mForbidden\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-64-509d0d57ae64>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     13\u001b[0m translation = translate_client.translate(\n\u001b[1;32m     14\u001b[0m     \u001b[0mtext\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m     target_language=target)\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mu'Text: {}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/google/cloud/translate/client.pyc\u001b[0m in \u001b[0;36mtranslate\u001b[0;34m(self, values, target_language, format_, source_language, customization_ids, model)\u001b[0m\n\u001b[1;32m    232\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    233\u001b[0m         response = self._connection.api_request(\n\u001b[0;32m--> 234\u001b[0;31m             method='GET', path='', query_params=query_params)\n\u001b[0m\u001b[1;32m    235\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    236\u001b[0m         \u001b[0mtranslations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'translations'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/site-packages/google/cloud/_http.pyc\u001b[0m in \u001b[0;36mapi_request\u001b[0;34m(self, method, path, query_params, data, content_type, headers, api_base_url, api_version, expect_json, _target_object)\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m             raise make_exception(response, content,\n\u001b[0;32m--> 303\u001b[0;31m                                  error_info=method + ' ' + url)\n\u001b[0m\u001b[1;32m    304\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m         \u001b[0mstring_or_bytes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbinary_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mForbidden\u001b[0m: 403 Daily Limit Exceeded (GET https://translation.googleapis.com/language/translate/v2?target=ru&q=Hello%2C+world%21)"
     ]
    }
   ],
   "source": [
    "# Imports the Google Cloud client library\n",
    "from google.cloud import translate\n",
    "\n",
    "# Instantiates a client\n",
    "translate_client = translate.Client()\n",
    "\n",
    "# The text to translate\n",
    "text = u'Hello, world!'\n",
    "# The target language\n",
    "target = 'ru'\n",
    "\n",
    "# Translates some text into Russian\n",
    "translation = translate_client.translate(\n",
    "    text,\n",
    "    target_language=target)\n",
    "\n",
    "print(u'Text: {}'.format(text))\n",
    "print(u'Translation: {}'.format(translation['translatedText']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:google.auth._default:No project ID could be determined from the Cloud SDK configuration. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translation: 你好\n"
     ]
    }
   ],
   "source": [
    "from google.cloud import translate\n",
    "\n",
    "translate_client = translate.Client()\n",
    "text = u'Hello'\n",
    "target = 'zh'\n",
    "\n",
    "translation = translate_client.translate(text,target_language=target)\n",
    "print(u'Translation: {}'.format(translation['translatedText']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'Why sometimes easy to use and sometimes easy to use'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text = u'为什么时而好用时而不好用'\n",
    "target = 'en'\n",
    "translation = translate_client.translate(text,target_language=target)\n",
    "translation['translatedText']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sourceLang = 'zh-Hans'\n",
    "targetLang = 'en'\n",
    "sourceText = '你好啊'\n",
    "\n",
    "url = \"https://translate.googleapis.com/translate_a/single?client=gtx&sl=\" + sourceLang + \"&tl=\" + targetLang + \"&dt=t&q=\" + sourceText"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://translate.googleapis.com/translate_a/single?client=gtx&sl=zh-Hans&tl=en&dt=t&q=\\xe4\\xbd\\xa0\\xe5\\xa5\\xbd\\xe5\\x95\\x8a'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "url = 'response'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "https://translate.googleapis.com/translate_a/single?client=gtx&sl=zh-Hans&tl=en&dt=t&q=%E4%BD%A0%E5%A5%BD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html><html lang=en><meta charset=utf-8><meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\"><title>Error 403 (Forbidden)!!1</title><style>*{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}</style><a href=//www.google.com/><span id=logo aria-label=Google></span></a><p><b>403.</b> <ins>That’s an error.</ins><p>Your client does not have permission to get URL <code>/translate_a/single?client=gtx&amp;sl=zh-Hans&amp;tl=en&amp;dt=t&amp;q=%E4%BD%A0%E5%A5%BD</code> from this server.  (Client IP address: 216.174.124.181)<br><br>\n",
      "\n",
      "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \r\n",
      "  <ins>That’s all we know.</ins>\n"
     ]
    }
   ],
   "source": [
    "import urllib\n",
    "\n",
    "f = urllib.urlopen(url)\n",
    "myfile = f.read()\n",
    "print myfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "No JSON object could be decoded",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-5b2368465ee0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0murllib\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murllib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0murlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;32mprint\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/json/__init__.pyc\u001b[0m in \u001b[0;36mloads\u001b[0;34m(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[0m\n\u001b[1;32m    336\u001b[0m             \u001b[0mparse_int\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mparse_float\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m \u001b[0;32mand\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    337\u001b[0m             parse_constant is None and object_pairs_hook is None and not kw):\n\u001b[0;32m--> 338\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_default_decoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    339\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcls\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    340\u001b[0m         \u001b[0mcls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mJSONDecoder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/json/decoder.pyc\u001b[0m in \u001b[0;36mdecode\u001b[0;34m(self, s, _w)\u001b[0m\n\u001b[1;32m    364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    365\u001b[0m         \"\"\"\n\u001b[0;32m--> 366\u001b[0;31m         \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_w\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    367\u001b[0m         \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_w\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    368\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/Tianqi/anaconda/lib/python2.7/json/decoder.pyc\u001b[0m in \u001b[0;36mraw_decode\u001b[0;34m(self, s, idx)\u001b[0m\n\u001b[1;32m    382\u001b[0m             \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscan_once\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    383\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"No JSON object could be decoded\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: No JSON object could be decoded"
     ]
    }
   ],
   "source": [
    "import urllib, json\n",
    "response = urllib.urlopen(url)\n",
    "data = json.loads(response.read())\n",
    "print data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "iput_text = \"不发育，严重的全球发育迟缓，身材矮小，高血压，小头畸形，皮层视觉障碍，趾骨平滑肌，重叠趾，球鼻，鼻腔不发达，延迟牙喷，额叶多毛症，巩膜葡萄膜炎，垂直性低张力，颞视盘，松散的食管括约肌，俱乐部脚，畸形特征，营养不良，先天性CMV，失神经注视，下旋转海马，胼，体，血管瘤，胎记，额头多毛症，眼睛，听觉\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "iput_text = \"不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育，不发育\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "iput_text = \"不发育\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non-developed, severe global growth retardation, short stature, high blood pressure, small head deformity, cortical visual impairment, phalanx smooth muscle, overlapping toes, ball nose, nasal inactivity, delayed toothbrush, frontal lobelia, scleral uveitis, \n"
     ]
    }
   ],
   "source": [
    "import urllib2\n",
    "site = \"https://translate.googleapis.com/translate_a/single?client=gtx&sl=zh-Hans&tl=en&dt=t&q=\"+iput_text\n",
    "hdr = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',\n",
    "        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',\n",
    "        'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.3',\n",
    "        'Accept-Encoding': 'none',\n",
    "        'Accept-Language': 'en-US,en;q=0.8',\n",
    "        'Connection': 'keep-alive'}\n",
    "req = urllib2.Request(site, headers=hdr)\n",
    "try:\n",
    "    page = urllib2.urlopen(req)\n",
    "except urllib2.HTTPError, e:\n",
    "    print e\n",
    "\n",
    "content = page.read()\n",
    "translate = json.loads(content)[0][0][0]\n",
    "print translate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[[u'Non-developed, severe global growth retardation, short stature, high blood pressure, small head deformity, cortical visual impairment, phalanx smooth muscle, overlapping toes, ball nose, nasal inactivity, delayed toothbrush, frontal lobelia, scleral uveitis, ',\n",
       "   u'\\u4e0d\\u53d1\\u80b2\\uff0c\\u4e25\\u91cd\\u7684\\u5168\\u7403\\u53d1\\u80b2\\u8fdf\\u7f13\\uff0c\\u8eab\\u6750\\u77ee\\u5c0f\\uff0c\\u9ad8\\u8840\\u538b\\uff0c\\u5c0f\\u5934\\u7578\\u5f62\\uff0c\\u76ae\\u5c42\\u89c6\\u89c9\\u969c\\u788d\\uff0c\\u8dbe\\u9aa8\\u5e73\\u6ed1\\u808c\\uff0c\\u91cd\\u53e0\\u8dbe\\uff0c\\u7403\\u9f3b\\uff0c\\u9f3b\\u8154\\u4e0d\\u53d1\\u8fbe\\uff0c\\u5ef6\\u8fdf\\u7259\\u55b7\\uff0c\\u989d\\u53f6\\u591a\\u6bdb\\u75c7\\uff0c\\u5de9\\u819c\\u8461\\u8404\\u819c\\u708e\\uff0c',\n",
       "   None,\n",
       "   None,\n",
       "   3],\n",
       "  [u'Vertical, low tension, temporal disc, loose esophageal sphincter, club foot, malformations, malnutrition, congenital CMV, denervated gyrus, lower hippocampus, corpuscles, body, hemangioma, birthmark, forehead hirsutism, eye, hearing',\n",
       "   u'\\u5782\\u76f4\\u6027\\u4f4e\\u5f20\\u529b\\uff0c\\u989e\\u89c6\\u76d8\\uff0c\\u677e\\u6563\\u7684\\u98df\\u7ba1\\u62ec\\u7ea6\\u808c\\uff0c\\u4ff1\\u4e50\\u90e8\\u811a\\uff0c\\u7578\\u5f62\\u7279\\u5f81\\uff0c\\u8425\\u517b\\u4e0d\\u826f\\uff0c\\u5148\\u5929\\u6027CMV\\uff0c\\u5931\\u795e\\u7ecf\\u6ce8\\u89c6\\uff0c\\u4e0b\\u65cb\\u8f6c\\u6d77\\u9a6c\\uff0c\\u80fc\\uff0c\\u4f53\\uff0c\\u8840\\u7ba1\\u7624\\uff0c\\u80ce\\u8bb0\\uff0c\\u989d\\u5934\\u591a\\u6bdb\\u75c7\\uff0c\\u773c\\u775b\\uff0c\\u542c\\u89c9',\n",
       "   None,\n",
       "   None,\n",
       "   3]],\n",
       " None,\n",
       " u'zh-CN']"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json.loads(content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "# from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "from collections import defaultdict\n",
    "from langdetect import detect\n",
    "import urllib2\n",
    "import json\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def read_input_pheno_file(input_phenotype):\n",
    "\tif not input_phenotype:\n",
    "\t\treturn '', '', ''\n",
    "\tlanguage = detect(unicode(input_phenotype))\n",
    "\tif language == \"zh-cn\":\n",
    "\t\tsite = \"https://translate.googleapis.com/translate_a/single?client=gtx&sl=zh-Hans&tl=en&dt=t&q=\"+input_phenotype\n",
    "\t\thdr = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',\n",
    "\t\t\t\t'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',\n",
    "\t\t\t\t'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.3',\n",
    "\t\t\t\t'Accept-Encoding': 'none',\n",
    "\t\t\t\t'Accept-Language': 'en-US,en;q=0.8',\n",
    "\t\t\t\t'Connection': 'keep-alive'}\n",
    "\t\treq = urllib2.Request(site, headers=hdr)\n",
    "\t\ttry:\n",
    "\t\t\tpage = urllib2.urlopen(req)\n",
    "\t\t\tcontent = page.read()\n",
    "\t\t\tinput_phenotype = json.loads(content)[0][0][0]\n",
    "\t\texcept urllib2.HTTPError, e:\n",
    "\t\t\treturn '', '', ''\n",
    "\ttext = StringIO(unicode(input_phenotype), newline=None)\n",
    "\tlines = text.readlines()\n",
    "\tlines = [line.strip() for line in lines]\n",
    "\tphenos = []\n",
    "\tfor line in lines:\n",
    "\t\tif not line:\n",
    "\t\t\tcontinue\n",
    "        phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "        phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "        phenos_each_line = [s.lower() for s in phenos_each_line if s]\n",
    "        phenos += phenos_each_line\n",
    "\n",
    "\tcorner_cases = dict()\n",
    "\toriginal_phenos = [_.strip() for _ in phenos]\n",
    "\tfor pheno in phenos:\n",
    "\t\tif re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "\t\t\tphenos.append('growth delay')\n",
    "\t\t\torner_cases['growth delay'] = pheno.strip()\n",
    "\t\tfor pheno in phenos:\n",
    "\t\t\tif re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "\t\t\t\tphenos.append('developmental delay')\n",
    "\t\t\t\tcorner_cases['developmental delay'] = pheno.strip()\n",
    "\t\tphenos = [_.strip() for _ in phenos]\n",
    "\t\tphenos = list(set(phenos))\n",
    "\t\treturn phenos, corner_cases, original_phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([u'short stature',\n",
       "  u'craniosynostosis',\n",
       "  u'ventriculomegaly',\n",
       "  u'metopic',\n",
       "  u'growth hormone deficiency',\n",
       "  u'anetoderma',\n",
       "  u'lamdoid',\n",
       "  u'macular atrophy',\n",
       "  u'developmental delay',\n",
       "  u'sagittal'],\n",
       " {},\n",
       " [u'lamdoid',\n",
       "  u'sagittal',\n",
       "  u'craniosynostosis',\n",
       "  u'metopic',\n",
       "  u'macular atrophy',\n",
       "  u'anetoderma',\n",
       "  u'short stature',\n",
       "  u'developmental delay',\n",
       "  u'ventriculomegaly',\n",
       "  u'growth hormone deficiency'])"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_input_pheno_file('Lamdoid, Sagittal, Craniosynostosis, Metopic, macular atrophy, anetoderma, short stature, developmental delay, Ventriculomegaly, growth hormone deficiency')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from nltk.corpus import wordnet as wn\n",
    "from collections import deque\n",
    "import re\n",
    "from nltk.stem.wordnet import WordNetLemmatizer\n",
    "from collections import Counter\n",
    "import os.path\n",
    "import numpy as np\n",
    "BASE = '/Users/Tianqi/Keep_Learning/bio-nlp/deepbrain/deepb/'\n",
    "\n",
    "\"\"\"\n",
    "For each input phenotype of the patient, we do preliminary mapping of it to hpo standard terms.\n",
    "We convert the input phenotype term and hpo standard terms to a word list and strip symbols in the word and lemmatize the word.\n",
    "The word list of hpo standard term and its synonyms and the word list of input phenotype are compared and if they are a match or partial match and they do not contain non-overlapped words with opposite meanings, we consider the hpo term as a candidate.\n",
    "The matched hpo terms are filtered on their common ancestors, which basically reflect semantics of the hpo terms. The hpo term with the highest similarity score is used as the gold standard. If another hpo candidate has a very high level common ancestor (not close in semantics) with it, then that hpo candidate will be removed. Matches after this filtering step are direct matches (from input phenotype).\n",
    "We also get indirect matches by mapping the synonyms of hpo term with very high similarity score to other hpo terms. \n",
    "The direct and indirect matches are combined and if they do not have non-overlapped opposite meaning words, they are taken into the final matches.\n",
    "\n",
    "The hpo dictionary is like a graph. It has a root node.\n",
    "hpo_subclass: dictionary with a hpo term (id) as key and its direct children as value.\n",
    "hpo_superclass: dictionary with a hpo term (id) as key and its direct parents as value.\n",
    "\"\"\"\n",
    "\n",
    "HPO_SUBCLASS = os.path.join(BASE, \"data/hpo_subclasses.txt\")\n",
    "HPO_SUPERCLASSES = os.path.join(BASE, \"data/hpo_superclasses.txt\")\n",
    "HPO_NAMES = os.path.join(BASE, \"data/hpo_names.txt\")\n",
    "HPO_FILE = os.path.join(BASE, \"data/all_hpo_terms_and_synonyms.txt\")\n",
    "PHENOTYPE_TO_GENE_FILE = os.path.join(BASE, \"data/Expanded_OMIM_ALL_FREQUENCIES_phenotype_to_genes_without_synonym.txt\")\n",
    "PHENOTYPE_TO_DISEASE_FILE = os.path.join(BASE, 'data/Expanded_ALL_SOURCES_ALL_FREQUENCIES_diseases_to_phenotypes.txt')\n",
    "\n",
    "## Get levels of each hpo term; e.g., HP:0000001 is the first level; its direct children are the second level.\n",
    "## Phenotypic abnormality is level 1; Abnormality of prenatal development or birth is level 2.\n",
    "## In the search for the common ancestors, we should stop at level 2; if two terms have the common ancestor below or equal at level 2, then we deem them to be semantically similar.\n",
    "\n",
    "PHENOTYPE_COUNT_FILE = os.path.join(BASE, 'data/hpoid_count_in_pubmed.txt')\n",
    "df_pheno_count = pd.read_csv(PHENOTYPE_COUNT_FILE, sep = '\\t', usecols = [0, 2])\n",
    "hpoidpheno_count = df_pheno_count.set_index('hpoid')['count'].to_dict()\n",
    "\n",
    "hpo_subclass = dict()\n",
    "with open(HPO_SUBCLASS, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        subclasses = parts[1].strip('[]').replace(\"'\", \"\")\n",
    "        subclasses = [_.strip() for _ in subclasses.split(',')]\n",
    "        hpo_subclass[hpoid] = subclasses\n",
    "\n",
    "hpo_levels = dict()\n",
    "level = 0\n",
    "root = 'HP:0000001'\n",
    "curr_level_nodes = [root] \n",
    "while curr_level_nodes:\n",
    "    next_level_nodes = []\n",
    "    for node in curr_level_nodes:\n",
    "        hpo_levels[node] = level\n",
    "        if node in hpo_subclass:\n",
    "            next_level_nodes += hpo_subclass[node]\n",
    "    curr_level_nodes = next_level_nodes\n",
    "    level += 1\n",
    "\n",
    "hpo_superclass = dict()\n",
    "with open(HPO_SUPERCLASSES, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        superclasses = parts[1].strip('[]').replace(\"'\", \"\")\n",
    "        superclasses = [_.strip() for _ in superclasses.split(',')]\n",
    "        hpo_superclass[hpoid] = superclasses\n",
    "\n",
    "hpo_name = dict()\n",
    "with open(HPO_NAMES, 'rb') as f:\n",
    "    # Skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid = parts[0]\n",
    "        hponame = parts[1]\n",
    "        hpo_name[hpoid] = hponame        \n",
    "\n",
    "def getAllAncestorsBFS(node, tree):\n",
    "    \"\"\" get all ancestors of a node using BFS\n",
    "\n",
    "    Args:\n",
    "        node (string): node in a tree\n",
    "        tree (dict): dictionary with node as key and its direct parents as value\n",
    "\n",
    "    Returns:\n",
    "        P (list): the ancestors of a node with the first element as the direct parent and so on\n",
    "\n",
    "    \"\"\"\n",
    "    P, Q = [node], deque([node])\n",
    "    while Q:\n",
    "        u = Q.popleft()\n",
    "        if u not in tree:\n",
    "            continue\n",
    "        for v in tree[u]:\n",
    "            if v in P: continue\n",
    "            if v not in tree: continue\n",
    "            P.append(v)\n",
    "            Q.append(v)\n",
    "    # Append All ('HP:0000001') as the highest ancestor \n",
    "    P.append('HP:0000001')\n",
    "    return P \n",
    "\n",
    "def findLowestCommonAncestor(node1, node2, tree):\n",
    "    \"\"\" find the lowest common ancestor of two nodes in a tree\n",
    "\n",
    "    Args:\n",
    "        node1 (string): hpoid of a phenotype\n",
    "        node2 (string): hpoid of another phenotype\n",
    "        tree (dict): dictionary with key as a node and value as the direct parents of the node\n",
    "\n",
    "    Returns:\n",
    "        lca (tuple): lowest common ancestor (hpoid, level, hponame)\n",
    "        common_ancestors (list): common ancestors from lowest common ancestor to top level common ancestor\n",
    "\n",
    "    \"\"\"\n",
    "    node1_ancestors = getAllAncestorsBFS(node1, tree) \n",
    "    node2_ancestors = getAllAncestorsBFS(node2, tree)\n",
    "    common_ancestors = [] \n",
    "    for ancestor in node2_ancestors:\n",
    "        if ancestor in node1_ancestors:\n",
    "            name = hpo_name[ancestor]\n",
    "            level = hpo_levels[ancestor]\n",
    "            common_ancestors.append((ancestor, level, name)) \n",
    "    lca = common_ancestors[0]\n",
    "    return lca, common_ancestors\n",
    "\n",
    "def get_antonyms(word):\n",
    "    \"\"\" generate antonyms of a word using NLTK and wordnet\n",
    "\n",
    "    Args:\n",
    "        word (string): input word\n",
    "\n",
    "    Returns:\n",
    "         antonyms (set): a set of antonyms of the input word\n",
    "\n",
    "    \"\"\"\n",
    "    antonyms = []\n",
    "    for synset in wn.synsets(word):\n",
    "        for lemma in synset.lemmas():\n",
    "            antonyms += lemma.antonyms()\n",
    "            similar_synsets = synset.similar_tos() \n",
    "            for similar_synset in similar_synsets:\n",
    "                for lemma_ in similar_synset.lemmas():\n",
    "                    antonyms += lemma_.antonyms()\n",
    "    direct_antonyms = list(set(antonyms))\n",
    "    # Get similar words of the direct antonyms\n",
    "    for antonym in direct_antonyms:\n",
    "        similar_synsets = antonym.synset().similar_tos()\n",
    "        for synset in similar_synsets:\n",
    "            for lemma in synset.lemmas():\n",
    "                antonyms.append(lemma)\n",
    "    antonyms = list(set(antonyms))\n",
    "    antonyms = set([_.name() for _ in antonyms])\n",
    "    return antonyms\n",
    "\n",
    "with open(HPO_FILE, 'rb') as infile:\n",
    "    hpo_name2id = dict()\n",
    "    hpo_id2synonyms = dict()\n",
    "    # skip the first line\n",
    "    infile.readline()\n",
    "    for line in infile.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        hpoid, hponame = parts[0], parts[1]\n",
    "        hpo_name2id[hponame] = hpoid\n",
    "        hpoid = hpoid.split('-')[0]\n",
    "        if hponame in hpo_id2synonyms:\n",
    "            hpo_id2synonyms[hpoid].append(hponame)\n",
    "        else:\n",
    "            hpo_id2synonyms[hpoid] = [hponame]\n",
    "    hpo_terms = hpo_name2id.keys()  \n",
    "\n",
    "processed_antonyms = dict()\n",
    "def isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "    \"\"\" check if the input two word lists contain words with opposite meaning\n",
    "\n",
    "    Args:\n",
    "        words_in_pheno_only (set): a set of words that appear in the phenotype word list only\n",
    "        word_in_hpo_only (set): a set of words that appear in the hpo term word list only\n",
    "\n",
    "    Returns:\n",
    "         bool: True or False\n",
    "\n",
    "    \"\"\"\n",
    "    global processed_antonyms\n",
    "    # If any word ends with ness, convert to its noun form\n",
    "    words_in_pheno_only |= set([_[0:-4] for _ in words_in_pheno_only if _.endswith('ness')])\n",
    "    words_in_hpo_only |= set([_[0:-4] for _ in words_in_hpo_only if _.endswith('ness')])\n",
    "    # corner case: gain\n",
    "    if 'gain' in words_in_hpo_only:\n",
    "        words_in_hpo_only.add('increased')\n",
    "        words_in_hpo_only.add('steady')\n",
    "    if 'gain' in words_in_pheno_only:\n",
    "        words_in_pheno_only.add('increased')\n",
    "        words_in_pheno_only.add('steady')\n",
    "    # Poor weight gain and weight gain are in opposite meaning\n",
    "    if (words_in_pheno_only == set(['poor']) and words_in_hpo_only == set()) or (words_in_hpo_only == set(['poor']) and words_in_pheno_only == set()):\n",
    "        return True\n",
    "    for word in words_in_pheno_only:\n",
    "        try:\n",
    "            if word in processed_antonyms:\n",
    "                antonyms = processed_antonyms[word]\n",
    "            else:\n",
    "                antonyms = get_antonyms(word)\n",
    "                processed_antonyms[word] = antonyms\n",
    "        except UnicodeDecodeError as e:\n",
    "            continue\n",
    "        if antonyms.intersection(words_in_hpo_only):\n",
    "            return True\n",
    "    for word in words_in_hpo_only:\n",
    "        try:\n",
    "            if word in processed_antonyms:\n",
    "                antonyms = processed_antonyms[word]\n",
    "            else:\n",
    "                antonyms = get_antonyms(word)\n",
    "                processed_antonyms[word] = antonyms\n",
    "        except UnicodeDecodeError as e:\n",
    "            continue\n",
    "        if antonyms.intersection(words_in_pheno_only):\n",
    "            return True\n",
    "    return False \n",
    "\n",
    "def processhpoterms(hpo_):\n",
    "    \"\"\" process each hpo term by transforming each term to a list of individual words, and strip ,()' in the word\n",
    "        and lemmatize each word. Use a global dictionary to record the mapping from original hpo term to the processed list.\n",
    "        Use re for split hpo terms\n",
    "        Use lemmatizer to lemmatize words e.g., disturbances --> distrubance\n",
    "\n",
    "    Args:\n",
    "        hpo_ (string); a hpo term\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    \"\"\"\n",
    "    global hpo_words_dict\n",
    "    hpo_words =  [_.strip(\",()'\") for _ in re.split(' |/', hpo_)]\n",
    "    try:\n",
    "        hpo_words = [lemmatizer.lemmatize(word) for word in hpo_words]\n",
    "    except UnicodeDecodeError as e:\n",
    "        pass\n",
    "    hpo_words_dict[hpo_] = hpo_words\n",
    "\n",
    "# If the opposite meanings are found for a hpoid, then its synonyms are all in opposite meanings\n",
    "unmatches_due_to_opposite_meanings = []\n",
    "hpo_words_dict = dict()\n",
    "# Use lemmatizer to lemmatize words e.g., disturbances --> distrubance\n",
    "lemmatizer = WordNetLemmatizer()\n",
    "\n",
    "def map2hpo(pheno):\n",
    "    \"\"\" preliminary mapping of each input phenotype of the patient to hpo standard terms\n",
    "\n",
    "    Args:\n",
    "        pheno (string): an input phenotype of the patient\n",
    "\n",
    "    Returns:\n",
    "        matches (list): each element is a tuple (hpoid, hponame, match_type, similarity_score)\n",
    "            e.g., [('HP:0012171', 'Stereotypical hand wringing', 'OVERLAP', 0.6666666666666666), ...]\n",
    "    \"\"\"\n",
    "    global unmatches_due_to_opposite_meanings, hpo_words_dict\n",
    "    matches = []\n",
    "    # Use lemmatizer to lemmatize words e.g., disturbances --> distrubance\n",
    "    pheno_ = pheno.lower()\n",
    "    pheno_words = [_.strip(\",()'\") for _ in re.split(' |/', pheno_)]\n",
    "    try:\n",
    "        pheno_words = [lemmatizer.lemmatize(word) for word in pheno_words]\n",
    "    except UnicodeDecodeError as e:\n",
    "        pass\n",
    "\n",
    "    for hpo in hpo_terms:\n",
    "        hpoid = hpo_name2id[hpo]\n",
    "        hpo_ = hpo.lower()\n",
    "        if hpo_.startswith('obsolete '):\n",
    "            continue \n",
    "        if pheno_ == hpo_:\n",
    "            matches.append((hpoid, hpo, 'EXACT', 1.0)) \n",
    "            continue\n",
    "        if hpo_ not in hpo_words_dict:\n",
    "            processhpoterms(hpo_)\n",
    "        hpo_words = hpo_words_dict[hpo_]\n",
    "\n",
    "        if pheno_words == hpo_words:\n",
    "            matches.append((hpoid, hpo, 'LEMMA_MATCH', 1.0)) \n",
    "            continue\n",
    "        if set(pheno_words) == set(hpo_words):\n",
    "            matches.append((hpoid, hpo, 'REORDER', 1.0)) \n",
    "            continue\n",
    "\n",
    "        common_words = set(pheno_words) & set(hpo_words)\n",
    "        if not common_words:\n",
    "            continue\n",
    "\n",
    "        words_in_pheno_only = set(pheno_words) - common_words\n",
    "        words_in_hpo_only = set(hpo_words) - common_words\n",
    "        # check if the pheno and hpo are in opposite meaning\n",
    "        # if any word in pheno is an antonym of a word in hpo, or the other way\n",
    "        # we skip this hpo\n",
    "        sim = float(len(common_words)) / max(len(hpo_words), len(pheno_words))  \n",
    "        if (sim >= 0.3333) and (pheno_ in hpo_) and (len(hpo_words) <= 10 * len(pheno_words)):\n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'SUB', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "            continue\n",
    "        if (sim >= 0.3333) and (hpo_ in pheno_) and (len(pheno_words) <= 10 * len(hpo_words)): \n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'SUPER', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "            continue\n",
    "        if (sim >= 0.5) and len(pheno_words) >= 2:\n",
    "            if not isOppositeMeaning(words_in_pheno_only, words_in_hpo_only):\n",
    "                matches.append((hpoid, hpo, 'OVERLAP', sim))\n",
    "            else:\n",
    "                unmatches_due_to_opposite_meanings.append(hpoid.split('-')[0])\n",
    "    matches_res = []\n",
    "    for match in matches:\n",
    "        if match[0].split('-')[0] not in unmatches_due_to_opposite_meanings:\n",
    "            matches_res.append(match)\n",
    "    matches = matches_res\n",
    "    return matches\n",
    "\n",
    "def filterMatchesOnCommonAncestors(matches):\n",
    "    \"\"\" this function filters matches on common ancestors. One input phenotype may map to multiple hpo standard terms\n",
    "        with different similarity score. We pick up the one with best match score, and check the common ancestor of\n",
    "        other hpo terms with it. The common ancestor is literally in the cengji jigou, e.g., abnormality of head -- abnormality\n",
    "        of eye -- baineichang. If the common ancestor of a hpo term and the best-match hpo term are very high level, then they are likely NOT\n",
    "        describing the same phenotype. This hpo term will be removed from the matches.\n",
    "\n",
    "    Args:\n",
    "        matches (list): each element is a tuple (hpoid, hponame, match_type, similarity_score)\n",
    "            e.g., [('HP:0012171', 'Stereotypical hand wringing', 'OVERLAP', 0.6666666666666666), ...]\n",
    "\n",
    "    Returns:\n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score).\n",
    "            e.g., [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "\n",
    "    \"\"\"\n",
    "    if not matches:\n",
    "        return matches \n",
    "    final_matches = []\n",
    "    # If the match is a disease not a phenotype, then we just add it to the final matches result\n",
    "    matches_tmp = []\n",
    "    for match in matches:\n",
    "        if not match[0].startswith('HP:'):\n",
    "            final_matches.append((match[1], match[0], match[3]))\n",
    "        else:\n",
    "            matches_tmp.append(match)\n",
    "    matches = matches_tmp\n",
    "    if not matches:\n",
    "        return final_matches\n",
    "    best_match_hpo = sorted(matches, key = lambda x: x[3], reverse = True)[0][0]\n",
    "    for match in matches:\n",
    "        term = match[0]\n",
    "        # Remove 'synonym' from the hpoid (term) e.g., HP:0004414-synonym --> HP:0004414\n",
    "        term = term.split('-')[0]\n",
    "        lca, common_ancestors = findLowestCommonAncestor(term, best_match_hpo.split('-')[0], hpo_superclass)\n",
    "        if lca[1] >= 2:\n",
    "            final_matches.append((match[1], match[0], match[3])) \n",
    "    return final_matches    \n",
    "\n",
    "def map2hpoWithPhenoSynonyms(pheno):\n",
    "    \"\"\" map each input phenotype of the patient to hpo standard terms\n",
    "\n",
    "    Args:\n",
    "        pheno (string): an input phenotype of the patient \n",
    "\n",
    "    Returns: \n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score)\n",
    "            e.g., [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "\n",
    "    \"\"\"\n",
    "    global unmatches_due_to_opposite_meanings\n",
    "    matches = map2hpo(pheno)\n",
    "    direct_matches = filterMatchesOnCommonAncestors(matches)  \n",
    "    final_matches = direct_matches \n",
    "    for match in direct_matches:\n",
    "        hponame = match[0]\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if sim >= 0.75:\n",
    "            hposynonyms = hpo_id2synonyms[hpoid]\n",
    "            for synonym in hposynonyms:\n",
    "                if synonym != hponame:\n",
    "                    matches_ = map2hpo(synonym)\n",
    "                    indirect_matches = filterMatchesOnCommonAncestors(matches_)\n",
    "                    final_matches = final_matches + indirect_matches\n",
    "    final_matches = list(set(final_matches))\n",
    "    matches_res = []\n",
    "    for match in final_matches:\n",
    "        if match[0].split('-')[0] not in unmatches_due_to_opposite_meanings:\n",
    "            matches_res.append(match)\n",
    "    final_matches = matches_res\n",
    "    unmatches_due_to_opposite_meanings = []\n",
    "    return final_matches\n",
    "\n",
    "hpoid2gene = dict()\n",
    "with open(PHENOTYPE_TO_GENE_FILE, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        geneid, genesymbol, hponame, hpoid = parts[0], parts[1], parts[2], parts[3]\n",
    "        if hpoid in hpoid2gene:\n",
    "            hpoid2gene[hpoid].append(genesymbol)   \n",
    "        else:\n",
    "            hpoid2gene[hpoid] = [genesymbol]\n",
    "\n",
    "def map2gene(final_matches, CANDIDATE_GENES):\n",
    "    \"\"\" Map to genes for each phenotype without similarity score (one phenotype at one time)\n",
    "\n",
    "    Args:\n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score)\n",
    "        CANDIDATE_GENES (list): input candidate genes of the patient\n",
    "       \n",
    "    Returns:\n",
    "        mapped_genes (list): genes associated with the phenotype \n",
    "\n",
    "    \"\"\"\n",
    "    # Pre-process the final_matches by remove '-synonym' from the hpoid and then unique the hpoids\n",
    "    final_matches = [_[1] for _ in final_matches] \n",
    "    final_matches = [_.split('-')[0] for _ in final_matches] \n",
    "    final_matches = list(set(final_matches))\n",
    "    mapped_genes = []\n",
    "    for hpoid in final_matches:\n",
    "        try:\n",
    "            mapped_genes += hpoid2gene[hpoid] \n",
    "        except KeyError:\n",
    "            pass\n",
    "    ## For each phenotype in patient record, we generated multiple phenotype keywords \n",
    "    ## to maximize the mapping to the genes. However, we only count it once for each \n",
    "    ## phenotype even if the multiple keywords lead to multiple mappings.\n",
    "    ## In practice, we don't need to screen through all genes. Only a few gene candidates are examined.\n",
    "    mapped_genes = list(set(mapped_genes) & set(CANDIDATE_GENES))\n",
    "    return mapped_genes\n",
    "\n",
    "def map2geneWithSim(final_matches, CANDIDATE_GENES):\n",
    "    \"\"\" Map to genes for each phenotype (one phenotype at one time) with similarity score \n",
    "    final_matches like: [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "\n",
    "    Args:\n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score)\n",
    "        CANDIDATE_GENES (list): input candidate genes of the patient\n",
    "\n",
    "    Returns:\n",
    "        mapped_genes_score (dict): a dict with gene symbol as key and similarity score as value \n",
    "        mapped_genes_score_phenospecificity (dict): a dict with gene symbol as key and similarity score \n",
    "            that is weighted by phenotype rareness as value \n",
    "\n",
    "    \"\"\"\n",
    "    global hpoidpheno_count\n",
    "    hpoid_sim = dict() \n",
    "    for match in final_matches:\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if hpoid not in hpoid_sim or sim > hpoid_sim[hpoid]:\n",
    "            hpoid_sim[hpoid] = sim\n",
    "    mapped_genes_score = dict()\n",
    "    mapped_genes_score_phenospecificity = dict()\n",
    "\n",
    "    for hpoid in hpoid_sim:\n",
    "        try:\n",
    "            mapped_genes = hpoid2gene[hpoid]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        sim = hpoid_sim[hpoid]\n",
    "        for gene in mapped_genes:\n",
    "            # In practice, we don't need to screen through all genes. \n",
    "            # Only a few gene candidates are examined.\n",
    "            if gene not in CANDIDATE_GENES:\n",
    "                continue\n",
    "            if gene not in mapped_genes_score or sim > mapped_genes_score[gene]:\n",
    "                mapped_genes_score[gene] = sim\n",
    "                if hpoid in hpoidpheno_count:\n",
    "                    pheno_weight = 6.0 / np.log(hpoidpheno_count[hpoid] + 2.7183) \n",
    "                else: \n",
    "                    pheno_weight = 6.0  \n",
    "                mapped_genes_score_phenospecificity[gene] = sim * pheno_weight\n",
    "    return mapped_genes_score, mapped_genes_score_phenospecificity \n",
    "\n",
    "hpoid2disease = dict()\n",
    "numphenosindisease = dict()\n",
    "with open(PHENOTYPE_TO_DISEASE_FILE, 'rb') as f:\n",
    "    # skip the first header line\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        line = line.rstrip()\n",
    "        parts = line.split('\\t')\n",
    "        diseaseid, hpoid, hponame, diseasename = parts[0], parts[1], parts[2], parts[3]\n",
    "        if diseasename in numphenosindisease:\n",
    "            numphenosindisease[diseasename] += 1\n",
    "        else:\n",
    "            numphenosindisease[diseasename] = 1\n",
    "        if hpoid in hpoid2disease:\n",
    "            hpoid2disease[hpoid].append(diseasename)\n",
    "        else:\n",
    "            hpoid2disease[hpoid] = [diseasename]\n",
    "\n",
    "def map2disease(final_matches):\n",
    "    \"\"\" Map to diseases for each phenotype without similarity score (one phenotype at one time)\n",
    "\n",
    "    Args:\n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score)\n",
    "       \n",
    "    Returns:\n",
    "        mapped_diseases (list): diseases associated with the phenotype \n",
    "\n",
    "    \"\"\"\n",
    "    # Pre-process the final_matches by remove '-synonym' from the hpoid and then unique the hpoids\n",
    "    final_matches = [_[1] for _ in final_matches] \n",
    "    final_matches = [_.split('-')[0] for _ in final_matches] \n",
    "    final_matches = list(set(final_matches))\n",
    "    mapped_diseases = []\n",
    "    for hpoid in final_matches:\n",
    "        try:\n",
    "            mapped_diseases += hpoid2disease[hpoid] \n",
    "        except KeyError:\n",
    "            pass\n",
    "    return mapped_diseases\n",
    "\n",
    "def map2diseaseWithSim(final_matches):\n",
    "    \"\"\" Map to diseases for each phenotype (one phenotype at one time) with similarity score \n",
    "    final_matches like: [('Postnatal macrocephaly','HP:0005490',0.5), ('Macrocephaly,relative','HP:0004482-synonym', 0.5), ...]\n",
    "\n",
    "    Args:\n",
    "        final_matches (list): each element is a tuple (hponame, hpoid, similarity_score)\n",
    "\n",
    "    Returns:\n",
    "        mapped_diseases_score (dict): a dict with diseasename as key and similarity score as value \n",
    "\n",
    "    \"\"\"\n",
    "    hpoid_sim = dict() \n",
    "    for match in final_matches:\n",
    "        hpoid = match[1].split('-')[0]\n",
    "        sim = match[2]\n",
    "        if hpoid not in hpoid_sim or sim > hpoid_sim[hpoid]:\n",
    "            hpoid_sim[hpoid] = sim\n",
    "    mapped_diseases_score = dict()\n",
    "    for hpoid in hpoid_sim:\n",
    "        try:\n",
    "            mapped_diseases = hpoid2disease[hpoid]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        sim = hpoid_sim[hpoid]\n",
    "        for disease in mapped_diseases:\n",
    "            if disease not in mapped_diseases_score or sim > mapped_diseases_score[disease]:\n",
    "                mapped_diseases_score[disease] = sim\n",
    "    return mapped_diseases_score \n",
    "\n",
    "\n",
    "def generate_score(phenos, CANDIDATE_GENES, corner_cases, original_phenos):\n",
    "    \"\"\" This is the main function in this .py file. It is called in the main.py file.\n",
    "\n",
    "    Args:\n",
    "        phenos (list): input patient's phenotypes\n",
    "        CANDIDATE_GENES (list): input candidate genes of the patient\n",
    "        corner_cases (dict): corner cases of input patient's phenotypes that need to be handled. Currently the corner cases are to make growth delay and development delay interchangebly. \n",
    "        original_phenos (list): input patient's phenotypes before handling the corner cases\n",
    "\n",
    "    Returns:\n",
    "        ranking_genes (list): each element in the list is a tuple (gene, score, hits, phenospecificity_score). \n",
    "            It records gene symbol, phenotype hit score for this gene, nubmer of phenotypes associated with this gene, \n",
    "            and the phenotype correlation score by weighing the rareness of each phenotype.  \n",
    "        ranking_diseases (list): each element in the list is a tuple (disease, score, hits). \n",
    "            It records the disease name, phenotype hit score for this disease, and number of phenotypes associated with this disease.\n",
    "        gene_associated_phenos (dict): phenotypes associated with each gene\n",
    "\n",
    "    \"\"\"\n",
    "    all_mapped_genes = []\n",
    "    all_mapped_genes_score = {}\n",
    "    gene_phenos = {} \n",
    "    all_mapped_diseases = []\n",
    "    all_mapped_diseases_score = {}\n",
    "    disease_phenos = {} \n",
    "\n",
    "    gene_associated_phenos = dict()\n",
    "    all_mapped_genes_score_phenospecificity = {}\n",
    "\n",
    "    for pheno in phenos:\n",
    "        final_matches = map2hpoWithPhenoSynonyms(pheno)\n",
    "        mapped_genes = map2gene(final_matches, CANDIDATE_GENES)\n",
    "\n",
    "        for gene in mapped_genes:\n",
    "            if pheno not in original_phenos:\n",
    "                continue\n",
    "            if gene in gene_associated_phenos:\n",
    "                gene_associated_phenos[gene].append(pheno)\n",
    "            else:\n",
    "                gene_associated_phenos[gene] = [pheno]\n",
    "\n",
    "        all_mapped_genes += mapped_genes\n",
    "        mapped_genes_score, mapped_genes_score_phenospecificity = map2geneWithSim(final_matches, CANDIDATE_GENES)\n",
    "        if pheno in corner_cases:\n",
    "            pheno = corner_cases[pheno]\n",
    "        for gene in mapped_genes_score:\n",
    "            if gene not in all_mapped_genes_score:\n",
    "                all_mapped_genes_score[gene] = mapped_genes_score[gene]\n",
    "                all_mapped_genes_score_phenospecificity[gene] = mapped_genes_score_phenospecificity[gene]\n",
    "            else:\n",
    "                all_mapped_genes_score[gene] += mapped_genes_score[gene]\n",
    "                all_mapped_genes_score_phenospecificity[gene] += mapped_genes_score_phenospecificity[gene]\n",
    "            if gene not in gene_phenos:\n",
    "                gene_phenos[gene] = [pheno]\n",
    "            else:\n",
    "                gene_phenos[gene].append(pheno)\n",
    "\n",
    "        mapped_diseases = map2disease(final_matches)\n",
    "        all_mapped_diseases += mapped_diseases\n",
    "        mapped_diseases_score = map2diseaseWithSim(final_matches)\n",
    "        for disease in mapped_diseases_score:\n",
    "            if disease not in all_mapped_diseases_score:\n",
    "                all_mapped_diseases_score[disease] = mapped_diseases_score[disease]\n",
    "            else:\n",
    "                all_mapped_diseases_score[disease] += mapped_diseases_score[disease]\n",
    "            if disease not in disease_phenos:\n",
    "                disease_phenos[disease] = [pheno]\n",
    "            else:\n",
    "                disease_phenos[disease].append(pheno)\n",
    "\n",
    "    num_patient_phenos = len(phenos)\n",
    "    for disease in all_mapped_diseases_score:\n",
    "        numphenos = numphenosindisease[disease]\n",
    "        disease_score = all_mapped_diseases_score[disease]\n",
    "        # algo to calculate the disease score\n",
    "        all_mapped_diseases_score[disease] = disease_score ** 2 / float(numphenos) / (1 + num_patient_phenos - disease_score)\n",
    "\n",
    "    all_mapped_genes_score = [(k,v) for k, v in all_mapped_genes_score.items()]   \n",
    "    all_mapped_genes_score = sorted(all_mapped_genes_score, key = lambda pair: pair[1], reverse = True)\n",
    "    all_mapped_diseases_score = [(k,v) for k, v in all_mapped_diseases_score.items()]   \n",
    "    all_mapped_diseases_score = sorted(all_mapped_diseases_score, key = lambda pair: pair[1], reverse = True)\n",
    "\n",
    "    count = Counter(all_mapped_genes)\n",
    "    count = sorted(count.items(), key = lambda pair: pair[1], reverse = True)\n",
    "    count_disease = Counter(all_mapped_diseases)\n",
    "    count_disease = sorted(count_disease.items(), key = lambda pair: pair[1], reverse = True)\n",
    "\n",
    "    ranking_genes = []\n",
    "    ranking_diseases = []\n",
    "    for id in xrange(len(all_mapped_genes_score)):\n",
    "        gene, score = all_mapped_genes_score[id]\n",
    "        hits = None\n",
    "        phenospecificity_score = all_mapped_genes_score_phenospecificity[gene]\n",
    "        for gene_hits in count:\n",
    "            if gene_hits[0] == gene:\n",
    "                hits = gene_hits[1] \n",
    "                break\n",
    "        if hits:\n",
    "            ranking_genes.append((gene, score, hits, phenospecificity_score))\n",
    "\n",
    "    for id in xrange(len(all_mapped_diseases_score)):\n",
    "        disease, score = all_mapped_diseases_score[id]\n",
    "        hits = None\n",
    "        for disease_hits in count_disease:\n",
    "            if disease_hits[0] == disease:\n",
    "                hits = disease_hits[1] \n",
    "                break\n",
    "        if hits:\n",
    "            ranking_diseases.append((disease, score, hits))\n",
    "\n",
    "    return ranking_genes, ranking_diseases, gene_associated_phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "reload(sys)\n",
    "sys.setdefaultencoding(\"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import urllib2\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from langdetect import detect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "input_phenotype = '甲状腺'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "language = detect(unicode(input_phenotype))\n",
    "if language == \"zh-cn\" or language == \"ko\":\n",
    "    site = \"https://translate.googleapis.com/translate_a/single?client=gtx&sl=zh-Hans&tl=en&dt=t&q=\"+input_phenotype\n",
    "    hdr = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',\n",
    "            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',\n",
    "            'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.3',\n",
    "            'Accept-Encoding': 'none',\n",
    "            'Accept-Language': 'en-US,en;q=0.8',\n",
    "            'Connection': 'keep-alive'}\n",
    "    req = urllib2.Request(site, headers=hdr)\n",
    "    try:\n",
    "        page = urllib2.urlopen(req)\n",
    "        content = page.read()\n",
    "        input_phenotype = json.loads(content)[0][0][0]\n",
    "    except urllib2.HTTPError, e:\n",
    "        print e\n",
    "        \n",
    "\n",
    "# match_result = map2hpoWithPhenoSynonyms(input_phenotype)\n",
    "# match_result = sorted(match_result, key = lambda x: x[2], reverse = True)\n",
    "# if match_result==[]:\n",
    "#     match_result = ''\n",
    "# elif match_result[0][0] == 'Familial  hyperprolactinemia':\n",
    "#     match_result = ''\n",
    "# else:\n",
    "#     if match_result[0][2] == 1.0:\n",
    "#         match_result = match_result[:1]\n",
    "#     match_id = [i[1] for i in match_result]\n",
    "#     for indx,i in enumerate(match_id):\n",
    "#         if i[-7:]=='synonym':\n",
    "#             match_id[indx] = match_id[indx][:-8]\n",
    "#     match_table = chpo[chpo['编号'].isin(match_id)].iloc[:,[3,2,1,0,5]].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'thyroid'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "match_result = map2hpoWithPhenoSynonyms(input_phenotype)\n",
    "match_result = sorted(match_result, key = lambda x: x[2], reverse = True)\n",
    "if match_result==[]:\n",
    "    match_result = ''\n",
    "elif match_result[0][0] == 'Familial  hyperprolactinemia':\n",
    "    match_result = ''\n",
    "else:\n",
    "    if match_result[0][2] == 1.0:\n",
    "        match_result = match_result[:1]\n",
    "    match_id = [i[1] for i in match_result]\n",
    "    for indx,i in enumerate(match_id):\n",
    "        if i[-7:]=='synonym':\n",
    "            match_id[indx] = match_id[indx][:-8]\n",
    "    match_table = chpo[chpo['编号'].isin(match_id)].iloc[:,[3,2,1,0,5]].reset_index(drop=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>表型中文名</th>\n",
       "      <th>表型英文名</th>\n",
       "      <th>HPO编号</th>\n",
       "      <th>类别</th>\n",
       "      <th>释义</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>心动过速</td>\n",
       "      <td>Tachycardia</td>\n",
       "      <td>HP:0001649</td>\n",
       "      <td>心血管系统的异常</td>\n",
       "      <td>心脏运动速率极快，通常在100次每秒以上。</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  表型中文名        表型英文名       HPO编号        类别                     释义\n",
       "0  心动过速  Tachycardia  HP:0001649  心血管系统的异常  心脏运动速率极快，通常在100次每秒以上。"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import google"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from google import google"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "num_page = 1\n",
    "search_results = google.search(\"afraid of light\", 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'Photophobia - Wikipedia'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_results[1].name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'Photophobia', u'Heliophobia']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set([i.name[:-12] for i in search_results if i.name[-9:]=='Wikipedia']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
