{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from main import master_function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_file_path = 'input/input_genes.txt'\n",
    "phenotype_file_path = 'input/input_phenotype.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final_res, df_genes, phenos = master_function(phenotype_file_path,gene_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>id</th>\n",
       "      <th>final_score</th>\n",
       "      <th>pathogenicity_score</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>hit_criteria</th>\n",
       "      <th>hpo_hit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CBS</td>\n",
       "      <td>c.833T&gt;C</td>\n",
       "      <td>chr21:g.44483184A&gt;G</td>\n",
       "      <td>7.783641</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>PS1|PM5|PP3|PP5</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXOSC8</td>\n",
       "      <td>c.815G&gt;C</td>\n",
       "      <td>chr13:g.37583420G&gt;C</td>\n",
       "      <td>7.708397</td>\n",
       "      <td>5.125000</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>PS1|PM5|PM1|PP3|PP2|PP5</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEGF8</td>\n",
       "      <td>c.8180G&gt;A</td>\n",
       "      <td>chr19:g.42880569G&gt;A</td>\n",
       "      <td>5.527655</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>10.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CLCNKA</td>\n",
       "      <td>c.910C&gt;T</td>\n",
       "      <td>chr1:g.16354556C&gt;T</td>\n",
       "      <td>5.132180</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PVS1|PP3</td>\n",
       "      <td>3.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGAP3</td>\n",
       "      <td>c.67G&gt;A</td>\n",
       "      <td>chr17:g.37844201C&gt;T</td>\n",
       "      <td>5.083242</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>7.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>c.344C&gt;T</td>\n",
       "      <td>chr18:g.57136761G&gt;A</td>\n",
       "      <td>4.154201</td>\n",
       "      <td>1.875000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>8.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GNAS</td>\n",
       "      <td>c.*293+1G&gt;A</td>\n",
       "      <td>chr20:g.57478847G&gt;A</td>\n",
       "      <td>4.062075</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>c.3611C&gt;T</td>\n",
       "      <td>chrX:g.153588552G&gt;A</td>\n",
       "      <td>3.922368</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>16.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RTTN</td>\n",
       "      <td>c.4275G&gt;C</td>\n",
       "      <td>chr18:g.67755252C&gt;G</td>\n",
       "      <td>3.420056</td>\n",
       "      <td>2.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WNT7A</td>\n",
       "      <td>c.299-4G&gt;A</td>\n",
       "      <td>chr3:g.13896304C&gt;T</td>\n",
       "      <td>3.376938</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM2|PP3</td>\n",
       "      <td>8.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CAMKMT</td>\n",
       "      <td>c.493G&gt;T</td>\n",
       "      <td>chr2:g.44934538G&gt;T</td>\n",
       "      <td>3.056788</td>\n",
       "      <td>1.300000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>9.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACACA</td>\n",
       "      <td>c.3496A&gt;G</td>\n",
       "      <td>chr17:g.35564641T&gt;C</td>\n",
       "      <td>3.031660</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WWOX</td>\n",
       "      <td>c.990C&gt;G</td>\n",
       "      <td>chr16:g.78466583C&gt;G</td>\n",
       "      <td>2.939219</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>9.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NPHP4</td>\n",
       "      <td>c.2246T&gt;C</td>\n",
       "      <td>chr1:g.5950986A&gt;G</td>\n",
       "      <td>2.728738</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>5.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WAS</td>\n",
       "      <td>c.995T&gt;C</td>\n",
       "      <td>chrX:g.48547112T&gt;C</td>\n",
       "      <td>2.490538</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>6.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ARMC5</td>\n",
       "      <td>c.2303A&gt;G</td>\n",
       "      <td>chr16:g.31477609A&gt;G</td>\n",
       "      <td>2.273745</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CACNA1D</td>\n",
       "      <td>c.3881C&gt;T</td>\n",
       "      <td>chr3:g.53796059C&gt;T</td>\n",
       "      <td>2.068106</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FANCL</td>\n",
       "      <td>c.979C&gt;G</td>\n",
       "      <td>chr2:g.58388713G&gt;C</td>\n",
       "      <td>2.031038</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>c.1402C&gt;T</td>\n",
       "      <td>chr13:g.110845240G&gt;A</td>\n",
       "      <td>1.880097</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>UBAC2</td>\n",
       "      <td>c.332T&gt;C</td>\n",
       "      <td>chr13:g.99896821T&gt;C</td>\n",
       "      <td>1.810618</td>\n",
       "      <td>1.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H6PD</td>\n",
       "      <td>c.1342G&gt;A</td>\n",
       "      <td>chr1:g.9323894G&gt;A</td>\n",
       "      <td>1.647918</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GP1BA</td>\n",
       "      <td>c.1763T&gt;C</td>\n",
       "      <td>chr17:g.4837662T&gt;C</td>\n",
       "      <td>1.559581</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MPO</td>\n",
       "      <td>c.1348G&gt;A</td>\n",
       "      <td>chr17:g.56352920C&gt;T</td>\n",
       "      <td>1.373265</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OSTM1</td>\n",
       "      <td>c.238C&gt;A</td>\n",
       "      <td>chr6:g.108395618G&gt;T</td>\n",
       "      <td>1.213008</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PM2|PP3|BP1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GSN</td>\n",
       "      <td>c.1663G&gt;A</td>\n",
       "      <td>chr9:g.124088928G&gt;A</td>\n",
       "      <td>1.213008</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|PP2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TMC8</td>\n",
       "      <td>c.913C&gt;T</td>\n",
       "      <td>chr17:g.76130571C&gt;T</td>\n",
       "      <td>1.030827</td>\n",
       "      <td>1.125000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LIPE</td>\n",
       "      <td>c.1210C&gt;T</td>\n",
       "      <td>chr19:g.42914668G&gt;A</td>\n",
       "      <td>0.735622</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|PP3|BP1</td>\n",
       "      <td>1.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SHROOM4</td>\n",
       "      <td>c.3364A&gt;C</td>\n",
       "      <td>chrX:g.50350778T&gt;G</td>\n",
       "      <td>0.649641</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BP4</td>\n",
       "      <td>2.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TP53</td>\n",
       "      <td>c.558T&gt;A</td>\n",
       "      <td>chr17:g.7578372A&gt;T</td>\n",
       "      <td>0.554518</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SALL2</td>\n",
       "      <td>c.2718G&gt;T</td>\n",
       "      <td>chr14:g.21991144C&gt;A</td>\n",
       "      <td>0.405465</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BP1|BP4</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PDSS1</td>\n",
       "      <td>c.1055G&gt;A</td>\n",
       "      <td>chr10:g.27031454G&gt;A</td>\n",
       "      <td>0.376019</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3</td>\n",
       "      <td>3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MYO7A</td>\n",
       "      <td>c.687C&gt;T</td>\n",
       "      <td>chr11:g.76868002C&gt;T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM2|BP7|BP6</td>\n",
       "      <td>2.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FANCM</td>\n",
       "      <td>c.2716A&gt;T</td>\n",
       "      <td>chr14:g.45644673A&gt;T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BP1|BP4</td>\n",
       "      <td>13.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NRXN1</td>\n",
       "      <td>c.2242C&gt;A</td>\n",
       "      <td>chr2:g.50765412G&gt;T</td>\n",
       "      <td>-0.101366</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3|BP1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>c.5438G&gt;A</td>\n",
       "      <td>chr8:g.144999070C&gt;T</td>\n",
       "      <td>-0.287823</td>\n",
       "      <td>-0.125000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|PM2|BS2|PP3|BP1</td>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ABCA4</td>\n",
       "      <td>c.5693G&gt;A</td>\n",
       "      <td>chr1:g.94476377C&gt;T</td>\n",
       "      <td>-0.410557</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3|PP5</td>\n",
       "      <td>4.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DYNC1H1</td>\n",
       "      <td>c.13808G&gt;T</td>\n",
       "      <td>chr14:g.102516531G&gt;T</td>\n",
       "      <td>-0.627741</td>\n",
       "      <td>-0.375000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PP3|BP6</td>\n",
       "      <td>4.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MYH2</td>\n",
       "      <td>c.3884G&gt;A</td>\n",
       "      <td>chr17:g.10430361C&gt;T</td>\n",
       "      <td>-0.687218</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>PM1|BS2|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NOD2</td>\n",
       "      <td>c.2377G&gt;A</td>\n",
       "      <td>chr16:g.50746199G&gt;A</td>\n",
       "      <td>-1.145363</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NOD2</td>\n",
       "      <td>c.1292C&gt;T</td>\n",
       "      <td>chr16:g.50745114C&gt;T</td>\n",
       "      <td>-1.145363</td>\n",
       "      <td>-1.250000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MMP1</td>\n",
       "      <td>c.105+2T&gt;C</td>\n",
       "      <td>chr11:g.102668717A&gt;G</td>\n",
       "      <td>-1.374436</td>\n",
       "      <td>-1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2|PP3</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GJC2</td>\n",
       "      <td>c.1193C&gt;T</td>\n",
       "      <td>chr1:g.228346652C&gt;T</td>\n",
       "      <td>-1.917842</td>\n",
       "      <td>-1.125000</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>PM1|BS2|PP3|BP6</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HFE</td>\n",
       "      <td>c.845G&gt;A</td>\n",
       "      <td>chr6:g.26093141G&gt;A</td>\n",
       "      <td>-2.079442</td>\n",
       "      <td>-1.500000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS1|PP3</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ATP2A2</td>\n",
       "      <td>c.464-70C&gt;A</td>\n",
       "      <td>chr12:g.110760727C&gt;A</td>\n",
       "      <td>-2.772589</td>\n",
       "      <td>-2.000000</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>BS2</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CDH23</td>\n",
       "      <td>c.3293A&gt;G</td>\n",
       "      <td>chr10:g.73472494A&gt;G</td>\n",
       "      <td>-3.003638</td>\n",
       "      <td>-2.166667</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>BS2|PP3|BP6</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene      variant                    id  final_score  \\\n",
       "0       CBS     c.833T>C   chr21:g.44483184A>G     7.783641   \n",
       "1    EXOSC8     c.815G>C   chr13:g.37583420G>C     7.708397   \n",
       "2     MEGF8    c.8180G>A   chr19:g.42880569G>A     5.527655   \n",
       "3    CLCNKA     c.910C>T    chr1:g.16354556C>T     5.132180   \n",
       "4     PGAP3      c.67G>A   chr17:g.37844201C>T     5.083242   \n",
       "5     CCBE1     c.344C>T   chr18:g.57136761G>A     4.154201   \n",
       "6      GNAS  c.*293+1G>A   chr20:g.57478847G>A     4.062075   \n",
       "7      FLNA    c.3611C>T   chrX:g.153588552G>A     3.922368   \n",
       "8      RTTN    c.4275G>C   chr18:g.67755252C>G     3.420056   \n",
       "9     WNT7A   c.299-4G>A    chr3:g.13896304C>T     3.376938   \n",
       "10   CAMKMT     c.493G>T    chr2:g.44934538G>T     3.056788   \n",
       "11    ACACA    c.3496A>G   chr17:g.35564641T>C     3.031660   \n",
       "12     WWOX     c.990C>G   chr16:g.78466583C>G     2.939219   \n",
       "13    NPHP4    c.2246T>C     chr1:g.5950986A>G     2.728738   \n",
       "14      WAS     c.995T>C    chrX:g.48547112T>C     2.490538   \n",
       "15    ARMC5    c.2303A>G   chr16:g.31477609A>G     2.273745   \n",
       "16  CACNA1D    c.3881C>T    chr3:g.53796059C>T     2.068106   \n",
       "17    FANCL     c.979C>G    chr2:g.58388713G>C     2.031038   \n",
       "18   COL4A1    c.1402C>T  chr13:g.110845240G>A     1.880097   \n",
       "19    UBAC2     c.332T>C   chr13:g.99896821T>C     1.810618   \n",
       "20     H6PD    c.1342G>A     chr1:g.9323894G>A     1.647918   \n",
       "21    GP1BA    c.1763T>C    chr17:g.4837662T>C     1.559581   \n",
       "22      MPO    c.1348G>A   chr17:g.56352920C>T     1.373265   \n",
       "23    OSTM1     c.238C>A   chr6:g.108395618G>T     1.213008   \n",
       "24      GSN    c.1663G>A   chr9:g.124088928G>A     1.213008   \n",
       "25     TMC8     c.913C>T   chr17:g.76130571C>T     1.030827   \n",
       "26     LIPE    c.1210C>T   chr19:g.42914668G>A     0.735622   \n",
       "27  SHROOM4    c.3364A>C    chrX:g.50350778T>G     0.649641   \n",
       "28     TP53     c.558T>A    chr17:g.7578372A>T     0.554518   \n",
       "29    SALL2    c.2718G>T   chr14:g.21991144C>A     0.405465   \n",
       "30    PDSS1    c.1055G>A   chr10:g.27031454G>A     0.376019   \n",
       "31    MYO7A     c.687C>T   chr11:g.76868002C>T     0.000000   \n",
       "32    FANCM    c.2716A>T   chr14:g.45644673A>T     0.000000   \n",
       "33    NRXN1    c.2242C>A    chr2:g.50765412G>T    -0.101366   \n",
       "34     PLEC    c.5438G>A   chr8:g.144999070C>T    -0.287823   \n",
       "35    ABCA4    c.5693G>A    chr1:g.94476377C>T    -0.410557   \n",
       "36  DYNC1H1   c.13808G>T  chr14:g.102516531G>T    -0.627741   \n",
       "37     MYH2    c.3884G>A   chr17:g.10430361C>T    -0.687218   \n",
       "38     NOD2    c.2377G>A   chr16:g.50746199G>A    -1.145363   \n",
       "39     NOD2    c.1292C>T   chr16:g.50745114C>T    -1.145363   \n",
       "40     MMP1   c.105+2T>C  chr11:g.102668717A>G    -1.374436   \n",
       "41     GJC2    c.1193C>T   chr1:g.228346652C>T    -1.917842   \n",
       "42      HFE     c.845G>A    chr6:g.26093141G>A    -2.079442   \n",
       "43   ATP2A2  c.464-70C>A  chr12:g.110760727C>A    -2.772589   \n",
       "44    CDH23    c.3293A>G   chr10:g.73472494A>G    -3.003638   \n",
       "\n",
       "    pathogenicity_score           pathogenicity             hit_criteria  \\\n",
       "0              4.000000       Likely pathogenic          PS1|PM5|PP3|PP5   \n",
       "1              5.125000              Pathogenic  PS1|PM5|PM1|PP3|PP2|PP5   \n",
       "2              2.250000  Uncertain significance              PM1|PM2|PP3   \n",
       "3              3.500000  Uncertain significance                 PVS1|PP3   \n",
       "4              2.333333  Uncertain significance              PM1|PM2|PP3   \n",
       "5              1.875000  Uncertain significance              PM1|PP3|PP2   \n",
       "6              1.500000  Uncertain significance                  PM2|PP3   \n",
       "7              1.375000  Uncertain significance                  PM2|PP3   \n",
       "8              2.125000  Uncertain significance              PM1|PM2|PP3   \n",
       "9              1.500000  Uncertain significance                  PM2|PP3   \n",
       "10             1.300000  Uncertain significance                  PM1|PP3   \n",
       "11             2.333333  Uncertain significance              PM1|PM2|PP3   \n",
       "12             1.250000  Uncertain significance                  PM1|PP3   \n",
       "13             1.500000  Uncertain significance                  PM1|PP3   \n",
       "14             1.250000  Uncertain significance                  PM1|PP3   \n",
       "15             1.750000  Uncertain significance          PM1|PM2|PP3|BP1   \n",
       "16             1.375000  Uncertain significance                  PM1|PP3   \n",
       "17             0.750000  Uncertain significance              PM1|PP3|BP1   \n",
       "18             1.250000  Uncertain significance                  PM1|PP3   \n",
       "19             1.125000  Uncertain significance                  PM1|PP3   \n",
       "20             1.500000  Uncertain significance                  PM1|PP3   \n",
       "21             2.250000  Uncertain significance              PM1|PM2|PP3   \n",
       "22             1.250000  Uncertain significance                  PM1|PP3   \n",
       "23             1.750000  Uncertain significance          PM1|PM2|PP3|BP1   \n",
       "24             1.750000  Uncertain significance              PM1|PP3|PP2   \n",
       "25             1.125000  Uncertain significance                  PM1|PP3   \n",
       "26             0.750000  Uncertain significance              PM1|PP3|BP1   \n",
       "27             0.500000  Uncertain significance                  PM1|BP4   \n",
       "28             0.400000  Uncertain significance                      PP3   \n",
       "29             1.000000           Likely benign          PM1|PM2|BP1|BP4   \n",
       "30             0.250000  Uncertain significance                      PP3   \n",
       "31             0.000000           Likely benign              PM2|BP7|BP6   \n",
       "32             0.000000           Likely benign              PM1|BP1|BP4   \n",
       "33            -0.250000  Uncertain significance                  PP3|BP1   \n",
       "34            -0.125000           Likely benign      PM1|PM2|BS2|PP3|BP1   \n",
       "35            -0.250000  Uncertain significance          PM1|BS2|PP3|PP5   \n",
       "36            -0.375000  Uncertain significance                  PP3|BP6   \n",
       "37            -0.750000  Uncertain significance              PM1|BS2|PP3   \n",
       "38            -1.250000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "39            -1.250000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "40            -1.500000  Uncertain significance                  BS2|PP3   \n",
       "41            -1.125000           Likely benign          PM1|BS2|PP3|BP6   \n",
       "42            -1.500000  Uncertain significance                  BS1|PP3   \n",
       "43            -2.000000  Uncertain significance                      BS2   \n",
       "44            -2.166667           Likely benign              BS2|PP3|BP6   \n",
       "\n",
       "    hpo_hit_score  \n",
       "0        6.000000  \n",
       "1        3.500000  \n",
       "2       10.666667  \n",
       "3        3.333333  \n",
       "4        7.833333  \n",
       "5        8.166667  \n",
       "6       14.000000  \n",
       "7       16.333333  \n",
       "8        4.000000  \n",
       "9        8.500000  \n",
       "10       9.500000  \n",
       "11       2.666667  \n",
       "12       9.500000  \n",
       "13       5.166667  \n",
       "14       6.333333  \n",
       "15       2.666667  \n",
       "16       3.500000  \n",
       "17      14.000000  \n",
       "18       3.500000  \n",
       "19       4.000000  \n",
       "20       2.000000  \n",
       "21       1.000000  \n",
       "22       2.000000  \n",
       "23       1.000000  \n",
       "24       1.000000  \n",
       "25       1.500000  \n",
       "26       1.666667  \n",
       "27       2.666667  \n",
       "28       3.000000  \n",
       "29       0.500000  \n",
       "30       3.500000  \n",
       "31       2.500000  \n",
       "32      13.000000  \n",
       "33       0.500000  \n",
       "34       9.000000  \n",
       "35       4.166667  \n",
       "36       4.333333  \n",
       "37       1.500000  \n",
       "38       1.500000  \n",
       "39       1.500000  \n",
       "40       1.500000  \n",
       "41       4.500000  \n",
       "42       3.000000  \n",
       "43       3.000000  \n",
       "44       3.000000  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open('result/pubmed_query_results.csv', 'rb') as f:\n",
    "    x_raw, y_test, genes, variants, titles, texts, impactfactors, years, journals = [], [], [], [], [], [], [], [], []\n",
    "    f.readline()\n",
    "    for line in f.readlines():\n",
    "        text = line.rstrip()\n",
    "        parts = text.split('\\t')\n",
    "        title, text, gene, variant, protein, impactfactor, year, journal = parts[3], parts[7], parts[0], parts[1], parts[2], parts[6], parts[5], parts[4]\n",
    "#         x_raw.append(dataloader.extractSentencesFromText(title, text, gene, variant, protein))\n",
    "#         y_test.append(1)\n",
    "#         genes.append(gene)\n",
    "#         variants.append(variant)\n",
    "#         titles.append(title)\n",
    "#         texts.append(text)\n",
    "#         impactfactors.append(impactfactor)\n",
    "#         years.append(year)\n",
    "#         journals.append(journal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Hereditary hemochromatosis masquerading as rheumatoid arthritis.'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Early erroneous diagnosis of rheumatic disease is common in subjects with arthropathy due to hereditary hemochromatosis. A 71-year-old male with chronic obstructive pulmonary disease and monoclonal gammopathy underwent hip replacement and was referred to our Department because of altered liver function tests. Test results were negative for hepatitis B surface antigen and hepatitis C virus, and positive for rheumatoid factor. A diagnosis of rheumatoid arthritis had been made on the basis of compatible joint involvement and laboratory data and steroid treatment prescribed. Since his serum ferritin was 3249 ng/mL, genetic testing for hereditary hemochromatosis was carried out and revealed homozygosity for Cys282Tyr mutation in the HFE gene. Liver biopsy disclosed cirrhosis compatible with hemochromatosis. Following a review of the patients' radiographs, the diagnosis of hemochromatosis arthropathy was made. Phlebotomies and family screening for hereditary hemochromatosis were done. The most logical explanation for the positive rheumatoid factor result in this subject are his age and the presence of two chronic diseases involving long-standing antigenic stimulation and monoclonal gammopathy of uncertain significance. It is important to distinguish rheumatoid arthritis from hemochromatosis arthropathy for several reasons: patients with hereditary hemochromatosis do not require corticosteroid treatment; in case of erroneous diagnosis of rheumatoid arthritis, phlebotomy is not started early, and familial genetic counseling is not considered. In male subjects with positive rheumatoid factor and joint and liver disease, hereditary hemochromatosis should be considered. More liberal use of genetic testing is justified in such cases.\""
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HFE'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'c.845G>A'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'p.Cys282Tyr'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.5'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "impactfactor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "year"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Annali italiani di medicina interna : organo ufficiale della Societa italiana di medicina interna'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "journal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '__file__' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-26-fbe53c644428>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mBASE\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdirname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabspath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name '__file__' is not defined"
     ]
    }
   ],
   "source": [
    "import os.path\n",
    "BASE = os.path.dirname(os.path.abspath(__file__))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eval_path = 'eval.py'\n",
    "pubmed_query_results_path = 'result/pubmed_query_results.csv'\n",
    "prediction_filters345_path = 'result/prediction_filters345.csv'\n",
    "prediction_filters246_path =  'result/prediction_filters246.csv'\n",
    "model345_3000_path = 'data/model/sentclass_filters345/model-3000'\n",
    "model246_3000_path = 'data/model/sentclass_filters246/model-3000'\n",
    "vocab345_path = 'data/model/sentclass_filters345/vocab'\n",
    "vocab246_path = 'data/model/sentclass_filters246/vocab'\n",
    "\n",
    "os.system('python %s --eval_data_file %s --outfile %s --checkpoint_file %s --vocab_file %s' % (eval_path, pubmed_query_results_path,prediction_filters345_path,model345_3000_path,vocab345_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('pubmed_data_variantnotnull.txt', sep='\\t', header=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "(gene varchar(20), protein_variant varchar(30), pmid varchar(15), title tinytext, journal tinytext, year char(4), month char(3), impact_factor float(4,2), abstract text, variant_id varchar(30), protein_domain varchar(500));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df.columns = [\n",
    "    'gene',\n",
    "    'protein_variant',\n",
    "    'pmid',\n",
    "    'title',\n",
    "    'journal',\n",
    "    'year',\n",
    "    'month',\n",
    "    'impact_factor',\n",
    "    'abstract',\n",
    "    'variant_id',\n",
    "    'protein_domain'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>protein_variant</th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>journal</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>impact_factor</th>\n",
       "      <th>abstract</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>protein_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>p.Leu153Phe</td>\n",
       "      <td>22040064</td>\n",
       "      <td>Frequencies of SLC44A2 alleles encoding human ...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.042</td>\n",
       "      <td>The human neutrophil antigen-3 (HNA-3) epitope...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLC44A2</td>\n",
       "      <td>p.Arg154Gln</td>\n",
       "      <td>22040064</td>\n",
       "      <td>Frequencies of SLC44A2 alleles encoding human ...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.042</td>\n",
       "      <td>The human neutrophil antigen-3 (HNA-3) epitope...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APOBEC-1</td>\n",
       "      <td>p.Glu181Gln</td>\n",
       "      <td>22326345</td>\n",
       "      <td>Hypermutation of ApoB mRNA by rat APOBEC-1 ove...</td>\n",
       "      <td>Journal of molecular biology</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>4.517</td>\n",
       "      <td>APOBEC-3 proteins induce C-to-U hypermutations...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP21A2</td>\n",
       "      <td>p.Val281Leu</td>\n",
       "      <td>22886582</td>\n",
       "      <td>Complement component 4 copy number variation a...</td>\n",
       "      <td>Human genetics</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>Dec</td>\n",
       "      <td>5.138</td>\n",
       "      <td>Congenital adrenal hyperplasia (CAH) due to 21...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD59</td>\n",
       "      <td>c.146delA</td>\n",
       "      <td>24383981</td>\n",
       "      <td>A new blood group antigen is defined by anti-C...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Jul</td>\n",
       "      <td>3.042</td>\n",
       "      <td>CD59 is a cell surface glycoprotein of approxi...</td>\n",
       "      <td>chr11:g.33738939del</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>c.1096C&gt;T</td>\n",
       "      <td>25814141</td>\n",
       "      <td>Caspase-8 Deficiency Presenting as Late-Onset ...</td>\n",
       "      <td>Journal of clinical immunology</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.094</td>\n",
       "      <td>Caspase-8 deficiency (CED) was originally desc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C4A</td>\n",
       "      <td>p.Cys4Ala</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>C4B</td>\n",
       "      <td>p.Cys4Asx</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C4A</td>\n",
       "      <td>p.Cys4Asx</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C4B</td>\n",
       "      <td>p.Cys4Ala</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>T</td>\n",
       "      <td>p.Asp0Asx</td>\n",
       "      <td>26526993</td>\n",
       "      <td>Human Obesity Associated with an Intronic SNP ...</td>\n",
       "      <td>Cell reports</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>Nov</td>\n",
       "      <td>7.870</td>\n",
       "      <td>Brain-derived neurotrophic factor (BDNF) plays...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>p.Val600Glu</td>\n",
       "      <td>26565903</td>\n",
       "      <td>AKT1 Activation Promotes Development of Melano...</td>\n",
       "      <td>Cell reports</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>Nov</td>\n",
       "      <td>7.870</td>\n",
       "      <td>Metastases are the major cause of melanoma-rel...</td>\n",
       "      <td>chr7:g.140453136A&gt;T</td>\n",
       "      <td>Protein kinase domain|Protein kinase-like doma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T</td>\n",
       "      <td>p.Thr1Asp</td>\n",
       "      <td>26578518</td>\n",
       "      <td>Antibody Response to Serpin B13 Induces Adapti...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Type 1 diabetes mellitus (T1D) is characterize...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>JAK2</td>\n",
       "      <td>p.Val617Phe</td>\n",
       "      <td>26650379</td>\n",
       "      <td>Polycythemia is associated with bone loss and ...</td>\n",
       "      <td>Osteoporosis international : a journal establi...</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>3.445</td>\n",
       "      <td>Increased fragility has been described in huma...</td>\n",
       "      <td>chr9:g.5073770G&gt;T</td>\n",
       "      <td>Protein kinase domain|Protein kinase-like doma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>p.Cys677Thr</td>\n",
       "      <td>26681493</td>\n",
       "      <td>Association between MTHFR C677T polymorphism a...</td>\n",
       "      <td>Psychology, health &amp; medicine</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>09</td>\n",
       "      <td>1.347</td>\n",
       "      <td>Depression is a worldwide public health issue,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AIDA</td>\n",
       "      <td>p.Thr1Asp</td>\n",
       "      <td>26692253</td>\n",
       "      <td>Interleukin-1 antagonism moderates the inflamm...</td>\n",
       "      <td>European journal of immunology</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>4.179</td>\n",
       "      <td>It was hypothesized that IL-1 antagonism would...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IL-10</td>\n",
       "      <td>p.Ser9Ala</td>\n",
       "      <td>26708990</td>\n",
       "      <td>Syk negatively regulates TLR4-mediated IFNβ an...</td>\n",
       "      <td>Biochimica et biophysica acta</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>0.5</td>\n",
       "      <td>While Syk has been shown to associate with TLR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TLR4</td>\n",
       "      <td>p.Cys3His</td>\n",
       "      <td>26771084</td>\n",
       "      <td>Cell-Permeable Peptide Blocks TLR4 Signaling a...</td>\n",
       "      <td>Cell transplantation</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.427</td>\n",
       "      <td>Toll-like receptor 4 (TLR4) activation in panc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EPHA2</td>\n",
       "      <td>p.Asp1235Lys</td>\n",
       "      <td>26910424</td>\n",
       "      <td>Dissociation of a Dynamic Protein Complex Stud...</td>\n",
       "      <td>Biophysical journal</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.632</td>\n",
       "      <td>The process of protein complex dissociation re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EPHA2</td>\n",
       "      <td>p.Arg957Asp</td>\n",
       "      <td>26910424</td>\n",
       "      <td>Dissociation of a Dynamic Protein Complex Stud...</td>\n",
       "      <td>Biophysical journal</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.632</td>\n",
       "      <td>The process of protein complex dissociation re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>EPHA2</td>\n",
       "      <td>p.Lys956Asp</td>\n",
       "      <td>26910424</td>\n",
       "      <td>Dissociation of a Dynamic Protein Complex Stud...</td>\n",
       "      <td>Biophysical journal</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.632</td>\n",
       "      <td>The process of protein complex dissociation re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EPHA2</td>\n",
       "      <td>p.Asp1223Arg</td>\n",
       "      <td>26910424</td>\n",
       "      <td>Dissociation of a Dynamic Protein Complex Stud...</td>\n",
       "      <td>Biophysical journal</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.632</td>\n",
       "      <td>The process of protein complex dissociation re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>p.Glu4242Gly</td>\n",
       "      <td>26910427</td>\n",
       "      <td>Distinct Components of Retrograde Ca(V)1.1-RyR...</td>\n",
       "      <td>Biophysical journal</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.632</td>\n",
       "      <td>The molecular basis for excitation-contraction...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>APOBEC3F</td>\n",
       "      <td>p.Ile231Val</td>\n",
       "      <td>26942578</td>\n",
       "      <td>Role of APOBEC3F Gene Variation in HIV-1 Disea...</td>\n",
       "      <td>PLoS genetics</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>6.661</td>\n",
       "      <td>Human APOBEC3 cytidine deaminases are intrinsi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>APOBEC3F</td>\n",
       "      <td>p.Ala3Gly</td>\n",
       "      <td>26942578</td>\n",
       "      <td>Role of APOBEC3F Gene Variation in HIV-1 Disea...</td>\n",
       "      <td>PLoS genetics</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>6.661</td>\n",
       "      <td>Human APOBEC3 cytidine deaminases are intrinsi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>APOBEC3F</td>\n",
       "      <td>p.Ala3Phe</td>\n",
       "      <td>26942578</td>\n",
       "      <td>Role of APOBEC3F Gene Variation in HIV-1 Disea...</td>\n",
       "      <td>PLoS genetics</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>6.661</td>\n",
       "      <td>Human APOBEC3 cytidine deaminases are intrinsi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SDS</td>\n",
       "      <td>p.Glu294Arg</td>\n",
       "      <td>26956237</td>\n",
       "      <td>Expression, characterization and mutagenesis o...</td>\n",
       "      <td>Biotechnology letters</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Jul</td>\n",
       "      <td>1.639</td>\n",
       "      <td>To search for a novel glutamate decarboxylase ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SDS</td>\n",
       "      <td>p.His467Ala</td>\n",
       "      <td>26956237</td>\n",
       "      <td>Expression, characterization and mutagenesis o...</td>\n",
       "      <td>Biotechnology letters</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Jul</td>\n",
       "      <td>1.639</td>\n",
       "      <td>To search for a novel glutamate decarboxylase ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SIK2</td>\n",
       "      <td>p.Ser89Ala</td>\n",
       "      <td>26983400</td>\n",
       "      <td>SIK2 regulates fasting-induced PPARα activity ...</td>\n",
       "      <td>Scientific reports</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>5.228</td>\n",
       "      <td>Fatty acid oxidation and subsequent ketogenesi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AGXT2</td>\n",
       "      <td>p.Val498Leu</td>\n",
       "      <td>26984639</td>\n",
       "      <td>Associations of functional alanine-glyoxylate ...</td>\n",
       "      <td>Scientific reports</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>5.228</td>\n",
       "      <td>Asymmetric and symmetric dimethylarginines (AD...</td>\n",
       "      <td>chr5:g.34998877C&gt;G</td>\n",
       "      <td>Pyridoxal phosphate-dependent transferase|Pyri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357012</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>p.Thr47Asp</td>\n",
       "      <td>21182832</td>\n",
       "      <td>Expression of estrogen and progesterone recept...</td>\n",
       "      <td>Chemico-biological interactions</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>May</td>\n",
       "      <td>2.618</td>\n",
       "      <td>Prolonged exposure to estrogens is a significa...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357013</th>\n",
       "      <td>OGG1</td>\n",
       "      <td>p.Lys751Gln</td>\n",
       "      <td>21183201</td>\n",
       "      <td>DNA repair gene polymorphisms at XRCC1, XRCC3,...</td>\n",
       "      <td>Chemosphere</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.698</td>\n",
       "      <td>Reduction in DNA repair capacity is associated...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357014</th>\n",
       "      <td>CAV2</td>\n",
       "      <td>p.Arg1349Gln</td>\n",
       "      <td>21183743</td>\n",
       "      <td>Recurrent stroke due to a novel voltage sensor...</td>\n",
       "      <td>Stroke</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>5.787</td>\n",
       "      <td>Familial hemiplegic migraine is characterized ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357015</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Asp530Asn</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357016</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Asp523Glu</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357017</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Asp523Asn</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357018</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Asp2049Asn</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357019</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Glu2072Gln</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357020</th>\n",
       "      <td>PKD2</td>\n",
       "      <td>p.Asp525Asn</td>\n",
       "      <td>21185261</td>\n",
       "      <td>The single pore residue Asp523 in PKD2L1 deter...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>2.371</td>\n",
       "      <td>The polycystic kidney disease 1-like 3 (PKD1L3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357021</th>\n",
       "      <td>KLF15</td>\n",
       "      <td>p.Thr2Asp</td>\n",
       "      <td>21187013</td>\n",
       "      <td>Bio-informatics analysis of a gene co-expressi...</td>\n",
       "      <td>BMC systems biology</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>Dec</td>\n",
       "      <td>2.213</td>\n",
       "      <td>Obesity causes insulin resistance in target ti...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357022</th>\n",
       "      <td>CA2</td>\n",
       "      <td>p.Glu412Gly</td>\n",
       "      <td>21187073</td>\n",
       "      <td>Characterization of sarcoplasmic reticulum Ca2...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>2.371</td>\n",
       "      <td>Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357023</th>\n",
       "      <td>CA2</td>\n",
       "      <td>p.Thr441Ala</td>\n",
       "      <td>21187073</td>\n",
       "      <td>Characterization of sarcoplasmic reticulum Ca2...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>2.371</td>\n",
       "      <td>Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357024</th>\n",
       "      <td>CA2</td>\n",
       "      <td>p.Cys561Ala</td>\n",
       "      <td>21187073</td>\n",
       "      <td>Characterization of sarcoplasmic reticulum Ca2...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>2.371</td>\n",
       "      <td>Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357025</th>\n",
       "      <td>CA2</td>\n",
       "      <td>p.Arg560Val</td>\n",
       "      <td>21187073</td>\n",
       "      <td>Characterization of sarcoplasmic reticulum Ca2...</td>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>2.371</td>\n",
       "      <td>Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357026</th>\n",
       "      <td>XRCC1</td>\n",
       "      <td>p.Leu751Gln</td>\n",
       "      <td>21188533</td>\n",
       "      <td>An association between DNA repair gene polymor...</td>\n",
       "      <td>Molecular biology reports</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Nov</td>\n",
       "      <td>1.698</td>\n",
       "      <td>DNA repair genetic polymorphisms have been stu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357027</th>\n",
       "      <td>XRCC1</td>\n",
       "      <td>p.Asp312Asn</td>\n",
       "      <td>21188533</td>\n",
       "      <td>An association between DNA repair gene polymor...</td>\n",
       "      <td>Molecular biology reports</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Nov</td>\n",
       "      <td>1.698</td>\n",
       "      <td>DNA repair genetic polymorphisms have been stu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357028</th>\n",
       "      <td>NF1</td>\n",
       "      <td>p.Val600Glu</td>\n",
       "      <td>21190184</td>\n",
       "      <td>Functional characterization of a BRAF insertio...</td>\n",
       "      <td>International journal of cancer</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Nov</td>\n",
       "      <td>5.531</td>\n",
       "      <td>Pilocytic astrocytoma (PA) is emerging as a tu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357029</th>\n",
       "      <td>CA3</td>\n",
       "      <td>p.Cys57Asx</td>\n",
       "      <td>21190219</td>\n",
       "      <td>Chronic corticosterone exposure alters postsyn...</td>\n",
       "      <td>Synapse (New York, N.Y.)</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Aug</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Animal models provide compelling evidence that...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357030</th>\n",
       "      <td>NTRK1</td>\n",
       "      <td>p.Val600Glu</td>\n",
       "      <td>21190444</td>\n",
       "      <td>Prevalence, clinicopathologic features, and so...</td>\n",
       "      <td>Thyroid : official journal of the American Thy...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>3.784</td>\n",
       "      <td>Although hereditary nonmedullary thyroid cance...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357031</th>\n",
       "      <td>SLITRK6</td>\n",
       "      <td>p.Asp2Arg</td>\n",
       "      <td>21192083</td>\n",
       "      <td>Region-specific gene expression in early postn...</td>\n",
       "      <td>The Journal of comparative neurology</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Previous studies in the developing mouse thala...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357032</th>\n",
       "      <td>ITGA2</td>\n",
       "      <td>p.Gly873Ala</td>\n",
       "      <td>21193198</td>\n",
       "      <td>Predictive role of polymorphisms in interleuki...</td>\n",
       "      <td>Atherosclerosis</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>3.942</td>\n",
       "      <td>Ischemic stroke is prevalent in type 2 diabete...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357033</th>\n",
       "      <td>IGF-1</td>\n",
       "      <td>p.Ser6Lys</td>\n",
       "      <td>21193562</td>\n",
       "      <td>Chronic heart failure reduces Akt phosphorylat...</td>\n",
       "      <td>Journal of applied physiology (Bethesda, Md. :...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Patients with chronic heart failure (HF) frequ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357034</th>\n",
       "      <td>CA3</td>\n",
       "      <td>p.Tyr2Arg</td>\n",
       "      <td>21195134</td>\n",
       "      <td>Vulnerability to nicotine abstinence-related s...</td>\n",
       "      <td>Neuroscience letters</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>2.107</td>\n",
       "      <td>An outbred rat model of the novelty-seeking ph...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357035</th>\n",
       "      <td>OGG1</td>\n",
       "      <td>p.Ile223Val</td>\n",
       "      <td>21195604</td>\n",
       "      <td>First evidence for digenic inheritance in here...</td>\n",
       "      <td>European journal of cancer (Oxford, England : ...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>May</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Biallelic mutations in the base excision repai...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357036</th>\n",
       "      <td>F7</td>\n",
       "      <td>p.Val158Met</td>\n",
       "      <td>21195697</td>\n",
       "      <td>The effects of catechol-O-methyl-transferase p...</td>\n",
       "      <td>Brain research</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>2.561</td>\n",
       "      <td>The catechol-O-methyl-transferase (COMT) gene ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357037</th>\n",
       "      <td>C3</td>\n",
       "      <td>p.Val158Met</td>\n",
       "      <td>21195697</td>\n",
       "      <td>The effects of catechol-O-methyl-transferase p...</td>\n",
       "      <td>Brain research</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Mar</td>\n",
       "      <td>2.561</td>\n",
       "      <td>The catechol-O-methyl-transferase (COMT) gene ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357038</th>\n",
       "      <td>CD27</td>\n",
       "      <td>p.Asx8Arg</td>\n",
       "      <td>21195803</td>\n",
       "      <td>Long term recall of memory CD8 T cells in mice...</td>\n",
       "      <td>Vaccine</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>3.413</td>\n",
       "      <td>Since long-term immunity is a critical compone...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357039</th>\n",
       "      <td>TMC8</td>\n",
       "      <td>p.Asn306Ile</td>\n",
       "      <td>21196704</td>\n",
       "      <td>Homozygosity for the c.917A→T (p.N306l) polymo...</td>\n",
       "      <td>Dermatology (Basel, Switzerland)</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Epidermodysplasia verruciformis Lewandowsky-Lu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357040</th>\n",
       "      <td>TMC8</td>\n",
       "      <td>p.Asn306l</td>\n",
       "      <td>21196704</td>\n",
       "      <td>Homozygosity for the c.917A→T (p.N306l) polymo...</td>\n",
       "      <td>Dermatology (Basel, Switzerland)</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>Feb</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Epidermodysplasia verruciformis Lewandowsky-Lu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357041</th>\n",
       "      <td>LOXL1</td>\n",
       "      <td>p.Tyr559Asp</td>\n",
       "      <td>21197115</td>\n",
       "      <td>Analysis of LOXL1 polymorphisms in a Saudi Ara...</td>\n",
       "      <td>Molecular vision</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>Dec</td>\n",
       "      <td>2.110</td>\n",
       "      <td>To investigate whether single nucleotide polym...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>357042 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene protein_variant      pmid  \\\n",
       "0              T     p.Leu153Phe  22040064   \n",
       "1        SLC44A2     p.Arg154Gln  22040064   \n",
       "2       APOBEC-1     p.Glu181Gln  22326345   \n",
       "3        CYP21A2     p.Val281Leu  22886582   \n",
       "4           CD59       c.146delA  24383981   \n",
       "5              T       c.1096C>T  25814141   \n",
       "6            C4A       p.Cys4Ala  26493816   \n",
       "7            C4B       p.Cys4Asx  26493816   \n",
       "8            C4A       p.Cys4Asx  26493816   \n",
       "9            C4B       p.Cys4Ala  26493816   \n",
       "10             T       p.Asp0Asx  26526993   \n",
       "11          BRAF     p.Val600Glu  26565903   \n",
       "12             T       p.Thr1Asp  26578518   \n",
       "13          JAK2     p.Val617Phe  26650379   \n",
       "14         MTHFR     p.Cys677Thr  26681493   \n",
       "15          AIDA       p.Thr1Asp  26692253   \n",
       "16         IL-10       p.Ser9Ala  26708990   \n",
       "17          TLR4       p.Cys3His  26771084   \n",
       "18         EPHA2    p.Asp1235Lys  26910424   \n",
       "19         EPHA2     p.Arg957Asp  26910424   \n",
       "20         EPHA2     p.Lys956Asp  26910424   \n",
       "21         EPHA2    p.Asp1223Arg  26910424   \n",
       "22          RYR1    p.Glu4242Gly  26910427   \n",
       "23      APOBEC3F     p.Ile231Val  26942578   \n",
       "24      APOBEC3F       p.Ala3Gly  26942578   \n",
       "25      APOBEC3F       p.Ala3Phe  26942578   \n",
       "26           SDS     p.Glu294Arg  26956237   \n",
       "27           SDS     p.His467Ala  26956237   \n",
       "28          SIK2      p.Ser89Ala  26983400   \n",
       "29         AGXT2     p.Val498Leu  26984639   \n",
       "...          ...             ...       ...   \n",
       "357012      ESR1      p.Thr47Asp  21182832   \n",
       "357013      OGG1     p.Lys751Gln  21183201   \n",
       "357014      CAV2    p.Arg1349Gln  21183743   \n",
       "357015      PKD2     p.Asp530Asn  21185261   \n",
       "357016      PKD2     p.Asp523Glu  21185261   \n",
       "357017      PKD2     p.Asp523Asn  21185261   \n",
       "357018      PKD2    p.Asp2049Asn  21185261   \n",
       "357019      PKD2    p.Glu2072Gln  21185261   \n",
       "357020      PKD2     p.Asp525Asn  21185261   \n",
       "357021     KLF15       p.Thr2Asp  21187013   \n",
       "357022       CA2     p.Glu412Gly  21187073   \n",
       "357023       CA2     p.Thr441Ala  21187073   \n",
       "357024       CA2     p.Cys561Ala  21187073   \n",
       "357025       CA2     p.Arg560Val  21187073   \n",
       "357026     XRCC1     p.Leu751Gln  21188533   \n",
       "357027     XRCC1     p.Asp312Asn  21188533   \n",
       "357028       NF1     p.Val600Glu  21190184   \n",
       "357029       CA3      p.Cys57Asx  21190219   \n",
       "357030     NTRK1     p.Val600Glu  21190444   \n",
       "357031   SLITRK6       p.Asp2Arg  21192083   \n",
       "357032     ITGA2     p.Gly873Ala  21193198   \n",
       "357033     IGF-1       p.Ser6Lys  21193562   \n",
       "357034       CA3       p.Tyr2Arg  21195134   \n",
       "357035      OGG1     p.Ile223Val  21195604   \n",
       "357036        F7     p.Val158Met  21195697   \n",
       "357037        C3     p.Val158Met  21195697   \n",
       "357038      CD27       p.Asx8Arg  21195803   \n",
       "357039      TMC8     p.Asn306Ile  21196704   \n",
       "357040      TMC8       p.Asn306l  21196704   \n",
       "357041     LOXL1     p.Tyr559Asp  21197115   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Frequencies of SLC44A2 alleles encoding human ...   \n",
       "1       Frequencies of SLC44A2 alleles encoding human ...   \n",
       "2       Hypermutation of ApoB mRNA by rat APOBEC-1 ove...   \n",
       "3       Complement component 4 copy number variation a...   \n",
       "4       A new blood group antigen is defined by anti-C...   \n",
       "5       Caspase-8 Deficiency Presenting as Late-Onset ...   \n",
       "6       Gene copy-number variations (CNVs) of compleme...   \n",
       "7       Gene copy-number variations (CNVs) of compleme...   \n",
       "8       Gene copy-number variations (CNVs) of compleme...   \n",
       "9       Gene copy-number variations (CNVs) of compleme...   \n",
       "10      Human Obesity Associated with an Intronic SNP ...   \n",
       "11      AKT1 Activation Promotes Development of Melano...   \n",
       "12      Antibody Response to Serpin B13 Induces Adapti...   \n",
       "13      Polycythemia is associated with bone loss and ...   \n",
       "14      Association between MTHFR C677T polymorphism a...   \n",
       "15      Interleukin-1 antagonism moderates the inflamm...   \n",
       "16      Syk negatively regulates TLR4-mediated IFNβ an...   \n",
       "17      Cell-Permeable Peptide Blocks TLR4 Signaling a...   \n",
       "18      Dissociation of a Dynamic Protein Complex Stud...   \n",
       "19      Dissociation of a Dynamic Protein Complex Stud...   \n",
       "20      Dissociation of a Dynamic Protein Complex Stud...   \n",
       "21      Dissociation of a Dynamic Protein Complex Stud...   \n",
       "22      Distinct Components of Retrograde Ca(V)1.1-RyR...   \n",
       "23      Role of APOBEC3F Gene Variation in HIV-1 Disea...   \n",
       "24      Role of APOBEC3F Gene Variation in HIV-1 Disea...   \n",
       "25      Role of APOBEC3F Gene Variation in HIV-1 Disea...   \n",
       "26      Expression, characterization and mutagenesis o...   \n",
       "27      Expression, characterization and mutagenesis o...   \n",
       "28      SIK2 regulates fasting-induced PPARα activity ...   \n",
       "29      Associations of functional alanine-glyoxylate ...   \n",
       "...                                                   ...   \n",
       "357012  Expression of estrogen and progesterone recept...   \n",
       "357013  DNA repair gene polymorphisms at XRCC1, XRCC3,...   \n",
       "357014  Recurrent stroke due to a novel voltage sensor...   \n",
       "357015  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357016  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357017  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357018  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357019  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357020  The single pore residue Asp523 in PKD2L1 deter...   \n",
       "357021  Bio-informatics analysis of a gene co-expressi...   \n",
       "357022  Characterization of sarcoplasmic reticulum Ca2...   \n",
       "357023  Characterization of sarcoplasmic reticulum Ca2...   \n",
       "357024  Characterization of sarcoplasmic reticulum Ca2...   \n",
       "357025  Characterization of sarcoplasmic reticulum Ca2...   \n",
       "357026  An association between DNA repair gene polymor...   \n",
       "357027  An association between DNA repair gene polymor...   \n",
       "357028  Functional characterization of a BRAF insertio...   \n",
       "357029  Chronic corticosterone exposure alters postsyn...   \n",
       "357030  Prevalence, clinicopathologic features, and so...   \n",
       "357031  Region-specific gene expression in early postn...   \n",
       "357032  Predictive role of polymorphisms in interleuki...   \n",
       "357033  Chronic heart failure reduces Akt phosphorylat...   \n",
       "357034  Vulnerability to nicotine abstinence-related s...   \n",
       "357035  First evidence for digenic inheritance in here...   \n",
       "357036  The effects of catechol-O-methyl-transferase p...   \n",
       "357037  The effects of catechol-O-methyl-transferase p...   \n",
       "357038  Long term recall of memory CD8 T cells in mice...   \n",
       "357039  Homozygosity for the c.917A→T (p.N306l) polymo...   \n",
       "357040  Homozygosity for the c.917A→T (p.N306l) polymo...   \n",
       "357041  Analysis of LOXL1 polymorphisms in a Saudi Ara...   \n",
       "\n",
       "                                                  journal    year month  \\\n",
       "0                                             Transfusion  2012.0   May   \n",
       "1                                             Transfusion  2012.0   May   \n",
       "2                            Journal of molecular biology  2012.0   Apr   \n",
       "3                                          Human genetics  2012.0   Dec   \n",
       "4                                             Transfusion  2014.0   Jul   \n",
       "5                          Journal of clinical immunology  2015.0   May   \n",
       "6                        Annals of the rheumatic diseases  2016.0   Sep   \n",
       "7                        Annals of the rheumatic diseases  2016.0   Sep   \n",
       "8                        Annals of the rheumatic diseases  2016.0   Sep   \n",
       "9                        Annals of the rheumatic diseases  2016.0   Sep   \n",
       "10                                           Cell reports  2015.0   Nov   \n",
       "11                                           Cell reports  2015.0   Nov   \n",
       "12                    The Journal of biological chemistry  2016.0   Jan   \n",
       "13      Osteoporosis international : a journal establi...  2016.0   Apr   \n",
       "14                          Psychology, health & medicine  2016.0    09   \n",
       "15                         European journal of immunology  2016.0   Apr   \n",
       "16                          Biochimica et biophysica acta  2016.0   Mar   \n",
       "17                                   Cell transplantation  2016.0   NaN   \n",
       "18                                    Biophysical journal  2016.0   Feb   \n",
       "19                                    Biophysical journal  2016.0   Feb   \n",
       "20                                    Biophysical journal  2016.0   Feb   \n",
       "21                                    Biophysical journal  2016.0   Feb   \n",
       "22                                    Biophysical journal  2016.0   Feb   \n",
       "23                                          PLoS genetics  2016.0   Mar   \n",
       "24                                          PLoS genetics  2016.0   Mar   \n",
       "25                                          PLoS genetics  2016.0   Mar   \n",
       "26                                  Biotechnology letters  2016.0   Jul   \n",
       "27                                  Biotechnology letters  2016.0   Jul   \n",
       "28                                     Scientific reports  2016.0   Mar   \n",
       "29                                     Scientific reports  2016.0   Mar   \n",
       "...                                                   ...     ...   ...   \n",
       "357012                    Chemico-biological interactions  2011.0   May   \n",
       "357013                                        Chemosphere  2011.0   Feb   \n",
       "357014                                             Stroke  2011.0   Feb   \n",
       "357015  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357016  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357017  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357018  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357019  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357020  Biochemical and biophysical research communica...  2011.0   Jan   \n",
       "357021                                BMC systems biology  2010.0   Dec   \n",
       "357022  Biochemical and biophysical research communica...  2011.0   Feb   \n",
       "357023  Biochemical and biophysical research communica...  2011.0   Feb   \n",
       "357024  Biochemical and biophysical research communica...  2011.0   Feb   \n",
       "357025  Biochemical and biophysical research communica...  2011.0   Feb   \n",
       "357026                          Molecular biology reports  2011.0   Nov   \n",
       "357027                          Molecular biology reports  2011.0   Nov   \n",
       "357028                    International journal of cancer  2011.0   Nov   \n",
       "357029                           Synapse (New York, N.Y.)  2011.0   Aug   \n",
       "357030  Thyroid : official journal of the American Thy...  2011.0   Apr   \n",
       "357031               The Journal of comparative neurology  2011.0   Feb   \n",
       "357032                                    Atherosclerosis  2011.0   Mar   \n",
       "357033  Journal of applied physiology (Bethesda, Md. :...  2011.0   Apr   \n",
       "357034                               Neuroscience letters  2011.0   Mar   \n",
       "357035  European journal of cancer (Oxford, England : ...  2011.0   May   \n",
       "357036                                     Brain research  2011.0   Mar   \n",
       "357037                                     Brain research  2011.0   Mar   \n",
       "357038                                            Vaccine  2011.0   Feb   \n",
       "357039                   Dermatology (Basel, Switzerland)  2011.0   Feb   \n",
       "357040                   Dermatology (Basel, Switzerland)  2011.0   Feb   \n",
       "357041                                   Molecular vision  2010.0   Dec   \n",
       "\n",
       "       impact_factor                                           abstract  \\\n",
       "0              3.042  The human neutrophil antigen-3 (HNA-3) epitope...   \n",
       "1              3.042  The human neutrophil antigen-3 (HNA-3) epitope...   \n",
       "2              4.517  APOBEC-3 proteins induce C-to-U hypermutations...   \n",
       "3              5.138  Congenital adrenal hyperplasia (CAH) due to 21...   \n",
       "4              3.042  CD59 is a cell surface glycoprotein of approxi...   \n",
       "5              3.094  Caspase-8 deficiency (CED) was originally desc...   \n",
       "6             12.384  Complement-mediated vasculopathy of muscle and...   \n",
       "7             12.384  Complement-mediated vasculopathy of muscle and...   \n",
       "8             12.384  Complement-mediated vasculopathy of muscle and...   \n",
       "9             12.384  Complement-mediated vasculopathy of muscle and...   \n",
       "10             7.870  Brain-derived neurotrophic factor (BDNF) plays...   \n",
       "11             7.870  Metastases are the major cause of melanoma-rel...   \n",
       "12               0.5  Type 1 diabetes mellitus (T1D) is characterize...   \n",
       "13             3.445  Increased fragility has been described in huma...   \n",
       "14             1.347  Depression is a worldwide public health issue,...   \n",
       "15             4.179  It was hypothesized that IL-1 antagonism would...   \n",
       "16               0.5  While Syk has been shown to associate with TLR...   \n",
       "17             3.427  Toll-like receptor 4 (TLR4) activation in panc...   \n",
       "18             3.632  The process of protein complex dissociation re...   \n",
       "19             3.632  The process of protein complex dissociation re...   \n",
       "20             3.632  The process of protein complex dissociation re...   \n",
       "21             3.632  The process of protein complex dissociation re...   \n",
       "22             3.632  The molecular basis for excitation-contraction...   \n",
       "23             6.661  Human APOBEC3 cytidine deaminases are intrinsi...   \n",
       "24             6.661  Human APOBEC3 cytidine deaminases are intrinsi...   \n",
       "25             6.661  Human APOBEC3 cytidine deaminases are intrinsi...   \n",
       "26             1.639  To search for a novel glutamate decarboxylase ...   \n",
       "27             1.639  To search for a novel glutamate decarboxylase ...   \n",
       "28             5.228  Fatty acid oxidation and subsequent ketogenesi...   \n",
       "29             5.228  Asymmetric and symmetric dimethylarginines (AD...   \n",
       "...              ...                                                ...   \n",
       "357012         2.618  Prolonged exposure to estrogens is a significa...   \n",
       "357013         3.698  Reduction in DNA repair capacity is associated...   \n",
       "357014         5.787  Familial hemiplegic migraine is characterized ...   \n",
       "357015         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357016         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357017         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357018         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357019         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357020         2.371  The polycystic kidney disease 1-like 3 (PKD1L3...   \n",
       "357021         2.213  Obesity causes insulin resistance in target ti...   \n",
       "357022         2.371  Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...   \n",
       "357023         2.371  Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...   \n",
       "357024         2.371  Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...   \n",
       "357025         2.371  Sarcoplasmic reticulum Ca(2+) ATPase (SERCA) i...   \n",
       "357026         1.698  DNA repair genetic polymorphisms have been stu...   \n",
       "357027         1.698  DNA repair genetic polymorphisms have been stu...   \n",
       "357028         5.531  Pilocytic astrocytoma (PA) is emerging as a tu...   \n",
       "357029           0.5  Animal models provide compelling evidence that...   \n",
       "357030         3.784  Although hereditary nonmedullary thyroid cance...   \n",
       "357031           0.5  Previous studies in the developing mouse thala...   \n",
       "357032         3.942  Ischemic stroke is prevalent in type 2 diabete...   \n",
       "357033           0.5  Patients with chronic heart failure (HF) frequ...   \n",
       "357034         2.107  An outbred rat model of the novelty-seeking ph...   \n",
       "357035           0.5  Biallelic mutations in the base excision repai...   \n",
       "357036         2.561  The catechol-O-methyl-transferase (COMT) gene ...   \n",
       "357037         2.561  The catechol-O-methyl-transferase (COMT) gene ...   \n",
       "357038         3.413  Since long-term immunity is a critical compone...   \n",
       "357039           0.5  Epidermodysplasia verruciformis Lewandowsky-Lu...   \n",
       "357040           0.5  Epidermodysplasia verruciformis Lewandowsky-Lu...   \n",
       "357041         2.110  To investigate whether single nucleotide polym...   \n",
       "\n",
       "                 variant_id                                     protein_domain  \n",
       "0                       NaN                                                NaN  \n",
       "1                       NaN                                                NaN  \n",
       "2                       NaN                                                NaN  \n",
       "3                       NaN                                                NaN  \n",
       "4       chr11:g.33738939del                                                NaN  \n",
       "5                       NaN                                                NaN  \n",
       "6                       NaN                                                NaN  \n",
       "7                       NaN                                                NaN  \n",
       "8                       NaN                                                NaN  \n",
       "9                       NaN                                                NaN  \n",
       "10                      NaN                                                NaN  \n",
       "11      chr7:g.140453136A>T  Protein kinase domain|Protein kinase-like doma...  \n",
       "12                      NaN                                                NaN  \n",
       "13        chr9:g.5073770G>T  Protein kinase domain|Protein kinase-like doma...  \n",
       "14                      NaN                                                NaN  \n",
       "15                      NaN                                                NaN  \n",
       "16                      NaN                                                NaN  \n",
       "17                      NaN                                                NaN  \n",
       "18                      NaN                                                NaN  \n",
       "19                      NaN                                                NaN  \n",
       "20                      NaN                                                NaN  \n",
       "21                      NaN                                                NaN  \n",
       "22                      NaN                                                NaN  \n",
       "23                      NaN                                                NaN  \n",
       "24                      NaN                                                NaN  \n",
       "25                      NaN                                                NaN  \n",
       "26                      NaN                                                NaN  \n",
       "27                      NaN                                                NaN  \n",
       "28                      NaN                                                NaN  \n",
       "29       chr5:g.34998877C>G  Pyridoxal phosphate-dependent transferase|Pyri...  \n",
       "...                     ...                                                ...  \n",
       "357012                  NaN                                                NaN  \n",
       "357013                  NaN                                                NaN  \n",
       "357014                  NaN                                                NaN  \n",
       "357015                  NaN                                                NaN  \n",
       "357016                  NaN                                                NaN  \n",
       "357017                  NaN                                                NaN  \n",
       "357018                  NaN                                                NaN  \n",
       "357019                  NaN                                                NaN  \n",
       "357020                  NaN                                                NaN  \n",
       "357021                  NaN                                                NaN  \n",
       "357022                  NaN                                                NaN  \n",
       "357023                  NaN                                                NaN  \n",
       "357024                  NaN                                                NaN  \n",
       "357025                  NaN                                                NaN  \n",
       "357026                  NaN                                                NaN  \n",
       "357027                  NaN                                                NaN  \n",
       "357028                  NaN                                                NaN  \n",
       "357029                  NaN                                                NaN  \n",
       "357030                  NaN                                                NaN  \n",
       "357031                  NaN                                                NaN  \n",
       "357032                  NaN                                                NaN  \n",
       "357033                  NaN                                                NaN  \n",
       "357034                  NaN                                                NaN  \n",
       "357035                  NaN                                                NaN  \n",
       "357036                  NaN                                                NaN  \n",
       "357037                  NaN                                                NaN  \n",
       "357038                  NaN                                                NaN  \n",
       "357039                  NaN                                                NaN  \n",
       "357040                  NaN                                                NaN  \n",
       "357041                  NaN                                                NaN  \n",
       "\n",
       "[357042 rows x 11 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df1 = df.iloc[0:10,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>protein_variant</th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>journal</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>impact_factor</th>\n",
       "      <th>abstract</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>protein_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>p.Leu153Phe</td>\n",
       "      <td>22040064</td>\n",
       "      <td>Frequencies of SLC44A2 alleles encoding human ...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.042</td>\n",
       "      <td>The human neutrophil antigen-3 (HNA-3) epitope...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLC44A2</td>\n",
       "      <td>p.Arg154Gln</td>\n",
       "      <td>22040064</td>\n",
       "      <td>Frequencies of SLC44A2 alleles encoding human ...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.042</td>\n",
       "      <td>The human neutrophil antigen-3 (HNA-3) epitope...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APOBEC-1</td>\n",
       "      <td>p.Glu181Gln</td>\n",
       "      <td>22326345</td>\n",
       "      <td>Hypermutation of ApoB mRNA by rat APOBEC-1 ove...</td>\n",
       "      <td>Journal of molecular biology</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>Apr</td>\n",
       "      <td>4.517</td>\n",
       "      <td>APOBEC-3 proteins induce C-to-U hypermutations...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP21A2</td>\n",
       "      <td>p.Val281Leu</td>\n",
       "      <td>22886582</td>\n",
       "      <td>Complement component 4 copy number variation a...</td>\n",
       "      <td>Human genetics</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>Dec</td>\n",
       "      <td>5.138</td>\n",
       "      <td>Congenital adrenal hyperplasia (CAH) due to 21...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD59</td>\n",
       "      <td>c.146delA</td>\n",
       "      <td>24383981</td>\n",
       "      <td>A new blood group antigen is defined by anti-C...</td>\n",
       "      <td>Transfusion</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Jul</td>\n",
       "      <td>3.042</td>\n",
       "      <td>CD59 is a cell surface glycoprotein of approxi...</td>\n",
       "      <td>chr11:g.33738939del</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T</td>\n",
       "      <td>c.1096C&gt;T</td>\n",
       "      <td>25814141</td>\n",
       "      <td>Caspase-8 Deficiency Presenting as Late-Onset ...</td>\n",
       "      <td>Journal of clinical immunology</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>May</td>\n",
       "      <td>3.094</td>\n",
       "      <td>Caspase-8 deficiency (CED) was originally desc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C4A</td>\n",
       "      <td>p.Cys4Ala</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>C4B</td>\n",
       "      <td>p.Cys4Asx</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C4A</td>\n",
       "      <td>p.Cys4Asx</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C4B</td>\n",
       "      <td>p.Cys4Ala</td>\n",
       "      <td>26493816</td>\n",
       "      <td>Gene copy-number variations (CNVs) of compleme...</td>\n",
       "      <td>Annals of the rheumatic diseases</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>12.384</td>\n",
       "      <td>Complement-mediated vasculopathy of muscle and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene protein_variant      pmid  \\\n",
       "0         T     p.Leu153Phe  22040064   \n",
       "1   SLC44A2     p.Arg154Gln  22040064   \n",
       "2  APOBEC-1     p.Glu181Gln  22326345   \n",
       "3   CYP21A2     p.Val281Leu  22886582   \n",
       "4      CD59       c.146delA  24383981   \n",
       "5         T       c.1096C>T  25814141   \n",
       "6       C4A       p.Cys4Ala  26493816   \n",
       "7       C4B       p.Cys4Asx  26493816   \n",
       "8       C4A       p.Cys4Asx  26493816   \n",
       "9       C4B       p.Cys4Ala  26493816   \n",
       "\n",
       "                                               title  \\\n",
       "0  Frequencies of SLC44A2 alleles encoding human ...   \n",
       "1  Frequencies of SLC44A2 alleles encoding human ...   \n",
       "2  Hypermutation of ApoB mRNA by rat APOBEC-1 ove...   \n",
       "3  Complement component 4 copy number variation a...   \n",
       "4  A new blood group antigen is defined by anti-C...   \n",
       "5  Caspase-8 Deficiency Presenting as Late-Onset ...   \n",
       "6  Gene copy-number variations (CNVs) of compleme...   \n",
       "7  Gene copy-number variations (CNVs) of compleme...   \n",
       "8  Gene copy-number variations (CNVs) of compleme...   \n",
       "9  Gene copy-number variations (CNVs) of compleme...   \n",
       "\n",
       "                            journal    year month impact_factor  \\\n",
       "0                       Transfusion  2012.0   May         3.042   \n",
       "1                       Transfusion  2012.0   May         3.042   \n",
       "2      Journal of molecular biology  2012.0   Apr         4.517   \n",
       "3                    Human genetics  2012.0   Dec         5.138   \n",
       "4                       Transfusion  2014.0   Jul         3.042   \n",
       "5    Journal of clinical immunology  2015.0   May         3.094   \n",
       "6  Annals of the rheumatic diseases  2016.0   Sep        12.384   \n",
       "7  Annals of the rheumatic diseases  2016.0   Sep        12.384   \n",
       "8  Annals of the rheumatic diseases  2016.0   Sep        12.384   \n",
       "9  Annals of the rheumatic diseases  2016.0   Sep        12.384   \n",
       "\n",
       "                                            abstract           variant_id  \\\n",
       "0  The human neutrophil antigen-3 (HNA-3) epitope...                  NaN   \n",
       "1  The human neutrophil antigen-3 (HNA-3) epitope...                  NaN   \n",
       "2  APOBEC-3 proteins induce C-to-U hypermutations...                  NaN   \n",
       "3  Congenital adrenal hyperplasia (CAH) due to 21...                  NaN   \n",
       "4  CD59 is a cell surface glycoprotein of approxi...  chr11:g.33738939del   \n",
       "5  Caspase-8 deficiency (CED) was originally desc...                  NaN   \n",
       "6  Complement-mediated vasculopathy of muscle and...                  NaN   \n",
       "7  Complement-mediated vasculopathy of muscle and...                  NaN   \n",
       "8  Complement-mediated vasculopathy of muscle and...                  NaN   \n",
       "9  Complement-mediated vasculopathy of muscle and...                  NaN   \n",
       "\n",
       "  protein_domain  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "5            NaN  \n",
       "6            NaN  \n",
       "7            NaN  \n",
       "8            NaN  \n",
       "9            NaN  "
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "357042"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dj1 = df1.to_json(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "output = '['\n",
    "for i in range(len(df1)):\n",
    "    row = df1.iloc[i:i+1,:].to_json(orient='records')[1:-1]\n",
    "    output = output + '{\"model\":\"deepb.pubmed\",\"pk\":%i,\"fields\":%s},'%(i,row)\n",
    "output = output[:-1]+']'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"model\":\"deepb.pubmed\",\"pk\":0,\"fields\":{\"gene\":\"T\",\"protein_variant\":\"p.Leu153Phe\",\"pmid\":22040064,\"title\":\"Frequencies of SLC44A2 alleles encoding human neutrophil antigen-3 variants in the African American population.\",\"journal\":\"Transfusion\",\"year\":2012.0,\"month\":\"May\",\"impact_factor\":\"3.042\",\"abstract\":\"The human neutrophil antigen-3 (HNA-3) epitopes reside on the choline transporter-like protein-2 (CTL2). A single-nucleotide substitution (461G>A; Arg154Gln) on the CTL2 gene (SLC44A2) defines the allele SLC44A2*1, which expresses HNA-3a, and SLC44A2*2, which expresses HNA-3b; an additional substitution (457C>T; Leu153Phe) in SLC44A2*1:2 may impact genotyping systems. People who only express HNA-3b may develop anti-HNA-3a. These alloantibodies have been linked to severe transfusion-related acute lung injury, which may be a reason to screen blood donors for SLC44A2*2 homozygosity. For Caucasian and Asian populations, SLC44A2 allele frequencies are known. Our primary objective was to determine the SLC44A2 allele frequencies in the African American population. Purified DNA from 334 individuals (202 male, 132 female; 241 African American, 93 Caucasian) was collected. Two real-time polymerase chain reaction assays were developed to genotype all samples; results were confirmed by nucleotide sequencing. In 241 African American donors, the allele frequency of SLC44A2*1 was 93% (85%-<100%; 95% confidence intervals, Poisson distribution) while SLC44A2*2 was 7% (5%-10%). In 93 Caucasian donors, the allele frequency of SLC44A2*1 was 83% (71%-98%) and SLC44A2*2 was 17% (11%-24%), matching previously reported data for Caucasians but differing from African Americans (p\\\\u2003<\\\\u20030.001, Fisher\\'s exact test). This study describes the allele frequencies of the three known HNA-3 variants in an African American population. We found that African Americans have a significantly lower probability of possessing the SLC44A2*2 allele and may thus be less likely to form the clinically relevant anti-HNA-3a.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":1,\"fields\":{\"gene\":\"SLC44A2\",\"protein_variant\":\"p.Arg154Gln\",\"pmid\":22040064,\"title\":\"Frequencies of SLC44A2 alleles encoding human neutrophil antigen-3 variants in the African American population.\",\"journal\":\"Transfusion\",\"year\":2012.0,\"month\":\"May\",\"impact_factor\":\"3.042\",\"abstract\":\"The human neutrophil antigen-3 (HNA-3) epitopes reside on the choline transporter-like protein-2 (CTL2). A single-nucleotide substitution (461G>A; Arg154Gln) on the CTL2 gene (SLC44A2) defines the allele SLC44A2*1, which expresses HNA-3a, and SLC44A2*2, which expresses HNA-3b; an additional substitution (457C>T; Leu153Phe) in SLC44A2*1:2 may impact genotyping systems. People who only express HNA-3b may develop anti-HNA-3a. These alloantibodies have been linked to severe transfusion-related acute lung injury, which may be a reason to screen blood donors for SLC44A2*2 homozygosity. For Caucasian and Asian populations, SLC44A2 allele frequencies are known. Our primary objective was to determine the SLC44A2 allele frequencies in the African American population. Purified DNA from 334 individuals (202 male, 132 female; 241 African American, 93 Caucasian) was collected. Two real-time polymerase chain reaction assays were developed to genotype all samples; results were confirmed by nucleotide sequencing. In 241 African American donors, the allele frequency of SLC44A2*1 was 93% (85%-<100%; 95% confidence intervals, Poisson distribution) while SLC44A2*2 was 7% (5%-10%). In 93 Caucasian donors, the allele frequency of SLC44A2*1 was 83% (71%-98%) and SLC44A2*2 was 17% (11%-24%), matching previously reported data for Caucasians but differing from African Americans (p\\\\u2003<\\\\u20030.001, Fisher\\'s exact test). This study describes the allele frequencies of the three known HNA-3 variants in an African American population. We found that African Americans have a significantly lower probability of possessing the SLC44A2*2 allele and may thus be less likely to form the clinically relevant anti-HNA-3a.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":2,\"fields\":{\"gene\":\"APOBEC-1\",\"protein_variant\":\"p.Glu181Gln\",\"pmid\":22326345,\"title\":\"Hypermutation of ApoB mRNA by rat APOBEC-1 overexpression mimics APOBEC-3 hypermutation.\",\"journal\":\"Journal of molecular biology\",\"year\":2012.0,\"month\":\"Apr\",\"impact_factor\":\"4.517\",\"abstract\":\"APOBEC-3 proteins induce C-to-U hypermutations in the viral genome of various viruses and have broad antiviral activity. Generally, only a small proportion of viral genomes (<10(-)(2)) are hypermutated by APOBEC-3s, but often many cytidines (up to 40%) are converted into uridine. The mechanism of this unique selective hypermutation remains unknown. We found that rat APOBEC-1 overexpression had a hypermutation pattern similar to that of APOBEC-3s on its substrate apolipoprotein B (apoB) mRNA. Transient plasmid transfection of rat APOBEC-1 resulted in 0.4% and 1.8% hypermutations with apoB mRNA in HepG2 and McA7777 cells, respectively. The low frequency of hypermutated apoB mRNA targets was enriched by differential DNA denaturation PCR at 72-76\\\\u00a0\\\\u00b0C, with hypermutation levels increasing up to 67%. Up to 69.6% of cytidines in HepG2 and up to 75.5% of cytidines in McA7777 cells were converted into uridines in the hypermutated apoB mRNA. When rat APOBEC-1 was overexpressed by adenovirus, the hypermutation frequency of apoB mRNA increased from 0.4% to \\\\u223c20% and was readily detected by regular PCR. However, this higher expression efficiency only increased the frequency of hypermutation, not the number of affected cytidines in hypermutated targets. Rat APOBEC-1 hypermutation was modulated by cofactors and eliminated by an E181Q mutation, indicating the role of cofactors in hypermutation. The finding of an APOBEC-3 hypermutation pattern with rat APOBEC-1 suggests that cofactors could also be involved in APOBEC-3 hypermutation. Using hepatitis B virus hypermutation, we found that KSRP increased APOBEC-3C and APOBEC-3B hypermutation. These data show that, like rat APOBEC-1 hypermutation, cellular factors may play a regulatory role in APOBEC-3 hypermutation.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":3,\"fields\":{\"gene\":\"CYP21A2\",\"protein_variant\":\"p.Val281Leu\",\"pmid\":22886582,\"title\":\"Complement component 4 copy number variation and CYP21A2 genotype associations in patients with congenital adrenal hyperplasia due to 21-hydroxylase deficiency.\",\"journal\":\"Human genetics\",\"year\":2012.0,\"month\":\"Dec\",\"impact_factor\":\"5.138\",\"abstract\":\"Congenital adrenal hyperplasia (CAH) due to 21-hydroxylase deficiency (21-OHD) is an autosomal recessive disorder of cortisol biosynthesis caused by CYP21A2 mutations. An increase in gene copy number variation (CNV) exists at the CYP21A2 locus. CNV of C4, a neighboring gene that encodes complement component 4, is associated with autoimmune disease susceptibility. In this study, we performed comprehensive genetic analysis of the RP-C4-CYP21-TNX (RCCX) region in 127 unrelated 21-OHD patients (100 classic, 27 nonclassic). C4 copy number was determined by Southern blot. C4 CNV and serum C4 levels were evaluated in relation to CYP21A2 mutations and relevant phenotypes. We found that the most common CYP21A2 mutation associated with the nonclassic form of CAH, V281L, was associated with high C4 copy number (p\\\\u00a0=\\\\u00a07.13\\\\u00a0\\\\u00d7\\\\u00a010(-16)). Large CYP21A2 deletion, a common mutation associated with the classic form of CAH, was associated with low C4 copy number (p\\\\u00a0=\\\\u00a01.61\\\\u00a0\\\\u00d7\\\\u00a010(-14)). Monomodular RCCX with a short C4 gene, a risk factor for autoimmune disease, was significantly less frequent in CAH patients compared to population estimates (2.8 vs. 10.6\\\\u00a0%; p\\\\u00a0=\\\\u00a01.08\\\\u00a0\\\\u00d7\\\\u00a010(-4)). In conclusion, CAH patients have increased C4 CNV, with mutation-specific associations that may be protective for autoimmune disease. The study of CYP21A2 in relation to neighboring genes provides insight into the genetics of CNV hotspots, an important determinant of human health.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":4,\"fields\":{\"gene\":\"CD59\",\"protein_variant\":\"c.146delA\",\"pmid\":24383981,\"title\":\"A new blood group antigen is defined by anti-CD59, detected in a CD59-deficient patient.\",\"journal\":\"Transfusion\",\"year\":2014.0,\"month\":\"Jul\",\"impact_factor\":\"3.042\",\"abstract\":\"CD59 is a cell surface glycoprotein of approximately 20\\\\u2009kDa limiting the lytic activity of the terminal complement complex C5b-9. Although CD59 is known as a red blood cell (RBC) antigen defined by monoclonal antibodies, it so far has not been identified as a blood group antigen, since the description of a human alloantibody was missing. In this study we show the presence of an anti-CD59 in a patient affected by a homozygous CD59 deficiency. RBC CD59 and CD55 were determined by flow cytometry or by the column agglutination technique using monoclonal antisera. Commercially available His-tagged recombinant soluble CD59 protein was used to inhibit anti-CD59. Seven cases of an isolated CD59 deficiency due to three distinct null alleles of the CD59 gene have been published so far. Recently we described the CD59-null allele c.146delA in a young child of heterozygous parents. Her plasma contained an alloantibody directed against the high-prevalence RBC antigen CD59. The antibody specificity was identified using soluble recombinant human CD59 protein, which blocked the reactivity of the patient\\'s antibody and of monoclonal anti-CD59 but not of monoclonal anti-CD55. In addition, RBC alloantibodies such as anti-K, anti-C, anti-c, or anti-Fy(a) remained unaffected. Therefore, inhibition by recombinant CD59 is a useful diagnostic tool to detect alloantibodies in the presence of anti-CD59. This is the first demonstration of a human anti-CD59 alloantibody, which defines CD59 as an RBC blood group antigen. CD59 represents a candidate for a new blood group system.\",\"variant_id\":\"chr11:g.33738939del\",\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":5,\"fields\":{\"gene\":\"T\",\"protein_variant\":\"c.1096C>T\",\"pmid\":25814141,\"title\":\"Caspase-8 Deficiency Presenting as Late-Onset Multi-Organ Lymphocytic Infiltration with Granulomas in two Adult Siblings.\",\"journal\":\"Journal of clinical immunology\",\"year\":2015.0,\"month\":\"May\",\"impact_factor\":\"3.094\",\"abstract\":\"Caspase-8 deficiency (CED) was originally described in 2002 in two pediatric patients presenting with clinical manifestations resembling autoimmune lymphoproliferative syndrome (ALPS) accompanied by infections, and T, B and NK cell defects. Since then, no new CED patients were published. Here we report two adult siblings (Pt1 and Pt2) presenting in their late thirties with pulmonary hypertension leading to lung transplant (Pt1), and a complex neurological disease leading to multiple cranial nerves palsies (Pt2) as their main manifestations. A thorough clinical and immunological evaluation was performed at the Primary Immunodeficiency Clinic at NIH, followed by whole exome sequencing. The patients had multiorgan lymphocytic infiltration and granulomas, as well as clinical signs of immune deficiency\\\\/ immune dysregulation. Both siblings carried homozygous mutations in CASP8, c.1096C > T, p.248R > W. This was the same mutation described on the previously published CED patients, to whom these new patients were likely distantly related. We report two new CED patients presenting during adulthood with life-threatening end-organ lymphocyte infiltrates affecting the lungs, liver, spleen, bone marrow and central nervous system. This phenotype broadens the clinical spectrum of manifestations associated with this disease and warrants the search of CASP8 mutations in other cohorts of patients.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":6,\"fields\":{\"gene\":\"C4A\",\"protein_variant\":\"p.Cys4Ala\",\"pmid\":26493816,\"title\":\"Gene copy-number variations (CNVs) of complement C4 and C4A deficiency in genetic risk and pathogenesis of juvenile dermatomyositis.\",\"journal\":\"Annals of the rheumatic diseases\",\"year\":2016.0,\"month\":\"Sep\",\"impact_factor\":\"12.384\",\"abstract\":\"Complement-mediated vasculopathy of muscle and skin are clinical features of juvenile dermatomyositis (JDM). We assess gene copy-number variations (CNVs) for complement C4 and its isotypes, C4A and C4B, in genetic risks and pathogenesis of JDM. The study population included 105 patients with JDM and 500 healthy European Americans. Gene copy-numbers (GCNs) for total C4, C4A, C4B and HLA-DRB1 genotypes were determined by Southern blots and qPCRs. Processed activation product C4d bound to erythrocytes (E-C4d) was measured by flow cytometry. Global gene-expression microarrays were performed in 19 patients with JDM and seven controls using PAXgene-blood RNA. Differential expression levels for selected genes were validated by qPCR. Significantly lower GCNs and differences in distribution of GCN groups for total C4 and C4A were observed in JDM versus controls. Lower GCN of C4A in JDM remained among HLA DR3-positive subjects (p=0.015). Homozygous or heterozygous C4A-deficiency was present in 40.0% of patients with JDM compared with 18.2% of controls (OR=3.00 (1.87 to 4.79), p=8.2\\\\u00d710(-6)). Patients with JDM had higher levels of E-C4d than controls (p=0.004). In JDM, C4A-deficient subjects had higher levels of E-C4d (p=0.0003) and higher frequency of elevated levels of multiple serum muscle enzymes at diagnosis (p=0.0025). Microarray profiling of blood RNA revealed upregulation of type I interferon-stimulated genes and lower abundance of transcripts for T-cell and chemokine function genes in JDM, but this was less prominent among C4A-deficient or DR3-positive patients. Complement C4A deficiency appears to be an important factor for the genetic risk and pathogenesis of JDM, particularly in patients with a DR3-positive background.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":7,\"fields\":{\"gene\":\"C4B\",\"protein_variant\":\"p.Cys4Asx\",\"pmid\":26493816,\"title\":\"Gene copy-number variations (CNVs) of complement C4 and C4A deficiency in genetic risk and pathogenesis of juvenile dermatomyositis.\",\"journal\":\"Annals of the rheumatic diseases\",\"year\":2016.0,\"month\":\"Sep\",\"impact_factor\":\"12.384\",\"abstract\":\"Complement-mediated vasculopathy of muscle and skin are clinical features of juvenile dermatomyositis (JDM). We assess gene copy-number variations (CNVs) for complement C4 and its isotypes, C4A and C4B, in genetic risks and pathogenesis of JDM. The study population included 105 patients with JDM and 500 healthy European Americans. Gene copy-numbers (GCNs) for total C4, C4A, C4B and HLA-DRB1 genotypes were determined by Southern blots and qPCRs. Processed activation product C4d bound to erythrocytes (E-C4d) was measured by flow cytometry. Global gene-expression microarrays were performed in 19 patients with JDM and seven controls using PAXgene-blood RNA. Differential expression levels for selected genes were validated by qPCR. Significantly lower GCNs and differences in distribution of GCN groups for total C4 and C4A were observed in JDM versus controls. Lower GCN of C4A in JDM remained among HLA DR3-positive subjects (p=0.015). Homozygous or heterozygous C4A-deficiency was present in 40.0% of patients with JDM compared with 18.2% of controls (OR=3.00 (1.87 to 4.79), p=8.2\\\\u00d710(-6)). Patients with JDM had higher levels of E-C4d than controls (p=0.004). In JDM, C4A-deficient subjects had higher levels of E-C4d (p=0.0003) and higher frequency of elevated levels of multiple serum muscle enzymes at diagnosis (p=0.0025). Microarray profiling of blood RNA revealed upregulation of type I interferon-stimulated genes and lower abundance of transcripts for T-cell and chemokine function genes in JDM, but this was less prominent among C4A-deficient or DR3-positive patients. Complement C4A deficiency appears to be an important factor for the genetic risk and pathogenesis of JDM, particularly in patients with a DR3-positive background.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":8,\"fields\":{\"gene\":\"C4A\",\"protein_variant\":\"p.Cys4Asx\",\"pmid\":26493816,\"title\":\"Gene copy-number variations (CNVs) of complement C4 and C4A deficiency in genetic risk and pathogenesis of juvenile dermatomyositis.\",\"journal\":\"Annals of the rheumatic diseases\",\"year\":2016.0,\"month\":\"Sep\",\"impact_factor\":\"12.384\",\"abstract\":\"Complement-mediated vasculopathy of muscle and skin are clinical features of juvenile dermatomyositis (JDM). We assess gene copy-number variations (CNVs) for complement C4 and its isotypes, C4A and C4B, in genetic risks and pathogenesis of JDM. The study population included 105 patients with JDM and 500 healthy European Americans. Gene copy-numbers (GCNs) for total C4, C4A, C4B and HLA-DRB1 genotypes were determined by Southern blots and qPCRs. Processed activation product C4d bound to erythrocytes (E-C4d) was measured by flow cytometry. Global gene-expression microarrays were performed in 19 patients with JDM and seven controls using PAXgene-blood RNA. Differential expression levels for selected genes were validated by qPCR. Significantly lower GCNs and differences in distribution of GCN groups for total C4 and C4A were observed in JDM versus controls. Lower GCN of C4A in JDM remained among HLA DR3-positive subjects (p=0.015). Homozygous or heterozygous C4A-deficiency was present in 40.0% of patients with JDM compared with 18.2% of controls (OR=3.00 (1.87 to 4.79), p=8.2\\\\u00d710(-6)). Patients with JDM had higher levels of E-C4d than controls (p=0.004). In JDM, C4A-deficient subjects had higher levels of E-C4d (p=0.0003) and higher frequency of elevated levels of multiple serum muscle enzymes at diagnosis (p=0.0025). Microarray profiling of blood RNA revealed upregulation of type I interferon-stimulated genes and lower abundance of transcripts for T-cell and chemokine function genes in JDM, but this was less prominent among C4A-deficient or DR3-positive patients. Complement C4A deficiency appears to be an important factor for the genetic risk and pathogenesis of JDM, particularly in patients with a DR3-positive background.\",\"variant_id\":null,\"protein_domain\":null}},{\"model\":\"deepb.pubmed\",\"pk\":9,\"fields\":{\"gene\":\"C4B\",\"protein_variant\":\"p.Cys4Ala\",\"pmid\":26493816,\"title\":\"Gene copy-number variations (CNVs) of complement C4 and C4A deficiency in genetic risk and pathogenesis of juvenile dermatomyositis.\",\"journal\":\"Annals of the rheumatic diseases\",\"year\":2016.0,\"month\":\"Sep\",\"impact_factor\":\"12.384\",\"abstract\":\"Complement-mediated vasculopathy of muscle and skin are clinical features of juvenile dermatomyositis (JDM). We assess gene copy-number variations (CNVs) for complement C4 and its isotypes, C4A and C4B, in genetic risks and pathogenesis of JDM. The study population included 105 patients with JDM and 500 healthy European Americans. Gene copy-numbers (GCNs) for total C4, C4A, C4B and HLA-DRB1 genotypes were determined by Southern blots and qPCRs. Processed activation product C4d bound to erythrocytes (E-C4d) was measured by flow cytometry. Global gene-expression microarrays were performed in 19 patients with JDM and seven controls using PAXgene-blood RNA. Differential expression levels for selected genes were validated by qPCR. Significantly lower GCNs and differences in distribution of GCN groups for total C4 and C4A were observed in JDM versus controls. Lower GCN of C4A in JDM remained among HLA DR3-positive subjects (p=0.015). Homozygous or heterozygous C4A-deficiency was present in 40.0% of patients with JDM compared with 18.2% of controls (OR=3.00 (1.87 to 4.79), p=8.2\\\\u00d710(-6)). Patients with JDM had higher levels of E-C4d than controls (p=0.004). In JDM, C4A-deficient subjects had higher levels of E-C4d (p=0.0003) and higher frequency of elevated levels of multiple serum muscle enzymes at diagnosis (p=0.0025). Microarray profiling of blood RNA revealed upregulation of type I interferon-stimulated genes and lower abundance of transcripts for T-cell and chemokine function genes in JDM, but this was less prominent among C4A-deficient or DR3-positive patients. Complement C4A deficiency appears to be an important factor for the genetic risk and pathogenesis of JDM, particularly in patients with a DR3-positive background.\",\"variant_id\":null,\"protein_domain\":null}}]'"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "[\n",
    "  {\n",
    "    \"model\": \"myapp.person\",\n",
    "    \"pk\": 1,\n",
    "    \"fields\": {\n",
    "      \"first_name\": \"John\",\n",
    "      \"last_name\": \"Lennon\"\n",
    "    }\n",
    "  },\n",
    "  {\n",
    "    \"model\": \"myapp.person\",\n",
    "    \"pk\": 2,\n",
    "    \"fields\": {\n",
    "      \"first_name\": \"Paul\",\n",
    "      \"last_name\": \"McCartney\"\n",
    "    }\n",
    "  }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "output = '['\n",
    "for i in range(len(df)):\n",
    "    row = df.iloc[i:i+1,:].to_json(orient='records')[1:-1]\n",
    "    output = output + '{\"model\":\"deepb.pubmed\",\"pk\":%i,\"fields\":%s},'%(i,row)\n",
    "output = output[:-1]+']'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open('pubmed_data.json', 'w') as f:\n",
    "    f.write(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1493161608.010493"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mACMG.py\u001b[m\u001b[m*                   \u001b[31mamino_acid_mapping.py\u001b[m\u001b[m*     \u001b[31mdataloader.py\u001b[m\u001b[m*             main.pyc                   pubmed.pyc                 tests.py\r\n",
      "ACMG.pyc                   amino_acid_mapping.pyc     dataloader.pyc             \u001b[31mmap_phenotype_to_gene.py\u001b[m\u001b[m*  \u001b[34mresult\u001b[m\u001b[m/                    urls.py\r\n",
      "README.md                  apps.py                    \u001b[31meval.py\u001b[m\u001b[m*                   map_phenotype_to_gene.pyc  \u001b[34mstatic\u001b[m\u001b[m/                    urls.pyc\r\n",
      "__init__.py                apps.pyc                   \u001b[34mfixtures\u001b[m\u001b[m/                  \u001b[34mmigrations\u001b[m\u001b[m/                tasks.py                   views.py\r\n",
      "__init__.pyc               \u001b[31mcollectVariantInfo.py\u001b[m\u001b[m*     \u001b[31mgetInterproDomain.py\u001b[m\u001b[m*      models.py                  tasks.pyc                  views.pyc\r\n",
      "admin.py                   collectVariantInfo.pyc     \u001b[34minput\u001b[m\u001b[m/                     models.pyc                 \u001b[34mtemplates\u001b[m\u001b[m/\r\n",
      "admin.pyc                  \u001b[34mdata\u001b[m\u001b[m/                      main.py                    \u001b[31mpubmed.py\u001b[m\u001b[m*                 test.ipynb\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open('input/', 'r') as myfile:\n",
    "    data=myfile.read().replace('\\n', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "raw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "input_gene = raw_input.raw_input_gene\n",
    "input_phenotype = raw_input.raw_input_phenotype\n",
    "\n",
    "# input_gene_list = [i.split('\\t') for i in input_gene.split('\\n')]\n",
    "# df_genes = pd.DataFrame(input_gene_list[1:-1], columns=input_gene_list[0])\n",
    "df_genes = pd.read_csv(StringIO(input_gene), sep = '\\t')\n",
    "CANDIDATE_GENES = pd.unique(df_genes.Gene.values).tolist()\n",
    "\n",
    "corner_cases = dict()\n",
    "phenos = []\n",
    "input_phenotype = input_phenotype.split('\\n')\n",
    "for line in input_phenotype:\n",
    "    if line.startswith('#') or not line.strip():\n",
    "        continue\n",
    "    line = line.rstrip()\n",
    "    phenos += line.split(',')\n",
    "    for pheno in phenos:\n",
    "        if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "            phenos.append('growth delay')\n",
    "            corner_cases['growth delay'] = pheno.strip()\n",
    "    for pheno in phenos:\n",
    "        if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "            phenos.append('developmental delay')\n",
    "            corner_cases['developmental delay'] = pheno.strip()\n",
    "phenos = [_.strip() for _ in phenos]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mACMG.py\u001b[m\u001b[m*                   \u001b[31mamino_acid_mapping.py\u001b[m\u001b[m*     \u001b[31mdataloader.py\u001b[m\u001b[m*             main.pyc                   pubmed.pyc                 tests.py\r\n",
      "ACMG.pyc                   amino_acid_mapping.pyc     dataloader.pyc             \u001b[31mmap_phenotype_to_gene.py\u001b[m\u001b[m*  \u001b[34mresult\u001b[m\u001b[m/                    urls.py\r\n",
      "README.md                  apps.py                    \u001b[31meval.py\u001b[m\u001b[m*                   map_phenotype_to_gene.pyc  \u001b[34mstatic\u001b[m\u001b[m/                    urls.pyc\r\n",
      "__init__.py                apps.pyc                   \u001b[34mfixtures\u001b[m\u001b[m/                  \u001b[34mmigrations\u001b[m\u001b[m/                tasks.py                   views.py\r\n",
      "__init__.pyc               \u001b[31mcollectVariantInfo.py\u001b[m\u001b[m*     \u001b[31mgetInterproDomain.py\u001b[m\u001b[m*      models.py                  tasks.pyc                  views.pyc\r\n",
      "admin.py                   collectVariantInfo.pyc     \u001b[34minput\u001b[m\u001b[m/                     models.pyc                 \u001b[34mtemplates\u001b[m\u001b[m/\r\n",
      "admin.pyc                  \u001b[34mdata\u001b[m\u001b[m/                      main.py                    \u001b[31mpubmed.py\u001b[m\u001b[m*                 test.ipynb\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dfg = pd.read_csv('input/input_genes.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "input_gene = dfg.to_json(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import map_phenotype_to_gene\n",
    "import collectVariantInfo\n",
    "import pubmed\n",
    "import ACMG\n",
    "import filterVariantOnPhenotype\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import myvariant\n",
    "from deepb.models import Main_table, Raw_input_table\n",
    "from io import StringIO\n",
    "from collections import Counter\n",
    "\n",
    "# input_phenotype = 'data/sample_patient_phenotype.txt'\n",
    "# input_genes = 'data/sample_genes.txt'\n",
    "\n",
    "def format_hgvs(chrom, pos, ref, alt):\n",
    "    '''get a valid hgvs name from VCF-style \"chrom, pos, ref, alt\" data.\n",
    "\n",
    "    Example:\n",
    "\n",
    "        >>> myvariant.format_hgvs(\"1\", 35366, \"C\", \"T\")\n",
    "        >>> myvariant.format_hgvs(\"2\", 17142, \"G\", \"GA\")\n",
    "        >>> myvariant.format_hgvs(\"MT\", 8270, \"CACCCCCTCT\", \"C\")\n",
    "        >>> myvariant.format_hgvs(\"X\", 107930849, \"GGA\", \"C\")\n",
    "\n",
    "    '''\n",
    "    chrom = str(chrom)\n",
    "    if chrom.lower().startswith('chr'):\n",
    "        # trim off leading \"chr\" if any\n",
    "        chrom = chrom[3:]\n",
    "    if len(ref) == len(alt) == 1:\n",
    "        # this is a SNP\n",
    "        hgvs = 'chr{0}:g.{1}{2}>{3}'.format(chrom, pos, ref, alt)\n",
    "    elif len(ref) > 1 and len(alt) == 1:\n",
    "        # this is a deletion:\n",
    "        if ref[0] == alt:\n",
    "            start = int(pos) + 1\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}del'.format(chrom, start, end)\n",
    "        else:\n",
    "            end = int(pos) + len(ref) - 1\n",
    "            hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    elif len(ref) == 1 and len(alt) > 1:\n",
    "        # this is a insertion\n",
    "        if alt[0] == ref:\n",
    "            hgvs = 'chr{0}:g.{1}_{2}ins'.format(chrom, pos, int(pos) + 1)\n",
    "            ins_seq = alt[1:]\n",
    "            hgvs += ins_seq\n",
    "        else:\n",
    "            hgvs = 'chr{0}:g.{1}delins{2}'.format(chrom, pos, alt)\n",
    "    elif len(ref) > 1 and len(alt) > 1:\n",
    "        end = int(pos) + len(alt) - 1\n",
    "        hgvs = 'chr{0}:g.{1}_{2}delins{3}'.format(chrom, pos, end, alt)\n",
    "    else:\n",
    "        raise ValueError(\"Cannot convert {} into HGVS id.\".format((chrom, pos, ref, alt)))\n",
    "    return hgvs\n",
    "\n",
    "def read_input_pheno_file(input_phenotype):\n",
    "        if not input_phenotype:\n",
    "            return '', ''\n",
    "        text = StringIO(unicode(input_phenotype), newline=None)\n",
    "        lines = text.readlines()\n",
    "        lines = [line.strip() for line in lines]\n",
    "        phenos = []\n",
    "        for line in lines:\n",
    "            if not line:\n",
    "                continue\n",
    "            phenos_each_line = re.split(r'  +|\\t+|,|;|\\.|\\|', line.strip())\n",
    "            phenos_each_line = [re.sub(r'^\\W+|\\W+$', '', s) for s in phenos_each_line]\n",
    "            phenos_each_line = [s for s in phenos_each_line if s]\n",
    "            phenos += phenos_each_line\n",
    "\n",
    "        corner_cases = dict()\n",
    "        for pheno in phenos:\n",
    "                if re.search('development', pheno) and re.search('delay', pheno) and not re.search('growth', pheno):\n",
    "                        phenos.append('growth delay')\n",
    "                        corner_cases['growth delay'] = pheno.strip()\n",
    "        for pheno in phenos:\n",
    "                if re.search('growth', pheno) and re.search('delay', pheno) and not re.search('development', pheno):\n",
    "                        phenos.append('developmental delay')\n",
    "                        corner_cases['developmental delay'] = pheno.strip()\n",
    "        return phenos, corner_cases\n",
    "\n",
    "def read_input_gene_file(input_gene):\n",
    "        candidate_vars = []\n",
    "        input_gene = input_gene.split('\\n')\n",
    "        header = input_gene[0]\n",
    "        sniffer = csv.Sniffer()\n",
    "        dialect = sniffer.sniff(header)\n",
    "        delimiter =  dialect.delimiter\n",
    "        field_names = header.split(delimiter)\n",
    "        chrom_idx, pos_idx, ref_idx, alt_idx, gene_idx = None, None, None, None, 0 \n",
    "\n",
    "        for idx in xrange(len(field_names)):\n",
    "            field = field_names[idx]\n",
    "            if re.match(r'chrom', field, re.I): chrom_idx = idx\n",
    "            if re.match(r'pos|start', field, re.I): pos_idx = idx\n",
    "            if re.match(r'ref', field, re.I): ref_idx = idx\n",
    "            if re.match(r'alt|allele 1', field, re.I): alt_idx = idx\n",
    "            if re.match(r'gene (gene)|gene', field, re.I): gene_idx = idx\n",
    "\n",
    "        input_gene_list = []\n",
    "        CANDIDATE_GENES = []\n",
    "        for line in input_gene[1:]:\n",
    "                if not line:\n",
    "                        continue\n",
    "                line = line.rstrip()\n",
    "                parts = re.split(r'%s' % delimiter, line)\n",
    "                input_gene_list.append(parts)\n",
    "                gene = parts[gene_idx]\n",
    "                CANDIDATE_GENES.append(gene)\n",
    "                transcript, variant, variant_id = '', '', ''\n",
    "                for part in parts:\n",
    "                    if re.search(r'_.*:c\\.', part):\n",
    "                        transcript, variant = part.split(':')\n",
    "                    if re.search(r'_.*:g\\.', part):\n",
    "                        variant_id = 'chr' + part.split(':')[0].split('.')[-1] + part.split(':')[-1]\n",
    "                    if re.search(r'chr.*:g\\.', part, re.I):\n",
    "                        variant_id = part\n",
    "                if not variant_id and (chrom_idx and pos_idx and ref_idx and alt_idx):\n",
    "                    chrome, pos, ref, alt = parts[chrom_idx], parts[pos_idx], parts[ref_idx], parts[alt_idx]\n",
    "                    variant_id = format_hgvs(chrome, pos, ref, alt)\n",
    "                candidate_vars.append((gene, variant, transcript, variant_id))\n",
    "\n",
    "        # remove lines in the input file which has wrong number of fields\n",
    "        field_nums = []\n",
    "        for line in input_gene_list:\n",
    "            field_nums.append(len(line))\n",
    "        count = Counter(field_nums)\n",
    "        correct_field_num = count.most_common()[0][0]\n",
    "        correct_input_gene_list = []\n",
    "        for line in input_gene_list:\n",
    "                if len(line) == correct_field_num:\n",
    "                        correct_input_gene_list.append(line)\n",
    "        df_genes = pd.DataFrame(correct_input_gene_list, columns = field_names)\n",
    "        return candidate_vars, CANDIDATE_GENES, df_genes, field_names \n",
    "\n",
    "def map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars):\n",
    "        ranking_genes, ranking_disease = map_phenotype_to_gene.generate_score(phenos, CANDIDATE_GENES, corner_cases)\n",
    "        # collect variant info\n",
    "        hpo_filtered_genes = np.unique([i[0] for i in ranking_genes]).tolist()\n",
    "\n",
    "        tmp_candidate_vars = []\n",
    "        for var in candidate_vars:\n",
    "                if var[0] in hpo_filtered_genes:\n",
    "                        tmp_candidate_vars.append(var)\n",
    "        candidate_vars = tmp_candidate_vars\n",
    "        return ranking_genes, candidate_vars\n",
    "\n",
    "raw_input = Raw_input_table.objects.get(id=raw_input_id)\n",
    "input_gene = open(\"/Users/Tianqi/Desktop/xiaonantest1_g.txt\",'rU').read()\n",
    "input_phenotype = open(\"/Users/Tianqi/Desktop/xiaonantest1_p.txt\",'rU').read()\n",
    "\n",
    "# Read input pheno file and generate phenos and corner_cases \n",
    "phenos, corner_cases = read_input_pheno_file(input_phenotype)\n",
    "\n",
    "# Read input gene file and generate candidate_vars. candidate_vars are (gene, variant, transcript, variant_id); CANDIDATE_GENES is a list of gene symbols; df_genes is a dataframe that keeps all the data that user uploaded; field_names are header of the input gene file \n",
    "candidate_vars, CANDIDATE_GENES, df_genes, field_names = read_input_gene_file(input_gene)\n",
    "\n",
    "# map phenotype to gene; the candidate_vars was filtered: if it is a gene associated with phenos, then keep it.\n",
    "if phenos:\n",
    "    ranking_genes, candidate_vars = map_phenotype2gene(CANDIDATE_GENES, phenos, corner_cases, candidate_vars)\n",
    "else:\n",
    "    ranking_genes = []\n",
    "    for gene in CANDIDATE_GENES:\n",
    "        ranking_genes.append((gene, 1.0, 1))\n",
    "\n",
    "# collect variant info\n",
    "mv = myvariant.MyVariantInfo()\n",
    "final_res, variants = collectVariantInfo.get_variants(candidate_vars)\n",
    "\n",
    "# pubmed\n",
    "df_pubmed = pubmed.queryPubmedDB(final_res)\n",
    "\n",
    "# ACMG\n",
    "df_hpo_ranking_genes = pd.DataFrame(ranking_genes, columns=['gene', 'score', 'hits'])\n",
    "df_hpo_ranking_genes = df_hpo_ranking_genes[['gene', 'score']]\n",
    "ACMG_result = ACMG.Get_ACMG_result(df_hpo_ranking_genes, variants, df_pubmed)\n",
    "\n",
    "# filter variant on phenotype\n",
    "if phenos:\n",
    "    df_final_res = filterVariantOnPhenotype.generateOutput(variants, ACMG_result, phenos)\n",
    "else:\n",
    "    df_final_res = ACMG_result\n",
    "\n",
    "return df_final_res, df_genes, phenos, field_names\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
